<!DOCTYPE html>












  


<html class="theme-next mist use-motion" lang="zh-CN">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">


























<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2">

<link rel="stylesheet" href="/css/main.css?v=7.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/Favico.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/Favico.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/Favico.png?v=7.0.0">


  <link rel="mask-icon" href="/images/logo.svg?v=7.0.0" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '7.0.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="1. 根据免疫组库TCRβ预测病人的CMV感染状态这个项目用到了641个样本（cohort 1），包括352 例CMV阴性(CMV-)和289例CMV阳性(CMV+) 外部验证用到了120个样本（cohort 2） 该机器学习的任务为：   讨论 TCRβ 免疫组的数据特点：   可能出现的TCRβ的集合非常大，而单个样本只能从中检测到稀疏的少数几个； 一个新样本中很可能会出现训练样本集合中未出现">
<meta name="keywords" content="Machine-Learning,Bioinformatics">
<meta property="og:type" content="article">
<meta property="og:title" content="机器学习在生物信息学中的应用">
<meta property="og:url" content="http://yoursite.com/2019/04/04/Application-of-MachineLearning-in-Bioinformatics/index.html">
<meta property="og:site_name" content="Lianm&#39;s Blog">
<meta property="og:description" content="1. 根据免疫组库TCRβ预测病人的CMV感染状态这个项目用到了641个样本（cohort 1），包括352 例CMV阴性(CMV-)和289例CMV阳性(CMV+) 外部验证用到了120个样本（cohort 2） 该机器学习的任务为：   讨论 TCRβ 免疫组的数据特点：   可能出现的TCRβ的集合非常大，而单个样本只能从中检测到稀疏的少数几个； 一个新样本中很可能会出现训练样本集合中未出现">
<meta property="og:locale" content="zh-CN">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-4.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-workflow.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-3.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-5.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-6.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-7.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-8.png">
<meta property="og:image" content="http://yoursite.com/images/T-cell-sequencing-in-cancers-colorectal-cancer-2.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-1.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-2.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-3.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-4.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-5.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-redraw-pictures-for-NGpapre-1.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-PhenotypeBurdenDot_0.05_A-B.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-5.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary_TCRβ_incidence_bias.0.05.A-B.png">
<meta property="og:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-supplementary-calculate-FDR.png">
<meta property="og:updated_time" content="2019-05-05T08:57:59.589Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="机器学习在生物信息学中的应用">
<meta name="twitter:description" content="1. 根据免疫组库TCRβ预测病人的CMV感染状态这个项目用到了641个样本（cohort 1），包括352 例CMV阴性(CMV-)和289例CMV阳性(CMV+) 外部验证用到了120个样本（cohort 2） 该机器学习的任务为：   讨论 TCRβ 免疫组的数据特点：   可能出现的TCRβ的集合非常大，而单个样本只能从中检测到稀疏的少数几个； 一个新样本中很可能会出现训练样本集合中未出现">
<meta name="twitter:image" content="http://yoursite.com/images/App-ML-in-Bioinfo-TCRbeta-classification-4.png">






  <link rel="canonical" href="http://yoursite.com/2019/04/04/Application-of-MachineLearning-in-Bioinformatics/">



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>机器学习在生物信息学中的应用 | Lianm's Blog</title>
  












  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Lianm's Blog</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">Focus on Bioinformatics and Machine-Learning</p>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="切换导航栏">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>首页</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br>关于</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">

    
    
    
      
    

    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br>标签</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">

    
    
    
      
    

    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br>分类</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>归档</a>

  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>搜索</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    
  
  
  
  

  

  <a href="https://github.com/Ming-Lian" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="80" height="80" viewbox="0 0 250 250" style="fill: #222; color: #fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a>



    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/04/04/Application-of-MachineLearning-in-Bioinformatics/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content="中国科学院北京基因组研究所研究生">
      <meta itemprop="image" content="/images/avatar.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">机器学习在生物信息学中的应用

              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2019-04-04 10:15:20" itemprop="dateCreated datePublished" datetime="2019-04-04T10:15:20+08:00">2019-04-04</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2019-05-05 16:57:59" itemprop="dateModified" datetime="2019-05-05T16:57:59+08:00">2019-05-05</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Machine-Learning/" itemprop="url" rel="index"><span itemprop="name">Machine-Learning</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="1-根据免疫组库TCRβ预测病人的CMV感染状态"><a href="#1-根据免疫组库TCRβ预测病人的CMV感染状态" class="headerlink" title="1. 根据免疫组库TCRβ预测病人的CMV感染状态"></a>1. 根据免疫组库TCRβ预测病人的CMV感染状态</h2><p>这个项目用到了641个样本（cohort 1），包括352 例CMV阴性(CMV<sub>-</sub>)和289例CMV阳性(CMV<sub>+</sub>)</p>
<p>外部验证用到了120个样本（cohort 2）</p>
<p>该机器学习的任务为：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-4.png" alt></p>

<p>讨论 TCRβ 免疫组的数据特点：</p>
<blockquote>
<ul>
<li>可能出现的TCRβ的集合非常大，而单个样本只能从中检测到稀疏的少数几个；</li>
<li>一个新样本中很可能会出现训练样本集合中未出现的TCRβ克隆类型；</li>
<li>对于一个给定的TCR，它对给定抗原肽的结合亲和力会受到HLA类型的调控，因此原始的用于判别分析的特征集合还受到隐变量——HLA类型的影响；</li>
</ul>
</blockquote>
<p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-workflow.png" alt></p>

<ul>
<li><p>特征选择：鉴定表型相关的TCRβs</p>
<p>  使用Fisher精确检验（单尾检验，具体实现过程请查看文末 <a href="#2-1-Fisher检验筛选CMV阳性（CMV-）相关克隆">*2.1. Fisher检验筛选CMV阳性（CMV<sup>+</sup>）相关克隆</a>）：</p>
  <p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-3.png" alt></p>

<p>  Fisher检验的阈值设为：$P&lt;1\times 10^{-4}$，FDR&lt;0.14（该FDR的计算方法见文末 <a href="#3-FDR的计算方法">*3. FDR的计算方法</a>，且阈值的选择问题本质上是一个模型选择问题 (model selection) ，该问题会在这部分靠后的位置进行讨论），且富集在CMV<sup>+</sup>样本中，从而得到与CMV<sup>+</sup>相关的CDR3克隆集合，共有164个</p>
<p>  通过下面的TCRβ克隆在两组中的发生率的散点图可以明显地看到，筛出的表型相关的TCRβ克隆的确显著地表达在CMV+组中</p>
  <p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-5.png" alt></p>
</li>
<li><p>计算表型负荷（phenotype burden）</p>
<p>  一个样本的表型负荷（phenotype burden）被定义为：</p>
<blockquote>
<p>该样本的所有unique TCRβs中与阳性表型相关的克隆的数量所占的比例</p>
<p>若阳性表型相关的克隆的集合记为CDR，样本i的unique克隆集合记为CDR<sub>i</sub>，则它的表型负荷为：</p>
<p>$$PB_i=\frac{||CDR_i \cap CDR||}{||CDR_i||}$$</p>
<p>其中$||·||$表示集合<code>·</code>中元素的数量</p>
</blockquote>
<p>  下图是将上面的表型负荷计算公式中的分子与分母分别作为纵轴和横轴，画成二维的散点图</p>
  <p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-6.png" alt></p>

<p>  可以明显地看出两类样本在这个层面上来看，有很好的区分度</p>
</li>
<li><p>基于二项分布的贝叶斯判别模型</p>
<p>  基本思想：</p>
<blockquote>
<p>对于$CMV^+$相关TCR数为$k’$，total unique TCR数为$n’$的样本，认为它一个概率为它的表型负荷$\theta’$（$\theta’$服从Beta分布），$n=n’$, $k=k’$的二项分布（伯努利分布），根据贝叶斯思想，构造最优贝叶斯分类器，即</p>
<p>$$h(k’,n’)=arg \max_{x \in \{+,\,-\}} p(c=x \mid k’,\,n’)$$</p>
<p>其中</p>
<p>$$p(c=x \mid k’,\,n’)=\frac{p(k’ \mid n’,\,c)p(c)}{p(k’)}$$</p>
<p>而$p(k’)$是一个常数，对分类器的结果没有影响，可以省略</p>
</blockquote>
<p>  那么就需要根据训练集估计：</p>
<ul>
<li>类先验概率$p(c)$</li>
<li><p>类条件概率（似然）$p(k’ \mid n’,\,c)$</p>
<p><strong>（1）首先根据概率图模型推出单个样本的概率表示公式</strong></p>
<p>概率图模型如下：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-7.png" alt></p>

<p>则对于j样本，我们可以算出它的 $\theta_i$ 的后验分布：</p>
<p>$$p(\theta_i \mid y_{ij},\,\alpha_i,\,\beta_i)=\frac{p(\theta_i \mid \alpha_i,\,\beta_i)p(y_{ij} \mid \theta_i)}{p(y_{ij})}$$</p>
<p>其中，</p>
<blockquote>
<ul>
<li>$p(y_{ij})$：表示事件$(k_{ij} \mid n_{ij})$发生的概率，即$p(k_{ij}\mid n_{ij},\,c_i)$</li>
<li>$p(y_{ij} \mid \theta_i)$：表示$Binomial(k_{ij} \mid n_{ij},\,\theta_i)=\left(\begin{matrix}n_{ij}\ k_{ij}\end{matrix}\right)\theta_i^{k_{ij}}(1-\theta_i)^{n_{ij}-k_{}ij}$</li>
<li>$p(\theta_i \mid \alpha_i,\,\beta_i)$：表示$\theta_i$的先验分布$Beta(\theta_i\mid\alpha_i,\,\beta_i)$</li>
</ul>
</blockquote>
<p>对上面的公式进一步推导</p>
<p>$$\begin{aligned}<br>&amp;p(\theta_i \mid y_{ij},\,\alpha_i,\,\beta_i) \\<br>&amp;=\frac{p(\theta_i \mid \alpha_i,\,\beta_i)p(y_{ij} \mid \theta_i)}{p(y_{ij})}\\<br>&amp;=\frac{1}{p(y_{ij})}Beta(\theta_i \mid \alpha_i,\,\beta_i)Binomial(k_{ij}\mid n_{ij},\,\theta_i) \\<br>&amp;=\frac{1}{p(y_{ij})} \quad {\frac{1}{B(\alpha_i,\,\beta_i)}\theta_i^{\alpha_i-1}(1-\theta_i)^{\beta_i-1}} \quad {\left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right)\theta_i^{k_{ij}}(1-\theta_i)^{n_{ij}-k_{ij}}} \\<br>&amp;=\frac{1}{p(y_{ij})} \quad \frac{1}{B(\alpha_i,\,\beta_i)} \left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right) \theta_i^{(\alpha_i+k_{ij})-1}(1-\theta_i)^{(\beta_i+n_{ij}-k_{ij})-1} \\<br>&amp;=\frac{1}{p(y_{ij})} \quad  \frac{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})}{B(\alpha_i,\,\beta_i)} \left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right) \frac{1}{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})} \theta_i^{(\alpha_i+k_{ij})-1}(1-\theta_i)^{(\beta_i+n_{ij}-k_{ij})-1} \\<br>&amp;=\frac{1}{p(y_{ij})} \quad  \frac{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})}{B(\alpha_i,\,\beta_i)} \left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right) \quad Beta(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})<br>\end{aligned}<br>$$</p>
<p>根据Beta分布的先验分布的，已知</p>
<p>$$p(\theta_i \mid y_{ij},\,\alpha_i,\,\beta_i)=Beta(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})$$</p>
<p>因此</p>
<p>$$p(k_{ij}\mid n_{ij},\,c_i)=p(y_{ij})=\left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right) \frac{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})}{B(\alpha_i,\,\beta_i)} $$</p>
<p>这样我们就得到单个样本的概率表示公式，其中$B(·)$是Beta函数，从上面的表达式中，我们可以看出$p(k_{ij}\mid n_{ij},\,c_i)$是$\alpha_i$和$\beta_i$的函数</p>
<p><strong>（2）优化每个组的表型负荷 $\theta_i$ 的先验分布的两个参数 $\alpha_i$ 和 $\theta_i$——最大似然法</strong></p>
<p>我们要最大化$c_i$组的样本集合它们的联合概率：</p>
<p>$$<br>\begin{aligned}<br>p(k_i \mid n_i,\,c_i) &amp;=\prod_{j,j \in c_i} \left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right) \frac{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})}{B(\alpha_i,\,\beta_i)}\\<br>&amp;= \prod_{j,j \in c_i} \left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right) \quad \prod_{j,j \in c_i} \frac{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})}{B(\alpha_i,\,\beta_i)}<br>\end{aligned}<br>$$</p>
<p>其中，$\prod_{j,\,j \in c_i} \left(\begin{matrix}n_{ij}\\ k_{ij}\end{matrix}\right)$是常数，可以省略，则</p>
<p>$$p(k_i \mid n_i,\,c_i)=\prod_{j,j \in c_i} \frac{B(\alpha_i+k_{ij},\,\beta_i+n_{ij}-k_{ij})}{B(\alpha_i,\,\beta_i)}$$</p>
<p>对它取对数，得到</p>
<p>$$l(\alpha_i,\,\beta_i)=\log p(k_i \mid n_i,\,c_i) \\ = \sum_{j,\,j \in c_i} \log B(\alpha_i+k_{ij}, \, \beta_i+n_{ij}-k_{ij})-N_i\log B(\alpha_i, \, \beta_i)$$</p>
<p>其中，$N_i$是属于$c_i$组的样本数</p>
<p>因此优化目标为：</p>
<p>$$(\alpha_i^*,\beta_i^*)=arg \, \max\limits_{\alpha_i,\,\beta_i}l(\alpha_i,\,\beta_i)$$</p>
<p>$l(\alpha_i,\,\beta_i)$分别对$\alpha_i$和$\beta_i$求偏导</p>
<p>$$\frac{\partial l}{\partial \alpha_i}=-N_i(\psi(\alpha_i)-\psi(\alpha_i+\beta_i))+\sum_{j,\,j \in c_i}(\psi(\alpha_i+k_{ij})-\psi(\alpha_i+\beta_i+n_{ij}))$$</p>
<p>$$\frac{\partial l}{\partial \beta_i}=-N_i(\psi(\beta_i)-\psi(\alpha_i+\beta_i))+\sum_{j,\,j \in c_i}(\psi(\beta_i+n_{ij}-k_{ij})-\psi(\alpha_i+\beta_i+n_{ij}))$$</p>
<p>其中，$\psi(·)$是伽马函数</p>
<p>使用梯度上升（gradient ascent）法来求解优化目标，其中梯度的公式为：</p>
<p>$$\alpha_i := \alpha_i+\alpha\frac{\partial l}{\partial \alpha_i}\\ \beta_i := \beta_i+\alpha\frac{\partial l}{\partial \beta_i}$$</p>
<p>最终得到的解记为$\alpha_i^*$和$\beta_i^*$，其中</p>
<p>$$<br>\alpha_+^*=4.0,\quad \beta_+^*=51,820.1\\<br>\alpha_-^*=26.7,\quad \beta_-^*=2,814,963.8<br>$$</p>
<p>（3）根据训练好的分类器对新样本进行分类</p>
<p>分类器为</p>
<p>$$<br>\begin{aligned}<br>h(k’,n’) &amp;= arg \max_{x \in \{CMV^+,\,CMV^-\}} p(c=x \mid n’,k’) \\<br>&amp;=arg \max_{x \in \{CMV^+,\,CMV^-\}} p(k’ \mid n’, c=x)p(c=x) \\<br>&amp;=arg \max_{x \in \{CMV^+,\,CMV^-\}}\left(\begin{matrix}n’ \\ k’\end{matrix}\right)\frac{B(\alpha_x^*+k’,\,\beta_x^*+n’-k’)}{B(\alpha_x^*,\,\beta_x^*)}\frac{N_x}{N}<br>\end{aligned}<br>$$</p>
<p>  （4）选择合适的的阈值：model selection</p>
<p>使用交叉验证法中的留一法 (leave-one-out)来进行模型选择</p>
<p>定义该问题的优化目标为最小化cross-entropy loss：</p>
<p>$$L(\phi)=-\frac{1}{N}\sum_{i=1}^N [c_i\log q_i(\phi)+(1-c_i)\log (1-q_i(\phi))]$$</p>
<p>其中，$c_i$表示样本$i$的所属类别，$q_i(\phi)$表示样本$i$被判为$CMV^+$的概率，N为训练样本数（N个训练样本数的训练集它总共有N种留一法组合形式）</p>
<p>分析优化目标为什么选择最小化cross-entropy loss：</p>
<blockquote>
<p>我们的目的是选择合适的阈值，那么什么叫合适的阈值？就是在该阈值下，能筛选出合适的features，基于这些features训练出的模型能有足够高的泛化性能，即模型的variance足够的小，也就是对于测试数据（这里就是留一法中的那一个样本）有足够高的预测准确性，定量化描述就是N次留一法的准确预测的概率的均值（几何平均数）最大化，即</p>
<p>$$\max_{\phi} \sqrt[N]{\prod_i^N q_i(\phi)^{c_i}(1-q_i(\phi)^{1-c_i}} \\ s.t.\quad c_i \in {0,\,1}$$</p>
<p>由于取对数不影响优化方向，所以</p>
<p>$$\begin{aligned}&amp;\log \left[\prod_i^N q_i(\phi)^{c_i}(1-q_i(\phi)^{1-c_i} \right]^{1/N} \\ &amp;=\frac{1}{N}\log \prod_i^N q_i(\phi)^{c_i}(1-q_i(\phi)^{1-c_i} \\ &amp;=\frac{1}{N}\sum_{i=1}^N [c_i\log q_i(\phi)+(1-c_i)\log (1-q_i(\phi))] \end{aligned}$$</p>
<p>从而得到上面的优化目标</p>
</blockquote>
<p>最后得到的分类器的性能非常高</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-8.png" alt></p>

</li>
</ul>
</li>
</ul>
<h2 id="2-SC3：单细胞表达谱的无监督聚类"><a href="#2-SC3：单细胞表达谱的无监督聚类" class="headerlink" title="2. SC3：单细胞表达谱的无监督聚类"></a>2. SC3：单细胞表达谱的无监督聚类</h2><p>该聚类方法名为SC3（Single-Cell Consensus Clustering）</p>
<p>该方法本质上就是K-means聚类，不过在执行K-means聚类的前后进行了一些特殊的操作：</p>
<blockquote>
<ul>
<li><strong>k-means聚类前</strong>：进行了数据预处理，即特征的构造，称为特征工程，该方法中是对输入的原始特征空间进行PCA变换或拉普拉斯矩阵变换，对变换后的新特征矩阵逐渐增加提取的主成分数，来构造一系列新特征；</li>
<li><strong>k-means聚类后</strong>：特征工程构造出来的一系列新特征集合，基于这些新特征集合通过k-means聚类能得到一系列不同的聚类结果，尝试对这些聚类结果总结出consensus clustering</li>
</ul>
</blockquote>
<p align="center"><img src="/images/T-cell-sequencing-in-cancers-colorectal-cancer-2.png" alt></p>

<p>本人比较好奇的地方是：<strong>怎么从一系列不同的聚类结果中总结出consensus clustering？</strong></p>
<blockquote>
<p>使用CSPA算法（cluster-based similarity partitioning algorithm）</p>
<p>（1）对每一个聚类结果按照以下方法构造二值相似度矩阵S：如果两个样本i和j在该聚类结果中被聚到同一个集合中，则它们之间的相似度为1，在二值相似度矩阵中对应的值 S<sub>i,j</sub> = 1，否则S<sub>i,j</sub> = 0；</p>
<p>（2）对所有的聚类结果的二值相似度矩阵S取平均，得到consensus matrix；</p>
<p>（3）基于consensus matrix进行层次聚类，得到最终的consensus clustering；</p>
</blockquote>
<h2 id="3-GATK的VQSR"><a href="#3-GATK的VQSR" class="headerlink" title="3. GATK的VQSR"></a>3. GATK的VQSR</h2><h2 id="补充知识"><a href="#补充知识" class="headerlink" title="补充知识"></a>补充知识</h2><h3 id="1-beta分布"><a href="#1-beta分布" class="headerlink" title="*1. beta分布"></a>*1. beta分布</h3><h4 id="1-1-什么是beta分布"><a href="#1-1-什么是beta分布" class="headerlink" title="*1.1. 什么是beta分布"></a>*1.1. 什么是beta分布</h4><p>对于硬币或者骰子这样的简单实验，我们事先能很准确地掌握系统成功的概率。</p>
<p>然而通常情况下，系统成功的概率是未知的，但是根据频率学派的观点，我们可以<strong>通过频率来估计概率</strong>。为了测试系统的成功概率，我们做n次试验，统计成功的次数k，于是很直观地就可以计算出。然而由于系统成功的概率是未知的，这个公式计算出的只是系统成功概率的<strong>最佳估计</strong>。也就是说实际上也可能为其它的值，只是为其它的值的概率较小。因此我们并不能完全确定硬币出现正面的概率就是该值，所以也是一个随机变量，它符合Beta分布，其取值范围为0到1</p>
<p>用一句话来说，beta分布可以看作一个<strong>概率的概率密度分布</strong>，当你不知道一个东西的具体概率是多少时，它可以给出了所有概率出现的可能性大小。</p>
<p>Beta分布有和两个参数α和β，其中α为成功次数加1，β为失败次数加1。</p>
<h4 id="1-2-理解beta分布"><a href="#1-2-理解beta分布" class="headerlink" title="*1.2. 理解beta分布"></a>*1.2. 理解beta分布</h4><p>举一个简单的例子，熟悉棒球运动的都知道有一个指标就是棒球击球率(batting average)，就是用一个运动员击中的球数除以击球的总数，我们一般认为0.266是正常水平的击球率，而如果击球率高达0.3就被认为是非常优秀的。</p>
<p>现在有一个棒球运动员，我们希望能够预测他在这一赛季中的棒球击球率是多少。你可能就会直接计算棒球击球率，用击中的数除以击球数，但是如果这个棒球运动员只打了一次，而且还命中了，那么他就击球率就是100%了，这显然是不合理的，因为根据棒球的历史信息，我们知道这个击球率应该是0.215到0.36之间才对啊。</p>
<p>对于这个问题，一个最好的方法来表示这些经验（在统计中称为先验信息）就是用beta分布，这表示在我们没有看到这个运动员打球之前，我们就有了一个大概的范围。beta分布的定义域是(0,1)这就跟概率的范围是一样的。<br>s<br>接下来我们将这些先验信息转换为beta分布的参数，我们知道一个击球率应该是平均0.27左右，而他的范围是0.21到0.35，那么根据这个信息，我们可以取α=81,β=219</p>
<p>之所以取这两个参数是因为：</p>
<blockquote>
<ul>
<li><p>beta分布的均值</p>
  <p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-1.png" alt></p>
</li>
<li><p>这个分布主要落在了(0.2,0.35)间，这是从经验中得出的合理的范围</p>
</li>
</ul>
</blockquote>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-2.png" alt></p>

<p align="center">Beta(81, 219)</p>

<p>在这个例子里，我们的x轴就表示各个击球率的取值，x对应的y值就是这个击球率所对应的概率密度。也就是说beta分布可以看作一个概率的概率密度分布。</p>
<p>那么有了先验信息后，现在我们考虑一个运动员只打一次球，那么他现在的数据就是“1中；1击”。这时候我们就可以更新我们的分布了，让这个曲线做一些移动去适应我们的新信息，移动的方法很简单</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-3.png" alt></p>

<p>其中α0和β0是一开始的参数，在这里是81和219。所以在这一例子里，α增加了1（击中了一次）。β没有增加(没有漏球)。这就是我们的新的beta分布 Beta(81+1,219)，我们跟原来的比较一下：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-4.png" alt></p>

<p>可以看到这个分布其实没多大变化，这是因为只打了1次球并不能说明什么问题。但是如果我们得到了更多的数据，假设一共打了300次，其中击中了100次，200次没击中，那么这一新分布就是 Beta(81+100,219+200) ： </p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-knowledge-beta-distribution-5.png" alt></p>

<p>注意到这个曲线变得更加尖，并且平移到了一个右边的位置，表示比平均水平要高</p>
<p>有趣的现象：</p>
<blockquote>
<p>根据这个新的beta分布，我们可以得出他的数学期望为：αα+β=82+10082+100+219+200=.303 ，这一结果要比直接的估计要小 100100+200=.333 。你可能已经意识到，我们事实上就是在这个运动员在击球之前可以理解为他已经成功了81次，失败了219次这样一个先验信息。</p>
</blockquote>
<h3 id="2-重复NG免疫组库TCRβ文章的图"><a href="#2-重复NG免疫组库TCRβ文章的图" class="headerlink" title="*2. 重复NG免疫组库TCRβ文章的图"></a>*2. 重复NG免疫组库TCRβ文章的图</h3><h4 id="2-1-Fisher检验筛选CMV阳性（CMV-）相关克隆"><a href="#2-1-Fisher检验筛选CMV阳性（CMV-）相关克隆" class="headerlink" title="*2.1. Fisher检验筛选CMV阳性（CMV+）相关克隆"></a>*2.1. Fisher检验筛选CMV阳性（CMV<sup>+</sup>）相关克隆</h4><p>目的是得到类似下面的结果：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-redraw-pictures-for-NGpapre-1.png" alt></p>

<p>提供的输入：</p>
<blockquote>
<ul>
<li>Observed matrix：行为CDR3克隆，列为样本，矩阵中的值只能取0或1，0表示该feature在该样本中未被观测到，1为被观测到了</li>
<li>样本的分组信息文件（该文件非必须，若样本名中包含分组信息，则可以不提供这个文件，只需提供分从样本名中提取分组信息的正则表达式）</li>
<li>当前感兴趣的两个组，因为考虑到可能提供的样本是多组的，而Fisher检验只能进行两组间的比较，所以多组时需要指定分析的两组是谁</li>
</ul>
</blockquote>
<p>实现的思路：</p>
<blockquote>
<p>原始提供的输入文件是profile matrix，设定一个阈值（默认为0），将大于阈值的设为1，等于阈值的设为0，从而得到Observed matrix。这一步操作不在下面脚本的功能当中，需要自己在执行下面的脚本之前完成这个操作；</p>
<p>将准备好的Observed matrix输入，行为TCR克隆（feature），列为sample，因为是要对TCR克隆分析它与分组的相关性，因此每次对矩阵的行执行Fisher检验，根据提供的样本的分组信息，得到该TCR克隆的2X2列联表</p>
<p>$$\begin{matrix}\hline&amp; Group \\ Observe &amp; A &amp; B \\ \hline Yes &amp; a &amp; b \\ \hline No &amp; c &amp; d \\ \hline \end{matrix}$$</p>
<p>然后再对这个列联表计算Fisher检验的p-value，将TCR克隆，列联表中的a，b和算出的p-value输出，就得到了我们想要的结果</p>
</blockquote>
<p>脚本名：<code>FisherTestForMatrix.R</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">####################################################################</span><br><span class="line"># 该脚本用于对样本的Oberved矩阵执行Fisher精确检验（单尾），用于 #</span><br><span class="line"># 检测feature与分组的相关性                                       #</span><br><span class="line">####################################################################</span><br><span class="line"></span><br><span class="line"># 参数说明：</span><br><span class="line"># - （1）样本的Oberved矩阵，行为特征，列为样本，矩阵中的值只能取0</span><br><span class="line">#	或1，0表示该feature在该样本中未被观测到，1为被观测到了</span><br><span class="line">#</span><br><span class="line"># - （2）是否提供样本的分组信息文件，1为是，0位否</span><br><span class="line">#</span><br><span class="line"># - （3）若上一个参数选择1（提供分组信息文件），则该参数应该设为</span><br><span class="line">#	分组信息文件的路径，分组文件要求至少包含两列——SampleId和Group；</span><br><span class="line">#	若上一个参数选择0（不提供分组信息文件），则认为分组信息已经包</span><br><span class="line">#	含的样本的命名中，则可以通过提供分组信息在样本名中匹配的正则</span><br><span class="line">#	表达式</span><br><span class="line">#</span><br><span class="line"># - （4）感兴趣的两组，例如是A组和B组，则写成&quot;A-B&quot;，即中间用连字</span><br><span class="line">#	符连接</span><br><span class="line">#</span><br><span class="line"># - （5）设置的p值阈值</span><br><span class="line">#</span><br><span class="line"># - （6）设置并行化线程数，若不设置这个参数，则默认不采用并行化计</span><br><span class="line">#	算方法</span><br><span class="line"></span><br><span class="line">library(stringr)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(parallel)</span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(TRUE)</span><br><span class="line">MatFile &lt;- Args[1]</span><br><span class="line">Bool_GroupFile &lt;- Args[2]</span><br><span class="line">GroupFile &lt;- Args[3]</span><br><span class="line">TargetGroups &lt;- Args[4]</span><br><span class="line">pvalue &lt;- Args[5]</span><br><span class="line"></span><br><span class="line"># 开启并行化</span><br><span class="line">if(!is.na(Args[6]))&#123;</span><br><span class="line">	no_cores &lt;- as.integer(Args[6])</span><br><span class="line">	cl &lt;- makeCluster(no_cores)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">Matrix &lt;- read.table(MatFile, header=T, row.names=1)</span><br><span class="line">TargetGroups &lt;- unlist(strsplit(TargetGroups, &apos;-&apos;))</span><br><span class="line"></span><br><span class="line">print(&quot;Loading Observed Matrix successfully&quot;)</span><br><span class="line"></span><br><span class="line">#################################################</span><br><span class="line"># 1. 获得Oberved矩阵样本对应的分组信息</span><br><span class="line">#################################################</span><br><span class="line"></span><br><span class="line">if(Bool_GroupFile %in% c(1, 0))&#123;</span><br><span class="line">	# 从提供的分组信息文件中获得</span><br><span class="line">	if(Bool_GroupFile==1)&#123;</span><br><span class="line">		Group &lt;- read.table(GroupFile, header=T)</span><br><span class="line">		# 考虑到可能存在Observe矩阵列名命名不规范，即以数字起始（会在开头</span><br><span class="line">		#	添加X字符），或其中包含连字符（将连字符替换为点）因此需要将</span><br><span class="line">		#	Group变量中的SampleId进行相应的替换，以保证一致</span><br><span class="line">		colname_matrix &lt;- ifelse(grepl(&apos;^X&apos;,colnames(Matrix)),sub(&apos;^X&apos;,&apos;&apos;,colnames(Matrix)),colnames(Matrix))</span><br><span class="line">		colname_matrix &lt;- gsub(&apos;\\.&apos;,&apos;-&apos;,colname_matrix)</span><br><span class="line">		SampleGroup &lt;- Group$Group[match(colname_matrix, Group$SampleId)]</span><br><span class="line">	# 从样本名中用正则提取</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		GroupPattern &lt;- GroupFile</span><br><span class="line">		SampleGroup &lt;- unlist(str_extract(colnames(Matrix), GroupPattern))</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	stop(&quot;参数指定错误！第二个参数必须为0或1&quot;)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">print(&quot;Load/get group info for coresponding samples in each col in Observed Matrix&quot;)</span><br><span class="line"></span><br><span class="line">#################################################</span><br><span class="line"># 2. 对每个feature执行Fisher检验，得到检验结果</span><br><span class="line">#################################################</span><br><span class="line"></span><br><span class="line"># 用于执行fisher检验的函数，最终返回的是TRUE或FALSE</span><br><span class="line">FisherTest &lt;- function(ObserveVec, GroupIndex, TargetGroups)&#123;</span><br><span class="line">	# 初始化列联表</span><br><span class="line">	FisherMat &lt;- matrix(c(0,0,0,0),</span><br><span class="line">						nrow = 2,</span><br><span class="line">						dimnames = list(Observe=c(&apos;Yes&apos;,&apos;No&apos;),</span><br><span class="line">										Group=TargetGroups</span><br><span class="line">										))</span><br><span class="line">	# 为列联表的每一项填上对应的值</span><br><span class="line">	FisherMat[1,1] &lt;- sum(ObserveVec==1&amp;GroupIndex==colnames(FisherMat)[1])</span><br><span class="line">	FisherMat[1,2] &lt;- sum(ObserveVec==1&amp;GroupIndex==colnames(FisherMat)[2])</span><br><span class="line">	FisherMat[2,1] &lt;- sum(ObserveVec==0&amp;GroupIndex==colnames(FisherMat)[1])</span><br><span class="line">	FisherMat[2,2] &lt;- sum(ObserveVec==0&amp;GroupIndex==colnames(FisherMat)[2])</span><br><span class="line">	# 进行fisher检验，得到p值</span><br><span class="line">	pvalue &lt;- fisher.test(FisherMat, alternative = &quot;two.sided&quot;)$p.value</span><br><span class="line">	# 返回向量：组1计数、组2计数、p值</span><br><span class="line">	c(FisherMat[1,1], FisherMat[1,2], pvalue)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"># 为每个feature（即矩阵的行）执行fisher检验，得到是每个feature的pvalue，每一个的返回值以列形式进行追加</span><br><span class="line">if(!is.na(Args[6]))&#123;</span><br><span class="line">	StatOut &lt;- parApply(cl,Matrix, 1, FisherTest, SampleGroup, TargetGroups)</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	StatOut &lt;- apply(Matrix, 1, FisherTest, SampleGroup, TargetGroups)</span><br><span class="line">&#125;</span><br><span class="line">StatOut_Table &lt;- data.frame(Feature=colnames(StatOut),Group1=StatOut[1,],Group2=StatOut[2,],Pval=StatOut[3,])</span><br><span class="line">colnames(StatOut_Table) &lt;- c(&apos;Feature&apos;, paste(&quot;Group_&quot;,TargetGroups[1],sep=&apos;&apos;), paste(&quot;Group_&quot;,TargetGroups[2],sep=&apos;&apos;), &apos;Pval&apos;)</span><br><span class="line"></span><br><span class="line">print(&quot;Finish Fisher&apos;s Exact Test for each feature&quot;)</span><br><span class="line"></span><br><span class="line"># 将Fisher检验结果写入文件</span><br><span class="line">write.table(StatOut_Table, paste(str_extract(MatFile,&apos;^(.*?)\\.&apos;),TargetGroups[1],&quot;-&quot;,TargetGroups[2],&quot;.stat&quot;,sep=&apos;&apos;),row.names=F,col.names=T,sep=&quot;\t&quot;,quote=F)</span><br><span class="line"></span><br><span class="line">print(&quot;Finish writing Fisher&apos;s Exact Test Output into file&quot;)</span><br><span class="line"></span><br><span class="line">#################################################</span><br><span class="line"># 3. 表型负荷（Phenotype Burden）相关的计算与画图</span><br><span class="line">#################################################</span><br><span class="line"></span><br><span class="line"># 计算每个样本的表型负荷相关的两个值，该样本中与分组1相关的features数，以及该样本中观测到的features数</span><br><span class="line">if(!is.na(Args[6]))&#123;</span><br><span class="line">	RelativeFeatures &lt;- parApply(cl, Matrix[,SampleGroup %in% TargetGroups], 2, function(x,y,p,z) sum(x==1&amp;y&lt;p&amp;z), </span><br><span class="line">							StatOut_Table$Pval, </span><br><span class="line">							pvalue, </span><br><span class="line">							StatOut_Table[,2]&gt;StatOut_Table[,3])</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	RelativeFeatures &lt;- apply(Matrix[,SampleGroup %in% TargetGroups], 2, function(x,y,p,z) sum(x==1&amp;y&lt;p&amp;z), </span><br><span class="line">							StatOut_Table$Pval, </span><br><span class="line">							pvalue, </span><br><span class="line">							StatOut_Table[,2]&gt;StatOut_Table[,3])</span><br><span class="line">&#125;</span><br><span class="line">stopCluster(cl)</span><br><span class="line">TotalFeatures &lt;- colSums(Matrix[,SampleGroup %in% TargetGroups])</span><br><span class="line">PhenotypeBurden &lt;- data.frame(Relative=RelativeFeatures, Total=TotalFeatures, Group=SampleGroup[SampleGroup %in% TargetGroups])</span><br><span class="line"></span><br><span class="line">print(&quot;Finish calculate statistics for Phenotype-Burden&quot;)</span><br><span class="line"></span><br><span class="line"># 画散点图</span><br><span class="line">print(&quot;Start dotplot&quot;)</span><br><span class="line">png(paste(&quot;PhenotypeBurdenDot_&quot;,pvalue,&quot;_&quot;,TargetGroups[1],&quot;-&quot;,TargetGroups[2],&quot;.png&quot;,sep=&apos;&apos;))</span><br><span class="line">ggplot(PhenotypeBurden)+geom_point(aes(x=Total, y=Relative, color=Group))</span><br><span class="line">dev.off()</span><br><span class="line"># 保存数据</span><br><span class="line">write.table(PhenotypeBurden,paste(str_extract(MatFile,&apos;^(.*?)\\.&apos;),pvalue,&quot;.&quot;, TargetGroups[1],&quot;-&quot;,TargetGroups[2],&quot;.data&quot;,sep=&apos;&apos;),row.names=T,col.names=T,sep=&quot;\t&quot;,quote=F)</span><br></pre></td></tr></table></figure>
<p>用法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ Rscript FisherTestForMatrix.R &lt;profile matrix&gt; &lt;1|0&gt; &lt;group info|regx&gt; &lt;2 interested group&gt; &lt;pval&gt;</span><br></pre></td></tr></table></figure>
<p>具体的参数说明，请查看脚本的注释信息</p>
<h4 id="2-2-绘制表型负荷相关散点图"><a href="#2-2-绘制表型负荷相关散点图" class="headerlink" title="*2.2. 绘制表型负荷相关散点图"></a>*2.2. 绘制表型负荷相关散点图</h4><h5 id="2-2-1-训练集"><a href="#2-2-1-训练集" class="headerlink" title="*2.2.1. 训练集"></a>*2.2.1. 训练集</h5><p>在上一步 <a href="#2-1-Fisher检验筛选CMV阳性（CMV-）相关克隆">*2.1. Fisher检验筛选CMV阳性（CMV<sup>+</sup>）相关克隆</a> 的操作中会同时得到这张图</p>
<p>画出的表型负荷相关散点图如下：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-PhenotypeBurdenDot_0.05_A-B.png" alt></p>

<h5 id="2-2-2-测试集"><a href="#2-2-2-测试集" class="headerlink" title="*2.2.2. 测试集"></a>*2.2.2. 测试集</h5><p>测试集的表型负荷相关散点图需要基于训练集的结果，需要执行另外的操作</p>
<p>需要提供的输入：</p>
<blockquote>
<ul>
<li><p>测试集的Observed matrix：行为CDR3克隆，列为样本，矩阵中的值只能取0或1，0表示该feature在该样本中未被观测到，1为被观测到了</p>
</li>
<li><p>样本的分组信息文件（该文件非必须，若样本名中包含分组信息，则可以不提供这个文件，只需提供分从样本名中提取分组信息的正则表达式）</p>
</li>
<li><p>当前感兴趣的两个组，因为考虑到可能提供的样本是多组的，而Fisher检验只能进行两组间的比较，所以多组时需要指定分析的两组是谁</p>
</li>
<li><p>训练集的Fisher检验输出结果</p>
</li>
</ul>
</blockquote>
<p>脚本名：<code>PhenotypeBurden2TestCohort.R</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">###################################################</span><br><span class="line"># 该脚本用于对测试集的每个样本计算表型负荷相关 #</span><br><span class="line"># 的两个值：该样本中与分组1相关的features数，以 #</span><br><span class="line"># 及该样本中观测到的features数                   #</span><br><span class="line">###################################################</span><br><span class="line"></span><br><span class="line"># 参数说明：</span><br><span class="line"># - （1）测试集Oberved矩阵，行为特征，列为样本，矩阵中的值只能取0</span><br><span class="line">#	或1，0表示该feature在该样本中未被观测到，1为被观测到了</span><br><span class="line">#</span><br><span class="line"># - （2）是否提供样本的分组信息文件，1为是，0位否</span><br><span class="line">#</span><br><span class="line"># - （3）若上一个参数选择1（提供分组信息文件），则该参数应该设为</span><br><span class="line">#	分组信息文件的路径，分组文件要求至少包含两列——SampleId和Group；</span><br><span class="line">#	若上一个参数选择0（不提供分组信息文件），则认为分组信息已经包</span><br><span class="line">#	含的样本的命名中，则可以通过提供分组信息在样本名中匹配的正则</span><br><span class="line">#	表达式</span><br><span class="line">#</span><br><span class="line"># - （4）感兴趣的两组，例如是A组和B组，则写成&quot;A-B&quot;，即中间用连字</span><br><span class="line">#	符连接</span><br><span class="line">#</span><br><span class="line"># - （5）设置的p值阈值</span><br><span class="line">#</span><br><span class="line"># - （6）上一步对训练集执行Fisher检验的检验结果输出文件</span><br><span class="line"></span><br><span class="line">library(stringr)</span><br><span class="line">library(ggplot2)</span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(TRUE)</span><br><span class="line">MatFile &lt;- Args[1]</span><br><span class="line">Bool_GroupFile &lt;- Args[2]</span><br><span class="line">GroupFile &lt;- Args[3]</span><br><span class="line">TargetGroups &lt;- Args[4]</span><br><span class="line">pvalue &lt;- Args[5]</span><br><span class="line">StatFile &lt;- Args[6]</span><br><span class="line"></span><br><span class="line">Matrix &lt;- read.table(MatFile, header=T, row.names=1)</span><br><span class="line">TargetGroups &lt;- unlist(strsplit(TargetGroups, &apos;-&apos;))</span><br><span class="line">StatOut &lt;- read.table(StatFile, header=T)</span><br><span class="line"></span><br><span class="line">print(&quot;Loading Observed Matrix and Fisher&apos;s Exact Test output successfully&quot;)</span><br><span class="line"></span><br><span class="line">#################################################</span><br><span class="line"># 1. 获得Oberved矩阵样本对应的分组信息</span><br><span class="line">#################################################</span><br><span class="line"></span><br><span class="line">if(Bool_GroupFile %in% c(1, 0))&#123;</span><br><span class="line">	# 从提供的分组信息文件中获得</span><br><span class="line">	if(Bool_GroupFile==1)&#123;</span><br><span class="line">		Group &lt;- read.table(GroupFile, header=T)</span><br><span class="line">		# 考虑到可能存在Observe矩阵列名命名不规范，即以数字起始（会在开头</span><br><span class="line">		#	添加X字符），或其中包含连字符（将连字符替换为点）因此需要将</span><br><span class="line">		#	Group变量中的SampleId进行相应的替换，以保证一致</span><br><span class="line">		colname_matrix &lt;- ifelse(grepl(&apos;^X&apos;,colnames(Matrix)),sub(&apos;^X&apos;,&apos;&apos;,colnames(Matrix)),colnames(Matrix))</span><br><span class="line">		colname_matrix &lt;- gsub(&apos;\\.&apos;,&apos;-&apos;,colname_matrix)</span><br><span class="line">		SampleGroup &lt;- Group$Group[match(colname_matrix, Group$SampleId)]</span><br><span class="line">	# 从样本名中用正则提取</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		GroupPattern &lt;- GroupFile</span><br><span class="line">		SampleGroup &lt;- unlist(str_extract(colnames(Matrix), GroupPattern))</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	stop(&quot;参数指定错误！第二个参数必须为0或1&quot;)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">#################################################</span><br><span class="line"># 2. 表型负荷（Phenotype Burden）相关的计算与画图</span><br><span class="line">#################################################</span><br><span class="line"></span><br><span class="line"># 计算每个样本的表型负荷相关的两个值，该样本中与分组1相关的features数，以及该样本中观测到的features数</span><br><span class="line">RelativeFeatures_list &lt;- StatOut$Feature[StatOut[,2]&gt;StatOut[,3]&amp;StatOut$Pval&lt;pvalue]</span><br><span class="line">RelativeFeatures &lt;- apply(Matrix[rownames(Matrix) %in% RelativeFeatures_list,SampleGroup %in% TargetGroups], 2, sum)</span><br><span class="line">TotalFeatures &lt;- colSums(Matrix[,SampleGroup %in% TargetGroups])</span><br><span class="line">PhenotypeBurden &lt;- data.frame(Relative=RelativeFeatures, Total=TotalFeatures, Group=SampleGroup[SampleGroup %in% TargetGroups])</span><br><span class="line"></span><br><span class="line">print(&quot;Finish calculate statistics for Phenotype-Burden&quot;)</span><br><span class="line"></span><br><span class="line"># 画散点图</span><br><span class="line">print(&quot;Start dotplot&quot;)</span><br><span class="line">png(paste(&quot;PhenotypeBurdenDot_&quot;,pvalue,&quot;_&quot;,TargetGroups[1],&quot;-&quot;,TargetGroups[2],&quot;.png&quot;,sep=&apos;&apos;))</span><br><span class="line">ggplot(PhenotypeBurden)+geom_point(aes(x=Total, y=Relative, color=Group))+</span><br><span class="line">	labs(title=&quot;CMV-associated vs. total-unique&quot;)</span><br><span class="line">dev.off()</span><br><span class="line"># 保存数据</span><br><span class="line">write.table(PhenotypeBurden,paste(str_extract(MatFile,&apos;^(.*?)\\.&apos;),pvalue,&quot;.&quot;, TargetGroups[1],&quot;-&quot;,TargetGroups[2],&quot;.data&quot;,sep=&apos;&apos;),row.names=T,col.names=T,sep=&quot;\t&quot;,quote=F)</span><br></pre></td></tr></table></figure>
<p>用法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ Rscript PhenotypeBurden2TestCohort.R &lt;profile matrix&gt; &lt;1|0&gt; &lt;group info|regx&gt; &lt;2 interested group&gt; &lt;pval&gt; &lt;stat&gt;</span><br></pre></td></tr></table></figure>
<h4 id="2-3-TCRβ在两组间分布偏好性的散点图"><a href="#2-3-TCRβ在两组间分布偏好性的散点图" class="headerlink" title="*2.3. TCRβ在两组间分布偏好性的散点图"></a>*2.3. TCRβ在两组间分布偏好性的散点图</h4><p>就是画这幅图：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-TCRbeta-classification-5.png" alt></p>

<p>注意：这幅图的横纵坐标都进行了log10变换</p>
<p>在上一步 <a href="#2-1-Fisher检验筛选CMV阳性（CMV-）相关克隆">*2.1. Fisher检验筛选CMV阳性（CMV<sup>+</sup>）相关克隆</a> 的操作中会同时得到这张图</p>
<p>画出的散点图如下：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary_TCRβ_incidence_bias.0.05.A-B.png" alt></p>

<h3 id="3-FDR的计算方法"><a href="#3-FDR的计算方法" class="headerlink" title="*3. FDR的计算方法"></a>*3. FDR的计算方法</h3><h4 id="3-1-回顾那些统计检验方法"><a href="#3-1-回顾那些统计检验方法" class="headerlink" title="*3.1. 回顾那些统计检验方法"></a>*3.1. 回顾那些统计检验方法</h4><h5 id="3-1-1-T-test与Moderated-t-Test"><a href="#3-1-1-T-test与Moderated-t-Test" class="headerlink" title="*3.1.1. T-test与Moderated t-Test"></a>*3.1.1. T-test与Moderated t-Test</h5><p>t-test的统计量：</p>
<p>$$<br>t= \frac{\overline X_1(i)-\overline X_2(i)}{S(i)}<br>$$</p>
<p>Moderated t-Test的统计量：</p>
<p>$$<br>d= \frac{\overline X_1(i)-\overline X_2(i)}{S(i)+S_0}<br>$$</p>
<p>Moderated t-Test的统计量d与t-test的t的计算方法很相似，差别就在于分母中方差的计算方法，</p>
<table>
<thead>
<tr>
<th style="text-align:center"><code></code></th>
<th style="text-align:center">T1</th>
<th style="text-align:center">T2</th>
<th style="text-align:center">T3</th>
<th style="text-align:center">C1</th>
<th style="text-align:center">C2</th>
<th style="text-align:center">C3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">Gene</td>
<td style="text-align:center">XT,1</td>
<td style="text-align:center">XT,2</td>
<td style="text-align:center">XT,3</td>
<td style="text-align:center">XC,1</td>
<td style="text-align:center">XC,2</td>
<td style="text-align:center">XC,3</td>
</tr>
</tbody>
</table>
<p>由上面展示的该基因的实际样本分组，计算出方差$S(i)=S_{X_1X_2}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}$</p>
<p>然后随机打乱上面的样本分组，得到：</p>
<table>
<thead>
<tr>
<th style="text-align:center"><code></code></th>
<th style="text-align:center">T1</th>
<th style="text-align:center">T2</th>
<th style="text-align:center">T3</th>
<th style="text-align:center">C1</th>
<th style="text-align:center">C2</th>
<th style="text-align:center">C3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">Gene</td>
<td style="text-align:center">XC,2</td>
<td style="text-align:center">XT,1</td>
<td style="text-align:center">XC,3</td>
<td style="text-align:center">XT,3</td>
<td style="text-align:center">XT,2</td>
<td style="text-align:center">XC,1</td>
</tr>
</tbody>
</table>
<p>根据打乱的结果算出$S_0$，进行n次这样的随机打乱，计算得到$d_1,d_2,…,d_n$</p>
<p>最后算出它的P值：</p>
<p>$$P=\frac{\# \{d_i \ge d,i=1,2,…,n\}}{n} $$</p>
<p>之所以不用t检验的统计量查表法，是因为Moderated t-Test的统计量已经不再符合某种统计分布了，而且这样算出来的P值也具有一定的统计意义</p>
<h4 id="3-2-多重假设检验的必要性"><a href="#3-2-多重假设检验的必要性" class="headerlink" title="*3.2. 多重假设检验的必要性"></a>*3.2. 多重假设检验的必要性</h4><p>统计学中的假设检验的基本思路是：</p>
<blockquote>
<p>设立零假设（null hypothesis）$H_0$，以及与零假设$H_0$相对应的非零假设（alternative hypothesis， or reject null hypothesis）$H_1$，在假设$H_0$成立的前提下，计算出$H_0$发生的概率，若$H_0$的发生概率很低，基于小概率事件几乎不可能发生，所以可以拒绝零假设</p>
</blockquote>
<p>但是这些传统的假设检验方法研究的对象，都是一次试验</p>
<p>在一次试验中（注意：是一次试验， 即single test），0.05 或0.01的cutoff足够严格了(想象一下，一个口袋有100个球，95个白的，5个红的, 只让你摸一次，你能摸到红的可能性是多大？)</p>
<p>但是对于多次试验，又称多重假设检验，再使用p值是不恰当的，下面来分析一下为什么：</p>
<p>大家都知道墨菲定律：如果事情有变坏的可能，不管这种可能性有多小，它总会发生</p>
<p>用统计的语言去描述墨菲定律：</p>
<blockquote>
<p>在数理统计中，有一条重要的统计规律：假设某意外事件在一次实验（活动）中发生的概率为p（p&gt;0），则在n次实验（活动）中至少有一次发生的概率为 $p_n=1-(1-p)^n$</p>
<p>由此可见，无论概率p多么小（即小概率事件），当n越来越大时，$p_n$越来越接近1</p>
</blockquote>
<p>这和我们的一句俗语非常吻合：常在河边走，哪有不湿鞋；夜路走多了，总能碰见鬼</p>
<p>在多重假设检验中，我们一般关注的不再是每一次假设检验的准确性，而是控制在作出的多个统计推断中犯错误的概率，即False Discover Rate（FDR），这对于医院的诊断情景下尤其重要：</p>
<blockquote>
<p>假如有一种诊断艾滋病(AIDS)的试剂，试验验证其准确性为99%（每100次诊断就有一次false positive）。对于一个被检测的人（single test)）来说，这种准确性够了；但对于医院 （multiple test)）来说，这种准确性远远不够</p>
<p>因为每诊断10 000个个体，就会有100个人被误诊为艾滋病(AIDS)，每一个误诊就意味着一个潜在的医疗事故和医疗纠纷，对于一些大型医院，一两个月的某一项诊断的接诊数就能达到这个级别，如果按照这个误诊率，医院恐怕得关门，所以医院需要严格控制误诊的数量，宁可错杀一万也不能放过一个，因为把一个没病的病人误判为有病，总比把一个有病的病人误判为没病更保险一些</p>
</blockquote>
<ul>
<li><p>100 independent genes. (We have 100 hypotheses to test)</p>
</li>
<li><p>No significant differences in gene expression between 2 classes (H0 is true). Thus, the probability that a particular test (say, for gene 1) is declared significant at level 0.05 is exactly 0.05. (Probability of reject H0  in one test if H0 is true = 0.05)</p>
</li>
<li><p>However, the probability of declaring at least one of the 100 hypotheses false (i.e. rejecting at least one, or finding at least one result significant) is: </p>
</li>
</ul>
<p>$$1-(1-0.05)^{100}\approx 0.994$$</p>
<h4 id="3-3-如何计算FDR？"><a href="#3-3-如何计算FDR？" class="headerlink" title="*3.3. 如何计算FDR？"></a>*3.3. 如何计算FDR？</h4><p>统计检验的混淆矩阵：</p>
<table>
<thead>
<tr>
<th style="text-align:center"><code></code></th>
<th style="text-align:center">$H_0$ is true</th>
<th style="text-align:center">$H_1$ is true</th>
<th style="text-align:center">Total</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">Significant</td>
<td style="text-align:center">V</td>
<td style="text-align:center">S</td>
<td style="text-align:center">R</td>
</tr>
<tr>
<td style="text-align:center">Not Significant</td>
<td style="text-align:center">U</td>
<td style="text-align:center">T</td>
<td style="text-align:center">m-R</td>
</tr>
<tr>
<td style="text-align:center">Total</td>
<td style="text-align:center">m<sub>0</sub></td>
<td style="text-align:center">m-m<sub>0</sub></td>
<td style="text-align:center">m</td>
</tr>
</tbody>
</table>
<ul>
<li><p><strong>FWER (Family Wise Error Rate)</strong></p>
<p>  作出一个或多个假阳性判断的概率</p>
<p>  $$FWER=Pr(V\ge 1)$$</p>
<p>  使用这种方法的统计学过程：</p>
<ul>
<li>The Bonferroni procedure</li>
<li>Tukey’s procedure</li>
<li>Holm’s step-down procedure </li>
</ul>
</li>
<li><p><strong>FDR (False Discovery Rate)</strong></p>
<p>  在所有的单检验中作出假阳性判断比例的期望</p>
<p>  $$FDR=E\left[\frac{V}{R}\right]$$</p>
<p>  使用这种方法的统计学过程：</p>
<ul>
<li>Benjamini–Hochberg procedure</li>
<li>Benjamini–Hochberg–Yekutieli procedure</li>
</ul>
</li>
</ul>
<h5 id="3-3-1-Benjamini-Hochberg-procedure-BH"><a href="#3-3-1-Benjamini-Hochberg-procedure-BH" class="headerlink" title="*3.3.1. Benjamini-Hochberg procedure (BH)"></a>*3.3.1. Benjamini-Hochberg procedure (BH)</h5><p>对于m个独立的假设检验，它们的P-value分别为：$p_i,i=1,2,…,m$</p>
<p>（1）按照升序的方法对这些P-value进行排序，得到：</p>
<p>$$p_{(1)} \le p_{(2)} \le … \le p_{(m)}$$</p>
<p>（2）对于给定是统计显著性值$\alpha \in (0,1)$，找到最大的k，使得</p>
<p>$$p_{(k)} \le \frac{\alpha * k}{m}$$</p>
<p>（3）对于排序靠前的k个假设检验，认为它们是真阳性 (positive )</p>
<p>即：$reject \, H_0^{(i)},\, 1 \le i \le k$ </p>
<p>$$<br>\begin{array}{c|l}<br>\hline<br>Gene &amp; p-value \\<br>\hline<br>G1 &amp; P1 =0.053 \\<br>\hline<br>G2 &amp; P2 =0.001 \\<br>\hline<br>G3    &amp; P3 =0.045 \\<br>\hline<br>G4    &amp; P4 =0.03 \\<br>\hline<br>G5 &amp; P5 =0.02 \\<br>\hline<br>G6 &amp; P6 =0.01 \\<br>\hline<br>\end{array}<br>\, \Rightarrow \,<br>\begin{array}{c|l}<br>\hline<br>Gene &amp; p-value \\<br>\hline<br>G2    &amp; P(1) =0.001 \\<br>\hline<br>G6    &amp; P(2) =0.01 \\<br>\hline<br>G5    &amp; P(3) =0.02 \\<br>\hline<br>G4    &amp; P(4) =0.03 \\<br>\hline<br>G3    &amp; P(5) =0.045 \\<br>\hline<br>G1    &amp; P(6) =0.053 \\<br>\hline<br>\end{array}<br>$$</p>
<p><br></p>
<p>$$\alpha = 0.05$$</p>
<blockquote>
<p>$P(4) =0.03&lt;0.05*\frac46=0.033$</p>
<p>$P(5) =0.045&gt;0.05*\frac56=0.041$</p>
<p>因此最大的k为4，此时可以得出：在FDR&lt;0.05的情况下，G2，G6，G5 和 G4 存在差异表达</p>
</blockquote>
<p>可以计算出q-value：</p>
<p>$$p_{(k)} \le \frac{\alpha*k}{m} \, \Rightarrow \, \frac{p_{(k)}*m}{k} \le \alpha$$</p>
<p><br></p>
<table>
<thead>
<tr>
<th style="text-align:center">Gene</th>
<th style="text-align:left">P</th>
<th style="text-align:left">q-value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">G2</td>
<td style="text-align:left">P(1) =0.001</td>
<td style="text-align:left">0.006</td>
</tr>
<tr>
<td style="text-align:center">G6</td>
<td style="text-align:left">P(2) =0.01</td>
<td style="text-align:left">0.03</td>
</tr>
<tr>
<td style="text-align:center">G5</td>
<td style="text-align:left">P(3) =0.02</td>
<td style="text-align:left">0.04</td>
</tr>
<tr>
<td style="text-align:center">G4</td>
<td style="text-align:left">P(4) =0.03</td>
<td style="text-align:left">0.045</td>
</tr>
<tr>
<td style="text-align:center">G3</td>
<td style="text-align:left">P(5) =0.045</td>
<td style="text-align:left">0.054</td>
</tr>
<tr>
<td style="text-align:center">G1</td>
<td style="text-align:left">P(6) =0.053</td>
<td style="text-align:left">0.053</td>
</tr>
</tbody>
</table>
<p>根据q-valuea的计算公式，我们可以很明显地看出：</p>
<p>$$q^{(i)}=p_{(k)}^{(i)}*\frac{Total \, Gene \, Number}{rank(p^{(i)})}=p_{(k)}^{(i)}*\frac{m}{k}$$</p>
<p>即，根据该基因p值的排序对它进行放大，越靠前放大的比例越大，越靠后放大的比例越小，排序最靠后的基因的p值不放大，等于它本身</p>
<p>我们也可以从可视化的角度来看待这个问题：</p>
<p>对于给定的$\alpha \in (0,1)$，设函数$y=\frac{\alpha}{m}x \quad (x=1,2,…,m)$，画出这条线，另外对于每个基因，它在图上的坐标为$(rank(p_{(k)}^{(i)}),p_{(k)}^{(i)})=(k,p_{(k)}^{(i)})$，图如下：</p>
<p align="center"><img src="/images/App-ML-in-Bioinfo-supplementary-calculate-FDR.png" alt></p>

<p>通过设置$\alpha$可以改变图中直线的斜率，$\alpha$越大，则直线的斜率越大，落在直线下方的点就越多，通过FDR检验的基因也就越多，反之，直线的斜率越小，落在直线下方的点就越少，通过FDR检验的基因也就越少</p>
<h5 id="3-4-2-Bonferroni-校正"><a href="#3-4-2-Bonferroni-校正" class="headerlink" title="*3.4.2. Bonferroni 校正"></a>*3.4.2. Bonferroni 校正</h5><p>Bonferroni 校正的基本思想：</p>
<blockquote>
<p>如果在同一数据集上同时检验n个独立的假设，那么用于每一假设的统计显著水平，应为仅检验一个假设时的显著水平的1/n</p>
<p>举个例子：如要在同一数据集上检验两个独立的假设，显著水平设为常见的0.05，此时用于检验该两个假设应使用更严格的0.025，即0.05* (1/2)</p>
</blockquote>
<p>序列化的 Bonferroni 校正步骤：</p>
<blockquote>
<p>对k个独立的检验，在给定的显著性水平α下，把每个检验对应的 P 值从小到大排列</p>
<p>$$p_{(1)} \le p_{(2)} \le … \le p_{(k)}$$</p>
<p>首先看最小的 P 值 $p_{(1)}$，如果$p_{(1)} \le \frac{\alpha}{k}$，就认为对应的检验在总体上（table wide）α水平上显著；如果不是，就认为所有的检验都不显著；</p>
<p>当且仅当 $p_{(1)} \le \frac{\alpha}{k}$ 时，再来看第二个P值$p_{(2)}$。如果$p_{(2)} \le \frac{\alpha}{k-1}$，就认为在总体水平上对应的检验在α水平上是显著的；</p>
<p>之后再进行下一个P值……一直进行这个过程，直到 $p_{(i)} \le \frac{\alpha}{k-i+1}$不成立；下结论i和以后的检验都不显著</p>
</blockquote>
<hr>
<p>参考资料：</p>
<p>(1)  Emerson R O , Dewitt W S , Vignali M , et al. Immunosequencing identifies signatures of cytomegalovirus exposure history and HLA-mediated effects on the T cell repertoire[J]. Nature Genetics, 2017, 49(5):659-665.</p>
<p>(2) Kiselev, V. Y. et al. SC3: consensus clustering of single-cell RNA-seq data[J]. Nat. Methods 14, 483–486 (2017).</p>
<p>(3) <a href="https://blog.csdn.net/sunmenggmail/article/details/17153053" target="_blank" rel="noopener">CSDN·chivalry《二项分布和Beta分布》</a></p>
<p>(4) <a href="https://blog.csdn.net/a358463121/article/details/52562940" target="_blank" rel="noopener">CSDN·Jie Qiao《带你理解beta分布》</a></p>
<p>(5) <a href="https://www.datalearner.com/blog/1051505532393058" target="_blank" rel="noopener">贝塔分布（Beta Distribution）简介及其应用</a></p>
<p>(6) <a href="https://www.statlect.com/probability-distributions/beta-distribution" target="_blank" rel="noopener">StatLect《Beta distribution》</a></p>
<p>(7)  Storey, J.D. &amp; Tibshirani, R. Statistical signifcance for genomewide studies.Proc. Natl. Acad. Sci. USA 100, 9440–9445 (2003)</p>
<p>(8) 国科大研究生课程《生物信息学》，陈小伟《基因表达分析》</p>
<p>(9) <a href="http://blog.sina.com.cn/s/blog_4af3f0d20100bzx9.html" target="_blank" rel="noopener">新浪博客·菜鸟《Bonferroni 校正》</a></p>
<p>(10) <a href="https://www.jianshu.com/p/a262cf3d18b9" target="_blank" rel="noopener">简书·Honglei_Ren《多重检验中的FDR错误控制方法与p-value的校正及Bonferroni校正》</a></p>

      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Machine-Learning/" rel="tag"># Machine-Learning</a>
          
            <a href="/tags/Bioinformatics/" rel="tag"># Bioinformatics</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/04/02/Linux-Advanced/" rel="next" title="Linux进阶笔记">
                <i class="fa fa-chevron-left"></i> Linux进阶笔记
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/04/06/Beauty-Mathmatics-Note/" rel="prev" title="《数学之美》阅读笔记">
                《数学之美》阅读笔记 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/images/avatar.jpg" alt="Lianm">
            
              <p class="site-author-name" itemprop="name">Lianm</p>
              <p class="site-description motion-element" itemprop="description">中国科学院北京基因组研究所研究生</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">31</span>
                    <span class="site-state-item-name">日志</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">6</span>
                    <span class="site-state-item-name">分类</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">31</span>
                    <span class="site-state-item-name">标签</span>
                  </a>
                </div>
              
            </nav>
          

          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://github.com/Ming-Lian" title="GitHub &rarr; https://github.com/Ming-Lian" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="mailto:lianming17m@big.ac.cn" title="E-Mail &rarr; mailto:lianming17m@big.ac.cn" rel="noopener" target="_blank"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://www.jianshu.com/u/42e711970fa8" title="简书 &rarr; https://www.jianshu.com/u/42e711970fa8" rel="noopener" target="_blank"><i class="fa fa-fw fa-jianshu"></i>简书</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://www.zhihu.com/people/understorm" title="知乎 &rarr; https://www.zhihu.com/people/understorm" rel="noopener" target="_blank"><i class="fa fa-fw fa-zhihu"></i>知乎</a>
                </span>
              
            </div>
          

          

          
          

          
            
          
          

        </div>
      </div>

      
      <!--noindex-->
        <div class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
            
            
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-根据免疫组库TCRβ预测病人的CMV感染状态"><span class="nav-text">1. 根据免疫组库TCRβ预测病人的CMV感染状态</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-SC3：单细胞表达谱的无监督聚类"><span class="nav-text">2. SC3：单细胞表达谱的无监督聚类</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3-GATK的VQSR"><span class="nav-text">3. GATK的VQSR</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#补充知识"><span class="nav-text">补充知识</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-beta分布"><span class="nav-text">*1. beta分布</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#1-1-什么是beta分布"><span class="nav-text">*1.1. 什么是beta分布</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-2-理解beta分布"><span class="nav-text">*1.2. 理解beta分布</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-重复NG免疫组库TCRβ文章的图"><span class="nav-text">*2. 重复NG免疫组库TCRβ文章的图</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#2-1-Fisher检验筛选CMV阳性（CMV-）相关克隆"><span class="nav-text">*2.1. Fisher检验筛选CMV阳性（CMV+）相关克隆</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-2-绘制表型负荷相关散点图"><span class="nav-text">*2.2. 绘制表型负荷相关散点图</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-2-1-训练集"><span class="nav-text">*2.2.1. 训练集</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-2-2-测试集"><span class="nav-text">*2.2.2. 测试集</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-3-TCRβ在两组间分布偏好性的散点图"><span class="nav-text">*2.3. TCRβ在两组间分布偏好性的散点图</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-FDR的计算方法"><span class="nav-text">*3. FDR的计算方法</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#3-1-回顾那些统计检验方法"><span class="nav-text">*3.1. 回顾那些统计检验方法</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-1-T-test与Moderated-t-Test"><span class="nav-text">*3.1.1. T-test与Moderated t-Test</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-2-多重假设检验的必要性"><span class="nav-text">*3.2. 多重假设检验的必要性</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-3-如何计算FDR？"><span class="nav-text">*3.3. 如何计算FDR？</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-3-1-Benjamini-Hochberg-procedure-BH"><span class="nav-text">*3.3.1. Benjamini-Hochberg procedure (BH)</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-2-Bonferroni-校正"><span class="nav-text">*3.4.2. Bonferroni 校正</span></a></li></ol></li></ol></li></ol></li></ol></div>
            

          </div>
        </div>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Lianm</span>

  

  
</div>


  <div class="powered-by">由 <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> 强力驱动 v3.8.0</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Mist</a> v7.0.0</div>




        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>














  
    
    
  
  <script color="0,0,255" opacity="0.5" zindex="-1" count="99" src="/lib/canvas-nest/canvas-nest.min.js"></script>



  
  











  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script src="/lib/three/three.min.js"></script>

  
  <script src="/lib/three/three-waves.min.js"></script>


  


  <script src="/js/src/utils.js?v=7.0.0"></script>

  <script src="/js/src/motion.js?v=7.0.0"></script>



  
  


  <script src="/js/src/schemes/muse.js?v=7.0.0"></script>




  
  <script src="/js/src/scrollspy.js?v=7.0.0"></script>
<script src="/js/src/post-details.js?v=7.0.0"></script>



  


  <script src="/js/src/bootstrap.js?v=7.0.0"></script>



  


  


  
  <script>
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x"></i></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x"></i></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  

  
  

  
    
      <script type="text/x-mathjax-config">
  

  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true,
      skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
    },
    TeX: {
      
      equationNumbers: {
        autoNumber: "AMS"
      }
    }
  });
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
      for (i = 0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';
      }
  });
</script>
<script src="//cdn.jsdelivr.net/npm/mathjax@2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<style>
.MathJax_Display {
  overflow: auto hidden;
}
</style>

    
  


  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js'; 
      }
      else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
      }
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  

  

  

  

  

  

  

</body>
</html>
