<!DOCTYPE html>












  


<html class="theme-next pisces use-motion" lang="zh-CN">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">


























<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2">

<link rel="stylesheet" href="/css/main.css?v=7.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/Favico.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/Favico.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/Favico.png?v=7.0.0">


  <link rel="mask-icon" href="/images/logo.svg?v=7.0.0" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '7.0.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="GATK4流程 准备配套数据要明确你的参考基因组版本了！！！ b36/b37/hg18/hg19/hg38，记住b37和hg19并不是完全一样的，有些微区别哦！！！ 1、下载hg19 这个下载地址非常多，常用的就是NCBI，ensembl和UCSC了，但是这里推荐用这个脚本下载（下载源为UCSC）： 1234567891011121314# 一个个地下载hg19的染色体for i in $(seq">
<meta property="og:type" content="article">
<meta property="og:title" content="深入了解snp-calling流程">
<meta property="og:url" content="http://yoursite.com/2019/02/08/call-snp/index.html">
<meta property="og:site_name" content="Lianm&#39;s Blog">
<meta property="og:description" content="GATK4流程 准备配套数据要明确你的参考基因组版本了！！！ b36/b37/hg18/hg19/hg38，记住b37和hg19并不是完全一样的，有些微区别哦！！！ 1、下载hg19 这个下载地址非常多，常用的就是NCBI，ensembl和UCSC了，但是这里推荐用这个脚本下载（下载源为UCSC）： 1234567891011121314# 一个个地下载hg19的染色体for i in $(seq">
<meta property="og:locale" content="zh-CN">
<meta property="og:image" content="http://yoursite.com/images/GATK4-logo.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-reason-of-duplicates.jpg">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-poisson-1.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-poisson-2.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-poisson-3.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-1.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-2.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-3.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-4.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-remove-duplicates-5.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-snp-calling-strategies.jpg">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-Germline-SNPs-Indels.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-hard-filter-QD.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-hard-filter-FS.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-hard-filter-SOR.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-hard-filter-MQ-1.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-hard-filter-MQ-2.png">
<meta property="og:image" content="http://yoursite.com/images/GATK4-pipeline-hard-filter-MQRankSum.png">
<meta property="og:updated_time" content="2019-02-17T06:12:08.116Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="深入了解snp-calling流程">
<meta name="twitter:description" content="GATK4流程 准备配套数据要明确你的参考基因组版本了！！！ b36/b37/hg18/hg19/hg38，记住b37和hg19并不是完全一样的，有些微区别哦！！！ 1、下载hg19 这个下载地址非常多，常用的就是NCBI，ensembl和UCSC了，但是这里推荐用这个脚本下载（下载源为UCSC）： 1234567891011121314# 一个个地下载hg19的染色体for i in $(seq">
<meta name="twitter:image" content="http://yoursite.com/images/GATK4-logo.png">






  <link rel="canonical" href="http://yoursite.com/2019/02/08/call-snp/">



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>深入了解snp-calling流程 | Lianm's Blog</title>
  












  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Lianm's Blog</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">Focus on Bioinformatics and Machine-Learning</p>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="切换导航栏">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>首页</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br>关于</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>归档</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-search">

    
    
    
      
    

    

    <a href="/search/" rel="section"><i class="menu-item-icon fa fa-fw fa-search"></i> <br>搜索</a>

  </li>

      
      
    </ul>
  

  

  
</nav>



  



</div>
    </header>

    
  
  
  
  

  

  <a href="https://github.com/Ming-Lian" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="80" height="80" viewbox="0 0 250 250" style="fill: #222; color: #fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a>



    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/08/call-snp/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content="中国科学院北京基因组研究所研究生">
      <meta itemprop="image" content="/images/avatar.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">深入了解snp-calling流程

              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2019-02-08 00:15:36" itemprop="dateCreated datePublished" datetime="2019-02-08T00:15:36+00:00">2019-02-08</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2019-02-17 06:12:08" itemprop="dateModified" datetime="2019-02-17T06:12:08+00:00">2019-02-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="GATK4流程"><a href="#GATK4流程" class="headerlink" title="GATK4流程"></a>GATK4流程</h2><p><img src="/images/GATK4-logo.png" alt></p>
<h3 id="准备配套数据"><a href="#准备配套数据" class="headerlink" title="准备配套数据"></a>准备配套数据</h3><p>要明确你的参考基因组版本了！！！ b36/b37/hg18/hg19/hg38，记住<strong>b37和hg19并不是完全一样</strong>的，有些微区别哦！！！</p>
<p><strong>1、下载hg19</strong></p>
<p>这个下载地址非常多，常用的就是NCBI，ensembl和UCSC了，但是这里推荐用这个脚本下载（下载源为UCSC）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># 一个个地下载hg19的染色体</span><br><span class="line">for i in $(seq 1 22) X Y M;</span><br><span class="line">do echo $i;</span><br><span class="line">wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/chr$&#123;i&#125;.fa.gz;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">gunzip *.gz</span><br><span class="line"></span><br><span class="line"># 用cat按照染色体的顺序拼接起来，因为GATK后面的一些步骤对染色体顺序要求非常变态，如果下载整个hg19，很难保证染色体顺序是1-22，X,Y,M</span><br><span class="line">for i in $(seq 1 22) X Y M;</span><br><span class="line">do cat chr$&#123;i&#125;.fa &gt;&gt; hg19.fasta;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">rm -fr chr*.fasta</span><br></pre></td></tr></table></figure>
<h3 id="BWA-Map-to-Reference"><a href="#BWA-Map-to-Reference" class="headerlink" title="BWA: Map to Reference"></a>BWA: Map to Reference</h3><ol>
<li><p>建立参考序列索引</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ bwa index -a bwtsw ref.fa</span><br></pre></td></tr></table></figure>
<p> 参数<code>-a</code>用于指定建立索引的算法：</p>
<blockquote>
<ul>
<li>bwtsw 适用于&gt;10M</li>
<li>is 适用于参考序列&lt;2G (默认-a is)</li>
</ul>
</blockquote>
<p> 可以不指定<code>-a</code>参数，bwa index会根据基因组大小来自动选择合适的索引方法</p>
</li>
<li><p>序列比对</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ bwa mem ref.fa sample_1.fq sample_2.fq -R &apos;@RG\tID:sample\tLB:sample\tSM:sample\tPL:ILLUMINA&apos; \</span><br><span class="line">	2&gt;sample_map.log | samtools sort -@ 20 -O bam -o sample.sorted.bam 1&gt;sample_sort.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<p> <code>-R</code> 选项为必须选项，用于定义头文件中的SAM/BAM文件中的read group和sample信息</p>
<blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; The file must have a proper bam header with read groups. Each read group must contain the platform (PL) and sample (SM) tags. </span><br><span class="line">&gt; For the platform value, we currently support 454, LS454, Illumina, Solid, ABI_Solid, and CG (all case-insensitive)</span><br><span class="line">&gt; </span><br><span class="line">&gt; The GATK requires several read group fields to be present in input files and will fail with errors if this requirement is not satisfied</span><br><span class="line">&gt; </span><br><span class="line">&gt; ID：输入reads集的ID号； LB： reads集的文库名； SM：样本名称； PL：测序平台</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<p> 查看SAM/BAM文件中的read group和sample信息：</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -H /path/to/my.bam | grep &apos;^@RG&apos;</span><br><span class="line">@RG ID:0    PL:solid    PU:Solid0044_20080829_1_Pilot1_Ceph_12414_B_lib_1_2Kb_MP_Pilot1_Ceph_12414_B_lib_1_2Kb_MP   LB:Lib1 PI:2750 DT:2008-08-28T20:00:00-0400 SM:NA12414  CN:bcm</span><br><span class="line">@RG ID:1    PL:solid    PU:0083_BCM_20080719_1_Pilot1_Ceph_12414_B_lib_1_2Kb_MP_Pilot1_Ceph_12414_B_lib_1_2Kb_MP    LB:Lib1 PI:2750 DT:2008-07-18T20:00:00-0400 SM:NA12414  CN:bcm</span><br><span class="line">@RG ID:2    PL:LS454    PU:R_2008_10_02_06_06_12_FLX01080312_retry  LB:HL#01_NA11881    PI:0    SM:NA11881  CN:454MSC</span><br><span class="line">@RG ID:3    PL:LS454    PU:R_2008_10_02_06_07_08_rig19_retry    LB:HL#01_NA11881    PI:0    SM:NA11881  CN:454MSC</span><br><span class="line">@RG ID:4    PL:LS454    PU:R_2008_10_02_17_50_32_FLX03080339_retry  LB:HL#01_NA11881    PI:0    SM:NA11881  CN:454MSC</span><br><span class="line">...</span><br></pre></td></tr></table></figure>
<p> read group信息不仅会加在头信息部分，也会在比对结果的每条记录里添加一个 <code>RG:Z:*</code> 标签</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view /path/to/my.bam | grep &apos;^@RG&apos;</span><br><span class="line">EAS139_44:2:61:681:18781    35  1   1   0   51M =   9   59  TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAA B&lt;&gt;;==?=?&lt;==?=?=&gt;&gt;?&gt;&gt;&lt;=&lt;?=?8&lt;=?&gt;?&lt;:=?&gt;?&lt;==?=&gt;:;&lt;?:= RG:Z:4  MF:i:18 Aq:i:0  NM:i:0  UQ:i:0  H0:i:85 H1:i:31</span><br><span class="line">EAS139_44:7:84:1300:7601    35  1   1   0   51M =   12  62  TAACCCTAAGCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAA G&lt;&gt;;==?=?&amp;=&gt;?=?&lt;==?&gt;?&lt;&gt;&gt;?=?&lt;==?&gt;?&lt;==?&gt;?1==@&gt;?;&lt;=&gt;&lt;; RG:Z:3  MF:i:18 Aq:i:0  NM:i:1  UQ:i:5  H0:i:0  H1:i:85</span><br><span class="line">EAS139_44:8:59:118:13881    35  1   1   0   51M =   2   52  TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAA @&lt;&gt;;&lt;=?=?==&gt;?&gt;?&lt;==?=&gt;&lt;=&gt;?-?;=&gt;?:&gt;&lt;==?7?;&lt;&gt;?5?&lt;&lt;=&gt;:; RG:Z:1  MF:i:18 Aq:i:0  NM:i:0  UQ:i:0  H0:i:85 H1:i:31</span><br><span class="line">EAS139_46:3:75:1326:2391    35  1   1   0   51M =   12  62  TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAA @&lt;&gt;==&gt;?&gt;@???B&gt;A&gt;?&gt;A?A&gt;??A?@&gt;?@A?@;??A&gt;@7&gt;?&gt;&gt;@:&gt;=@;@ RG:Z:0  MF:i:18 Aq:i:0  NM:i:0  UQ:i:0  H0:i:85 H1:i:31</span><br><span class="line">...</span><br></pre></td></tr></table></figure>
<p> 若原始SAM/BAM文件没有read group和sample信息，可以通过AddOrReplaceReadGroups添加这部分信息</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">$ java -jar picard.jar AddOrReplaceReadGroups \</span><br><span class="line">      I=input.bam \</span><br><span class="line">      O=output.bam \</span><br><span class="line">      RGID=4 \</span><br><span class="line">      RGLB=lib1 \</span><br><span class="line">      RGPL=illumina \</span><br><span class="line">      RGPU=unit1 \</span><br><span class="line">      RGSM=20</span><br></pre></td></tr></table></figure>
</li>
</ol>
<h3 id="前期处理"><a href="#前期处理" class="headerlink" title="前期处理"></a>前期处理</h3><p>在进行本部分的操作之前先要做好以下两部的准备工作</p>
<p>1、创建GATK索引。用Samtools为参考序列创建一个索引，这是为方便GATK能够快速地获取fasta上的任何序列做准备</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools faidx database/chr17.fa	# 该命令会在chr17.fa所在目录下创建一个chr17.fai索引文件</span><br></pre></td></tr></table></figure>
<p>2、生成.dict文件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ gatk CreateSequenceDictionary -R database/chr17.fa -O database/chr17.dict</span><br></pre></td></tr></table></figure>
<p>GATK4的调用语法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gatk [--java-options &quot;-Xmx4G&quot;] ToolName [GATK args]</span><br></pre></td></tr></table></figure>
<h4 id="去除PCR重复"><a href="#去除PCR重复" class="headerlink" title="去除PCR重复"></a>去除PCR重复</h4><h5 id="duplicates的产生原因"><a href="#duplicates的产生原因" class="headerlink" title="duplicates的产生原因"></a>duplicates的产生原因</h5><p><img src="/images/GATK4-pipeline-remove-duplicates-reason-of-duplicates.jpg" alt></p>
<ul>
<li><strong>PCR duplicates（PCR重复）</strong></li>
</ul>
<p>PCR扩增时，同一个DNA片段会产生多个相同的拷贝，第4步测序的时候，这些来源于同！一！个！拷贝的DNA片段会结合到Fellowcell的不同位置上，生成完全相同的测序cluster，然后被测序出来，这些相同的序列就是duplicate</p>
<ul>
<li><strong>Cluster duplicates</strong></li>
</ul>
<p>生成测序cluster的时候，某一个cluster中的DNA序列可能搭到旁边的另一个cluster的生成位点上，又再重新长成一个相同的cluster，这也是序列duplicate的另一个来源，这个现象在Illumina HiSeq4000之后的Flowcell中会有这类Cluster duplicates</p>
<ul>
<li><strong>Optical duplicates（光学重复）</strong></li>
</ul>
<p>某些cluster在测序的时候，捕获的荧光亮点由于光波的衍射，导致形状出现重影（如同近视散光一样），导致它可能会被当成两个荧光点来处理。这也会被读出为两条完全相同的reads</p>
<ul>
<li><strong>Sister duplicates</strong></li>
</ul>
<p>它是文库分子的两条互补链同时都与Flowcell上的引物结合分别形成了各自的cluster被测序，最后产生的这对reads是完全反向互补的。比对到参考基因组时，也分别在正负链的相同位置上，在有些分析中也会被认为是一种duplicates。</p>
<h5 id="用泊松分布解释-NGS-测序数据的-duplication-问题"><a href="#用泊松分布解释-NGS-测序数据的-duplication-问题" class="headerlink" title="用泊松分布解释 NGS 测序数据的 duplication 问题"></a>用泊松分布解释 NGS 测序数据的 duplication 问题</h5><p>我曾经有这样的疑惑，为什么文库构建过程中的 PCR 将每个文库分子都扩增了上千倍，以 PCR 10个循环为例 2<sup>10</sup>= 1024 ，但是实际测序数据中 duplication 率并不高（低于20%）</p>
<p>有一篇文章从统计概率的角度详细探讨了一下 duplication 率的影响因素</p>
<p>PCR 的过程中不同长度的文库分子被扩增的效率不同（GC 太高或 AT 含量太高都会影响扩增效率），PCR 更倾向于扩增短片段的文库分子，这里先不考虑文库片段扩增效率的差异，把问题简化一下，<strong>假设所有文库分子扩增效率都相同</strong>。PCR duplicate 的主要来源是<strong>同一个文库分子的不同拷贝都在 flowcell 上生成了可以被测序的 cluster</strong> ，导致同一个分子的序列被测序仪读取多次。那么为何在每个分子都有上千个拷贝的情况下，实际却很少出现同一分子的多个拷贝被测序的情况呢？主要原因就是文库中 unique 分子的数量比被 flowcell 上引物捕获的分子数量多很多，直白点说就是 <strong>flowcell 上用于捕获文库分子的引物数量太少了，两者不在同一个数量级</strong>，导致很少出现同一个文库分子的多个拷贝被 flowcell 上引物捕获生成 cluster。</p>
<p>假设文库中所有分子与引物的结合都是随机的，简化一下就相当于，一个箱子中有 n 种颜色的球（文库中的 n 种 unique 分子），每种颜色有 1000 个（PCR 扩增的，随 cycle 数变化），从这个箱子中随机拿出来 k 个球（最终测序得到 k 条 reads），其中出现相同颜色的球就是 duplicate，那么 duplication 率就可以根据有多少种颜色的球被取出 0,1,2,3…… 次的概率计算，可以近似用泊松分布模型来描述。</p>
<p>以人全基因组重测序 30X 为例，PE150 需要约 3x10<sup>8</sup>条 reads ，文库中 unique 分子数其实可以通过上机文库的浓度和体积（外加 PCR 循环数）计算出来，这里用近似值 3.5x10<sup>10</sup> 个 unique 分子。每个 unique 分子期望被测序的次数是 3x10<sup>8</sup>/3.5x10<sup>10</sup> = 0.0085 ，每个 unique 分子被测 0,1,2,3… 次的概率如下图：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; x &lt;- seq(0,10,1)</span><br><span class="line">&gt; xnames &lt;- as.character(x)</span><br><span class="line">&gt; xlab &lt;- &quot;一个文库分子的所有拷贝被测序的次数&quot;</span><br><span class="line">&gt; ylab &lt;- &quot;概率&quot;</span><br><span class="line">&gt; barplot(dpois(x,lambda = 0.0085),</span><br><span class="line">+ names.arg = xnames,</span><br><span class="line">+ xlab = xlab,</span><br><span class="line">+ ylab = ylab)</span><br></pre></td></tr></table></figure>
<p><img src="/images/GATK4-pipeline-remove-duplicates-poisson-1.png" alt></p>
<p>由于 unique 分子数量太多，被测 0 次的概率远高于 1 和 2 次，我们去除 0 次的看一下：</p>
<p><img src="/images/GATK4-pipeline-remove-duplicates-poisson-2.png" alt></p>
<p>unique 分子被测序 1 次的概率远大于 2次及以上，即便一个 unique 分子被测序 2 次，我们去除 duplicate 时候还会保留其中一条 reads。</p>
<p>如果降低文库中 unique 分子数量到 4.5x10<sup>9</sup>个，PCR 循环数增加以便浓度达到跟上面模拟的情况相同，测序 reads 数还是 3x10<sup>8</sup> 条，每个 unique 分子预期被测序的次数是 3x10<sup>8</sup>/4.5x10<sup>9</sup> = 0.067 。</p>
<p><img src="/images/GATK4-pipeline-remove-duplicates-poisson-3.png" alt></p>
<p>unique 分子数量减少，被测序 2次的概率增大，duplication 率显然也会增高。</p>
<p>到这里已经可以很明白的看出 <strong>duplication 率主要与文库中 unique 分子数量有关</strong>，所以建库过程中最大化 unique 分子数是降低 duplication 率的关键。文库中 unique 分子数越多，说明建库起始量越高，需要 PCR 的循环数越少，而文库中 unique 分子数越少，说明建库起始量越低，需要 PCR 的循环数越多，因此提高建库起始量是关键。</p>
<h5 id="PCR-bias的影响"><a href="#PCR-bias的影响" class="headerlink" title="PCR bias的影响"></a>PCR bias的影响</h5><ol>
<li><p>DNA在打断的那一步会发生一些损失， 主要表现是会引发一些碱基发生颠换变换（嘌呤-变嘧啶或者嘧啶变嘌呤） ， 带来假的变异。 PCR过程会扩大这个信号， 导致最后的检测结果中混入了假的结果；</p>
</li>
<li><p>PCR反应过程中也会带来新的碱基错误。 发生在前几轮的PCR扩增发生的错误会在后续的PCR过程中扩大， 同样带来假的变异；</p>
</li>
<li><p>对于真实的变异， PCR反应可能会对包含某一个碱基的DNA模版扩增更加剧烈（这个现象称为PCR Bias） 。 因此， 如果反应体系是对含有reference allele的模板扩增偏向强烈， 那么变异碱基的信息会变小， 从而会导致假阴。</p>
</li>
</ol>
<p><img src="/images/GATK4-pipeline-remove-duplicates-1.png" alt></p>
<h5 id="探究samtools和picard去除read-duplicates的方法"><a href="#探究samtools和picard去除read-duplicates的方法" class="headerlink" title="探究samtools和picard去除read duplicates的方法"></a>探究samtools和picard去除read duplicates的方法</h5><p><strong>1、samtools</strong></p>
<p>samtools 去除 duplicates 使用 <strong>rmdup</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools rmdup [-sS] &lt;input.srt.bam&gt; &lt;out.bam&gt;</span><br></pre></td></tr></table></figure>
<p>只需要开始<code>-s</code>的标签， 就可以对单端测序进行去除PCR重复。其实对单端测序去除PCR重复很简单的，因为比对flag情况只有0,4,16，只需要它们比对到染色体的起始终止坐标一致即可，flag很容易一致。</p>
<blockquote>
<p>Remove potential PCR duplicates: if multiple read pairs have identical external coordinates, only retain the pair with highest mapping quality. I</p>
</blockquote>
<p align="center">&lt;img src=./picture/GATK4-pipeline-remove-duplicates-samtools-principle-1.png width=900/&gt;</p>

<p>但是对于双端测序就有点复杂了~</p>
<blockquote>
<p>In the paired-end mode, this command ONLY works with FR orientation and requires ISIZE is correctly set. It does not work for unpaired reads (e.g. two ends mapped to different chromosomes or orphan reads).</p>
</blockquote>
<p align="center">&lt;img src=./picture/GATK4-pipeline-remove-duplicates-samtools-principle-2.png width=900/&gt;</p>

<p>很明显可以看出，去除PCR重复不仅仅需要它们比对到染色体的起始终止坐标一致，尤其是<strong>flag</strong>，在双端测序里面一大堆的flag情况，所以我们的94741坐标的5个reads，一个都没有去除！</p>
<p>这样的话，双端测序数据，用samtools rmdup效果就很差，所以很多人建议用picard工具的MarkDuplicates 功能</p>
<p><strong>2、picard</strong></p>
<p>picard对于单端或者双端测序数据并没有区分参数，可以用同一个命令！</p>
<p>对应单端测序，picard的处理结果与samtools rmdup没有差别，不过这个java软件的缺点就是<strong>奇慢无比</strong></p>
<h5 id="操作：排序及标记重复"><a href="#操作：排序及标记重复" class="headerlink" title="操作：排序及标记重复"></a>操作：排序及标记重复</h5><p><img src="/images/GATK4-pipeline-remove-duplicates-2.png" alt></p>
<p><img src="/images/GATK4-pipeline-remove-duplicates-3.png" alt></p>
<p>1、排序（SortSam）</p>
<ul>
<li>对sam文件进行排序并生成bam文件，将sam文件中同一染色体对应的条目按照坐标顺序从小到大进行排序</li>
<li>GATK4的排序功能是通过<code>picard SortSam</code>工具实现的。虽然<code>samtools sort</code>工具也可以实现该功能，但是在GATK流程中还是推荐用picard实现，因为SortSam会在输出文件的头信息部分添加一个SO标签用于说明文件已经被成功排序，且<strong>这个标签是必须的</strong>，GATK需要检查这个标签以保证后续分析可以正常进行</li>
<li><code>https://software.broadinstitute.org/gatk/documentation/tooldocs/current/picard_sam_SortSam.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"># 使用GATK命令</span><br><span class="line">$ gatk SortSam -I mapping/T.chr17.sam -O preprocess/T.chr17.sort.bam -R database/chr17.fa -SO coordinate --CREATE_INDEX</span><br><span class="line"># 使用picard命令</span><br><span class="line">$ java -jar picard.jar SortSam \</span><br><span class="line">      I=input.bam \</span><br><span class="line">      O=sorted.bam \</span><br><span class="line">      SORT_ORDER=coordinate</span><br></pre></td></tr></table></figure>
<p><img src="/images/GATK4-pipeline-remove-duplicates-4.png" alt></p>
<p>如何检查是否成功排序？</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -H /path/to/my.bam</span><br><span class="line">@HD     VN:1.0  GO:none SO:coordinate</span><br><span class="line">@SQ     SN:1    LN:247249719</span><br><span class="line">@SQ     SN:2    LN:242951149</span><br><span class="line">@SQ     SN:3    LN:199501827</span><br><span class="line">@SQ     SN:4    LN:191273063</span><br><span class="line">@SQ     SN:5    LN:180857866</span><br><span class="line">@SQ     SN:6    LN:170899992</span><br><span class="line">@SQ     SN:7    LN:158821424</span><br><span class="line">@SQ     SN:8    LN:146274826</span><br><span class="line">@SQ     SN:9    LN:140273252</span><br><span class="line">@SQ     SN:10   LN:135374737</span><br><span class="line">@SQ     SN:11   LN:134452384</span><br><span class="line">@SQ     SN:12   LN:132349534</span><br><span class="line">@SQ     SN:13   LN:114142980</span><br><span class="line">@SQ     SN:14   LN:106368585</span><br><span class="line">@SQ     SN:15   LN:100338915</span><br><span class="line">@SQ     SN:16   LN:88827254</span><br><span class="line">@SQ     SN:17   LN:78774742</span><br><span class="line">@SQ     SN:18   LN:76117153</span><br><span class="line">@SQ     SN:19   LN:63811651</span><br><span class="line">@SQ     SN:20   LN:62435964</span><br><span class="line">@SQ     SN:21   LN:46944323</span><br><span class="line">@SQ     SN:22   LN:49691432</span><br><span class="line">@SQ     SN:X    LN:154913754</span><br><span class="line">@SQ     SN:Y    LN:57772954</span><br><span class="line">@SQ     SN:MT   LN:16571</span><br><span class="line">@SQ     SN:NT_113887    LN:3994</span><br><span class="line">...</span><br></pre></td></tr></table></figure>
<p>若随后的比对记录中的contig那一列的顺序与头文件的顺序一致，且在头信息中包含<code>SO:coordinate</code>这个标签，则说明，该文件是排序过的</p>
<p>2、标记重复（Markduplicates）</p>
<ul>
<li>标记文库中的重复</li>
<li><code>https://software.broadinstitute.org/gatk/documentation/tooldocs/current/picard_sam_markduplicates_MarkDuplicates.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gatk MarkDuplicates -I preprocess/T.chr17.sort.bam -O preprocess/T.chr17.markdup.bam -M preprocess/T.chr17.metrics --CREATE_INDEX</span><br></pre></td></tr></table></figure>
<p><img src="/images/GATK4-pipeline-remove-duplicates-5.png" alt></p>
<h4 id="质量值校正"><a href="#质量值校正" class="headerlink" title="质量值校正"></a>质量值校正</h4><p>检测碱基质量分数中的系统错误，需要用到 GATK4 中的 BaseRecalibrator 工具</p>
<p><strong>碱基质量分数重校准（Base quality score recalibration，BQSR)</strong>，就是利用机器学习的方式调整原始碱基的质量分数。它分为两个步骤:</p>
<blockquote>
<ul>
<li>利用已有的snp数据库，建立相关性模型，产生重校准表( recalibration table)</li>
<li>根据这个模型对原始碱基进行调整，只会调整非已知SNP区域。</li>
</ul>
</blockquote>
<p>注：如果不是人类基因组，并且也缺少相应的已知SNP数据库，可以通过严格SNP筛选过程（例如结合GATK和samtools）建立一个snp数据库。</p>
<p><strong>注意：Base Recalibration是以read group为单位进行的，因此当一个BAM文件中包含了多个read group时，BaseRecalibrator会对每个read group分别建立校正模型</strong></p>
<p>1、建立较正模型</p>
<p>质量值校正，这一步需要用到variants的known-sites，所以需要先准备好已知的snp，indel的VCF文件：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># 下载known-site的VCF文件，到Ensembl上下载</span><br><span class="line">$ wget -c -P Ref/mouse/mm10/vcf ftp://ftp.ensembl.org/pub/release-93/variation/vcf/mus_musculus/mus_musculus.vcf.gz &gt;download.log &amp;</span><br><span class="line">$ cd Ref/mouse/mm10/vcf &amp;&amp; gunzip mus_musculus.vcf.gz &amp;&amp; mv mus_musculus.vcf dbsnp_150.mm10.vcf</span><br><span class="line"># 建好vcf文件的索引，需要用到GATK工具集中的IndexFeatureFile，该命令会在指定的vcf文件的相同路径下生成一个以&quot;.idx&quot;为后缀的文件</span><br><span class="line">$ gatk IndexFeatureFile -F dbsnp_150.mm10.vcf</span><br><span class="line"></span><br><span class="line"># 建立较正模型</span><br><span class="line">$ gatk BaseRecalibrator -R Ref/mouse/mm10/bwa/mm10.fa -I PharmacogenomicsDB/mouse/SAM/ERR118300.enriched.markdup.bam -O \</span><br><span class="line">PharmacogenomicsDB/mouse/SAM/ERR118300.recal.table --known-sites Ref/mouse/mm10/vcf/dbsnp_150.mm10.vcf</span><br></pre></td></tr></table></figure>
<p>2、质量值校准</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"># 质量校正</span><br><span class="line">$ gatk ApplyBQSR -R Ref/mouse/mm10/bwa/mm10.fa -I PharmacogenomicsDB/mouse/SAM/ERR118300.enriched.markdup.bam -bqsr \</span><br><span class="line">PharmacogenomicsDB/mouse/SAM/ERR118300.recal.table -O PharmacogenomicsDB/mouse/SAM/ERR118300.recal.bam</span><br></pre></td></tr></table></figure>
<h3 id="SNP、-INDEL位点识别"><a href="#SNP、-INDEL位点识别" class="headerlink" title="SNP、 INDEL位点识别"></a>SNP、 INDEL位点识别</h3><h4 id="SNP-calling-策略的选择"><a href="#SNP-calling-策略的选择" class="headerlink" title="SNP calling 策略的选择"></a>SNP calling 策略的选择</h4><p>当你有多个samples，然后你call snp时候，你是应该将所有sample分开call完之后，再merge在一起。还是直接将所有samples，同时用作input然后call snp呢？</p>
<p>这里需要知道有哪些snp calling的策略：</p>
<ul>
<li><p><strong>single sample calling</strong>：每一个sample的bam file都进行单独的snp calling，然后每个sample单独snp calling结果再合成一个总的snp calling的结果。</p>
</li>
<li><p><strong>batch calling</strong>： 一定数目群集的bamfiles 一起calling snps，然后再merge在一起</p>
</li>
<li><p><strong>joint calling</strong>： 所有samples的BAM files一起call 出一个包含所有samples 变异信息的output</p>
</li>
</ul>
<p>一般来说，如果条件允许（computational power等），使用joint calling ，即将所有samples同时call是比较优的选择</p>
<p>原因：</p>
<p><strong>1、对于低频率的变异具有更高更好的检测sensitivity</strong></p>
<p>在joint calling中，由于所有samples中所有的位点都是同时call的，换句话说就是，所有位点的信息都是share的，因此可以如果某些samples中个别位点是低频率的，但是可能在其他samples中，该位点的频率比较高，因此可以准确的对低频位点有更加好的calling 效果。</p>
<p><img src="/images/GATK4-pipeline-snp-calling-strategies.jpg" alt></p>
<p>如左图，在1到n的samples中，碱基G只出现在其中两个samples中。如果我们将这些samples单独的call snps，这个低频率G的位点将会被忽略。但是joint calling 却可以允许将这些碱基G出现的频率进行累加，将该低频率的突变也call出来。</p>
<p>在右图中，上面的sample是一个跟ref一样具有纯合位点，下面的sample在一些位点中有部分数据缺少的情况，例如rs429358.如果将这两个samples分开call snp，然后merge一起，这样会错误地将这两个samples，在这个位点上看作具有类似的变异频率，但是其实由于下面的samples在改位点的区域存在数据缺少，只能看作non-informative。</p>
<p><strong>2、更好的过滤掉假阳性的变异callling的结果</strong></p>
<p>现在使用过滤变异的方法例如VQSR等利用的统计模型，都基于一个比较大的samples size。joint calling 这种方法可以提供足够的数据，确保过滤这一步是统一应用于所有samples的。</p>
<h4 id="Germline-SNPs-Indels"><a href="#Germline-SNPs-Indels" class="headerlink" title="Germline SNPs + Indels"></a>Germline SNPs + Indels</h4><p>将一个或多个个体放在一起call snp，得到一个 joint callset，该snp calling的策略称为<strong>joint calling</strong></p>
<p><img src="/images/GATK4-pipeline-Germline-SNPs-Indels.png" alt></p>
<ul>
<li><p>第一步，单独为每个样本生成后续分析所需的中间文件——gVCF文件。这一步中包含了对原始fastq数据的质控、比对、排序、标记重复序列、BQSR和HaplotypeCaller gVCF等过程。这些过程全部都适合在单样本维度下独立完成。值得注意的是，与单样本模式不同，该模式中每个样本的gVCF应该成为这类流程的标配，在后续的步骤中我们可以通过gVCF很方便地完成群体的Joint Calling；</p>
</li>
<li><p>第二步，依据第一步完成的gVCF对这个群体进行Joint Calling，从而得到这个群体的变异结果和每个人准确的基因型（Genotype），最后使用VQSR完成变异的质控。</p>
</li>
</ul>
<h5 id="SNP、-INDEL位点识别-1"><a href="#SNP、-INDEL位点识别-1" class="headerlink" title="SNP、 INDEL位点识别"></a>SNP、 INDEL位点识别</h5><p>1、单独为每个样本生成后续分析所需的中间文件——gVCF文件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk HaplotypeCaller -R Ref/chr17.fa -I sam/T.chr17.recal.bam -ERC GVCF --dbsnp ../Ref/VCF/dbsnp_138.hg19.vcf \</span><br><span class="line">	-O calling/T.chr17.raw.snps.indels.vcf -L chr17:7400000-7800000</span><br></pre></td></tr></table></figure>
<p>HaplotypeCaller可以同时call snp 和indel，通过局部 de-novo 组装而非基于mapping的结果。一旦程序在某个区域发现了变异信号，它会忽略已有的mapping信息，在该区域执行reads重组装</p>
<p>如何对指定的区域call snp？</p>
<blockquote>
<p>使用 <code>-L</code> 参数指定识别突变位点的区域</p>
<ol>
<li><p>提供字符串指定单个interval，如<code>-L chr17:7400000-7800000</code> 只识别17号染色体7400000-7800000 区域的突变位点。或者使用 <code>-XL</code> 参数排除某个区域</p>
</li>
<li><p>提供一个保存interval lists的文件。GATK支持多种格式的interval lists文件</p>
<ul>
<li><p>Picard-style <code>.interval_list</code></p>
<p>  由 SAM-like header 和 intervals 信息组成，坐标系统为 1-based</p>
  <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">&gt; 		@HD     VN:1.0  SO:coordinate</span><br><span class="line">&gt; 		@SQ     SN:1    LN:249250621    AS:GRCh37       UR:http://www.broadinstitute.org/ftp/pub/seq/references/Homo_sapiens_assembly19.fasta   M5:1b22b98cdeb4a9304cb5d48026a85128     SP:Homo Sapiens</span><br><span class="line">&gt; 		@SQ     SN:2    LN:243199373    AS:GRCh37       UR:http://www.broadinstitute.org/ftp/pub/seq/references/Homo_sapiens_assembly19.fasta   M5:a0d9851da00400dec1098a9255ac712e     SP:Homo Sapiens</span><br><span class="line">&gt; 		1       30366   30503   +       target_1</span><br><span class="line">&gt; 		1       69089   70010   +       target_2</span><br><span class="line">&gt; 		1       367657  368599  +       target_3</span><br><span class="line">&gt; 		1       621094  622036  +       target_4</span><br><span class="line">&gt; 		1       861320  861395  +       target_5</span><br><span class="line">&gt; 		1       865533  865718  +       target_6</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</li>
</ul>
</li>
</ol>
</blockquote>
<blockquote>
<pre><code>- GATK-style `.list` or `.intervals`

    这种格式很简单，intervals需要写成这种格式：`&lt;chr&gt;:&lt;start&gt;-&lt;stop&gt;`，坐标系统为 1-based

- BED files `.bed`

    BED3格式：`&lt;chr&gt; &lt;start&gt; &lt;stop&gt;`，坐标系统为 0-based，GATK只接受 1-based 坐标系统，因此GATK会根据文件后缀 `.bed` 识别bed文件格式，然后会将 0-based 转换为 1-based

**注意：**
&gt; - intervals 必须按照 reference 的坐标进行排序
&gt; - 可以提供多个intervals集合，文件格式不必一致：`-L intervals1.interval_list -L intervals2.list -L intervals3.bed ...`
</code></pre></blockquote>
<p>2、合并多个GVCF文件得到GenomicsDB，为joint genotyping做准备。若是单样本跳过该步骤</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ gatk --java-options &quot;-Xmx4g -Xms4g&quot; GenomicsDBImport \</span><br><span class="line">      -V data/gvcfs/mother.g.vcf.gz \</span><br><span class="line">      -V data/gvcfs/father.g.vcf.gz \</span><br><span class="line">      -V data/gvcfs/son.g.vcf.gz \</span><br><span class="line">      --genomicsdb-workspace-path my_database \</span><br><span class="line">      -L 20</span><br></pre></td></tr></table></figure>
<p>也可以使用CombineGVCFs实现该功能。CombineGVCFs继承自GATK4之前的版本，但是它的运行速度不如GenomicsDB</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ gatk CombineGVCFs \</span><br><span class="line">   -R reference.fasta \</span><br><span class="line">   --variant sample1.g.vcf.gz \</span><br><span class="line">   --variant sample2.g.vcf.gz \</span><br><span class="line">   -O cohort.g.vcf.gz</span><br></pre></td></tr></table></figure>
<p>3、联合call snp，得到 joint-called SNP&amp;indel</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 单样本</span><br><span class="line">$ gatk --java-options &quot;-Xmx4g&quot; GenotypeGVCFs \</span><br><span class="line">   -R Homo_sapiens_assembly38.fasta \</span><br><span class="line">   -V input.g.vcf.gz \</span><br><span class="line">   -O output.vcf.gz</span><br><span class="line"></span><br><span class="line"># 多样本</span><br><span class="line">$ gatk --java-options &quot;-Xmx4g&quot; GenotypeGVCFs \</span><br><span class="line">   -R Homo_sapiens_assembly38.fasta \</span><br><span class="line">   -V gendb://my_database \</span><br><span class="line">   -O output.vcf.gz</span><br></pre></td></tr></table></figure>
<h5 id="SNP、-INDEL位点过滤"><a href="#SNP、-INDEL位点过滤" class="headerlink" title="SNP、 INDEL位点过滤"></a>SNP、 INDEL位点过滤</h5><ul>
<li><p>方法一：通过质量校正来过滤（Filter Variants by Variant (Quality Score) Recalibration）</p>
<p>  1、用机器学习的方法基于已知的变异位点对caller给出的原始 variant quality score 进行校正  (VQSR)，包含两步：（1）构建校正模型；（2）应用模型进行质量值校正</p>
<p>  由于评价SNP和Indel质量高低的标准不同，因此需要分SNP和Indel两种不同的模式，分别进行校正</p>
  <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line"># SNP mode</span><br><span class="line">## 构建校正模型</span><br><span class="line">$ gatk VariantRecalibrator \</span><br><span class="line">   -R Homo_sapiens_assembly38.fasta \</span><br><span class="line">   -V input.vcf.gz \</span><br><span class="line">   --resource hapmap,known=false,training=true,truth=true,prior=15.0:hapmap_3.3.hg38.sites.vcf.gz \</span><br><span class="line">   --resource omni,known=false,training=true,truth=false,prior=12.0:1000G_omni2.5.hg38.sites.vcf.gz \</span><br><span class="line">   --resource 1000G,known=false,training=true,truth=false,prior=10.0:1000G_phase1.snps.high_confidence.hg38.vcf.gz \</span><br><span class="line">   --resource dbsnp,known=true,training=false,truth=false,prior=2.0:Homo_sapiens_assembly38.dbsnp138.vcf.gz \</span><br><span class="line">   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \</span><br><span class="line">   -mode SNP \</span><br><span class="line">   -O output.snps.recal \</span><br><span class="line">   --tranches-file output.snps.tranches \</span><br><span class="line">   --rscript-file output.snps.plots.R</span><br><span class="line">## 应用模型进行质量值校正</span><br><span class="line">$ gatk ApplyVQSR \</span><br><span class="line">   -R Homo_sapiens_assembly38.fasta \</span><br><span class="line">   -V input.vcf.gz \</span><br><span class="line">   -O output.snps.VQSR.vcf.gz \</span><br><span class="line">   --truth-sensitivity-filter-level 99.0 \</span><br><span class="line">   --tranches-file output.snps.tranches \</span><br><span class="line">   --recal-file output.snps.recal \</span><br><span class="line">   -mode SNP</span><br><span class="line"></span><br><span class="line"># Indel mode</span><br><span class="line">## 上一步SNP mode产生的输出作为这一步的输入</span><br><span class="line">## 构建校正模型</span><br><span class="line">$ gatk VariantRecalibrator \</span><br><span class="line">   -R Homo_sapiens_assembly38.fasta \</span><br><span class="line">   -V input.snps.VQSR.vcf.gz \</span><br><span class="line">   --resource hapmap,known=false,training=true,truth=true,prior=15.0:hapmap_3.3.hg38.sites.vcf.gz \</span><br><span class="line">   --resource omni,known=false,training=true,truth=false,prior=12.0:1000G_omni2.5.hg38.sites.vcf.gz \</span><br><span class="line">   --resource 1000G,known=false,training=true,truth=false,prior=10.0:1000G_phase1.snps.high_confidence.hg38.vcf.gz \</span><br><span class="line">   --resource dbsnp,known=true,training=false,truth=false,prior=2.0:Homo_sapiens_assembly38.dbsnp138.vcf.gz \</span><br><span class="line">   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \</span><br><span class="line">   -mode INDEL \</span><br><span class="line">   -O output.indels.recal \</span><br><span class="line">   --tranches-file output.indels.tranches \</span><br><span class="line">   --rscript-file output.indels.plots.R</span><br><span class="line">## 应用模型进行质量值校正</span><br><span class="line">$ gatk ApplyVQSR \</span><br><span class="line">   -R Homo_sapiens_assembly38.fasta \</span><br><span class="line">   -V input.snps.VQSR.vcf.gz \</span><br><span class="line">   -O output.snps.indels.VQSR.vcf.gz \</span><br><span class="line">   --truth-sensitivity-filter-level 99.0 \</span><br><span class="line">   --tranches-file output.indels.tranches \</span><br><span class="line">   --recal-file output.indels.recal \</span><br><span class="line">   -mode INDEL</span><br></pre></td></tr></table></figure>
<p>  执行变异校正需要满足两个条件：</p>
<blockquote>
<ul>
<li>大量高质量的已知variants作为训练集，而这对于许多的物种是不满足的</li>
<li>数据集不能太小，因为它需要足够的数据集来识别 good vs. bad variants</li>
</ul>
</blockquote>
<p>  因此对于只有一两个样本、靶向测序、RNA-seq、非模式动物的数据，都不推荐进行变异质量值校正。若以上两个条件都不满足，则需要采取直接过滤 (hard-filtering) 的策略</p>
</li>
<li><p>方法二：直接过滤 (hard-filtering)</p>
<p>  <strong>Steps</strong></p>
<blockquote>
<ul>
<li>Extract the SNPs from the call set</li>
<li>Apply the filter to the SNP call set</li>
<li>Extract the Indels from the call set</li>
<li>Apply the filter to the Indel call set</li>
<li>Combine SNP and indel call set</li>
<li>Get passed call set</li>
</ul>
</blockquote>
<ul>
<li><p><strong>提取SNP位点</strong></p>
  <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk SelectVariants -R Ref/chr17.fa -V calling/T.chr17.raw.snps.indels.genotype.vcf \</span><br><span class="line">--select-type-to-include SNP -O filter/T.chr17.raw.snps.genotype.vcf</span><br></pre></td></tr></table></figure>
</li>
</ul>
</li>
</ul>
<pre><code>    参数说明：

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">--select-type-to-include,-select-type:Type</span><br><span class="line">                              Select only a certain type of variants from the input file  This argument may be specified</span><br><span class="line">                              0 or more times. Default value: null. Possible values: &#123;NO_VARIATION, SNP, MNP, INDEL,</span><br><span class="line">                              SYMBOLIC, MIXED&#125;</span><br></pre></td></tr></table></figure>


- **提取INDEL位点**

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk SelectVariants -R Ref/chr17.fa -V calling/T.chr17.raw.snps.indels.genotype.vcf \</span><br><span class="line">--select-type-to-include INDEL -O filter/T.chr17.raw.indels.genotype.vcf</span><br></pre></td></tr></table></figure>


- **SNP位点过滤**

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk VariantFiltration -R Ref/chr17.fa -V filter/T.chr17.raw.snps.genotype.vcf --filter-expression &quot;QD &lt; 2.0 || FS &gt; 60.0 || MQ &lt; 40.0 || SOR &gt; 3.0 || MQRankSum &lt; -12.5 || \</span><br><span class="line">ReadPosRankSum &lt; -8.0&quot; --filter-name &quot;SNP_FILTER&quot; -O filter/T.chr17.filter.snps.genotype.vcf</span><br></pre></td></tr></table></figure>


    参数说明：

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">--filter-expression,-filter:String</span><br><span class="line">                              One or more expression used with INFO fields to filter  This argument may be specified 0</span><br><span class="line">                              or more times. Default value: null.</span><br><span class="line"></span><br><span class="line">--filter-name:String          Names to use for the list of filters  This argument may be specified 0 or more times.</span><br><span class="line">                              Default value: null.</span><br></pre></td></tr></table></figure>


- **INDEL位点过滤**

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk VariantFiltration -R Ref/chr17.fa -V filter/T.chr17.raw.indels.genotype.vcf --filter-expression &quot;QD &lt; 2.0 || FS &gt; 200.0 || SOR &gt; 10.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; \</span><br><span class="line">-20.0&quot; --filter-name &quot;INDEL_FILTER&quot; -O filter/T.chr17.filter.indels.genotype.vcf</span><br></pre></td></tr></table></figure>


- **合并过滤后SNP、 INDEL文件**

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk MergeVcfs -I filter/T.chr17.filter.snps.genotype.vcf -I \</span><br><span class="line">filter/T.chr17.filter.indels.genotype.vcf -O filter/T.chr17.filter.snps.indels.genotype.vcf</span><br></pre></td></tr></table></figure>


- **提取PASS突变位点**

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gatk SelectVariants -R Ref/chr17.fa -V filter/T.chr17.filter.snps.indels.genotype.vcf -O</span><br><span class="line">T.chr17.pass.snps.indels.genotype.vcf -select &quot;vc.isNotFiltered()&quot;</span><br></pre></td></tr></table></figure>


    参数说明：

    <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">--selectExpressions,-select:String</span><br><span class="line">                              One or more criteria to use when selecting the data  This argument may be specified 0 or</span><br><span class="line">                              more times. Default value: null.</span><br></pre></td></tr></table></figure>
</code></pre><h5 id="Hard-filter阈值探究"><a href="#Hard-filter阈值探究" class="headerlink" title="Hard-filter阈值探究"></a>Hard-filter阈值探究</h5><p>GATK4官网给出的推荐阈值：</p>
<blockquote>
<p>For SNPs:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; QD &lt; 2.0</span><br><span class="line">&gt; MQ &lt; 40.0</span><br><span class="line">&gt; FS &gt; 60.0</span><br><span class="line">&gt; SOR &gt; 3.0</span><br><span class="line">&gt; MQRankSum &lt; -12.5</span><br><span class="line">&gt; ReadPosRankSum &lt; -8.0</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<p>For indels:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">&gt; QD &lt; 2.0</span><br><span class="line">&gt; ReadPosRankSum &lt; -20.0</span><br><span class="line">&gt; InbreedingCoeff &lt; -0.8</span><br><span class="line">&gt; FS &gt; 200.0</span><br><span class="line">&gt; SOR &gt; 10.0</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<p><a href="https://software.broadinstitute.org/gatk/documentation/article?id=11097" target="_blank" rel="noopener">点此</a> 查看GATK4原始网页</p>
<p>该阈值选择来自于GATK4官网的推荐，阈值依据于比较真 vs. 假 snp的特征值（annotation values）统计分布</p>
<blockquote>
<p>One of the most helpful ways to approach hard-filtering is to visualize the distribution of annotation values for a truth set called using a particular pipeline. These distributions are sharped by both the pipeline methodology and the underlying physical properties of the sequence data; so for a given pairing of data generation technology + analysis pipeline, you can derive filtering thresholds based on what the distributions look like for the truth set</p>
</blockquote>
<p>评估数据来源：1000Genomes 中的 whole genome trio</p>
<p>获得真/假变异的方法：</p>
<blockquote>
<p>将whole genome trio用HaplotypeCaller的GVCF mode得到各样本单独的GVCF文件，然后再用GenotypeGVCFs进行joint-calling得到未过滤的VCF文件，最后用VQSR（基于统计机器学习方法）得到过滤的变异集合</p>
<p>将PASS的变异认为是真的，将被过滤的变异当作是假的</p>
</blockquote>
<p>比较真变异与假变异的特征值（annotation values）的分布，进行评估的特征值有：QD, FS, SOR, MQ, MQRankSum 和 ReadPosRankSum</p>
<ul>
<li><strong>QualByDepth (QD)</strong></li>
</ul>
<p>变异质量值（variant confidence (from the QUAL field)）除以该位点覆盖的非同源的depth（与参考基因组不同的），即对质量值按照depth进行了标准化</p>
<p><img src="/images/GATK4-pipeline-hard-filter-QD.png" alt></p>
<p>大多数被过滤掉的变异的落在低QD区域</p>
<p>推荐：QD≥2</p>
<ul>
<li><strong>FisherStrand (FS)</strong></li>
</ul>
<p>一个衡量strand bias的量</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Phred-scaled probability that there is strand bias at the site</span><br></pre></td></tr></table></figure>
<p>当某一个位点的strand bias的程度很小或没有，那么 FS 值会接近于0</p>
<p><img src="/images/GATK4-pipeline-hard-filter-FS.png" alt></p>
<p>大多数过滤掉的变异的落在FS大于55的区域</p>
<p>推荐：FS≤60</p>
<ul>
<li><strong>StrandOddsRatio (SOR)</strong></li>
</ul>
<p>另一种评估strand bias的量</p>
<p><img src="/images/GATK4-pipeline-hard-filter-SOR.png" alt></p>
<p>Notice most of the variants that have an SOR value greater than 3 fail the VQSR filter</p>
<p>推荐：SOR≥3</p>
<ul>
<li><strong>RMSMappingQuality (MQ)</strong></li>
</ul>
<p>覆盖改位点的reads的mapping quality的均方根</p>
<table>
<thead>
<tr>
<th style="text-align:center">原图</th>
<th style="text-align:center">局部放大</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center"><img src="/images/GATK4-pipeline-hard-filter-MQ-1.png" alt></td>
<td style="text-align:center"><img src="/images/GATK4-pipeline-hard-filter-MQ-2.png" alt></td>
</tr>
</tbody>
</table>
<p>推荐：MQ≥40</p>
<ul>
<li><strong>MappingQualityRankSumTest (MQRankSum)</strong></li>
</ul>
<p><img src="/images/GATK4-pipeline-hard-filter-MQRankSum.png" alt></p>
<p>推荐：MQRankSum≥-2.5</p>
<h2 id="SAMtools-BCFtools流程"><a href="#SAMtools-BCFtools流程" class="headerlink" title="SAMtools-BCFtools流程"></a>SAMtools-BCFtools流程</h2><h2 id="关于SNP的过滤"><a href="#关于SNP的过滤" class="headerlink" title="关于SNP的过滤"></a>关于SNP的过滤</h2><p>SNP过滤的意义：</p>
<blockquote>
<ul>
<li><p>第一，过滤到一些低质量的SNP可以防止calling一些假阳性的SNP，这些假阳性的SNP会很大程度影响到后续的一系列的分析，例如GWAS等的分析，最后影响相关生物学问题的解答；</p>
</li>
<li><p>第二，如果你有很多的个体，往往你的call完SNP后，VCF文件的大小的会比较大，如果不经过过滤，对下游的群体结构分析，PCA等相关的计算，都会对计算机产生巨大的负担（运行速度还有需要的内存等）</p>
</li>
</ul>
</blockquote>
<h3 id="使用vcftools进行SNP过滤"><a href="#使用vcftools进行SNP过滤" class="headerlink" title="使用vcftools进行SNP过滤"></a>使用vcftools进行SNP过滤</h3><p>VCFtools能干什么</p>
<blockquote>
<ul>
<li>过滤特定的变异</li>
<li>比较文件</li>
<li>变异注释</li>
<li>文件格式转换</li>
<li>校验和合并文件</li>
<li>对变异位点集合进行坐标运算</li>
</ul>
</blockquote>
<hr>
<p>参考资料：</p>
<p>(1) <a href="http://www.bio-info-trainee.com/838.html" target="_blank" rel="noopener">生信菜鸟团：GATK使用注意事项</a></p>
<p>(2) <a href="./supplement/%E5%85%A8%E5%A4%96%E6%98%BE%E5%AD%90%E7%BB%84%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90-GATK4.pdf">小天师兄《全外显子组测序分析》</a></p>
<p>(3) <a href="https://mp.weixin.qq.com/s/RfEt-O-R2njje5Asu3WpwQ" target="_blank" rel="noopener">RNA-Seq是否可以替代WES完成外显子的变异检测?二代测序的四种Read重复是如何产生的?</a></p>
<p>(4) <a href="http://www.bio-info-trainee.com/2003.html" target="_blank" rel="noopener">生信菜鸟团：仔细探究samtools的rmdup是如何行使去除PCR重复reads功能的</a></p>
<p>(5) <a href="http://www.biotrainee.com/thread-1402-1-1.html" target="_blank" rel="noopener">生信技能树论坛：GATK之BaseRecalibrator</a></p>
<p>(6) <a href="https://gatkforums.broadinstitute.org/gatk/discussion/1317/collected-faqs-about-bam-files" target="_blank" rel="noopener">GATK Forum: Collected FAQs about input files for sequence read data (BAM/CRAM)</a></p>
<p>(7) <a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=1204" target="_blank" rel="noopener">GATK Forum: What input files does the GATK accept / require?</a></p>
<p>(8) <a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=1319" target="_blank" rel="noopener">GATK Forum: Collected FAQs about interval lists</a></p>
<p>(9) <a href="https://software.broadinstitute.org/gatk/best-practices/" target="_blank" rel="noopener">GATK Forum：GATK Best Practices</a></p>
<p>(10) <a href="https://mp.weixin.qq.com/s/Sa019WuSg8fRQgkWAIG4pQ" target="_blank" rel="noopener">公众号碱基矿工：GATK4全基因组数据分析最佳实践 ，我以这篇文章为标志，终结当前WGS系列数据分析的流程主体问题 | 完全代码</a></p>
<p>(11) <a href="https://mp.weixin.qq.com/s/XVkFuU2zWLY5r6grDwzkcA" target="_blank" rel="noopener">生信菜鸟团：生信笔记：call snp是应该一起call还是分开call？</a></p>
<p>(12) <a href="https://software.broadinstitute.org/gatk/documentation/article?id=11069" target="_blank" rel="noopener">GATK Forum：Hard-filtering germline short variants</a></p>
<p>(13) <a href="https://mp.weixin.qq.com/s/Juf7LklWkaCmUT_pusOuCA" target="_blank" rel="noopener">生信杂谈：用泊松分布解释 NGS 测序数据的 duplication 问题</a></p>
<p>(14) <a href="https://www.jianshu.com/p/e05ff3cace56" target="_blank" rel="noopener">【简书】关于SNP的过滤（1）：如何使用vcftools进行SNP过滤</a></p>
<p>(15) <a href="https://vcftools.github.io/index.html" target="_blank" rel="noopener">VCFtools官网</a></p>

      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/02/07/Andrew-Ng-Coursera-Homework/" rel="next" title="Andrew NG Coursera 课程编程作业">
                <i class="fa fa-chevron-left"></i> Andrew NG Coursera 课程编程作业
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/02/08/Stat-on-RNAseq/" rel="prev" title="RNA-seq中的那些统计学问题">
                RNA-seq中的那些统计学问题 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/images/avatar.jpg" alt="Lianm">
            
              <p class="site-author-name" itemprop="name">Lianm</p>
              <p class="site-description motion-element" itemprop="description">中国科学院北京基因组研究所研究生</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">15</span>
                    <span class="site-state-item-name">日志</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">5</span>
                    <span class="site-state-item-name">分类</span>
                  
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">24</span>
                    <span class="site-state-item-name">标签</span>
                  
                </div>
              
            </nav>
          

          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://github.com/Ming-Lian" title="GitHub &rarr; https://github.com/Ming-Lian" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="mailto:lianming17m@big.ac.cn" title="E-Mail &rarr; mailto:lianming17m@big.ac.cn" rel="noopener" target="_blank"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                </span>
              
            </div>
          

          

          
          

          
            
          
          

        </div>
      </div>

      
      <!--noindex-->
        <div class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
            
            
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#GATK4流程"><span class="nav-number">1.</span> <span class="nav-text">GATK4流程</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#准备配套数据"><span class="nav-number">1.1.</span> <span class="nav-text">准备配套数据</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#BWA-Map-to-Reference"><span class="nav-number">1.2.</span> <span class="nav-text">BWA: Map to Reference</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#前期处理"><span class="nav-number">1.3.</span> <span class="nav-text">前期处理</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#去除PCR重复"><span class="nav-number">1.3.1.</span> <span class="nav-text">去除PCR重复</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#duplicates的产生原因"><span class="nav-number">1.3.1.1.</span> <span class="nav-text">duplicates的产生原因</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#用泊松分布解释-NGS-测序数据的-duplication-问题"><span class="nav-number">1.3.1.2.</span> <span class="nav-text">用泊松分布解释 NGS 测序数据的 duplication 问题</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#PCR-bias的影响"><span class="nav-number">1.3.1.3.</span> <span class="nav-text">PCR bias的影响</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#探究samtools和picard去除read-duplicates的方法"><span class="nav-number">1.3.1.4.</span> <span class="nav-text">探究samtools和picard去除read duplicates的方法</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#操作：排序及标记重复"><span class="nav-number">1.3.1.5.</span> <span class="nav-text">操作：排序及标记重复</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#质量值校正"><span class="nav-number">1.3.2.</span> <span class="nav-text">质量值校正</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#SNP、-INDEL位点识别"><span class="nav-number">1.4.</span> <span class="nav-text">SNP、 INDEL位点识别</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#SNP-calling-策略的选择"><span class="nav-number">1.4.1.</span> <span class="nav-text">SNP calling 策略的选择</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#Germline-SNPs-Indels"><span class="nav-number">1.4.2.</span> <span class="nav-text">Germline SNPs + Indels</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#SNP、-INDEL位点识别-1"><span class="nav-number">1.4.2.1.</span> <span class="nav-text">SNP、 INDEL位点识别</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#SNP、-INDEL位点过滤"><span class="nav-number">1.4.2.2.</span> <span class="nav-text">SNP、 INDEL位点过滤</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#Hard-filter阈值探究"><span class="nav-number">1.4.2.3.</span> <span class="nav-text">Hard-filter阈值探究</span></a></li></ol></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#SAMtools-BCFtools流程"><span class="nav-number">2.</span> <span class="nav-text">SAMtools-BCFtools流程</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#关于SNP的过滤"><span class="nav-number">3.</span> <span class="nav-text">关于SNP的过滤</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#使用vcftools进行SNP过滤"><span class="nav-number">3.1.</span> <span class="nav-text">使用vcftools进行SNP过滤</span></a></li></ol></li></ol></div>
            

          </div>
        </div>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Lianm</span>

  

  
</div>


  <div class="powered-by">由 <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> 强力驱动 v3.8.0</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a> v7.0.0</div>




        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>


  


  <script src="/js/src/utils.js?v=7.0.0"></script>

  <script src="/js/src/motion.js?v=7.0.0"></script>



  
  


  <script src="/js/src/affix.js?v=7.0.0"></script>

  <script src="/js/src/schemes/pisces.js?v=7.0.0"></script>



  
  <script src="/js/src/scrollspy.js?v=7.0.0"></script>
<script src="/js/src/post-details.js?v=7.0.0"></script>



  


  <script src="/js/src/bootstrap.js?v=7.0.0"></script>



  


  


  




  

  

  

  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js'; 
      }
      else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
      }
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  

  

  

  

  

  

  

</body>
</html>
