<!DOCTYPE html>












  


<html class="theme-next muse use-motion" lang>
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">


























<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2">

<link rel="stylesheet" href="/css/main.css?v=7.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=7.0.0">


  <link rel="mask-icon" href="/images/logo.svg?v=7.0.0" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    version: '7.0.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta property="og:type" content="website">
<meta property="og:title" content="Lianm&#39;s Blog">
<meta property="og:url" content="http://yoursite.com/index.html">
<meta property="og:site_name" content="Lianm&#39;s Blog">
<meta property="og:locale" content="default">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Lianm&#39;s Blog">






  <link rel="canonical" href="http://yoursite.com/">



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>Lianm's Blog</title>
  












  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Lianm's Blog</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">Focus on Bioinformatics and Machine-Learning</p>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home menu-item-active">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>Home</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>Archives</a>

  </li>

      
      
    </ul>
  

  
    

  

  
</nav>



  



</div>
    </header>

    
  
  
  
  

  

  <a href="https://github.com/Ming-Lian" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="80" height="80" viewbox="0 0 250 250" style="fill: #222; color: #fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a>



    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/24/Dual-Parallelization-for-shell-R/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/24/Dual-Parallelization-for-shell-R/" class="post-title-link" itemprop="url">【编程技巧】shell+R双重并行化——加速分析过程</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-24 14:14:36 / Modified: 14:19:04" itemprop="dateCreated datePublished" datetime="2019-02-24T14:14:36+00:00">2019-02-24</time>
            

            
              

              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/编程/" itemprop="url" rel="index"><span itemprop="name">编程</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="问题描述"><a href="#问题描述" class="headerlink" title="问题描述"></a>问题描述</h2><p>在进行宏基因组shotgun数据分析时，数据集分case和control两组，每组11个样本，通过Gene Profiling得到400多万行的profile（以RPKM方式进行定量），得到的proflie文件有600+MB</p>
<p>接着想用wilcox检验来筛选两组gene abundance存在差异的基因，结果跑了一天一夜没跑出结果</p>
<p>在排除脚本问题后，基本确定问题是数据量太大，计算效率太低导致的，在解决这个问题过程中总结出了一种加速技巧：<strong>shell+R的双重并行化</strong></p>
<h2 id="shell并行化"><a href="#shell并行化" class="headerlink" title="shell并行化"></a>shell并行化</h2><p>我目前所知道的在linux环境下的并行化实现方式有两种：</p>
<p>（1）利用Slurm排队系统</p>
<p>可以使用<code>--array=1-11%4</code>参数来指定任务队列编号为1-11，每次最多并行执行4个任务</p>
<p>有两种书写格式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">1. 作为sbatch的参数</span><br><span class="line">$ sbatch --array=1-11%4 tasks.sh</span><br><span class="line"></span><br><span class="line">2. 写在shell脚本内部</span><br><span class="line">如：</span><br><span class="line">	#!/bin/bash</span><br><span class="line">	#SBATCH --array=1-11%4</span><br><span class="line">	...</span><br></pre></td></tr></table></figure>
<p>（2）使用<code>do{...} &amp;;done;wait</code>语句</p>
<p>该语句的格式为：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">while read i</span><br><span class="line">do</span><br><span class="line">	&#123;</span><br><span class="line">	...</span><br><span class="line">	&#125; &amp;</span><br><span class="line">done</span><br><span class="line">wait</span><br><span class="line"></span><br><span class="line">或</span><br><span class="line"></span><br><span class="line">for((i=0;i&lt;n;i++)</span><br><span class="line">do</span><br><span class="line">	&#123;</span><br><span class="line">	...</span><br><span class="line">	&#125; &amp;</span><br><span class="line">done</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>
<p>例如：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">for((i=0;i&lt;10;i++))</span><br><span class="line">do</span><br><span class="line">	&#123;</span><br><span class="line">	echo $i</span><br><span class="line">	sleep 5</span><br><span class="line">	&#125; &amp;</span><br><span class="line">done</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>
<p>上面的命令如果不使用并行化，则总的执行时间为10*5=50秒，而并行之后只需要5秒</p>
<p>可以在该命令模块外面再套一层循环，来<strong>控制每次并行化的任务数量</strong>，这个需求在大多数情况下是非常必要的：若总的任务很多，如果不对每一次并行化的任务数进行控制，而是一股脑地全部提交并行任务，非常容易到达服务器的负荷上限，那么等着你的很可能就是其他用户的骂声一片和服务器管理员的警告了！</p>
<p>例如，从1到100，每次输出5个数，每次输出后间隔5秒后再继续下一轮的输出：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">for((i=0;5*i&lt;100;i++))</span><br><span class="line">do</span><br><span class="line">	for((j=5*i+1;j&lt;5*i+6;j++))</span><br><span class="line">	do</span><br><span class="line">		&#123;</span><br><span class="line">		echo $j</span><br><span class="line">		sleep 5</span><br><span class="line">		&#125; &amp;</span><br><span class="line">	done</span><br><span class="line">	wait</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<h2 id="R并行化"><a href="#R并行化" class="headerlink" title="R并行化"></a>R并行化</h2><p>R的并行化是在 <code>*apply</code> 系列函数的基础上产生的，所以在介绍R的并行化之前，有必要对 <code>*apply</code> 系列函数作一个简单的说明，下面只对<code>apply( )</code>进行说明</p>
<blockquote>
<p>函数语法格式：<code>apply(x, margin, fun, ...)</code></p>
<ul>
<li><p><code>x</code>：一个data.frame或者是一个matrix</p>
</li>
<li><p><code>margin</code>：选择1或者2，1表示行，2表示列</p>
</li>
<li><p><code>fun</code>：一个函数对象，可以是R自带的，也可以是用户自定义的函数</p>
</li>
<li><p><code>...</code>：传递给函数fun的其他变量</p>
</li>
</ul>
<p>例如：<code>apply(x,1,sum)</code>，将变量x逐行传递给函数sum，进行求和，得到的是变量x每行的和的列表</p>
</blockquote>
<p>一个任务之所以能够进行并行化处理，是因为该任务可以被拆分成许多个相互独立的小任务，每个小任务的求解对其他任务没有任何影响，因此可以对它们进行分而治之，最后将每个小任务求解结果进行汇总，即简单地合并，apply函数实现的就是这样的任务，将一个比较大的变量X按照行（margin=1）或列（margin=2）进行拆分，然后对每个行分别独立地进行求解</p>
<p>但是<code>*apply</code>系列函数的分治策略并没有进行并行化，是一个只利用了一个线程的串行任务，但是由于它本身的分治属性，对它进行简单地改造和封装，就可以实现高效地并行化，这便是<code>parallel</code>包</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">library(parallel)</span><br><span class="line"></span><br><span class="line"># 初始化一个并行化的集群</span><br><span class="line">## 设置集群使用的核心数</span><br><span class="line">if (is.na(Args[4]))&#123;</span><br><span class="line">	no_cores &lt;- detectCores() - 10</span><br><span class="line">&#125; else &#123;</span><br><span class="line">	no_cores &lt;- as.integer(Args[4])</span><br><span class="line">&#125;</span><br><span class="line">## 按照指定的核心数，初始化一个集群对象</span><br><span class="line">cl &lt;- makeCluster(no_cores)</span><br><span class="line"></span><br><span class="line"># 调用parApply函数，执行并行化任务</span><br><span class="line">out &lt;- parApply(cl,data,1,fun)</span><br><span class="line"></span><br><span class="line"># 任务结束后，关闭集群对象</span><br><span class="line">stopCluster(cl)</span><br></pre></td></tr></table></figure>
<h2 id="shell-R双重并行化实现案例"><a href="#shell-R双重并行化实现案例" class="headerlink" title="shell+R双重并行化实现案例"></a>shell+R双重并行化实现案例</h2><p>有了上面提到的shell与R的并行化的实现方法，那么我们就可以将其运用于本文一开始提到的问题</p>
<p>并行化的思路：</p>
<blockquote>
<p>（1）先将原始的大profile文件进行拆分，若每m行拆成一个subprofile，可以拆成n个subprofiles</p>
<p>（2）对这n个subprofiles，按照每i个执行一次并行化任务（<strong>shell并行化</strong>），给每个并行化任务j个核心（<strong>R并行化</strong>）</p>
</blockquote>
<ol>
<li><p><strong>shell脚本</strong></p>
<p> 脚本名：<code>ParWilcox.sh</code></p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"># 参数说明：</span><br><span class="line"># - Profile files</span><br><span class="line"># - rows per sub-profile</span><br><span class="line"># - maximum tasks per run</span><br><span class="line"># - outdir</span><br><span class="line"></span><br><span class="line">profiles=$1</span><br><span class="line">nrow_per=$2</span><br><span class="line">num_tasks=$3</span><br><span class="line">outdir=$4</span><br><span class="line">if [ ! -d $outdir ]</span><br><span class="line">then</span><br><span class="line">	mkdir -p $outdir</span><br><span class="line">fi</span><br><span class="line">if [ -f $outdir/SubProfiles.list ]</span><br><span class="line">then</span><br><span class="line">	rm $outdir/SubProfiles.list</span><br><span class="line">fi</span><br><span class="line"></span><br><span class="line">##################################需要修改的变量##################################</span><br><span class="line">workdir=/Path/To/Workdir</span><br><span class="line">################################################################################</span><br><span class="line"></span><br><span class="line"># 分割原始Profiles</span><br><span class="line">nline=$(wc -l $profiles | awk &apos;&#123;printf $1&#125;&apos;)</span><br><span class="line">for((i=0;i*nrow_per&lt;nline;i++))</span><br><span class="line">do</span><br><span class="line">	awk -v j=$i -v n=$nrow_per &apos;NR==1 || (NR&gt;j*n+1 &amp;&amp; NR&lt;=(j+1)*n+1)&apos; $profiles &gt;$outdir/SubProfiles-$&#123;i&#125;</span><br><span class="line">	echo &quot;SubProfiles-$&#123;i&#125;&quot; &gt;&gt;$outdir/SubProfiles.list</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 并行化执行多个sub-profile的wilcox检验</span><br><span class="line">num_SubProfiles=$(wc -l $outdir/SubProfiles.list | awk &apos;&#123;printf $1&#125;&apos;)</span><br><span class="line">for((i=0;i*num_tasks&lt;num_SubProfiles;i++))</span><br><span class="line">do</span><br><span class="line">	awk -v j=$i -v n=$num_tasks &apos;NR&gt;j*n &amp;&amp; NR&lt;=(j+1)*n&apos; $outdir/SubProfiles.list &gt; $outdir/current_subprofiles.List</span><br><span class="line">	while read subprofile</span><br><span class="line">	do</span><br><span class="line">		&#123;</span><br><span class="line">		Rscript $workdir/Script/wilcox-sided-parallel.R $outdir/$subprofile $workdir/Gene_abundance/SampleGroup.txt $outdir/statOut.$subprofile 2</span><br><span class="line">		&#125; &amp;</span><br><span class="line">	done &lt; $outdir/current_subprofiles.List</span><br><span class="line">	wait</span><br><span class="line">done</span><br><span class="line">cat $outdir/statOut.* &gt;$outdir/statOut</span><br><span class="line">rm $outdir/statOut.*</span><br></pre></td></tr></table></figure>
</li>
<li><p><strong>被shell脚本调用的执行wilcox检验的R脚本</strong></p>
<p> 脚本名：<code>wilcox-sided-parallel.R</code></p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><span class="line"># 三个参数，按顺序分别为：</span><br><span class="line"># - profiles matrix file</span><br><span class="line"># - sample group file</span><br><span class="line"># - outfile name</span><br><span class="line"># - threads number（默认为服务器总线程-10）</span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(T)</span><br><span class="line"></span><br><span class="line">library(parallel)</span><br><span class="line"></span><br><span class="line"># Initiate cluster</span><br><span class="line">if (is.na(Args[4]))&#123;</span><br><span class="line">	no_cores &lt;- detectCores() - 10</span><br><span class="line">&#125; else &#123;</span><br><span class="line">	no_cores &lt;- as.integer(Args[4])</span><br><span class="line">&#125;</span><br><span class="line">cl &lt;- makeCluster(no_cores)</span><br><span class="line"></span><br><span class="line">data &lt;- read.table(Args[1],head=T,row.names=1,sep=&apos;\t&apos;)</span><br><span class="line">group &lt;- read.table(Args[2],head=T,sep=&apos;\t&apos;)</span><br><span class="line"></span><br><span class="line"># 找出profiles中每列代表的样本所属的组</span><br><span class="line">col_group1 &lt;- colnames(data) %in% sub(&apos;-&apos;,&apos;.&apos;,paste(group$Sample[group$Group==1],&apos;.rpkm&apos;,sep=&apos;&apos;))</span><br><span class="line">col_group2 &lt;- colnames(data) %in% sub(&apos;-&apos;,&apos;.&apos;,paste(group$Sample[group$Group==2],&apos;.rpkm&apos;,sep=&apos;&apos;))</span><br><span class="line"></span><br><span class="line"># 对单个基因进行wilcox检验，并输出统计检验结果，输出格式如下：</span><br><span class="line"># 	geneName,pvalue,group1Median,group2Median,direction</span><br><span class="line">wilcox_fun &lt;- function(data,col_group1,col_group2)&#123;</span><br><span class="line">	g1 &lt;- as.numeric(data[col_group1])</span><br><span class="line">	g2 &lt;- as.numeric(data[col_group2])</span><br><span class="line">	# 执行wilcox检验</span><br><span class="line">	stat &lt;- wilcox.test(g1,g2,paired = FALSE, exact=NULL, correct=TRUE, alternative=&quot;two.sided&quot;)</span><br><span class="line">	if(is.na(stat$p.value))&#123;</span><br><span class="line">		stat$p.value &lt;- 1</span><br><span class="line">	&#125;</span><br><span class="line">	# 对有统计学意义的基因进行判断，是上调&quot;up&quot;还是下调&quot;down&quot;或者是不变&quot;-&quot;（对于组2，即不吃药组）</span><br><span class="line">	if(stat$p.value &lt; 0.1  &amp; median(g1) &lt; median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;down&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else if(stat$p.value &lt; 0.1  &amp; median(g1) &gt; median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;up&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else if(stat$p.value &lt; 0.1  &amp; median(g1) == median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;-&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else&#123;</span><br><span class="line">		G12 &lt;- c()</span><br><span class="line">	&#125;</span><br><span class="line">	G12</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">statOut &lt;- parApply(cl,data,1,wilcox_fun,col_group1,col_group2)</span><br><span class="line">stopCluster(cl)</span><br><span class="line"></span><br><span class="line"># 删除为NULL的列表元素</span><br><span class="line">for(i in names(statOut))&#123;</span><br><span class="line">	if(is.null(statOut[[i]]))&#123;</span><br><span class="line">		statOut[[i]] &lt;- NULL</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"># 将结果写入文件中</span><br><span class="line">if(!is.null(statOut))&#123;</span><br><span class="line">	# 转换成数据框</span><br><span class="line">	statOut &lt;- as.data.frame(t(as.matrix(as.data.frame(statOut))))</span><br><span class="line">	colnames(statOut) &lt;- c(&apos;pvalue&apos;,&apos;group1Median&apos;,&apos;group2Median&apos;,&apos;direction&apos;)</span><br><span class="line">	# 写入文件</span><br><span class="line">	write.table(statOut,Args[3],sep = &apos;\t&apos;,row.names = T,col.names = T,quote = F)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
</li>
</ol>
<p>执行方法如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ bash ParWilcox.sh &lt;raw profile&gt; &lt;rows per sub-profile&gt; &lt;tasks per run&gt; &lt;threads&gt;</span><br></pre></td></tr></table></figure>
<p>脚本中用到的样本分组文件，如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">Sample	Group</span><br><span class="line">NP-003A	1</span><br><span class="line">NP-017A	1</span><br><span class="line">NP-018A	1</span><br><span class="line">...</span><br><span class="line">NP-007A	2</span><br><span class="line">NP-010A	2</span><br><span class="line">NP-011A	2</span><br><span class="line">...</span><br></pre></td></tr></table></figure>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/24/Follow-LYL-learning-math-physics/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/24/Follow-LYL-learning-math-physics/" class="post-title-link" itemprop="url">跟着李永乐老师学数学/物理</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-24 14:05:50 / Modified: 14:09:33" itemprop="dateCreated datePublished" datetime="2019-02-24T14:05:50+00:00">2019-02-24</time>
            

            
              

              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/MathIsFunGame/" itemprop="url" rel="index"><span itemprop="name">MathIsFunGame</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="数学部分"><a href="#数学部分" class="headerlink" title="数学部分"></a>数学部分</h2><h3 id="1-祖暅原理：我国古代数学家算球体积的方法"><a href="#1-祖暅原理：我国古代数学家算球体积的方法" class="headerlink" title="1. 祖暅原理：我国古代数学家算球体积的方法"></a>1. 祖暅原理：我国古代数学家算球体积的方法</h3><p><img src="/images/Follow-LYL-1.jpg" alt></p>
<h3 id="2-容斥原理：春节集卡活动集齐的概率有多大"><a href="#2-容斥原理：春节集卡活动集齐的概率有多大" class="headerlink" title="2. 容斥原理：春节集卡活动集齐的概率有多大"></a>2. 容斥原理：春节集卡活动集齐的概率有多大</h3><p><img src="/images/Follow-LYL-2.jpg" alt></p>
<hr>
<p>参考资料：</p>
<p>(1) 李永乐老师B站视频</p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/17/Practical-Scripts/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/17/Practical-Scripts/" class="post-title-link" itemprop="url">实用小脚本</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-17 08:14:59" itemprop="dateCreated datePublished" datetime="2019-02-17T08:14:59+00:00">2019-02-17</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-02-24 13:57:02" itemprop="dateModified" datetime="2019-02-24T13:57:02+00:00">2019-02-24</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/编程/" itemprop="url" rel="index"><span itemprop="name">编程</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="Perl"><a href="#Perl" class="headerlink" title="Perl"></a>Perl</h2><h3 id="1-拆分FASTA文件"><a href="#1-拆分FASTA文件" class="headerlink" title="1. 拆分FASTA文件"></a>1. 拆分FASTA文件</h3><p>将FASTA文件按照用户指定的序列条数进行拆分，即每n条序列写到一个文件中，或者按照指定的输出文件数进行拆分，则每个文件中包含的序列数相同</p>
<p>脚本名：<code>splitFasta.pl</code><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line">use POSIX;</span><br><span class="line"></span><br><span class="line"># 帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">	This script is used to split fasta file, which is too large with thosands of sequence</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">	$0 -i &lt;input&gt; -o &lt;output_dir&gt; [-n &lt;seq_num_per_file&gt;] [-m &lt;output_file_num&gt;]</span><br><span class="line">	</span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">	-i	[str]	Input raw fasta file</span><br><span class="line">	-o	[str]	Output file to which directory</span><br><span class="line">	-n	[int]	Sequence number per file, alternate chose paramerter &quot;-n&quot; or &quot;-m&quot;, if set &quot;-n&quot; and &quot;-m&quot; at the same time, only take &quot;-n&quot; parameter</span><br><span class="line">	-m	[int]	Output file number (default:100)</span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($input,$output_dir,$seq_num,$file_num);</span><br><span class="line">GetOptions(</span><br><span class="line">	&quot;i:s&quot;=&gt;\$input,</span><br><span class="line">	&quot;o:s&quot;=&gt;\$output_dir,</span><br><span class="line">	&quot;n:i&quot;=&gt;\$seq_num,</span><br><span class="line">	&quot;m:i&quot;=&gt;\$file_num</span><br><span class="line">	);</span><br><span class="line"></span><br><span class="line">die `pod2text $0` if ((!$input) or (!$output_dir));</span><br><span class="line"></span><br><span class="line"># 设置每个文件的序列条数</span><br><span class="line">if(!defined($seq_num))&#123;</span><br><span class="line">	if(!defined($file_num))&#123;</span><br><span class="line">		$file_num=100;</span><br><span class="line">		my $total_seq_num=`awk &apos;BEGIN&#123;n=0&#125; /^&gt;/&#123;n++&#125; END&#123;print n&#125;&apos; $input`;</span><br><span class="line">		chomp $total_seq_num;</span><br><span class="line">		$seq_num=ceil($total_seq_num/$file_num);</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		my $total_seq_num=`awk &apos;BEGIN&#123;n=0&#125; /^&gt;/&#123;n++&#125; END&#123;print n&#125;&apos; $input`;</span><br><span class="line">		chomp $total_seq_num;</span><br><span class="line">		$seq_num=ceil($total_seq_num/$file_num);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">open IN,&quot;&lt;$input&quot; or die &quot;Cann&apos;t open $input\n&quot;;</span><br><span class="line"></span><br><span class="line">my $n_seq=0;	# 该变量用于记录当前扫描到的序列数</span><br><span class="line">my $n_file=1;	# 该变量用于记录当前真正写入的文件的计数</span><br><span class="line">my $input_base=`basename $input`;</span><br><span class="line">chomp $input_base;</span><br><span class="line"></span><br><span class="line">open OUT,&quot;&gt;$output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;&quot; or die &quot;Cann&apos;t create $output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;\n&quot;;</span><br><span class="line"></span><br><span class="line">while(&lt;IN&gt;)&#123;</span><br><span class="line">	next if (/^\s+$/);	# 跳过空行</span><br><span class="line">	chomp;</span><br><span class="line">	if (/^&gt;/)&#123;</span><br><span class="line">		$n_seq++;</span><br><span class="line">		# 判断目前已经扫描到的序列数，若大于设定的split的序列数，则创建新文件</span><br><span class="line">		if ($n_seq&gt;$seq_num)&#123;</span><br><span class="line">			$n_seq=1;</span><br><span class="line">			$n_file++;</span><br><span class="line">			close OUT;</span><br><span class="line">			open OUT,&quot;&gt;$output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;&quot; or die &quot;Cann&apos;t create $output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;\n&quot;;</span><br><span class="line">			print OUT &quot;$_\n&quot;;</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			print OUT &quot;$_\n&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		print OUT &quot;$_\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close IN;</span><br></pre></td></tr></table></figure></p>
<p>执行方法： </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl splitFasta.pl -i &lt;input&gt; -o &lt;output_dir&gt; [-n &lt;seq_num_per_file&gt;] [-m &lt;output_file_num&gt;]</span><br></pre></td></tr></table></figure>
<p>具体的参数使用说明可以执行 <code>perl splitFasta.pl</code>，查看脚本使用文档</p>
<h3 id="2-从双端FASTQ文件中抽取指定数据量-bp-的序列"><a href="#2-从双端FASTQ文件中抽取指定数据量-bp-的序列" class="headerlink" title="2. 从双端FASTQ文件中抽取指定数据量(bp)的序列"></a>2. 从双端FASTQ文件中抽取指定数据量(bp)的序列</h3><p>按照用户指定的数据量，即多少bp，从原始的双端FASTQ文件中随机抽取序列，要求双端FASTQ文件中的序列必须配对</p>
<p>脚本名：<code>extractFastq.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line"></span><br><span class="line"># 脚本帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">	Thise script is used to extract a number of fastq recorders from original input fastq file</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">	$0 -n &lt;totalReads&gt; -e &lt;base-pairs to extract&gt; -l &lt;readLength&gt; -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br><span class="line"></span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">	-n	[int]	Total reads number of input.fastq</span><br><span class="line">	-e	[int]	Base-pairs per end you want to extract</span><br><span class="line">	-l	[int]	The reads length</span><br><span class="line">	-1	[str]	Input 1st-end fastq file</span><br><span class="line">	-2	[str]	Input 2nd-end fastq file</span><br><span class="line">	-o	[str]	Output directory [default: current folder]</span><br><span class="line"></span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($totalReads,$bpNum,$length,$Input1,$Input2,$Outdir);</span><br><span class="line">GetOptions(</span><br><span class="line">	&quot;n:i&quot;=&gt;\$totalReads,</span><br><span class="line">	&quot;e:i&quot;=&gt;\$bpNum,</span><br><span class="line">	&quot;l:i&quot;=&gt;\$length,</span><br><span class="line">	&quot;1:s&quot;=&gt;\$Input1,</span><br><span class="line">	&quot;2:s&quot;=&gt;\$Input2,</span><br><span class="line">	&quot;o:s&quot;=&gt;\$Outdir</span><br><span class="line">	);</span><br><span class="line"></span><br><span class="line">$Outdir=`pwd` unless (defined($Outdir));</span><br><span class="line">die `pod2text $0` if ((!$totalReads) or (!$bpNum)) or (!$length) or (!$Input1) or (!$Input2);</span><br><span class="line"></span><br><span class="line">open FQ1,&quot;&lt;$Input1&quot; or die &quot;$!\n&quot;;</span><br><span class="line">open FQ2,&quot;&lt;$Input2&quot; or die &quot;$!\n&quot;;</span><br><span class="line"></span><br><span class="line">my $Input1_basename=`basename $Input1`;</span><br><span class="line">chomp $Input1_basename;</span><br><span class="line">my $Input2_basename=`basename $Input2`;</span><br><span class="line">chomp $Input2_basename;</span><br><span class="line">open OUT1,&quot;&gt;$Outdir/$&#123;Input1_basename&#125;.extract&quot; or die &quot;$!\n&quot;;</span><br><span class="line">open OUT2,&quot;&gt;$Outdir/$&#123;Input2_basename&#125;.extract&quot; or die &quot;$!\n&quot;;</span><br><span class="line"></span><br><span class="line">my $readsRemain=$bpNum/$length;</span><br><span class="line"></span><br><span class="line">my $remainCount=0;</span><br><span class="line"></span><br><span class="line">while(! eof($FQ1))&#123;</span><br><span class="line">	# 读入1st-end fastq 文件的四行</span><br><span class="line">	my $fq1_1=&lt;FQ1&gt;;</span><br><span class="line">	my $fq1_2=&lt;FQ1&gt;;</span><br><span class="line">	my $fq1_3=&lt;FQ1&gt;;</span><br><span class="line">	my $fq1_4=&lt;FQ1&gt;;</span><br><span class="line">	chomp($fq1_1,$fq1_2,$fq1_3,$fq1_4);</span><br><span class="line">	# 读入2nd-end fastq 文件的四行</span><br><span class="line">	my $fq2_1=&lt;FQ2&gt;;</span><br><span class="line">	my $fq2_2=&lt;FQ2&gt;;</span><br><span class="line">	my $fq2_3=&lt;FQ2&gt;;</span><br><span class="line">	my $fq2_4=&lt;FQ2&gt;;</span><br><span class="line">	chomp($fq2_1,$fq2_2,$fq2_3,$fq2_4);</span><br><span class="line"></span><br><span class="line">	# 随机抽取</span><br><span class="line">	if (rand()&lt;$readsRemain/$totalReads)&#123;</span><br><span class="line">		$remainCount++;</span><br><span class="line">		print OUT1 &quot;$fq1_1\n$fq1_2\n$fq1_3\n$fq1_4\n&quot;;</span><br><span class="line">		print OUT2 &quot;$fq2_1\n$fq2_2\n$fq2_3\n$fq2_4\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">print &quot;Total reads: $totalReads\n&quot;;</span><br><span class="line">print &quot;Theorical remained reads: $readsRemain\n&quot;;</span><br><span class="line">print &quot;Practical remained reads: $remainCount\n&quot;;</span><br><span class="line"></span><br><span class="line">close FQ1;</span><br><span class="line">close FQ2;</span><br><span class="line">close OUT1;</span><br><span class="line">close OUT2;</span><br></pre></td></tr></table></figure>
<p>执行方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl extractFastq.pl -n &lt;totalReads&gt; -e &lt;base-pairs to extract&gt; -l &lt;readLength&gt; -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br></pre></td></tr></table></figure>
<p>具体的参数使用说明可以执行 <code>perl extractFastq.pl</code>，查看脚本使用文档</p>
<h3 id="3-双端FASTQ文件进行双端配对"><a href="#3-双端FASTQ文件进行双端配对" class="headerlink" title="3. 双端FASTQ文件进行双端配对"></a>3. 双端FASTQ文件进行双端配对</h3><p>脚本名：<code>PairsMate.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line"></span><br><span class="line"># 脚本帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">Thise script is used to match the pair-end reads from one sample</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">$0 -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br><span class="line"></span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">-1	[str]	Input 1st-end fastq file</span><br><span class="line">-2	[str]	Input 2nd-end fastq file</span><br><span class="line">-o	[str]	Output directory [default: current folder]</span><br><span class="line"></span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($Input1,$Input2,$Outdir);</span><br><span class="line">GetOptions(</span><br><span class="line">&quot;1:s&quot;=&gt;\$Input1,</span><br><span class="line">&quot;2:s&quot;=&gt;\$Input2,</span><br><span class="line">&quot;o:s&quot;=&gt;\$Outdir</span><br><span class="line">);</span><br><span class="line"></span><br><span class="line">$Outdir=`pwd` unless (defined($Outdir));</span><br><span class="line">die `pod2text $0` if (!$Input1) or (!$Input2);</span><br><span class="line"></span><br><span class="line">open FQ1,&quot;&lt;$Input1&quot; or die &quot;Cann&apos;t open $Input1\n&quot;;</span><br><span class="line">open FQ2,&quot;&lt;$Input2&quot; or die &quot;Cann&apos;t open $Input2\n&quot;;</span><br><span class="line"></span><br><span class="line">my $Input1_basename=`basename $Input1`;</span><br><span class="line">chomp $Input1_basename;</span><br><span class="line">my $Input2_basename=`basename $Input2`;</span><br><span class="line">chomp $Input2_basename;</span><br><span class="line">open OUT1,&quot;&gt;$Outdir/$&#123;Input1_basename&#125;.match&quot; or die &quot;Cann&apos;t create $Outdir/$&#123;Input1_basename&#125;.match\n&quot;;</span><br><span class="line">open OUT2,&quot;&gt;$Outdir/$&#123;Input2_basename&#125;.match&quot; or die &quot;Cann&apos;t create $Outdir/$&#123;Input2_basename&#125;.match\n&quot;;</span><br><span class="line"></span><br><span class="line"># 载入两个fastq文件，保存成哈希</span><br><span class="line">my (%fq1_seq,%fq1_qua,%fq2_seq,%fq2_qua);</span><br><span class="line">while (!eof(FQ1))&#123;</span><br><span class="line">        my ($id,)=split(/\s/,&lt;FQ1&gt;);</span><br><span class="line">        $fq1_seq&#123;$id&#125;=&lt;FQ1&gt;;</span><br><span class="line">        &lt;FQ1&gt;;</span><br><span class="line">        $fq1_qua&#123;$id&#125;=&lt;FQ1&gt;;</span><br><span class="line">        chomp ($fq1_seq&#123;$id&#125;,$fq1_qua&#123;$id&#125;);</span><br><span class="line">&#125;</span><br><span class="line">while (!eof(FQ2))&#123;</span><br><span class="line">        my ($id,)=split(/\s/,&lt;FQ2&gt;);</span><br><span class="line">        $fq2_seq&#123;$id&#125;=&lt;FQ2&gt;;</span><br><span class="line">        &lt;FQ2&gt;;</span><br><span class="line">        $fq2_qua&#123;$id&#125;=&lt;FQ2&gt;;</span><br><span class="line">        chomp ($fq2_seq&#123;$id&#125;,$fq2_qua&#123;$id&#125;);</span><br><span class="line">&#125;</span><br><span class="line">close FQ1;</span><br><span class="line">close FQ2;</span><br><span class="line"></span><br><span class="line"># 双端配对</span><br><span class="line">foreach my $key (sort keys %fq1_seq)&#123;</span><br><span class="line">        if(defined($fq2_seq&#123;$key&#125;))&#123;</span><br><span class="line">                print OUT1 &quot;$key\n$fq1_seq&#123;$key&#125;\n+\n$fq1_qua&#123;$key&#125;\n&quot;;</span><br><span class="line">                print OUT2 &quot;$key\n$fq2_seq&#123;$key&#125;\n+\n$fq2_qua&#123;$key&#125;\n&quot;;</span><br><span class="line">        &#125;</span><br><span class="line">&#125;</span><br><span class="line">close OUT1;</span><br><span class="line">close OUT2;</span><br></pre></td></tr></table></figure>
<p>执行方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl PairsMate.pl -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br></pre></td></tr></table></figure>
<p>具体的参数使用说明可以执行 <code>perl PairsMate.pl</code>，查看脚本使用文档</p>
<h3 id="4-根据序列Id提取FASTA序列"><a href="#4-根据序列Id提取FASTA序列" class="headerlink" title="4. 根据序列Id提取FASTA序列"></a>4. 根据序列Id提取FASTA序列</h3><p>提供FASTA文件，和包含序列Id的文件（有多列，用制表符隔开，其中一列为序列Id），提取该序列Id对应的FASTA序列</p>
<h3 id="5-统计FastQC输出"><a href="#5-统计FastQC输出" class="headerlink" title="5. 统计FastQC输出"></a>5. 统计FastQC输出</h3><p>统计FastQC结果，注意脚本需要在FastQC输出结果所在目录下运行。脚本的统计项目及输出如下：</p>
<table>
<thead>
<tr>
<th style="text-align:left">Sample</th>
<th style="text-align:left">Total Reads</th>
<th style="text-align:left">GC Content</th>
<th style="text-align:left">Q20</th>
<th style="text-align:left">Q30</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">NP006_RRS05401_1.clean.fq.extract_fastqc</td>
<td style="text-align:left">39998924</td>
<td style="text-align:left">49</td>
<td style="text-align:left">0.999430935</td>
<td style="text-align:left">0.946251454</td>
</tr>
<tr>
<td style="text-align:left">NP006_RRS05401_2.clean.fq.extract_fastqc</td>
<td style="text-align:left">39998924</td>
<td style="text-align:left">49</td>
<td style="text-align:left">0.986292131</td>
<td style="text-align:left">0.883245809</td>
</tr>
<tr>
<td style="text-align:left">NP007_RRS05402_1.clean.fq.extract_fastqc</td>
<td style="text-align:left">40003323</td>
<td style="text-align:left">49</td>
<td style="text-align:left">0.999429311</td>
<td style="text-align:left">0.946197936</td>
</tr>
</tbody>
</table>
<p>脚本名：<code>fastqc_stat.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">opendir (DIR, &quot;./&quot;) or die &quot;can&apos;t open the directory!&quot;;</span><br><span class="line">@dir = readdir DIR;</span><br><span class="line">foreach $file ( sort  @dir) </span><br><span class="line">&#123;</span><br><span class="line"></span><br><span class="line"># 跳过不需要的文件/文件夹，留下需要的文件夹</span><br><span class="line">next unless -d $file;</span><br><span class="line">next if $file eq &apos;.&apos;;</span><br><span class="line">next if $file eq &apos;..&apos;;</span><br><span class="line"></span><br><span class="line"># 提取total reads</span><br><span class="line">$total_reads=  `grep &apos;^Total&apos; ./$file/fastqc_data.txt`;</span><br><span class="line">$total_reads=(split(/\s+/,$total_reads))[2];</span><br><span class="line"># 提取%GC</span><br><span class="line">$GC= `grep &apos;%GC&apos; ./$file/fastqc_data.txt`;</span><br><span class="line">$GC=(split(/\s+/,$GC))[1];</span><br><span class="line">chomp $GC;</span><br><span class="line"></span><br><span class="line"># 提取Q20，Q30</span><br><span class="line">## 读入Per sequence quality scores部分的信息，保存成哈希</span><br><span class="line">open FH , &quot;&lt;./$file/fastqc_data.txt&quot;;</span><br><span class="line">while (&lt;FH&gt;)</span><br><span class="line">    &#123;</span><br><span class="line">    next unless /#Quality/;</span><br><span class="line">    while (&lt;FH&gt;)</span><br><span class="line">        &#123;</span><br><span class="line">        @F=split;</span><br><span class="line">        $hash&#123;$F[0]&#125;=$F[1];</span><br><span class="line">        last if /&gt;&gt;END_MODULE/;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">## 统计Q20，Q30</span><br><span class="line">$all=0;$Q20=0;$Q30=0;</span><br><span class="line">$all+=$hash&#123;$_&#125; foreach keys %hash;</span><br><span class="line">$Q20+=$hash&#123;$_&#125; foreach 0..20;</span><br><span class="line">$Q30+=$hash&#123;$_&#125; foreach 0..30;</span><br><span class="line">$Q20=1-$Q20/$all;</span><br><span class="line">$Q30=1-$Q30/$all;</span><br><span class="line">print &quot;$file\t$total_reads\t$GC\t$Q20\t$Q30\n&quot;;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>执行方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl fastqc_stat.pl</span><br></pre></td></tr></table></figure>
<h2 id="Python"><a href="#Python" class="headerlink" title="Python"></a>Python</h2><h3 id="1-格式化FASTA文件"><a href="#1-格式化FASTA文件" class="headerlink" title="1. 格式化FASTA文件"></a>1. 格式化FASTA文件</h3><p>在规范的Fasta文件中，你看到的一条序列记录包括两部分：</p>
<ul>
<li>以<code>&gt;</code>起始的序列名称，占一行；</li>
<li>由核苷酸ATCGN（核酸序列）或氨基酸字符组成的字符串，一般每60个字符一行，若一条序列很长，那么它可能会占多行；</li>
</ul>
<p>有的时候因为一些原因（一般都是自己在上游分析时图方便生成的）你得到的fasta文件中的序列部分没有按照60个字符一行的形式进行组织，而是将整条序列放在一行里，虽然一般来说这并不会对你的分析产生太大的影响，但是进行查看的时候会有一些不方便，比如</p>
<p>这个时候如果想将它调整组成规范的格式，要怎么实现呢？</p>
<p>用BioPython将原始FASTA文件读入，然后在写出，就能得到你想要的效果</p>
<p>这个脚本很简单，只有四行代码</p>
<p>脚本名：<code>formatFasta.py</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">from Bio import SeqIO</span><br><span class="line">import sys</span><br><span class="line"></span><br><span class="line">Seq = [ seq for seq in SeqIO.parse(sys.argv[1],&apos;fasta&apos;)]</span><br><span class="line"></span><br><span class="line">SeqIO.write(Seq,sys.argv[2],&apos;fasta&apos;)</span><br></pre></td></tr></table></figure>
<p>用法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ python formatFasta.py &lt;in.fa&gt; &lt;out.fa&gt;</span><br></pre></td></tr></table></figure>
<h2 id="R"><a href="#R" class="headerlink" title="R"></a>R</h2><h3 id="1-对profiles进行wilcox检验（并行化）"><a href="#1-对profiles进行wilcox检验（并行化）" class="headerlink" title="1. 对profiles进行wilcox检验（并行化）"></a>1. 对profiles进行wilcox检验（并行化）</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br></pre></td><td class="code"><pre><span class="line"># 四个参数，按顺序分别为：</span><br><span class="line"># - profiles matrix file</span><br><span class="line"># - sample group file </span><br><span class="line"># - outfile name</span><br><span class="line"># - threads number（默认为服务器总线程-10）</span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(T)</span><br><span class="line"></span><br><span class="line">library(parallel)</span><br><span class="line"></span><br><span class="line"># Initiate cluster</span><br><span class="line">if(is.na(Args[4])) &#123;</span><br><span class="line">	no_cores &lt;- detectCores() - 10</span><br><span class="line">&#125; else &#123;</span><br><span class="line">	no_cores &lt;- as.integer(Args[4])</span><br><span class="line">&#125;</span><br><span class="line">cl &lt;- makeCluster(no_cores)</span><br><span class="line"></span><br><span class="line">data &lt;- read.table(Args[1],head=T,row.names=1,sep=&apos;\t&apos;)</span><br><span class="line">group &lt;- read.table(Args[2],head=T,sep=&apos;\t&apos;)</span><br><span class="line"></span><br><span class="line"># 找出profiles中每列代表的样本所属的组</span><br><span class="line">col_group1 &lt;- colnames(data) %in% sub(&apos;-&apos;,&apos;.&apos;,paste(group$Sample[group$Group==1],&apos;.rpkm&apos;,sep=&apos;&apos;))</span><br><span class="line">col_group2 &lt;- colnames(data) %in% sub(&apos;-&apos;,&apos;.&apos;,paste(group$Sample[group$Group==2],&apos;.rpkm&apos;,sep=&apos;&apos;))</span><br><span class="line"></span><br><span class="line"># 对单个基因进行wilcox检验，并输出统计检验结果，输出格式如下：</span><br><span class="line"># 	geneName,pvalue,group1Median,group2Median,direction</span><br><span class="line">wilcox_fun &lt;- function(data,col_group1,col_group2)&#123;</span><br><span class="line">	g1 &lt;- as.numeric(data[col_group1])</span><br><span class="line">	g2 &lt;- as.numeric(data[col_group2])</span><br><span class="line">	# 执行wilcox检验</span><br><span class="line">	stat &lt;- wilcox.test(g1,g2,paired = FALSE, exact=NULL, correct=TRUE, alternative=&quot;two.sided&quot;)</span><br><span class="line">	if(is.na(stat$p.value))&#123;</span><br><span class="line">		stat$p.value &lt;- 1</span><br><span class="line">	&#125;</span><br><span class="line">	# 对有统计学意义的基因进行判断，是上调&quot;up&quot;还是下调&quot;down&quot;或者是不变&quot;-&quot;（对于组2，即不吃药组）</span><br><span class="line">	if(stat$p.value &lt; 0.1  &amp; median(g1) &lt; median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;down&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else if(stat$p.value &lt; 0.1  &amp; median(g1) &gt; median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;up&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else if(stat$p.value &lt; 0.1  &amp; median(g1) == median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;-&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else&#123;</span><br><span class="line">		G12 &lt;- c()</span><br><span class="line">	&#125;</span><br><span class="line">	G12</span><br><span class="line">&#125;</span><br><span class="line">	</span><br><span class="line">statOut &lt;- parApply(cl,data,1,wilcox_fun,col_group1,col_group2)</span><br><span class="line">stopCluster(cl)</span><br><span class="line"># 删除为NULL的列表元素</span><br><span class="line">for(i in names(statOut))&#123;</span><br><span class="line">	if(is.null(statOut[[i]]))&#123;</span><br><span class="line">		statOut[[i]] &lt;- NULL</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"># 将结果写入文件中</span><br><span class="line">if(!is.null(statOut))&#123;</span><br><span class="line">	# 转换成数据框</span><br><span class="line">	statOut &lt;- as.data.frame(t(as.matrix(as.data.frame(statOut))))</span><br><span class="line">	colnames(statOut) &lt;- c(&apos;pvalue&apos;,&apos;group1Median&apos;,&apos;group2Median&apos;,&apos;direction&apos;)</span><br><span class="line">	# 写入文件</span><br><span class="line">	write.table(statOut,Args[3],sep = &apos;\t&apos;,row.names = T,col.names = T,quote = F)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/17/MeRIP-seq/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/17/MeRIP-seq/" class="post-title-link" itemprop="url">MeRIP/ChIP-seq分析流程</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-17 07:47:05 / Modified: 08:11:32" itemprop="dateCreated datePublished" datetime="2019-02-17T07:47:05+00:00">2019-02-17</time>
            

            
              

              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p><img src="/images/MeRIP-seq.jpg" alt title="MeRIP-seq"></p>
<p align="center"><strong>Schematic diagram of the MeRIP-seq protocol</strong></p>

<p><img src="/images/MeRIP-seq-workflow.jpg" alt title="Workflow"></p>
<p>由于m6A-seq数据分析的原理与过程和ChIP-seq十分相似，所以这里略过前面的质控，简单说明比对和peak calling步骤，具体内容可以参考<a href="https://github.com/Ming-Lian/Memo/blob/master/ChIP-seq-pipeline.md" target="_blank" rel="noopener"><strong>ChIP-seq分析流程</strong></a></p>
<h2 id="m6A背景知识"><a href="#m6A背景知识" class="headerlink" title="m6A背景知识"></a>m6A背景知识</h2><p>目前已知有100多种RNA修饰，涉及到mRNAs、tRNAs、rRNAs、small nuclear RNA (snRNAs) 以及 small nucleolar RNAs (snoRNAs)等。其中甲基化修饰是一种非常广泛的修饰，N6-methyl adenosine (m6A)是真核生物mRNA上最为广泛的甲基化修饰之一，并存在于多种多样的物种中。</p>
<p>腺嘌呤可以被<strong>编码器</strong>METTL3、METTL14和WTAP及其他组分组成的甲基转移酶复合体甲基化，甲基化的腺嘌呤可以被<strong>读码器</strong>（目前发现m6A读码器主要有四个，定位于细胞核内的YTHDC1以及定位在细胞质中的YTHDF1、YTHDF2、YTHDF3、YTHDC2）识别，同时m6A可以被<strong>擦除器</strong>FTO和ALKBH5这两个去甲基化酶催化去甲基化。</p>
<p>在哺乳动物mRNA中，m6A修饰存在于7000多个基因中，<strong>保守基序为RRACH</strong> (R = G, A; H = A, C, U)。m6A修饰富集在mRNA<strong>终止密码子</strong>附近。</p>
<h2 id="比对参考基因组"><a href="#比对参考基因组" class="headerlink" title="比对参考基因组"></a>比对参考基因组</h2><p>在 ChIP-seq 中一般用 BWA 或者 Bowtie 进行完全比对就可以了，但是在 MeRIP-seq 中，由于分析的 RNA ，那么就存在<strong>可变剪切</strong>，对于存在可变剪切的 mapping 用 <strong>Tophat</strong> 或者 Tophat 的升级工具 <strong>HISAT2</strong> 更合适</p>
<h3 id="Tophat"><a href="#Tophat" class="headerlink" title="Tophat"></a>Tophat</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"># build reference index</span><br><span class="line">## &lt;1&gt; build genome index</span><br><span class="line">$ bowtie2-build  hg19.fa hg19</span><br><span class="line">## &lt;2&gt; build transcriptome index</span><br><span class="line">$ tophat -p 8 -G hg19.gtf --transcriptome-index=Ref/hg19/hg19_trans/know hg19</span><br><span class="line"></span><br><span class="line"># mapping</span><br><span class="line">$ tophat -p 8 --transcriptome-index=Ref/hg19/hg19_trans/know -o outdir hg19 reads1_1.fastq reads1_2.fastq</span><br><span class="line"># Only uniquely mapped reads with mapping quality score ≥20 were kept for the subsequent analysis for each sample</span><br><span class="line">$ samtools view -q 20 -O bam -o outdir/accepted_hits.highQual.bam outdir/accepted_hits.bam</span><br></pre></td></tr></table></figure>
<p>Tophat参数</p>
<blockquote>
<ul>
<li>-p Number of threads to use</li>
<li>-G Supply TopHat with a set of gene model annotations and/or known transcripts, as a GTF 2.2 or GFF3 formatted file. </li>
<li>–transcriptome-index TopHat should be first run with the -G/–GTF option together with the –transcriptome-index option pointing to a directory and a name prefix which will indicate where the transcriptome data files will be stored. Then subsequent TopHat runs using the same –transcriptome-index option value will directly use the transcriptome data created in the first run (no -G option needed after the first run). </li>
<li>-o Sets the name of the directory in which TopHat will write all of its output</li>
</ul>
</blockquote>
<p>samtools view 参数</p>
<blockquote>
<ul>
<li>-q only include reads with mapping quality &gt;= INT [0]</li>
</ul>
</blockquote>
<h3 id="HISAT2"><a href="#HISAT2" class="headerlink" title="HISAT2"></a>HISAT2</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"># build reference index</span><br><span class="line">##  using the python scripts included in the HISAT2 package, extract splice-site and exon information from the gene</span><br><span class="line">annotation fle</span><br><span class="line">$ extract_splice_sites.py chrX_data/genes/chrX.gtf &gt;chrX.ss</span><br><span class="line">$ extract_exons.py chrX_data/genes/chrX.gtf &gt;chrX.exon</span><br><span class="line">##  build a HISAT2 index</span><br><span class="line">$ hisat2-build --ss chrX.ss --exon chrX.exon chrX_data/genome/chrX.fa chrX_tran</span><br><span class="line"></span><br><span class="line"># mapping</span><br><span class="line">$ hisat2 -p 10 --dta -x chrX_tran -1 reads1_1.fastq -2 reads1_2.fastq | samtools sort -@ 8 -O bam -o reads1.sort.bam 1&gt;map.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<p><code>Usage: hisat2 [options]* -x &lt;ht2-idx&gt; {-1 &lt;m1&gt; -2 &lt;m2&gt; | -U &lt;r&gt;} [-S &lt;sam&gt;]</code></p>
<blockquote>
<ul>
<li>-p Number of threads to use</li>
<li>–dta reports alignments tailored for transcript assemblers</li>
<li>-x Hisat2 index</li>
<li>-1 The 1st input fastq file of paired-end reads</li>
<li>-2 The 2nd input fastq file of paired-end reads</li>
<li>-S File for SAM output (default: stdout)</li>
</ul>
</blockquote>
<h2 id="Peak-calling"><a href="#Peak-calling" class="headerlink" title="Peak calling"></a>Peak calling</h2><hr>
<h3 id="MACS2"><a href="#MACS2" class="headerlink" title="MACS2"></a>MACS2</h3><p>参考ChIP-seq分析流程中的<a href="https://github.com/Ming-Lian/Memo/blob/master/ChIP-seq-pipeline.md#peak-calling" target="_blank" rel="noopener">peak calling</a>过程</p>
<h3 id="PeakRanger"><a href="#PeakRanger" class="headerlink" title="PeakRanger"></a>PeakRanger</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">peakranger ccat --format bam SRR1042594.sorted.bam SRR1042593.sorted.bam  \</span><br><span class="line">Xu_MUT_rep1_ccat_report --report --gene_annot_file hg19refGene.txt -q 0.05 -t 4</span><br></pre></td></tr></table></figure>
<h2 id="Peaks注释"><a href="#Peaks注释" class="headerlink" title="Peaks注释"></a>Peaks注释</h2><hr>
<h3 id="CEAS"><a href="#CEAS" class="headerlink" title="CEAS"></a>CEAS</h3><p>哈佛刘小乐实验室出品的软件，可以跟MACS软件call到的peaks文件无缝连接，实现peaks的注释以及可视化分析</p>
<p>CEAS需要三种输入文件：</p>
<blockquote>
<ul>
<li>Gene annotation table file (sqlite3)<blockquote>
<p>可以到CEAS官网上下载：<a href="http://liulab.dfci.harvard.edu/CEAS/src/hg18.refGene.gz" target="_blank" rel="noopener">http://liulab.dfci.harvard.edu/CEAS/src/hg18.refGene.gz</a> ，也可以自己构建：到UCSC上下载，然后用<code>build_genomeBG</code>脚本转换成split3格式</p>
</blockquote>
</li>
<li>BED file with ChIP regions (TXT)<blockquote>
<p>需要包含chromosomes, start, and end locations，这样文件可以由 peak-caller （如MACS2）得到</p>
</blockquote>
</li>
<li>WIG file with ChiP enrichment signal (TXT)<blockquote>
<p> 如何得到wig文件可以参考<a href="https://github.com/Ming-Lian/NGS-analysis/blob/master/samtools%E6%93%8D%E4%BD%9C%E6%8C%87%E5%8D%97.md#output-wig" target="_blank" rel="noopener">samtools操作指南：以WIG文件输出测序深度</a></p>
</blockquote>
</li>
</ul>
</blockquote>
<p>CEAS的使用方法很简单：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ceas --name=H3K36me3_ceas --pf-res=20 --gn-group-names=&apos;Top 10%,Bottom 10%&apos;  \</span><br><span class="line">-g hg19.refGene -b  ../paper_results/GSM1278641_Xu_MUT_rep1_BAF155_MUT.peaks.bed \</span><br><span class="line">-w ../rawData/SRR1042593.wig</span><br></pre></td></tr></table></figure></p>
<blockquote>
<ul>
<li>–name Experiment name. This will be used to name the output files.</li>
<li>–pf-res Wig profiling resolution, DEFAULT: 50bp</li>
<li>–gn-group-names The names of the gene groups in –gn-groups. The gene group names are separated by commas. (eg, –gn-group-names=’top 10%,bottom 10%’).</li>
<li>-g Gene annotation table</li>
<li>-b BED file of ChIP regions</li>
<li>-w WIG file for either wig profiling or genome background annotation.</li>
</ul>
</blockquote>
<h2 id="Motif识别"><a href="#Motif识别" class="headerlink" title="Motif识别"></a>Motif识别</h2><h3 id="HOMER"><a href="#HOMER" class="headerlink" title="HOMER"></a>HOMER</h3><p>安装旧版本的HOMER比较复杂，因为旧版依赖于调用其他几个工具：</p>
<blockquote>
<ul>
<li>blat</li>
<li>Ghostscript</li>
<li>weblogo<br><strong>Does NOT work with version 3.0!!!!</strong></li>
</ul>
</blockquote>
<p>新版HOMER安装很简单，主要是通过<code>configureHomer.pl</code>脚本来安装和管理HOMER<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">cd ~/biosoft</span><br><span class="line">mkdir homer &amp;&amp;  cd homer</span><br><span class="line">wget http://homer.salk.edu/homer/configureHomer.pl</span><br><span class="line"></span><br><span class="line"># Installing the basic HOMER software</span><br><span class="line">perl configureHomer.pl -install</span><br><span class="line"></span><br><span class="line"># Download the hg19 version of the human genome</span><br><span class="line">perl configureHomer.pl -install hg19</span><br></pre></td></tr></table></figure></p>
<p>安装好后可以进行 Motif Identification<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># 提取对应的列给HOMER作为输入文件</span><br><span class="line"># change </span><br><span class="line">#		chr1	1454086	1454256	MACS_peak_1	59.88 </span><br><span class="line">#to   </span><br><span class="line">#		MACS_peak_1	chr1	1454086	1454256	+</span><br><span class="line">$ awk &apos;&#123;print $4&quot;\t&quot;$1&quot;\t&quot;$2&quot;\t&quot;$3&quot;\t+&quot;&#125;&apos; macs_peaks.bed &gt;homer_peaks.bed</span><br><span class="line"></span><br><span class="line"># MeRIP-seq 中 motif 的长度为6个 nt</span><br><span class="line">$ findMotifsGenome.pl homer_peaks.bed hg19 motifDir -size 200 -len 8,10,12</span><br><span class="line"></span><br><span class="line"># 自己指定background sequences，用bedtools shuffle构造随机的suffling peaks</span><br><span class="line">$ bedtools shuffle -i peaks.bed -g &lt;GENOME&gt; &gt;peaks_shuffle.bed</span><br><span class="line"># 用参数&quot;-bg&quot;指定background sequences</span><br><span class="line">$ findMotifsGenome.pl homer_peaks.bed hg19 motifDir -bg peaks_shuffle.bed -size 200 -len 8,10,12</span><br></pre></td></tr></table></figure></p>
<p><code>Usage: findMotifsGenome.pl &lt;pos file&gt; &lt;genome&gt; &lt;output directory&gt; [additional options]</code></p>
<p>注意：</p>
<blockquote>
<ul>
<li><code>&lt;genome&gt;</code> 参数只需要写出genome的序号，不需要写出具体路径</li>
<li><code>bedtools shuffle</code>中的genome文件的格式要求：<chromname><tab><chromsize><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">&gt; For example, Human (hg19):</span><br><span class="line">&gt;        chr1    249250621</span><br><span class="line">&gt;        chr2    243199373</span><br><span class="line">&gt;        ...</span><br><span class="line">&gt;        chr18_gl000207_random   4262</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</chromsize></tab></chromname></li>
</ul>
</blockquote>
<blockquote>
<p>可以使用 UCSC Genome Browser’s MySQL database 来获取 chromosome sizes 信息并构建genome文件<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e &quot;select chrom, size from hg19.chromInfo&quot; &gt;hg19.genome</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure></p>
</blockquote>
<p>最后得到的文件夹里面有一个详细的网页版报告</p>
<p><img src="http://homer.ucsd.edu/homer/ngs/peakMotifs.output.png" alt></p>
<h3 id="MEME"><a href="#MEME" class="headerlink" title="MEME"></a>MEME</h3><h4 id="算法原理"><a href="#算法原理" class="headerlink" title="算法原理"></a>算法原理</h4><p><img src="/images/MeRIP-seq-meme-principle-1.png" alt></p>
<p><img src="/images/MeRIP-seq-meme-principle-2.png" alt></p>
<p><img src="/images/MeRIP-seq-meme-principle-3.png" alt></p>
<h4 id="操作"><a href="#操作" class="headerlink" title="操作"></a>操作</h4><p>下载安装MEME</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">cd ~/biosoft</span><br><span class="line">mkdir MEMEsuite &amp;&amp;  cd MEMEsuite</span><br><span class="line">## http://meme-suite.org/doc/download.html</span><br><span class="line">wget  http://meme-suite.org/meme-software/4.11.2/meme_4.11.2_1.tar.gz</span><br><span class="line">tar zxvf meme_4.11.2_1.tar.gz </span><br><span class="line">cd meme_4.11.2/</span><br><span class="line">./configure --prefix=$HOME/my-bin/meme --with-url=&quot;http://meme-suite.org&quot;</span><br><span class="line">make </span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># 先提取peaks区域所对应的序列</span><br><span class="line">bedtools getfasta -fi input.fasta -bed input.bed -fo output.fasta</span><br><span class="line"># Motif identification</span><br><span class="line">meme output.fasta -dna -mod oops -pal</span><br></pre></td></tr></table></figure>
<p><code>Usage: meme &lt;dataset&gt; [optional arguments]</code></p>
<blockquote>
<ul>
<li>&lt; dataset &gt; File containing sequences in FASTA format</li>
<li>-dna Sequences use DNA alphabet</li>
<li>-mod Distribution of motifs,3 options: oops | zoops | anr</li>
<li>-pal Force palindromes (requires -dna)</li>
</ul>
</blockquote>
<h2 id="Differential-binding-analysis"><a href="#Differential-binding-analysis" class="headerlink" title="Differential binding analysis"></a>Differential binding analysis</h2><h3 id="Merge-peaks"><a href="#Merge-peaks" class="headerlink" title="Merge peaks"></a>Merge peaks</h3><p>当ChIP-seq数据中有多分组，多样本以及多个重复时，需要进行样本间peaks的merge</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">bedtools intersect -a Mcf7H3k27acUcdAlnRep1_peaks.filtered.bed -b Mcf7H3k27acUcdAlnRep2_peaks.filtered.bed -wa | cut -f1-3 | sort | uniq &gt; Mcf7Rep1_peaks.bed</span><br><span class="line">bedtools intersect -a Mcf7H3k27acUcdAlnRep1_peaks.filtered.bed -b Mcf7H3k27acUcdAlnRep2_peaks.filtered.bed -wb | cut -f1-3 | sort | uniq &gt; Mcf7Rep2_peaks.bed</span><br><span class="line">bedtools intersect -a Panc1H3k27acUcdAlnRep1_peaks.filtered.bed -b Panc1H3k27acUcdAlnRep2_peaks.filtered.bed -wa | cut -f1-3 | sort | uniq &gt; Panc1Rep1_peaks.bed</span><br><span class="line">bedtools intersect -a Panc1H3k27acUcdAlnRep1_peaks.filtered.bed -b Panc1H3k27acUcdAlnRep2_peaks.filtered.bed -wb | cut -f1-3 | sort | uniq &gt; Panc1Rep2_peaks.bed</span><br><span class="line"></span><br><span class="line">rm *filtered*</span><br><span class="line"></span><br><span class="line">cat *bed | sort -k1,1 -k2,2n | bedtools merge &gt; merge.bed</span><br></pre></td></tr></table></figure>
<h3 id="Preparing-ChIP-seq-count-table"><a href="#Preparing-ChIP-seq-count-table" class="headerlink" title="Preparing ChIP-seq count table"></a>Preparing ChIP-seq count table</h3><p>用<strong>bedtools</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># Make a bed file adding peak id as the fourth colum</span><br><span class="line">$ awk &apos;&#123;$3=$3&quot;\t&quot;&quot;peak_&quot;NR&#125;1&apos; OFS=&quot;\t&quot; merge.bed &gt; bed_for_multicov.bed</span><br><span class="line"># 输入的bam文件要提前做好index，可同时提供多个bam文件</span><br><span class="line">$ samtools sort -@ 8 -O BAM -o input1.sort.bam input1.bam </span><br><span class="line">$ samtools index -@ 8 input1.sort.bam input1.sort.bam.bai # 注意：生成的索引文件的文件名必须为在原bam文件名后追加&quot;.bai&quot;，否则bedtools multicov无法识别</span><br><span class="line">$ bedtools multicov -bams input1.bam input2.bam ... -bed bed_for_multicov.bed &gt; counts_multicov.txt</span><br></pre></td></tr></table></figure>
<blockquote>
<ul>
<li>NR 表示awk开始执行程序后所读取的数据行数</li>
<li>OFS Out of Field Separator，输出字段分隔符</li>
</ul>
</blockquote>
<p>用<strong>featureCounts</strong> (subread工具包中的组件）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># Make a saf(simplified annotation format) file for featureCount in the subread package,</span><br><span class="line"></span><br><span class="line">shown below:</span><br><span class="line">GeneID	Chr	Start	End	Strand</span><br><span class="line">497097	chr1	3204563	3207049	-</span><br><span class="line">497097	chr1	3411783	3411982	-</span><br><span class="line">497097	chr1	3660633	3661579	-</span><br><span class="line">...</span><br><span class="line"></span><br><span class="line">$ awk -F &quot;\t&quot; &apos;&#123;$1=&quot;peak_&quot;NR FS$1;$4=$4FS&quot;.&quot;&#125;1&apos; merge.bed &gt; subread.saf</span><br><span class="line">$ featureCounts -T 4 -a subread.saf -F SAF -o counts_subread.txt ../../data/*bam</span><br></pre></td></tr></table></figure>
<p><code>Usage: featureCounts [options] -a &lt;annotation_file&gt; -o &lt;output_file&gt; input_file1 [input_file2] ...</code></p>
<blockquote>
<ul>
<li>-a Name of an annotation file</li>
<li>-F Specify format of the provided annotation file. Acceptable formats include ‘GTF’ (or compatible GFF format) and ‘SAF’. ‘GTF’ by default</li>
<li>-o Name of the output file including read counts</li>
<li>-T Number of the threads</li>
</ul>
</blockquote>
<h3 id="Differential-binding-by-DESeq2"><a href="#Differential-binding-by-DESeq2" class="headerlink" title="Differential binding by DESeq2"></a>Differential binding by DESeq2</h3><ul>
<li>对 contrast 构建一个 counts 矩阵，第一组的每个样本占据一列，紧接着的是第二组的样本，也是每个样本一列。</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 做好表达矩阵</span><br><span class="line">count_table&lt;-read.delim(&quot;count_multicov.txt&quot;,header=F)</span><br><span class="line">count_matrix&lt;-as.matrix(count_table[,c(-1,-2,-3,-4)])</span><br><span class="line">rownames(count_matrix)&lt;-count_table$V4</span><br><span class="line"># 做好分组因子即可</span><br><span class="line">group_list&lt;-factor(c(&quot;control&quot;,&quot;control&quot;,&quot;control&quot;,&quot;treat&quot;,&quot;treat&quot;,&quot;treat&quot;))</span><br><span class="line">colData &lt;- data.frame(row.names=colnames(count_matrix), group_list=group_list))</span><br><span class="line"># 构建 DESeqDataSet 对象</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData = count_matrix,</span><br><span class="line">	colData = colData,</span><br><span class="line">	design = ~ group_list)</span><br></pre></td></tr></table></figure>
<ul>
<li>接着计算每个样本的 library size 用于后续的标准化。library size 等于落在该样本所有peaks上的reads的总数，即 counts 矩阵中每列的加和。如果bFullLibrarySize设为TRUE，则会使用library的总reads数（根据原BAM/BED文件统计获得）。然后使用<code>estimateDispersions</code>估计统计分布，需要将参数<code>fitType</code>设为’local’</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">dds&lt;-estimateSizeFactors(dds)</span><br><span class="line">dds&lt;-estimateDispersions(dds,fitType=&quot;local&quot;)</span><br></pre></td></tr></table></figure>
<ul>
<li>对负二项分布进行显著性检验（Negative Binomial GLM fitting and Wald statistics）</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">dds &lt;- nbinomWaldTest(dds)</span><br><span class="line">res &lt;- results(dds)</span><br><span class="line">res</span><br><span class="line"></span><br><span class="line">## log2 fold change (MLE): condition treated vs untreated </span><br><span class="line">## Wald test p-value: condition treated vs untreated </span><br><span class="line">## DataFrame with 9921 rows and 6 columns</span><br><span class="line">##                baseMean log2FoldChange     lfcSE        stat     pvalue</span><br><span class="line">##               &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;</span><br><span class="line">## FBgn0000008   95.144292    0.002276428 0.2237292  0.01017493 0.99188172</span><br><span class="line">## FBgn0000014    1.056523   -0.495113878 2.1431096 -0.23102593 0.81729466</span><br><span class="line">## FBgn0000017 4352.553569   -0.239918945 0.1263378 -1.89902705 0.05756092</span><br><span class="line">## FBgn0000018  418.610484   -0.104673913 0.1484903 -0.70492106 0.48085936</span><br><span class="line">## FBgn0000024    6.406200    0.210848562 0.6895923  0.30575830 0.75978868</span><br><span class="line">## ...                 ...            ...       ...         ...        ...</span><br><span class="line">## FBgn0261570 3208.388610     0.29553289 0.1273514  2.32061001  0.0203079</span><br><span class="line">## FBgn0261572    6.197188    -0.95882276 0.7753130 -1.23669125  0.2162017</span><br><span class="line">## FBgn0261573 2240.979511     0.01271946 0.1133028  0.11226079  0.9106166</span><br><span class="line">## FBgn0261574 4857.680373     0.01539243 0.1925619  0.07993497  0.9362890</span><br><span class="line">## FBgn0261575   10.682520     0.16356865 0.9308661  0.17571663  0.8605166</span><br><span class="line">##                  padj</span><br><span class="line">##             &lt;numeric&gt;</span><br><span class="line">## FBgn0000008 0.9972093</span><br><span class="line">## FBgn0000014        NA</span><br><span class="line">## FBgn0000017 0.2880108</span><br><span class="line">## FBgn0000018 0.8268644</span><br><span class="line">## FBgn0000024 0.9435005</span><br><span class="line">## ...               ...</span><br><span class="line">## FBgn0261570 0.1442486</span><br><span class="line">## FBgn0261572 0.6078453</span><br><span class="line">## FBgn0261573 0.9826550</span><br><span class="line">## FBgn0261574 0.9881787</span><br><span class="line">## FBgn0261575 0.9679223</span><br></pre></td></tr></table></figure>
<p>若存在多个分组需要进行两两比较，则需要提取指定的两个分组之间的比较结果<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">## 提取你想要的差异分析结果，我们这里是treated组对untreated组进行比较</span><br><span class="line">res &lt;-  results(dds2, contrast=c(&quot;group_list&quot;,&quot;treated&quot;,&quot;untreated&quot;))</span><br><span class="line">resOrdered &lt;- res[order(res$padj),]</span><br><span class="line">resOrdered=as.data.frame(resOrdered)</span><br></pre></td></tr></table></figure></p>
<p>注意两个概念：</p>
<blockquote>
<ul>
<li><p>Full library size: bam文件中reads的总数</p>
</li>
<li><p>Effective library size: 落在peaks区域的reads的总数</p>
</li>
</ul>
<p>DESeq2中默认使用 Full library size bam (bFullLibrarySize =TRUE)，在ChIP-seq中使用 Effective library size 更合适，所有应该设置为bFullLibrarySize =FALSE</p>
</blockquote>
<h3 id="4-Differential-binding-peaks-annotation"><a href="#4-Differential-binding-peaks-annotation" class="headerlink" title="4. Differential binding peaks annotation"></a>4. Differential binding peaks annotation</h3><p>在<code>Differential binding analysis: 2. Preparing ChIP-seq count table</code>得到的<code>bed_for_multicov.bed</code>中找DiffBind peaks的bed格式信息</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 用R进行取交集操作</span><br><span class="line">merge_bed&lt;-read.delim(&quot;m6A_seq/CallPeak/bed_for_multicov.bed&quot;,header=F)</span><br><span class="line">peaks_diffbind&lt;-read.delim(&quot;m6A_seq/CallPeak/res_diffBind.txt&quot;,header=T)</span><br><span class="line">peaks_diffbind_bed&lt;-merge_bed[merge_bed$V4 %in% rownames(peaks_diffbind),]</span><br><span class="line"># 保存文件</span><br><span class="line">write.table(peaks_diffbind_bed,&quot;m6A_seq/CallPeak/peaks_diffBind.bed&quot;,sep=&quot;\t&quot;,col.names=F,row.names=F,quote=F)</span><br></pre></td></tr></table></figure>
<p>接着，用基因组注释文件GFF/GTF注释peaks</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ bedtools intersect -wa -wb -a m6A_seq/CallPeak/peaks_diffBind.bed -b Ref/mm10/mm10_trans/Mus_musculus.GRCm38.91.gtf &gt;m6A_seq/CallPeak/peaks_diffbind.anno.bed</span><br><span class="line"></span><br><span class="line">$ head m6A_seq/CallPeak/peaks_diffbind.anno.bed</span><br><span class="line"></span><br><span class="line">## 1       7120312 7120526 peak_9  1       ensembl_havana  exon    7120194 7120615 .       +       .       gene_id &quot;ENSMUSG00000051285&quot;; gene_version &quot;17&quot;; transcript_id &quot;ENSMUST00000061280&quot;; transcript_version &quot;16&quot;; exon_number &quot;2&quot;; gene_name &quot;Pcmtd1&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTMUSG00000043373&quot;; havana_gene_version &quot;5&quot;; transcript_name &quot;Pcmtd1-201&quot;; transcript_source &quot;ensembl_havana&quot;; transcript_biotype &quot;protein_coding&quot;; tag &quot;CCDS&quot;; ccds_id &quot;CCDS35508&quot;; havana_transcript &quot;OTTMUST00000113805&quot;; havana_transcript_version &quot;2&quot;; exon_id &quot;ENSMUSE00000553965&quot;; exon_version &quot;2&quot;; tag &quot;basic&quot;; transcript_support_level &quot;1&quot;;</span><br><span class="line">## 1       7120312 7120526 peak_9  1       ensembl_havana  CDS     7120309 7120615 .       +       0       gene_id &quot;ENSMUSG00000051285&quot;; gene_version &quot;17&quot;; transcript_id &quot;ENSMUST00000061280&quot;; transcript_version &quot;16&quot;; exon_number &quot;2&quot;; gene_name &quot;Pcmtd1&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTMUSG00000043373&quot;; havana_gene_version &quot;5&quot;; transcript_name &quot;Pcmtd1-201&quot;; transcript_source &quot;ensembl_havana&quot;; transcript_biotype &quot;protein_coding&quot;; tag &quot;CCDS&quot;; ccds_id &quot;CCDS35508&quot;; havana_transcript &quot;OTTMUST00000113805&quot;; havana_transcript_version &quot;2&quot;; protein_id &quot;ENSMUSP00000059261&quot;; protein_version &quot;9&quot;; tag &quot;basic&quot;; transcript_support_level &quot;1&quot;;</span><br><span class="line">## 1       7120312 7120526 peak_9  1       ensembl_havana  gene    7088920 7173628 .       +       .       gene_id &quot;ENSMUSG00000051285&quot;; gene_version &quot;17&quot;; gene_name &quot;Pcmtd1&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTMUSG00000043373&quot;; havana_gene_version &quot;5&quot;;</span><br><span class="line">## 1       7120312 7120526 peak_9  1       ensembl_havana  transcript      7088920 7173628 .       +       .       gene_id &quot;ENSMUSG00000051285&quot;; gene_version &quot;17&quot;; transcript_id &quot;ENSMUST00000061280&quot;; transcript_version &quot;16&quot;; gene_name &quot;Pcmtd1&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTMUSG00000043373&quot;; havana_gene_version &quot;5&quot;; transcript_name &quot;Pcmtd1-201&quot;; transcript_source &quot;ensembl_havana&quot;; transcript_biotype &quot;protein_coding&quot;; tag &quot;CCDS&quot;; ccds_id &quot;CCDS35508&quot;; havana_transcript &quot;OTTMUST00000113805&quot;; havana_transcript_version &quot;2&quot;; tag &quot;basic&quot;; transcript_support_level &quot;1&quot;;</span><br><span class="line">## 1       7120312 7120526 peak_9  1       havana  transcript      7088930 7169598 .       +       .       gene_id &quot;ENSMUSG00000051285&quot;; gene_version &quot;17&quot;; transcript_id &quot;ENSMUST00000182114&quot;; transcript_version &quot;7&quot;; gene_name &quot;Pcmtd1&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTMUSG00000043373&quot;; havana_gene_version &quot;5&quot;; transcript_name &quot;Pcmtd1-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;protein_coding&quot;; havana_transcript &quot;OTTMUST00000113813&quot;; havana_transcript_version &quot;2&quot;; tag &quot;cds_end_NF&quot;; tag &quot;mRNA_end_NF&quot;; transcript_support_level &quot;1&quot;;</span><br></pre></td></tr></table></figure>
<h2 id="peaks注释基因的富集分析"><a href="#peaks注释基因的富集分析" class="headerlink" title="peaks注释基因的富集分析"></a>peaks注释基因的富集分析</h2><h3 id="clusterProfiler-GO-enrichment-analysis"><a href="#clusterProfiler-GO-enrichment-analysis" class="headerlink" title="clusterProfiler: GO enrichment analysis"></a>clusterProfiler: GO enrichment analysis</h3><p>多数人进行GO富集分析时喜欢使用DAVID，但是由于DAVID的最新版本是在2016年更新的，数据并不是最新的，所以不推荐使用DAVID。</p>
<p>推荐使用bioconductor包<strong>clusterProfiler</strong></p>
<p><strong>准备geneList</strong></p>
<p>从<code>DiffBind peaks annotation</code>中获得的peaks_diffbind.anno.bed文件中提取geneID</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ cut -f 13 m6A_seq/CallPeak/peaks_diffbind.anno.bed | \</span><br><span class="line">	awk &apos;BEGIN&#123;FS=&quot;;&quot;&#125; &#123;print $1&#125;&apos; | \</span><br><span class="line">	perl -ane &apos;chomp;$F[1]=~ s/\&quot;//g;print &quot;$F[1]\n&quot;&apos; | sort | uniq &gt;m6A_seq/CallPeak/peaks_diffbind_geneList.txt</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 载入geneList，最终提供给enrichGO的gene list应该是一个geneID的向量</span><br><span class="line">gene_list &lt;- read.table(&quot;peaks_diffbind_geneList.txt&quot;,header=F)</span><br><span class="line"></span><br><span class="line"># ID转换，转换需要的信息来自于对应物种的数据包（orgDB），所以使用前需要提前安装好数据包</span><br><span class="line">eg = bitr(gene_list$V1, fromType=&quot;SYMBOL&quot;, toType=&quot;ENTREZID&quot;, OrgDb=&quot;org.Hs.eg.db&quot;)</span><br><span class="line"></span><br><span class="line">## ID转换支持的ID类型可以通过以下方式查看</span><br><span class="line">keytypes(org.Hs.eg.db)</span><br><span class="line"></span><br><span class="line">## 当ID转换涉及KEGG ID时需要使用特殊的函数bitr_kegg，只能在kegg，ncbi-geneid，ncbi-proteinid和uniprot之间进行转换</span><br><span class="line">eg2np &lt;- bitr_kegg(hg, fromType=&apos;kegg&apos;, toType=&apos;ncbi-proteinid&apos;, organism=&apos;hsa&apos;)</span><br></pre></td></tr></table></figure>
<p><strong>GO over-representation test</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># 需要使用到相应物种的基因组注释数据包（org.Xx.eg.db），请提前安装好，下面以老鼠为例</span><br><span class="line">ego &lt;- enrichGO(gene         = gene_list$V1,</span><br><span class="line">                OrgDb         = org.Mm.eg.db,</span><br><span class="line">        		keytype       = &apos;ENSEMBL&apos;,</span><br><span class="line">                ont           = &quot;CC&quot;,</span><br><span class="line">                pAdjustMethod = &quot;BH&quot;,</span><br><span class="line">                pvalueCutoff  = 0.01,</span><br><span class="line">                qvalueCutoff  = 0.05,</span><br><span class="line">                readable      = TRUE)</span><br></pre></td></tr></table></figure>
<p>参数说明：</p>
<blockquote>
<ul>
<li>gene: a vector of entrez gene id</li>
<li>OrgDb: OrgDb</li>
<li>keytype: keytype of input gene</li>
<li>ont: One of “MF”, “BP”, and “CC” subontologies</li>
<li>pvalueCutoff: Cutoff value of pvalue</li>
<li>pAdjustMethod: one of “holm”, “hochberg”, “hommel”, “bonferroni”, “BH”, “BY”, “fdr”, “none”</li>
<li>qvalueCutoff: qvalue cutoff</li>
<li>readable: whether mapping gene ID to gene Name</li>
</ul>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">head(ego)</span><br><span class="line"></span><br><span class="line">##                    ID                              Description GeneRatio</span><br><span class="line">## GO:0005819 GO:0005819                                  spindle    25/198</span><br><span class="line">## GO:0000779 GO:0000779 condensed chromosome, centromeric region    15/198</span><br><span class="line">## GO:0000775 GO:0000775           chromosome, centromeric region    18/198</span><br><span class="line">## GO:0000776 GO:0000776                              kinetochore    15/198</span><br><span class="line">## GO:0000793 GO:0000793                     condensed chromosome    18/198</span><br><span class="line">## GO:0005876 GO:0005876                      spindle microtubule    10/198</span><br><span class="line">##              BgRatio       pvalue     p.adjust       qvalue</span><br><span class="line">## GO:0005819 238/11745 2.090374e-13 5.518588e-11 4.950886e-11</span><br><span class="line">## GO:0000779  90/11745 2.241220e-11 2.958411e-09 2.654077e-09</span><br><span class="line">## GO:0000775 152/11745 7.936845e-11 6.984424e-09 6.265930e-09</span><br><span class="line">## GO:0000776 103/11745 1.649359e-10 8.919109e-09 8.001593e-09</span><br><span class="line">## GO:0000793 159/11745 1.689225e-10 8.919109e-09 8.001593e-09</span><br><span class="line">## GO:0005876  45/11745 2.821374e-09 1.241405e-07 1.113700e-07</span><br><span class="line">##                                                                                                                                                         geneID</span><br><span class="line">## GO:0005819 CDCA8/CDC20/KIF23/CENPE/ASPM/DLGAP5/SKA1/NUSAP1/TPX2/TACC3/NEK2/CDK1/MAD2L1/KIF18A/BIRC5/KIF11/TTK/AURKB/PRC1/KIFC1/KIF18B/KIF20A/AURKA/CCNB1/KIF4A</span><br><span class="line">## GO:0000779                                                            CENPE/NDC80/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/CDT1/BIRC5/NCAPG/AURKB/AURKA/CCNB1</span><br><span class="line">## GO:0000775                                           CDCA8/CENPE/NDC80/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/NCAPG/AURKB/AURKA/CCNB1</span><br><span class="line">## GO:0000776                                                             CENPE/NDC80/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/AURKB/CCNB1</span><br><span class="line">## GO:0000793                                          CENPE/NDC80/TOP2A/NCAPH/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/CDT1/BIRC5/NCAPG/AURKB/CHEK1/AURKA/CCNB1</span><br><span class="line">## GO:0005876                                                                                         SKA1/NUSAP1/CDK1/KIF18A/KIF11/AURKB/PRC1/KIF18B/AURKA/KIF4A</span><br><span class="line">##            Count</span><br><span class="line">## GO:0005819    25</span><br><span class="line">## GO:0000779    15</span><br><span class="line">## GO:0000775    18</span><br><span class="line">## GO:0000776    15</span><br><span class="line">## GO:0000793    18</span><br><span class="line">## GO:0005876    10</span><br></pre></td></tr></table></figure>
<p>富集分析结果可视化</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># 绘制气泡图</span><br><span class="line"></span><br><span class="line">## 可以使用clusterProfiler中提供的绘图函数dotplot</span><br><span class="line">dotplot(ego)</span><br><span class="line"></span><br><span class="line">## 也可以使用ggplot2进行自定义绘图</span><br><span class="line">library(ggplot2)</span><br><span class="line">p&lt;-ggplot(ego)+geom_point(aes(x=)</span><br></pre></td></tr></table></figure>
<p><img src="/images/MeRIP-seq-GOenrich-dotplot.png" alt></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 绘制GOterm拓扑关系网，依赖topGO和Rgraphviz</span><br><span class="line">library(topGO)</span><br><span class="line">library(Rgraphviz)</span><br><span class="line"># 画出来的拓扑图中节点的文字太小，看不清，这是一个问题</span><br><span class="line">plotGOgraph(ego)</span><br></pre></td></tr></table></figure>
<p><img src="/images/MeRIP-seq-GOtop-graph.png" alt></p>
<hr>
<p>参考资料：</p>
<p>(1) Zhang C, Chen Y, Sun B, et al. m(6)A modulates haematopoietic stem and progenitor cell specification[J]. Nature, 2017, 549(7671):273.</p>
<p>(2) <a href="http://mp.weixin.qq.com/s/TKwWKEgfZWskP9NWjYO_Ug" target="_blank" rel="noopener">BIG科研 | 细胞质内的m6A结合蛋白YTHDF3促进mRNA的翻译</a></p>
<p>(3) Pertea M, Kim D, Pertea G, et al. Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie, and Ballgown[J]. Nature Protocols, 2016, 11(9):1650. </p>
<p>(4) <a href="https://github.com/Ming-Lian/Memo/blob/master/ChIP-seq-pipeline.md" target="_blank" rel="noopener">ChIP-seq-pipeline</a></p>
<p>(5) <a href="https://github.com/jmzeng1314/NGS-pipeline/tree/master/CHIPseq" target="_blank" rel="noopener">ChIPseq pipeline on jmzeng1314’s github</a></p>
<p>(6) <a href="https://github.com/crazyhottommy/ChIP-seq-analysis" target="_blank" rel="noopener">ChIPseq pipeline on crazyhottommy’s github</a></p>
<p>(7) <a href="https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part3.1_Differential_binding_DiffBind_lib_size.md" target="_blank" rel="noopener">library size and normalization for ChIP-seq</a></p>
<p>(8) <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" target="_blank" rel="noopener">Bioconductor tutorial: Analyzing RNA-seq data with DESeq2</a></p>
<p>(9) <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#introduction" target="_blank" rel="noopener">Bioconductor tutorial: clusterProfiler</a></p>
<p>(10) 国科大-韩春生《生物信息学应用 - 模序搜索》ppt</p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/17/InAction-PHP-MySQL-lincRNA-Database/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/17/InAction-PHP-MySQL-lincRNA-Database/" class="post-title-link" itemprop="url">PHP+MySQL实战：小鼠与人类lincRNA数据库</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-17 06:51:33 / Modified: 07:43:42" itemprop="dateCreated datePublished" datetime="2019-02-17T06:51:33+00:00">2019-02-17</time>
            

            
              

              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/前后端技术/" itemprop="url" rel="index"><span itemprop="name">前后端技术</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="Part1：lincRNA注释信息数据库"><a href="#Part1：lincRNA注释信息数据库" class="headerlink" title="Part1：lincRNA注释信息数据库"></a>Part1：lincRNA注释信息数据库</h1><h2 id="实现目标"><a href="#实现目标" class="headerlink" title="实现目标"></a>实现目标</h2><ul>
<li>将lincRNA的gtf格式的注释信息写到MySQL数据库中</li>
<li>用户需要登录才具有数据库访问权限</li>
<li>提供数据库的检索功能</li>
</ul>
<h2 id="思路"><a href="#思路" class="headerlink" title="思路"></a>思路</h2><ol>
<li><p>从 <strong>UCSC genome browser</strong> 或 <strong>Ensemble</strong> 等数据库中下载人类（homo sapiens）和小鼠 （Mus musculus）的全基因组注释信息文件(GTF或者GFF3格式文件）</p>
</li>
<li><p>提取出其中的lincRNA的注释信息记录，然后将其写入MySQL数据库系统</p>
</li>
<li><p>写好PHP脚本与之前建好的MySQL数据库系统进行数据交互</p>
</li>
</ol>
<h2 id="准备需要写入数据库中的数据"><a href="#准备需要写入数据库中的数据" class="headerlink" title="准备需要写入数据库中的数据"></a>准备需要写入数据库中的数据</h2><h3 id="下载注释信息文件"><a href="#下载注释信息文件" class="headerlink" title="下载注释信息文件"></a>下载注释信息文件</h3><p>获取GTF文件的下载地址：</p>
<p>打开Ensemble数据库：<a href="http://asia.ensembl.org/index.html" target="_blank" rel="noopener">http://asia.ensembl.org/index.html</a></p>
<p><img src="/images/InAction-PHP-MySQL-download-1.png" alt></p>
<p><img src="/images/InAction-PHP-MySQL-arrow.jpg" alt></p>
<p><img src="/images/InAction-PHP-MySQL-download-2.png" alt></p>
<p><img src="/images/InAction-PHP-MySQL-arrow.jpg" alt></p>
<p><img src="/images/InAction-PHP-MySQL-download-3.png" alt></p>
<p><img src="/images/InAction-PHP-MySQL-arrow.jpg" alt></p>
<p>以下列出了可供下载的基因组注释信息版本<code>release-XX</code>，目前的最新版本为<code>release-91</code></p>
<p><img src="/images/InAction-PHP-MySQL-download-4.png" alt></p>
<p>得到下载链接，开始下载（下载至Linux服务器中）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 下载GRCh38.91（人类）注释文件</span><br><span class="line">$ wget -c -P /Path/To/dir ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz</span><br><span class="line"></span><br><span class="line"># 下载GRCm38.91（小鼠）注释文件</span><br><span class="line">$ wget -c -P /Path/To/dir ftp://ftp.ensembl.org/pub/release-91/gtf/mus_musculus/Mus_musculus.GRCm38.91.gtf.gz</span><br></pre></td></tr></table></figure>
<h3 id="提取lincRNA部分记录"><a href="#提取lincRNA部分记录" class="headerlink" title="提取lincRNA部分记录"></a>提取lincRNA部分记录</h3><p>首先解压前面下载的压缩文件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gunzip /Path/To/dir/Homo_sapiens.GRCh38.91.gtf.gz</span><br><span class="line">$ gunzip /Path/To/dir/Mus_musculus.GRCm38.91.gtf.gz</span><br></pre></td></tr></table></figure>
<p>GTF格式说明</p>
<p><img src="/images/InAction-PHP-MySQL-gtf-format.png" alt></p>
<p>看一下我们的GTF文件的具体情况</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Chrom	Resource	Feature	 Start	 End	Score	Strand	Frame 	Attributes</span><br><span class="line">1       havana  	gene    3073253 3074322 .       +       .       gene_id &quot;ENSMUSG00000102693&quot;; gene_version &quot;1&quot;; gene_name &quot;4933401J01Rik&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;TEC&quot;; havana_gene &quot;OTTMUSG00000049935&quot;; havana_gene_version &quot;1&quot;;</span><br><span class="line">1       havana  	transcript      3073253 3074322 .       +       .       gene_id &quot;ENSMUSG00000102693&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENSMUST00000193812&quot;; transcript_version &quot;1&quot;; gene_name &quot;4933401J01Rik&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;TEC&quot;; havana_gene &quot;OTTMUSG00000049935&quot;; havana_gene_version &quot;1&quot;; transcript_name &quot;4933401J01Rik-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;TEC&quot;; havana_transcript &quot;OTTMUST00000127109&quot;; havana_transcript_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;NA&quot;;</span><br><span class="line">1       havana  	exon    3073253 3074322 .       +       .       gene_id &quot;ENSMUSG00000102693&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENSMUST00000193812&quot;; transcript_version &quot;1&quot;; exon_number &quot;1&quot;; gene_name &quot;4933401J01Rik&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;TEC&quot;; havana_gene &quot;OTTMUSG00000049935&quot;; havana_gene_version &quot;1&quot;; transcript_name &quot;4933401J01Rik-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;TEC&quot;; havana_transcript &quot;OTTMUST00000127109&quot;; havana_transcript_version &quot;1&quot;; exon_id &quot;ENSMUSE00001343744&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;NA&quot;;</span><br></pre></td></tr></table></figure>
<p>可以看到，最后一列的<strong>gene_biotype</strong>中保存的是基因类型，例如protein_coding，lincRNA，miRNA等，所以针对这一列来提取</p>
<p>用Perl单行命令提取lincRNA注释记录</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl -ne &apos;chomp;next if (/^\#/);@gtf=split /\t/;if($gtf[8] =~ /gene_biotype\s\&quot;lincRNA\&quot;/) &#123;print &quot;$_\n&quot;;&#125;&apos; /Path/To/dir/Mus_musculus.GRCm38.91.gtf &gt;/Path/To/dir/lincRNA_GRCm38.91.gtf</span><br></pre></td></tr></table></figure>
<p>想获取该数据库实战中的示例数据，请点 <a href="https://github.com/Ming-Lian/Setup-Database/tree/master/txt-supply" target="_blank" rel="noopener"><strong>这里</strong></a></p>
<p>提取出来的lincRNA注释记录为以下形式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Chrom	Resource	Feature	 Start	 End	Score	Strand	Frame 	Attributes</span><br><span class="line">1       havana  	gene    29554   31109   .       +       .       gene_id &quot;ENSG00000243485&quot;; gene_version &quot;5&quot;; gene_name &quot;MIR1302-2HG&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;;</span><br><span class="line">1       havana  	transcript      29554   31097   .       +       .       gene_id &quot;ENSG00000243485&quot;; gene_version &quot;5&quot;; transcript_id &quot;ENST00000473358&quot;; transcript_version &quot;1&quot;; gene_name &quot;MIR1302-2HG&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;MIR1302-2HG-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; tag &quot;basic&quot;; transcript_support_level &quot;5&quot;;</span><br><span class="line">1       havana  	exon    29554   30039   .       +       .       gene_id &quot;ENSG00000243485&quot;; gene_version &quot;5&quot;; transcript_id &quot;ENST00000473358&quot;; transcript_version &quot;1&quot;; exon_number &quot;1&quot;; gene_name &quot;MIR1302-2HG&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;MIR1302-2HG-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; exon_id &quot;ENSE00001947070&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;5&quot;;</span><br></pre></td></tr></table></figure>
<p>接下来，将以上格式的信息格式化成以下形式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Chrom	Biotype	Feature	 	Start	 End	 GeneId		GeneName	TranscriptId		ExonNumber</span><br><span class="line"> 1	lincRNA	gene		29554	31109	ENSG00000243485	MIR1302-2HG	-			-</span><br><span class="line"> 1	lincRNA	transcript	29554	31097	ENSG00000243485	MIR1302-2HG	ENST00000473358		-</span><br><span class="line"> 1	lincRNA	exon		29554	30039	ENSG00000243485	MIR1302-2HG	ENST00000473358		1</span><br></pre></td></tr></table></figure>
<p>利用Perl的正则匹配提取目标字符串进行文本格式化，想了解正则表达式，请点 <a href="http://www.runoob.com/perl/perl-regular-expressions.html" target="_blank" rel="noopener">这里</a></p>
<blockquote>
<p>对于第三列Feature的不同，可以分别构造不同的正则表达式：</p>
<ul>
<li>gene：</li>
</ul>
<p>原格式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; gene_id &quot;ENSG00000243485&quot;; gene_version &quot;5&quot;; gene_name &quot;MIR1302-2HG&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;;</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<p>正则表达式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; /gene_id\s\&quot;(\w+)\&quot;;.&#123;18,19&#125;\sgene_name\s\&quot;([\w-\.]+)\&quot;/</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<ul>
<li>transcript：</li>
</ul>
<p>原格式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; gene_id &quot;ENSG00000243485&quot;; gene_version &quot;5&quot;; transcript_id &quot;ENST00000473358&quot;; transcript_version &quot;1&quot;; gene_name &quot;MIR1302-2HG&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;MIR1302-2HG-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; tag &quot;basic&quot;; transcript_support_level &quot;5&quot;;</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<p>正则表达式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; /gene_id\s\&quot;(\w+)\&quot;;.&#123;18,19&#125;\stranscript_id\s\&quot;(\w+)\&quot;;.&#123;24,25&#125;\sgene_name\s\&quot;([\w-\.]+)\&quot;/</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<ul>
<li>exon：</li>
</ul>
<p>原格式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; gene_id &quot;ENSG00000243485&quot;; gene_version &quot;5&quot;; transcript_id &quot;ENST00000473358&quot;; transcript_version &quot;1&quot;; exon_number &quot;1&quot;; gene_name &quot;MIR1302-2HG&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;MIR1302-2HG-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; exon_id &quot;ENSE00001947070&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;5&quot;;</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<p>正则表达式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; /gene_id\s\&quot;(\w+)\&quot;;.&#123;18,19&#125;\stranscript_id\s\&quot;(\w+)\&quot;;.&#123;24,25&#125;\sexon_number\s\&quot;(\d+)\&quot;;\sgene_name\s\&quot;([\w-\.]+)\&quot;/</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"># 正则匹配的正则表达式可能有点复杂，这里为了方便阅读分为多行且加上缩进，在实际命令行中请写成一行</span><br><span class="line"></span><br><span class="line">$ perl -ne &apos;chomp;next if (/^\#/);@linc=split /\t/;</span><br><span class="line">if($linc[2] =~ /gene/)&#123;</span><br><span class="line">	if ($linc[8] =~ /gene_id\s\&quot;(\w+)\&quot;;.&#123;18,19&#125;\sgene_name\s\&quot;([\w-\.]+)\&quot;/)&#123;</span><br><span class="line">		print &quot;$linc[0]\tlincRNA\t$linc[2]\t$linc[3]\t$linc[4]\t$1\t$2\t-\t-\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;elsif($linc[2] =~ /transcript/)&#123;</span><br><span class="line">	if($linc[8] =~ /gene_id\s\&quot;(\w+)\&quot;;.&#123;18,19&#125;\stranscript_id\s\&quot;(\w+)\&quot;;.&#123;24,25&#125;\sgene_name\s\&quot;([\w-\.]+)\&quot;/)</span><br><span class="line">	&#123;</span><br><span class="line">		print &quot;$linc[0]\tlincRNA\t$linc[2]\t$linc[3]\t$linc[4]\t$1\t$3\t$2\t-\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	if($linc[8] =~ /gene_id\s\&quot;(\w+)\&quot;;.&#123;18,19&#125;\stranscript_id\s\&quot;(\w+)\&quot;;.&#123;24,25&#125;\sexon_number\s\&quot;(\d+)\&quot;;\sgene_name\s\&quot;([\w-\.]+)\&quot;/)</span><br><span class="line">	&#123;</span><br><span class="line">		print &quot;$linc[0]\tlincRNA\t$linc[2]\t$linc[3]\t$linc[4]\t$1\t$4\t$2\t$3\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;&apos;</span><br><span class="line"></span><br><span class="line">/Path/To/dir/lincRNA_GRCm38.91.gtf &gt;/Path/To/dir/lincRNA_GRCm38.91.gtf.format</span><br></pre></td></tr></table></figure>
<h2 id="将数据写入数据库中"><a href="#将数据写入数据库中" class="headerlink" title="将数据写入数据库中"></a>将数据写入数据库中</h2><p>以下以将<code>lincRNA_GRCh38.91.gtf.format</code>的数据导入<code>lincRNA_h</code>表中为例</p>
<h3 id="方法一：使用MySQLi"><a href="#方法一：使用MySQLi" class="headerlink" title="方法一：使用MySQLi"></a>方法一：使用MySQLi</h3><p>MySQLi：PHP中用于与MySQL数据库系统进行数据库交互的扩展</p>
<p>使用MySQLi的预处理与参数绑定方法，进行数据的批量导入</p>
<p>预备知识：</p>
<blockquote>
<ul>
<li>想了解 MySQLi 连接数据库的方法，请点 <a href="https://github.com/Ming-Lian/Memo/blob/master/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9APHP%E4%B8%80%E5%91%A8%E9%80%9F%E6%88%90.md#connect-mysql-mysqli" target="_blank" rel="noopener">这里</a></li>
<li>想了解 MySQLi 的预处理方法，请点 <a href="https://github.com/Ming-Lian/Memo/blob/master/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9APHP%E4%B8%80%E5%91%A8%E9%80%9F%E6%88%90.md#use-prepare-sentence" target="_blank" rel="noopener">这里</a></li>
<li>想了解 PHP 的文件处理方法，请点 <a href="https://github.com/Ming-Lian/Memo/blob/master/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9APHP%E4%B8%80%E5%91%A8%E9%80%9F%E6%88%90.md#file-process" target="_blank" rel="noopener">这里</a></li>
</ul>
</blockquote>
<p>1. 首先，创建好表格</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"># 登录MySQL</span><br><span class="line">$ mysql -u username -p</span><br><span class="line"></span><br><span class="line"># 选定要操作的数据库，即打算将表格创建在哪个数据库底下，例如我们要操作的数据库为lincRNAdb</span><br><span class="line">mysql&gt; use lincRNAdb;</span><br><span class="line"></span><br><span class="line"># 创建表格，表格名为lincRNA_h</span><br><span class="line">mysql&gt; CREATE TABLE lincRNA_h (</span><br><span class="line">  `id` int(11) NOT NULL AUTO_INCREMENT,</span><br><span class="line">  `chrom` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;染色体&apos;,</span><br><span class="line">  `biotype` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;基因类型&apos;,</span><br><span class="line">  `feature` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;结构单元&apos;,</span><br><span class="line">  `start` int(11) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;起始位置&apos;,</span><br><span class="line">  `end` int(11) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;终止位置&apos;,</span><br><span class="line">  `geneid` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;基因ENsembleId&apos;,</span><br><span class="line">  `genename` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;基因名&apos;,</span><br><span class="line">  `transcriptid` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;转录本ENsembleId&apos;,</span><br><span class="line">  `exon` int(11) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;转录本exon序号&apos;,</span><br><span class="line">  PRIMARY KEY (`id`) </span><br><span class="line">) ;</span><br></pre></td></tr></table></figure>
<ul>
<li>编辑php脚本<code>lincRNA_Ano_data_batch_import.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">$host=&quot;localhost&quot;;</span><br><span class="line"></span><br><span class="line">// 请设置好正确的用户名和密码，注意必须为具有数据库写权限的用户</span><br><span class="line">$username=&quot;root&quot;;</span><br><span class="line">$password=&quot;password&quot;;</span><br><span class="line"></span><br><span class="line">$dbname=&quot;testdb&quot;;</span><br><span class="line"></span><br><span class="line">// 连接数据库</span><br><span class="line">$conn=new mysqli($host,$username,$password,$dbname);</span><br><span class="line"></span><br><span class="line">// 检查连接</span><br><span class="line">if($conn-&gt;connect_error)&#123;</span><br><span class="line">	die(&quot;连接失败：&quot;.$conn-&gt;connect_error);</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	echo &quot;连接成功！\n&quot;;</span><br><span class="line">	echo &quot;正在批量导入数据，请耐心等待...\n&quot;;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">// 预处理及绑定</span><br><span class="line">$stmt=$conn_prepare(&quot;INSERT INTO $argv[1] (chrom,biotype,feature,start,end,geneid,genename,transcriptid,exon) VALUES (?,?,?,?,?,?,?,?,?)&quot;) ;</span><br><span class="line">$stmt-&gt;bind_param(&quot;sssiisssi&quot;,$chrom,$biotype,$feature,$start,$end,$geneid,$genename,$transcriptid,$exon);</span><br><span class="line"></span><br><span class="line">// 打开文件</span><br><span class="line">$ file=fopen(&quot;$argv[2]&quot;,&apos;r&apos;) or exit(&quot;Unable to open file!&quot;);</span><br><span class="line"></span><br><span class="line">// 逐行读取文件，并写入数据库</span><br><span class="line">while(!feof($file))&#123;</span><br><span class="line">	$data=explode(&quot;\t&quot;,fgets($file)); // 以tab为间隔标识将字符串打散</span><br><span class="line">	</span><br><span class="line">	// 设置参数</span><br><span class="line">	$chrom=$data[0];</span><br><span class="line">	$biotype=$data[1];</span><br><span class="line">	$feature=$data[2];</span><br><span class="line">	$start=$data[3];</span><br><span class="line">	$end=$data[4];</span><br><span class="line">	$geneid=$data[5];</span><br><span class="line">	$genename=$data[6];</span><br><span class="line">	$transcriptid=$data[7];</span><br><span class="line">	$exon=$data[8];</span><br><span class="line">	// 执行</span><br><span class="line">	$stmt-&gt;execute();</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">echo &quot;新记录插入成功&quot;;</span><br><span class="line"></span><br><span class="line">fcolse($file);</span><br><span class="line">$stmt-&gt;close();</span><br><span class="line">$conn-&gt;close();</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li>执行php脚本，完成数据的批量导入</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ php -f lincRNA_Ano_data_batch_import.php &lt;tablename&gt; &lt;data-to-import&gt;</span><br></pre></td></tr></table></figure>
<p>想了解如何在 Linux 命令行中使用和执行 PHP 代码，请点 <a href="https://github.com/Ming-Lian/Memo/blob/master/Beginning-SQL.md#run-php-code-in-linux" target="_blank" rel="noopener">这里</a></p>
<h3 id="方法二：使用sql脚本"><a href="#方法二：使用sql脚本" class="headerlink" title="方法二：使用sql脚本"></a>方法二：使用sql脚本</h3><h4 id="准备sql脚本"><a href="#准备sql脚本" class="headerlink" title="准备sql脚本"></a>准备sql脚本</h4><p>要往MySQL数据库中写入数据，需要用<strong>INSERT INTO 语句</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">INSERT INTO table_name (column1,column2,column3,...)</span><br><span class="line">VALUES (value1,value2,value3,...);</span><br></pre></td></tr></table></figure>
<p>一般情况下我们往数据库中写入少数几条数据时，直接在MySQL的交互环境中执行INSERT INTO 语句即可，但是当要批量导入时，再采用在MySQL的交互环境中执行INSERT INTO 语句，明显是低效而不现实的，这个时候我们可以写一个sql脚本，来执行数据库的批量操作。</p>
<p>简单来说sql脚本就是把多个INSERT INTO 语句写到一个文件里</p>
<p>sql脚本的文件起始部分（为人类和小鼠的数据分别准备一个sql文件，创建的对应表名分别为 lincRNA_h 和 lincRNA_m ，以下以人类的为例）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">SET NAMES utf8;</span><br><span class="line">SET FOREIGN_KEY_CHECKS = 0;</span><br><span class="line"></span><br><span class="line">-- ----------------------------</span><br><span class="line">--  Table structure for `lincRNA_h`</span><br><span class="line">-- ----------------------------</span><br><span class="line"></span><br><span class="line">DROP TABLE IF EXISTS `lincRNA_h`;</span><br><span class="line"></span><br><span class="line">CREATE TABLE lincRNA_h (</span><br><span class="line">  `id` int(11) NOT NULL AUTO_INCREMENT,</span><br><span class="line">  `chrom` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;染色体&apos;,</span><br><span class="line">  `biotype` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;基因类型&apos;,</span><br><span class="line">  `feature` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;结构单元&apos;,</span><br><span class="line">  `start` int(11) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;起始位置&apos;,</span><br><span class="line">  `end` int(11) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;终止位置&apos;,</span><br><span class="line">  `geneid` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;基因ENsembleId&apos;,</span><br><span class="line">  `genename` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;基因名&apos;,</span><br><span class="line">  `transcriptid` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;转录本ENsembleId&apos;,</span><br><span class="line">  `exon` int(11) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;转录本exon序号&apos;,</span><br><span class="line">  PRIMARY KEY (`id`) </span><br><span class="line">) ENGINE=InnoDB AUTO_INCREMENT=6 DEFAULT CHARSET=utf8;</span><br><span class="line"></span><br><span class="line">-- ----------------------------</span><br><span class="line">--  Records of `lincRNA_h`</span><br><span class="line">-- ----------------------------</span><br></pre></td></tr></table></figure>
<p>然后，用perl单行命令实现以下转换：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">1	lincRNA	gene	29554	31109	ENSG00000243485	MIR1302-2HG	-	-</span><br></pre></td></tr></table></figure>
<p><img src="/images/InAction-PHP-MySQL-arrow.jpg" alt></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">INSERT INTO lincRNA_h （chrom,biotype,feature,start,end,geneid,genename,transcriptid,exon) VALUES (&quot;1&quot;,&quot;lincRNA&quot;,&quot;gene&quot;,&quot;29554&quot;,&quot;31109&quot;,&quot;ENSG00000243485&quot;,&quot;MIR1302-2HG&quot;,&quot;-&quot;,&quot;-&quot;);</span><br></pre></td></tr></table></figure>
<p>并将转换后的结果追加到上一步已经写好起始部分的sql脚本的后面</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl -ane &apos;chomp;print &quot;INSERT INTO lincRNA_h (chrom,biotype,feature,start,end,geneid,genename,transcriptid,exon) VALUES (\&quot;$F[0]\&quot;,\&quot;$F[1]\&quot;,\&quot;$F[2]\&quot;,\&quot;$F[3]\&quot;,\&quot;$F[4]\&quot;,\&quot;$F[5]\&quot;,\&quot;$F[6]\&quot;,\&quot;$F[7]\&quot;,\&quot;$F[8]\&quot;);\n&quot;&apos; /Path/To/lincRNA_GRCh38.91.gtf.format &gt;&gt;/Path/To/lincRNA_h.sql</span><br></pre></td></tr></table></figure>
<h4 id="写入数据"><a href="#写入数据" class="headerlink" title="写入数据"></a>写入数据</h4><p>登录，进入MySQL的交互环境，才能进行以下操作</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"># 输入密码，登录MySQL</span><br><span class="line">$ mysql -u username -p</span><br><span class="line"></span><br><span class="line"># 查看已有的数据库</span><br><span class="line">mysql&gt; show databases;</span><br><span class="line"></span><br><span class="line"># 选择要进行操作的数据库，比如我们要操作的数据库为lincRNAdb</span><br><span class="line">## 注意：请确认你具有该数据库的写入权限，否则无法写入</span><br><span class="line">mysql&gt; use lincRNAdb;</span><br><span class="line"></span><br><span class="line"># 执行sql脚本，向数据库中批量导入数据</span><br><span class="line">mysql&gt; source /Path/To/lincRNA_h.sql;</span><br><span class="line"></span><br><span class="line"># 查看当前使用的数据库中已有表格，确保数据已成功导入：是否有lincRNA_h表格？</span><br><span class="line">mysql&gt; show tables;</span><br><span class="line"></span><br><span class="line"># 查看表格中的数据是否正确，为了避免将表中的数据全部打印出来，请使用where子句，只打印出前100条记录</span><br><span class="line">mysql&gt; select * from lincRNA_h where id&lt;100;</span><br></pre></td></tr></table></figure>
<h2 id="编辑php脚本"><a href="#编辑php脚本" class="headerlink" title="编辑php脚本"></a>编辑php脚本</h2><h3 id="注册"><a href="#注册" class="headerlink" title="注册"></a>注册</h3><h4 id="创建注册页面"><a href="#创建注册页面" class="headerlink" title="创建注册页面"></a>创建注册页面</h4><p><img src="/images/InAction-PHP-MySQL-register.png" alt></p>
<p>表单部分的代码（想了解表单的相关知识，请点 <a href="https://github.com/Ming-Lian/Memo/blob/master/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9APHP%E4%B8%80%E5%91%A8%E9%80%9F%E6%88%90.md#form-and-input" target="_blank" rel="noopener">这里</a>）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br></pre></td><td class="code"><pre><span class="line">&lt;form action=&quot;registeraction.php&quot; method=&quot;post&quot;&gt;</span><br><span class="line">&lt;table border=&quot;0&quot;&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;用户名：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; id=&quot;id_name&quot; name=&quot;username&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;密&amp;nbsp;&amp;nbsp;&amp;nbsp;码：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;password&quot; id=&quot;password&quot; name=&quot;password&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;重复密码：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;password&quot; id=&quot;re_password&quot; name=&quot;re_password&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;性别：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;</span><br><span class="line">		&lt;input type=&quot;radio&quot; id=&quot;sex&quot; name=&quot;sex&quot; value=&quot;男&quot;&gt;男</span><br><span class="line">		&lt;input type=&quot;radio&quot; id=&quot;sex&quot; name=&quot;sex&quot; value=&quot;女&quot;&gt;女</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;QQ：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; id=&quot;qq&quot; name=&quot;qq&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;Email：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;email&quot; id=&quot;email&quot; name=&quot;email&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;电话：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; id=&quot;phone&quot; name=&quot;phone&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;地址：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; id=&quot;address&quot; name=&quot;address&quot; required=&quot;required&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot; style=&quot;color:red;font-size:10px;&quot;&gt;</span><br><span class="line">&lt;!--提示信息--&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">$err=isset($_GET[&quot;err&quot;])?$_GET[&quot;err&quot;]:&quot;&quot;;</span><br><span class="line">switch($err) &#123;</span><br><span class="line">	case 1:</span><br><span class="line">		echo &quot;用户名已存在！&quot;;</span><br><span class="line"> 		break;</span><br><span class="line">	case 2:</span><br><span class="line">		echo &quot;密码与重复密码不一致！&quot;;</span><br><span class="line">		break;</span><br><span class="line">	case 3:</span><br><span class="line">		echo &quot;注册成功！&quot;;</span><br><span class="line">		break;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br><span class="line">&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot;&gt;</span><br><span class="line">		&lt;input type=&quot;submit&quot; id=&quot;register&quot; name=&quot;register&quot; value=&quot;注册&quot;&gt;</span><br><span class="line">		&lt;input type=&quot;reset&quot; id=&quot;reset&quot; name=&quot;reset&quot; value=&quot;重置&quot;&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot;&gt;如果已有账号，快去&lt;a href=&quot;login.php&quot;&gt;登录&lt;/a&gt;吧！&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;/table&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<p>表单信息提交的对象为同一目录下的<code>registeraction.php</code>脚本</p>
<h4 id="注册信息处理脚本"><a href="#注册信息处理脚本" class="headerlink" title="注册信息处理脚本"></a>注册信息处理脚本</h4><p><code>registeraction.php</code>脚本</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">// 声明变量</span><br><span class="line"></span><br><span class="line">$username = isset($_POST[&apos;username&apos;])?$_POST[&apos;username&apos;]:&quot;&quot;;</span><br><span class="line">$password = isset($_POST[&apos;password&apos;])?$_POST[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line">$re_password = isset($_POST[&apos;re_password&apos;])?$_POST[&apos;re_password&apos;]:&quot;&quot;;</span><br><span class="line">$sex = isset($_POST[&apos;sex&apos;])?$_POST[&apos;sex&apos;]:&quot;&quot;;</span><br><span class="line">$qq = isset($_POST[&apos;qq&apos;])?$_POST[&apos;qq&apos;]:&quot;&quot;;</span><br><span class="line">$email = isset($_POST[&apos;email&apos;])?$_POST[&apos;email&apos;]:&quot;&quot;;</span><br><span class="line">$phone = isset($_POST[&apos;phone&apos;])?$_POST[&apos;phone&apos;]:&quot;&quot;;</span><br><span class="line">$address = isset($_POST[&apos;address&apos;])?$_POST[&apos;address&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">if($password == $re_password) &#123;</span><br><span class="line">	// 建立连接，需要以root的身份登录MySQL，因为只有root用户才具有数据库的写权限</span><br><span class="line">	// 使用面向对象的MySQLi方法</span><br><span class="line">	$conn = new mysqli(&apos;localhost&apos;,&apos;root&apos;,&apos;&apos;,&apos;php&apos;); // 第三项需填入root用户的密码，若设置为免密登录，则可以为空</span><br><span class="line">	// 准备SQL语句,查询用户名</span><br><span class="line">	$sql_select=&quot;SELECT username FROM User WHERE username = &apos;$username&apos;&quot;;</span><br><span class="line">	// 执行SQL语句</span><br><span class="line">	$result = $conn-&gt;query($sql_select);</span><br><span class="line">	$row = $result-&gt;fetch_array();</span><br><span class="line">	// 判断用户名是否已存在</span><br><span class="line">	if($username == $row[&apos;username&apos;]) &#123;</span><br><span class="line">		//用户名已存在，显示提示信息</span><br><span class="line">		header(&quot;Location:register.php?err=1&quot;);</span><br><span class="line">	&#125; else &#123;</span><br><span class="line"></span><br><span class="line">		//用户名不存在，插入数据</span><br><span class="line">		//准备SQL语句</span><br><span class="line">		$sql_insert = &quot;INSERT INTO User(username,password,sex,qq,email,phone,address) VALUES(&apos;$username&apos;,&apos;$password&apos;,&apos;$sex&apos;,&apos;$qq&apos;,&apos;$email&apos;,&apos;$phone&apos;,&apos;$address&apos;)&quot;;</span><br><span class="line">		//执行SQL语句</span><br><span class="line">		$conn-&gt;query($sql_insert);</span><br><span class="line">		header(&quot;Location:register.php?err=3&quot;);</span><br><span class="line">	&#125;</span><br><span class="line"></span><br><span class="line">	//关闭数据库</span><br><span class="line">	$conn-&gt;close($conn);</span><br><span class="line">&#125; else &#123;</span><br><span class="line">	header(&quot;Location:register.php?err=2&quot;);</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h3 id="登录"><a href="#登录" class="headerlink" title="登录"></a>登录</h3><h4 id="登录页面"><a href="#登录页面" class="headerlink" title="登录页面"></a>登录页面</h4><p><img src="/images/InAction-PHP-MySQL-login.png" alt></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">&lt;form id=&quot;loginform&quot; action=&quot;loginaction.php&quot; method=&quot;post&quot;&gt;</span><br><span class="line">&lt;table border=&quot;0&quot;&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;用户名：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; id=&quot;name&quot; name=&quot;username&quot; required=&quot;required&quot; value=&quot;&lt;?php echo isset($_COOKIE[&quot;wang&quot;])?$_COOKIE[&quot;wang&quot;]:&quot;&quot;;?&gt;&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;密&amp;nbsp;&amp;nbsp;&amp;nbsp;码：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;password&quot; id=&quot;password&quot; name=&quot;password&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot;&gt;&lt;input type=&quot;checkbox&quot; name=&quot;remember&quot; value=&quot;on&quot;&gt;&lt;small&gt;记住我&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot; style=&quot;color:red;font-size:10px;&quot;&gt;</span><br><span class="line">&lt;!--提示信息--&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">$err=isset($_GET[&quot;err&quot;])?$_GET[&quot;err&quot;]:&quot;&quot;;</span><br><span class="line">switch($err) &#123;</span><br><span class="line">case 1:</span><br><span class="line">	echo &quot;用户名或密码错误！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 2:</span><br><span class="line">	echo &quot;用户名或密码不能为空！&quot;;</span><br><span class="line">	break;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot;&gt;</span><br><span class="line">		&lt;input type=&quot;submit&quot; id=&quot;login&quot; name=&quot;login&quot; value=&quot;登录&quot;&gt;</span><br><span class="line">		&lt;input type=&quot;reset&quot; id=&quot;reset&quot; name=&quot;reset&quot; value=&quot;重置&quot;&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot;&gt;还没有账号，快去&lt;a href=&quot;register.php&quot;&gt;注册&lt;/a&gt;吧！&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;/table&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<h4 id="登录脚本"><a href="#登录脚本" class="headerlink" title="登录脚本"></a>登录脚本</h4><p><a name="login-php-1">方法一：</a></p>
<p>该方法为，用root用户权限从User表中提取相应的用户名和密码信息，与用户填写的用户名和密码进行比较，来判断用户用户是否拥有数据库的访问权限</p>
<p>该方法需要拥有root用户的账户密码，若没有，请使用第二种方法</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">// 声明变量</span><br><span class="line">$username = isset($_POST[&apos;username&apos;])?$_POST[&apos;username&apos;]:&quot;&quot;;</span><br><span class="line">$password = isset($_POST[&apos;password&apos;])?$_POST[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line">$remember = isset($_POST[&apos;remember&apos;])?$_POST[&apos;remember&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 判断用户名和密码是否为空</span><br><span class="line">if(!empty($username)&amp;&amp;!empty($password)) &#123;</span><br><span class="line">	// 建立连接</span><br><span class="line">	// 使用面向对象的方法</span><br><span class="line">	$conn = new mysqli(&apos;localhost&apos;,&apos;root&apos;,&apos;&apos;,&apos;php&apos;);</span><br><span class="line">	// 准备SQL语句</span><br><span class="line">	$sql_select = &quot;SELECT user,password FROM User WHERE user = &apos;$username&apos; AND password = &apos;$password&apos;&quot;;</span><br><span class="line">	//执行SQL语句</span><br><span class="line">	$result = $conn-&gt;query($sql_select);</span><br><span class="line"></span><br><span class="line">	$row = $result-&gt;fetch_assoc();</span><br><span class="line"></span><br><span class="line">	//判断用户名或密码是否正确</span><br><span class="line">	if($username==$row[&apos;username&apos;]&amp;&amp;$password==$row[&apos;password&apos;]) &#123;</span><br><span class="line">		//选中“记住我”</span><br><span class="line">		if($remember==&quot;on&quot;) &#123;</span><br><span class="line">			//创建cookie</span><br><span class="line">			setcookie(&quot;wang&quot;, $username, time()+7*24*3600);</span><br><span class="line">		&#125;</span><br><span class="line">		//开启session</span><br><span class="line">		session_start();</span><br><span class="line">		//创建session</span><br><span class="line">		$_SESSION[&apos;user&apos;]=$username;</span><br><span class="line">		$_SESSION[&apos;password&apos;]=$password;</span><br><span class="line">		//写入日志</span><br><span class="line">		$ip = $_SERVER[&apos;REMOTE_ADDR&apos;];</span><br><span class="line">		$date = date(&apos;Y-m-d H:m:s&apos;);</span><br><span class="line"></span><br><span class="line">		$info = sprintf(&quot;当前访问用户：%s,IP地址：%s,时间：%s \n&quot;,$username, $ip, $date);</span><br><span class="line">		$sql_logs = &quot;INSERT INTO Logs(username,ip,date) VALUES(&apos;$username&apos;,&apos;$ip&apos;,&apos;$date&apos;)&quot;;</span><br><span class="line"></span><br><span class="line">		//日志写入文件，如实现此功能，需要创建文件目录logs</span><br><span class="line">		$f = fopen(&apos;./logs/&apos;.date(&apos;Ymd&apos;).&apos;.log&apos;,&apos;a+&apos;);</span><br><span class="line"></span><br><span class="line">		fwrite($f,$info);</span><br><span class="line">		fclose($f);</span><br><span class="line"></span><br><span class="line">		//跳转到loginsucc.php页面</span><br><span class="line">		header(&quot;Location:loginsucc.php&quot;);</span><br><span class="line">		//关闭数据库</span><br><span class="line">		$conn-&gt;close();</span><br><span class="line">	&#125;else &#123;</span><br><span class="line">		//用户名或密码错误，赋值err为1</span><br><span class="line">		header(&quot;Location:login.php?err=1&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else &#123;</span><br><span class="line">	//用户名或密码为空，赋值err为2</span><br><span class="line">	header(&quot;Location:login.php?err=2&quot;);</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<p><a name="login-php-2">方法二：</a></p>
<p>该方法通过使用用户填写的用户名和密码直接尝试登陆MySQL，通过登录结果来判断用户是否拥有数据库的使用权限</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">// 声明变量</span><br><span class="line">$username = isset($_POST[&apos;username&apos;])?$_POST[&apos;username&apos;]:&quot;&quot;;</span><br><span class="line">$password = isset($_POST[&apos;password&apos;])?$_POST[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line">$remember = isset($_POST[&apos;remember&apos;])?$_POST[&apos;remember&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 判断用户名和密码是否为空</span><br><span class="line">if(!empty($username)&amp;&amp;!empty($password)) &#123;</span><br><span class="line">	// 建立连接</span><br><span class="line">	// 使用面向对象的方法，使用用户填写的用户名和密码直接尝试登陆MySQL</span><br><span class="line">	$conn = new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">	</span><br><span class="line">	if ($conn-&gt;connect_error) &#123;</span><br><span class="line">		// 连接失败，用户名或密码错误，赋值err为1</span><br><span class="line">		header(&quot;Location:login.php?err=1&quot;);</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		//选中“记住我”</span><br><span class="line">		if($remember==&quot;on&quot;) &#123;</span><br><span class="line">			//创建cookie</span><br><span class="line">			setcookie(&quot;wang&quot;, $username, time()+7*24*3600);</span><br><span class="line">		&#125;</span><br><span class="line">		//开启session</span><br><span class="line">		session_start();</span><br><span class="line">		//创建session</span><br><span class="line">		$_SESSION[&apos;user&apos;]=$username;</span><br><span class="line">		$_SESSION[&apos;password&apos;]=$password;</span><br><span class="line"></span><br><span class="line">		//跳转到loginsucc.php页面</span><br><span class="line">		header(&quot;Location:loginsucc.php&quot;);</span><br><span class="line">		//关闭数据库</span><br><span class="line">		$conn-&gt;close();</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else &#123;</span><br><span class="line">	//用户名或密码为空，赋值err为2</span><br><span class="line">	header(&quot;Location:login.php?err=2&quot;);</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h4 id="登录成功"><a href="#登录成功" class="headerlink" title="登录成功"></a>登录成功</h4><p><img src="/images/InAction-PHP-MySQL-loginsucc.png" alt></p>
<p>通过判断全局变量<code>$_SESSION[&#39;user&#39;]</code>是否定义来判断是否成功登录</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">//声明变量</span><br><span class="line">$username= isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">//判断session是否为空</span><br><span class="line">if(!empty($username))&#123;</span><br><span class="line">?&gt;</span><br><span class="line">	&lt;h1&gt;登录成功！&lt;/h1&gt;</span><br><span class="line">	欢迎您！</span><br><span class="line">&lt;?php</span><br><span class="line">	echo $username;</span><br><span class="line">?&gt;</span><br><span class="line">	// 继续检索数据库</span><br><span class="line">	&lt;br/&gt;</span><br><span class="line">	&lt;a href=&quot;databaseQuery_Ano.php&quot;&gt;检索数据库&lt;/a&gt;</span><br><span class="line">	&lt;br/&gt;</span><br><span class="line">	&lt;a href=&quot;logout.php&quot;&gt;退出&lt;/a&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">&#125;else &#123;</span><br><span class="line">	//未登录，无权访问</span><br><span class="line">	?&gt;</span><br><span class="line">	&lt;h1&gt;你无权访问！！！&lt;/h1&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h4 id="退出登录"><a href="#退出登录" class="headerlink" title="退出登录"></a>退出登录</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">//撤销session</span><br><span class="line">session_unset();</span><br><span class="line">session_destroy();</span><br><span class="line">//跳转到login.php</span><br><span class="line">header(&quot;Location:login.php&quot;);</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h3 id="检索数据库"><a href="#检索数据库" class="headerlink" title="检索数据库"></a>检索数据库</h3><p>检索数据库，需要用到数据库用户的<strong>用户名</strong>和<strong>密码</strong>，用于在之前的登录过程中，我们已经将用户名和密码存储在 $_SESSION 中，所以可以通过 <code>$_SESSION[&#39;user&#39;]</code> 和 <code>$_SESSION[&#39;password&#39;]</code> 获得</p>
<p>这里将表单部分与PHP命令写在一个PHP脚本中，脚本命名为<code>databaseQuery_Ano.php</code></p>
<h4 id="表单部分"><a href="#表单部分" class="headerlink" title="表单部分"></a>表单部分</h4><p><img src="/images/InAction-PHP-MySQL-query-form.png" alt></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br></pre></td><td class="code"><pre><span class="line">&lt;form action=&quot;#&quot; method=&quot;post&quot;&gt;</span><br><span class="line">&lt;table&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td align=&quot;left&quot;&gt;物&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp种：&lt;/td&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;left&quot;&gt;</span><br><span class="line">		&lt;select name=&quot;specie&quot;&gt;</span><br><span class="line">			&lt;option value=&quot;&quot;&gt;选择一个物种:&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;human&quot;&gt;Homo Sapiens&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;mouse&quot;&gt;Mus musculus&lt;/option&gt;</span><br><span class="line">		&lt;/select&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td align=&quot;left&quot;&gt;基因名/基因ID:&lt;/td&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;left&quot;&gt;&lt;input type=&quot;text&quot; name=&quot;geneName&quot; value=&quot;MIR1302-2HG&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td align=&quot;left&quot;&gt;基&amp;nbsp;&amp;nbsp;&amp;nbsp因&amp;nbsp;&amp;nbsp;&amp;nbsp位&amp;nbsp;&amp;nbsp;&amp;nbsp置：&lt;/td&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;left&quot;&gt;</span><br><span class="line">		&lt;select name=&quot;chrom&quot;&gt;</span><br><span class="line">			&lt;option value=&quot;&quot;&gt;选择一条染色体：&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;1&quot;&gt;chr1&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;2&quot;&gt;chr2&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;3&quot;&gt;chr3&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;4&quot;&gt;chr4&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;5&quot;&gt;chr5&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;6&quot;&gt;chr6&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;7&quot;&gt;chr7&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;8&quot;&gt;chr8&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;9&quot;&gt;chr9&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;10&quot;&gt;chr10&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;11&quot;&gt;chr11&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;12&quot;&gt;chr12&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;13&quot;&gt;chr13&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;14&quot;&gt;chr14&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;15&quot;&gt;chr15&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;16&quot;&gt;chr16&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;17&quot;&gt;chr17&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;18&quot;&gt;chr18&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;19&quot;&gt;chr19&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;20&quot;&gt;chr20&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;21&quot;&gt;chr21&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;22&quot;&gt;chr22&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;X&quot;&gt;chrX&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;Y&quot;&gt;chrY&lt;/option&gt;</span><br><span class="line">		&lt;/select&gt;	</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;</span><br><span class="line">		区间：&lt;input type=&quot;text&quot; name=&quot;start&quot;&gt; - &lt;input type=&quot;text&quot; name=&quot;end&quot;&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td colspan=&quot;2&quot; align=&quot;center&quot; style=&quot;color:red;font-size:10px;&quot;&gt;</span><br><span class="line">&lt;!--提示信息--&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">$err=isset($_GET[&quot;err&quot;])?$_GET[&quot;err&quot;]:&quot;&quot;;</span><br><span class="line">switch($err) &#123;</span><br><span class="line">case 1:</span><br><span class="line">	echo &quot;表单填写错误:物种必须选，基因名和基因位置两项至少选一项填写！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 2:</span><br><span class="line">	echo &quot;表单填写错误：区间起始和终止位置需为整数，且起始位置小于终止位置！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 3:</span><br><span class="line">	echo &quot;未登陆，无权访问！&lt;br&gt;&quot;;</span><br><span class="line">	echo &quot;快去&lt;a href=&apos;login.php&apos;&gt;登录&lt;/a&gt;吧&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 4:</span><br><span class="line">	echo &quot;Update或Delete操作异常！未在数据库中找到对应记录！&quot;;</span><br><span class="line">	break;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td align=&quot;center&quot;&gt;&lt;input type=&quot;submit&quot; name=&quot;submit&quot; value=&quot;Submit&quot;&gt;&lt;/td&gt;</span><br><span class="line">	&lt;td align=&quot;center&quot;&gt;&lt;input type=&quot;reset&quot; value=&quot;Reset&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;/table&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<h4 id="PHP脚本"><a href="#PHP脚本" class="headerlink" title="PHP脚本"></a>PHP脚本</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 获取表单提交数据</span><br><span class="line">$specie=isset($_POST[&apos;specie&apos;])?$_POST[&apos;specie&apos;]:&quot;&quot;;</span><br><span class="line">$geneName=isset($_POST[&apos;geneName&apos;])?$_POST[&apos;geneName&apos;]:&quot;&quot;;</span><br><span class="line">$chrom=isset($_POST[&apos;chrom&apos;])?$_POST[&apos;chrom&apos;]:&quot;&quot;;</span><br><span class="line">$start=isset($_POST[&apos;start&apos;])?$_POST[&apos;start&apos;]:&quot;&quot;;</span><br><span class="line">$end=isset($_POST[&apos;end&apos;])?$_POST[&apos;end&apos;]:&quot;&quot;;</span><br><span class="line">$submit=isset($_POST[&apos;submit&apos;])?$_POST[&apos;submit&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">if($submit)&#123;</span><br><span class="line">	// 判断提交的检索信息是否满足要求：物种必须选，基因名和基因位置两项至少选一项填写</span><br><span class="line">	if(empty($specie)||(empty($geneName)&amp;&amp;empty($chrom)))&#123;</span><br><span class="line">		header(&quot;Location:databaseQuery.php?err=1&quot;);</span><br><span class="line"></span><br><span class="line">	// 判断用户名和密码是否已经设置</span><br><span class="line">	&#125;elseif(!empty($username)&amp;&amp;!empty($password))&#123;</span><br><span class="line">		// 连接数据库</span><br><span class="line">		$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">		if($conn-&gt;connect_error)&#123;</span><br><span class="line">			header(&quot;Location:databaseQuery_Ano.php?err=3&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">		// 设置要检索的数据库表格名</span><br><span class="line">		$table=&quot;&quot;;</span><br><span class="line">		if($specie==&quot;human&quot;)&#123;</span><br><span class="line">			$table=&quot;lincRNA_h&quot;;</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			$table=&quot;lincRNA_m&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">		</span><br><span class="line">		</span><br><span class="line">		// 创建SQL查询语句</span><br><span class="line">		$sql=&quot;SELECT * from $table where&quot;;</span><br><span class="line">		// 根据实际表单填写情况追加查询条件</span><br><span class="line">		$bool_and=false;</span><br><span class="line">		if(!empty($geneName))&#123;</span><br><span class="line">			$sql .= &quot; genename=&apos;$geneName&apos; or geneid=&apos;$geneName&apos;&quot;;</span><br><span class="line">			$bool_and=true;</span><br><span class="line">		&#125;</span><br><span class="line">		if(!empty($chrom))&#123;</span><br><span class="line">			if($bool_and)&#123;</span><br><span class="line">				$sql .= &quot; and&quot;;</span><br><span class="line">			&#125;</span><br><span class="line">			$sql .= &quot; chrom=&apos;$chrom&apos;&quot;;</span><br><span class="line">			if(!empty($start)&amp;&amp;!empty($end))&#123;</span><br><span class="line">				// 判断提交的起始点与终止点是否正确，即是否为整数，且起始点小于终止点</span><br><span class="line">				if(is_int(intval($start))&amp;&amp;is_int(intval($end))&amp;&amp;intval($start)&lt;intval($end))&#123;</span><br><span class="line">					$sql .= &quot; and start&gt;=$start and end&lt;=$end&quot;;</span><br><span class="line">				&#125;else&#123;</span><br><span class="line">					header(&quot;Location:databaseQuery_Ano.php?err=2&quot;);</span><br><span class="line">				&#125;</span><br><span class="line">			&#125;				</span><br><span class="line">		&#125;</span><br><span class="line">		$sql .= &quot;;&quot;;</span><br><span class="line">	</span><br><span class="line">		// 执行查询语句</span><br><span class="line">		$result=$conn-&gt;query($sql);</span><br><span class="line">	</span><br><span class="line">		</span><br><span class="line">	</span><br><span class="line">		if($result-&gt;num_rows &gt; 0)&#123;</span><br><span class="line">			// 输出查询结果</span><br><span class="line">			echo &quot;&lt;table &gt;&lt;tr&gt;&lt;th&gt;chrom&lt;/th&gt;&lt;th&gt;Biotype&lt;/th&gt;&lt;th&gt;Feature&lt;/th&gt;&lt;th&gt;Start&lt;/th&gt;&lt;th&gt;End&lt;/th&gt;&lt;th&gt;GeneId&lt;/th&gt;&lt;th&gt;GeneName&lt;/th&gt;&lt;th&gt;TranscriptId&lt;/th&gt;&lt;th&gt;ExonNumber&lt;/th&gt;&lt;th&gt;操作&lt;/th&gt;&lt;/tr&gt;&quot;;</span><br><span class="line">			</span><br><span class="line">			while($row = $result-&gt;fetch_array())&#123;</span><br><span class="line">				echo &quot;&lt;tr&gt;&lt;td&gt;&quot;.$row[&apos;chrom&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;biotype&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;feature&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;start&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;end&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;geneid&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;genename&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;transcriptid&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;exon&apos;].&quot;&lt;/td&gt;&lt;td&gt;&lt;a style=\&quot;padding:2px;\&quot; href=\&quot;update_form.php?id=&quot;.$row[&apos;id&apos;].&quot;&amp;dbname=&quot;.$table.&quot;\&quot;&gt;Update&lt;/a&gt;&lt;/tr&gt;&quot;;</span><br><span class="line">    		&#125;</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">    		echo &quot;未检索到满足条件的记录!&quot;;</span><br><span class="line">    	&#125;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		// 跳转到当前页面，并为err赋值1</span><br><span class="line">		header(&quot;Location:databaseQuery_Ano.php?err=3&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h1 id="Part2：内容与功能升级"><a href="#Part2：内容与功能升级" class="headerlink" title="Part2：内容与功能升级"></a>Part2：内容与功能升级</h1><h2 id="实现目标-1"><a href="#实现目标-1" class="headerlink" title="实现目标"></a>实现目标</h2><ul>
<li>增加 blast 功能：将用户提交的序列与 lncRNA 的序列库进行 blast 比对</li>
<li>创建 lncRNA 表达数据库，除了 part1 中提供的基础检索功能外，还要提供绘图功能</li>
<li>增加数据库的删除、追加（单条记录追加和批量追加）与编辑功能</li>
</ul>
<h2 id="思路-1"><a href="#思路-1" class="headerlink" title="思路"></a>思路</h2><p><strong>1. 增加 blast 功能</strong></p>
<p>目标序列的指定方式：</p>
<ul>
<li>用户在输入框中输入</li>
<li>用户提交目标序列的 fasta 文件</li>
</ul>
<p>创建lncRNA的 blast 序列库（即lncRNA的 blast 索引)：</p>
<ul>
<li>根据 part1 <a href="#prepare-data">《准备需要写入数据库中的数据》</a> 中得到的文件<code>lincRNA_GRCm38.91.gtf.format</code>和<code>lincRNA_GRCh38.91.gtf.format</code>，从中提取unique的lincRNA的Ensembl gene id</li>
<li>利用lincRNA的Ensembl gene id，从Ensembl的BioMart中下载lincRNA对应的序列</li>
<li>对这些序列建立blast索引</li>
</ul>
<p><strong>2. 创建 lncRNA 表达数据库</strong></p>
<p>从公共数据库（如GEO）中直接下载RNA-seq表达谱文件，或者下载原始fastq文件，自己跑RNA-seq分析流程来获取表达谱，然后提取其中的lincRNA部分，进行差异表达分析后将差异表达结果写进MySQL数据库中，同时保留表达谱文件，以供后期绘图时访问</p>
<ul>
<li><p>表达数据来源：从GEO数据库中下载某个RNA-seq实验的表达谱文件</p>
</li>
<li><p>差异表达分析：使用DESeq2进行差异表达分析，DESeq2使用方法请点 <a href="https://github.com/Ming-Lian/NGS-analysis/blob/master/RNA-seq.md#deseq2" target="_blank" rel="noopener">这里</a></p>
</li>
<li><p>将差异表达结果写进MySQL数据库中：过程与 part1 <a href="#use-mysqli">《将数据写入数据库中——方法一：使用MySQLi》</a>一样</p>
</li>
<li><p>绘图：即在检索获得差异表达结果列表后，每条记录后提供一个绘图按钮，用户点击后即可绘制出该基因在不同samples的表达柱状图，可通过调用R脚本实现</p>
</li>
</ul>
<p><strong>3. 增加数据库的删除、追加与编辑功能</strong></p>
<p>这部分应该是整个part2中最容易实现的部分</p>
<ul>
<li><p>删除 —— <strong>注意：在真正删除之前，请要求用户进行再次确认</strong></p>
<ul>
<li>删除一条记录：用SQL的DELETE语句</li>
<li>删除多条记录：在每条记录前有一个复选框，选中多条记录前的复选框，然后点击页面某个角落的“Delete”按钮，提交删除操作请求</li>
</ul>
</li>
<li><p>追加</p>
<ul>
<li>追加一条记录</li>
<li>批量追加：可以让用户提交文件来实现批量追加，不过真正进行追加操作之前，务必进行文件格式的检查</li>
</ul>
</li>
<li><p>编辑 —— 用UPDATE语句</p>
</li>
</ul>
<h2 id="添加blast功能"><a href="#添加blast功能" class="headerlink" title="添加blast功能"></a>添加blast功能</h2><h3 id="获得chrX上lincRNA序列"><a href="#获得chrX上lincRNA序列" class="headerlink" title="获得chrX上lincRNA序列"></a>获得chrX上lincRNA序列</h3><ul>
<li>利用Gene stable ID批量下载序列</li>
</ul>
<p>第一次写作业时，在mysql数据库中添加了lncRNA的位置信息、ID等。这次从之前建立数据得到的<code>lincRNA_GRCh38.91.gtf.format</code>和<code>lincRNA_GRCm38.91.gtf.format</code>文件中提取<code>Gene stable ID</code>，批量下载FASTA格式的序列信息。</p>
<ul>
<li>得到lnRNA的ID</li>
</ul>
<p>先以<code>lincRNA_GRCh38.91.gtf.format</code>（人类）为例</p>
<p>1. 通过<code>less lincRNA_GRCh38.91.gtf.format</code>查看文件</p>
<p>实例如下，第6列的就是基因ID（ENSG00000243485开始）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">1       lincRNA gene    29554   31109   ENSG00000243485 MIR1302-2HG     -       -</span><br><span class="line">1       lincRNA transcript      29554   31097   ENSG00000243485 MIR1302-2HG     ENST00000473358 -</span><br><span class="line">1       lincRNA exon    29554   30039   ENSG00000243485 MIR1302-2HG     ENST00000473358 1</span><br><span class="line">1       lincRNA exon    30564   30667   ENSG00000243485 MIR1302-2HG     ENST00000473358 2</span><br><span class="line">1       lincRNA exon    30976   31097   ENSG00000243485 MIR1302-2HG     ENST00000473358 3</span><br><span class="line">1       lincRNA transcript      30267   31109   ENSG00000243485 MIR1302-2HG     ENST00000469289 -</span><br><span class="line">1       lincRNA exon    30267   30667   ENSG00000243485 MIR1302-2HG     ENST00000469289 1</span><br><span class="line">1       lincRNA exon    30976   31109   ENSG00000243485 MIR1302-2HG     ENST00000469289 2</span><br><span class="line">1       lincRNA gene    34554   36081   ENSG00000237613 FAM138A -       -</span><br><span class="line">1       lincRNA transcript      34554   36081   ENSG00000237613 FAM138A ENST00000417324 -</span><br><span class="line">1       lincRNA exon    35721   36081   ENSG00000237613 FAM138A ENST00000417324 1</span><br><span class="line">1       lincRNA exon    35277   35481   ENSG00000237613 FAM138A ENST00000417324 2</span><br><span class="line">1       lincRNA exon    34554   35174   ENSG00000237613 FAM138A ENST00000417324 3</span><br><span class="line">1       lincRNA transcript      35245   36073   ENSG00000237613 FAM138A ENST00000461467 -</span><br><span class="line">1       lincRNA exon    35721   36073   ENSG00000237613 FAM138A ENST00000461467 1</span><br><span class="line">1       lincRNA exon    35245   35481   ENSG00000237613 FAM138A ENST00000461467 2</span><br><span class="line">1       lincRNA gene    89295   133723  ENSG00000238009 AL627309.1      -       -</span><br><span class="line">1       lincRNA transcript      89295   120932  ENSG00000238009 AL627309.1      ENST00000466430 -</span><br><span class="line">1       lincRNA exon    120775  120932  ENSG00000238009 AL627309.1      ENST00000466430 1</span><br><span class="line">1       lincRNA exon    112700  112804  ENSG00000238009 AL627309.1      ENST00000466430 2</span><br><span class="line">1       lincRNA exon    92091   92240   ENSG00000238009 AL627309.1      ENST00000466430 3</span><br><span class="line">1       lincRNA exon    89295   91629   ENSG00000238009 AL627309.1      ENST00000466430 4</span><br></pre></td></tr></table></figure>
<p>2. 用<code>cut -f 6  lincRNA_GRCh38.91.gtf.format|sort|uniq &gt; lincRNA_GRCh38.91.geneid.txt</code>命令截取第6列ID信息，并输出到out文件中</p>
<p>3. 打开ensembl的<a href="http://asia.ensembl.org/biomart/martview/274417045864fe1b2b7636ae8c6c67d8" target="_blank" rel="noopener">biomart</a>，按照图片调整参数，最后上传ID信息文件即可。</p>
<p><strong>（由于全基因组数据太大，所以我在REGION中只选择了x染色体）</strong></p>
<p><img src="/images/InAction-PHP-MySQL-part2-getdata-from-BioMart-1.png" alt></p>
<p><img src="/images/InAction-PHP-MySQL-part2-getdata-from-BioMart-2.png" alt></p>
<p><img src="/images/InAction-PHP-MySQL-part2-getdata-from-BioMart-3.png" alt></p>
<p><img src="/images/InAction-PHP-MySQL-part2-getdata-from-BioMart-4.png" alt></p>
<p>4. 点击<code>results</code>,跳转页面点GO生成FASTA序列。</p>
<p>5. 小鼠数据如法炮制，只是在第一张照片处把Human参数改成Mouse。</p>
<h3 id="构建BLAST索引"><a href="#构建BLAST索引" class="headerlink" title="构建BLAST索引"></a>构建BLAST索引</h3><ul>
<li>删除课堂教学用的数据</li>
</ul>
<p>每人配额只有2G，所以清空一下磁盘，方便做作业</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">#查看文件大小</span><br><span class="line">du -h</span><br><span class="line">#删除文件夹</span><br><span class="line">rm -rf filename</span><br><span class="line">-r ：循环删掉文件夹中的文件</span><br><span class="line">-f ：删除文件时不提示</span><br></pre></td></tr></table></figure>
<ul>
<li>构建BLAST索引</li>
</ul>
<p>在biomart下载完FASTA格式序列后，会得到<code>mart_export.txt</code>文件，修改文件名为<code>lincRNA_chrX.fa</code>将其导入服务器。</p>
<p>调用BLAST命令构建索引</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -in /Path/To/lincRNA_chrX.fa -input_type fasta -dbtype nucl -title lncRNA_h -parse_seqids -out /Path/To/lncRNA_h -logfile File_h</span><br></pre></td></tr></table></figure>
<p>参数介绍</p>
<ul>
<li>-in:FASTA文件位置</li>
<li>-input_type:输入格式</li>
<li>-dbtype:数据库格式</li>
<li>-title:索引名称</li>
<li>-parse_seqids:不知道啥意思</li>
<li>-out:输出文件名（title感觉没啥用，最后blast调用索引都是调用输出文件）</li>
<li>-logfile:软件日志</li>
</ul>
<p>命令完成，打开软件日志File_h查看：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Building a new DB, current time: 05/26/2018 15:17:13</span><br><span class="line">New DB name:   /home/201728016715029/blastdb/lncRNA_h</span><br><span class="line">New DB title:  lncRNA_h</span><br><span class="line">Sequence type: Nucleotide</span><br><span class="line">Keep MBits: T</span><br><span class="line">Maximum file size: 1000000000B</span><br><span class="line">Adding sequences from FASTA; added 246 sequences in 0.00809789 seconds.</span><br></pre></td></tr></table></figure>
<p>成功！</p>
<h3 id="执行BLAST"><a href="#执行BLAST" class="headerlink" title="执行BLAST"></a>执行BLAST</h3><ul>
<li>提交BLAST检索申请，脚本保存为<code>blastQuery.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line">&lt;form action=&quot;localblast.php&quot; method=&quot;post&quot; enctype=&quot;multipart/form-data&quot;&gt;</span><br><span class="line">&lt;p&gt;输入你的序列&lt;/p&gt;</span><br><span class="line">&lt;textarea style=&quot;width:500px;height:200px;&quot; name=&quot;seq&quot;&gt;&gt;ENSG00000226530|ENST00000418775</span><br><span class="line">CACGCTCCTCCAGGGCGCAGAGAGCATCAAAGCATTATCACCGACCCAACACACACCGGC</span><br><span class="line">GCTGCGTAGTTTCTGCAGCAGGTGTGGGGTGAAGAGTTGCCACACAGCTGTCTGACACCG</span><br><span class="line">GGTGTACGGCGTCACCCCTCACTTCCTGGACAAGGCAACATTTTCCTAAAGGTCAAAGGA</span><br><span class="line">GAAAAATCCTATCATCTGAGGTGCTGAACAAAGAATTCAATAAATTCAATCAATCTGATG</span><br><span class="line">CACTCTGCAGTCGTTCTAAGTTCTTGTCTACTAAAATCCTCGACTACTACACATCCACAA</span><br><span class="line">TCGCTTCTGTGAAACTCTTGGGACCAAATGGTTCAGAATTTGTCTCCCATACATACATAA</span><br><span class="line">TATCTATGTCAAGGCCTGTAGTAATACTGTGGAATCTAATACAATGATATTTCTGTAGTG</span><br><span class="line">AAACATACAAATGTTCACTCCAAGATCTAAAGATTACAAATAGCATCATGTCAGTTCAAA</span><br><span class="line">TCAGGTTTTATGGATAAATGAGTCAGGAAAAGTATGTTTTTCAAAGCATTTTGAATGTTA</span><br><span class="line">GACTTGTGCACAAGGGATTGTGAGTCTGAATCATATTTCCGTCAAC&lt;/textarea&gt;</span><br><span class="line">&lt;table&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;提交你的序列文件&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;</span><br><span class="line">		&lt;label for=&quot;file&quot;&gt;文件名：&lt;/label&gt;</span><br><span class="line">		&lt;input type=&quot;file&quot; style=&quot;width:800px;&quot; name=&quot;file&quot; id=&quot;file&quot;&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;submit&quot; value=&quot;Submit&quot;&gt;&lt;input type=&quot;submit&quot; value=&quot;Submit&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;!-- 提示信息 --&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td style=&quot;color:red;font-size:10px;&quot;&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">$err=isset($_GET[&quot;err&quot;])?$_GET[&quot;err&quot;]:&quot;&quot;;</span><br><span class="line">switch($err) &#123;</span><br><span class="line">	case 1:</span><br><span class="line">		echo &quot;请在输入序列或提交序列文件！&quot;;</span><br><span class="line">		break;</span><br><span class="line">	case 2:</span><br><span class="line">		echo &quot;无法提交输入序列！&quot;;</span><br><span class="line">		break;</span><br><span class="line">	case 3:</span><br><span class="line">		echo &quot;未登陆，无权操作！请&lt;a href=\&quot;login.php\&quot;&gt;登陆&lt;/a&gt;&quot;;</span><br><span class="line">		break;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;/table&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li>执行BLAST操作，保存为<code>localhost.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 通过POST方法获取表单提交信息</span><br><span class="line">$seq=isset($_POST[&apos;seq&apos;])?$_POST[&apos;seq&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 若既未输入序列也未提交序列文件，则返回BLAST申请界面，并显示错误信息</span><br><span class="line">if ($_FILES[&quot;file&quot;][&quot;error&quot;] &gt; 0 &amp;&amp; empty($seq))&#123;</span><br><span class="line">	header(&quot;Location:blastQuery.php?err=1&quot;);</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">// 若输入了序列，则将序列写入upload文件夹下的文本中</span><br><span class="line">$seqfile=&quot;upload/query.fa&quot;;</span><br><span class="line">if(!empty($seq))&#123;</span><br><span class="line">	if(!($file=fopen($seqfile, &quot;w&quot;)))&#123;</span><br><span class="line">        header(&quot;Location:blastQuery.php?err=2&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">	fwrite($file,$seq);</span><br><span class="line">	fclose($file);</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">if(!empty($username)&amp;&amp;!empty($password))&#123;</span><br><span class="line">	// 连接数据库</span><br><span class="line">	$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">	if($conn-&gt;connect_error)&#123;</span><br><span class="line">		header(&quot;Location:blastQuery.php?err=3&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	// 执行BLAST，优先用输入序列进行BLAST</span><br><span class="line">	if(!empty($seq))&#123;</span><br><span class="line">		system(&apos;/home/Ben/software/blast/ncbi-blast-2.6.0+/bin/blastn -query upload/query.fa -db blastdb/lncRNA_h -html -evalue 1  -num_descriptions 10 -num_alignments 10 -out blast.out&apos;);</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		system(&apos;/home/Ben/software/blast/ncbi-blast-2.6.0+/bin/blastn -query upload/&apos;.$_FILES[&quot;file&quot;][&quot;name&quot;].&apos;  -db blastdb/lncRNA_h -html -evalue 1  -num_descriptions 10 -num_alignments 10 -out blast.out&apos;);</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	// 打印BLAST结果</span><br><span class="line">	$file_out=fopen(&quot;blast.out&quot;,&quot;r&quot;) or exit(&quot;Unable to open file!&quot;);</span><br><span class="line">	while(!feof($file_out))&#123;</span><br><span class="line">		echo fgets($file_out);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	header(&quot;Location:blastQuery.php?err=3&quot;);</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h2 id="添加lincRNA表达谱"><a href="#添加lincRNA表达谱" class="headerlink" title="添加lincRNA表达谱"></a>添加lincRNA表达谱</h2><h3 id="获取在两个样本组的表达谱"><a href="#获取在两个样本组的表达谱" class="headerlink" title="获取在两个样本组的表达谱"></a>获取在两个样本组的表达谱</h3><h4 id="数据集简单介绍与下载"><a href="#数据集简单介绍与下载" class="headerlink" title="数据集简单介绍与下载"></a>数据集简单介绍与下载</h4><p>该笔记中使用的数据集为 <a href="#nat-pro">Nature Protocols 文章</a>中用到的数据集</p>
<table>
<thead>
<tr>
<th style="text-align:left">sample id</th>
<th style="text-align:left">Sex</th>
<th style="text-align:left">population</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">ERR188245</td>
<td style="text-align:left">Female</td>
<td style="text-align:left">GBR</td>
</tr>
<tr>
<td style="text-align:left">ERR188428</td>
<td style="text-align:left">Female</td>
<td style="text-align:left">GBR</td>
</tr>
<tr>
<td style="text-align:left">ERR188337</td>
<td style="text-align:left">Female</td>
<td style="text-align:left">GBR</td>
</tr>
<tr>
<td style="text-align:left">ERR188401</td>
<td style="text-align:left">Male</td>
<td style="text-align:left">GBR</td>
</tr>
<tr>
<td style="text-align:left">ERR188257</td>
<td style="text-align:left">Male</td>
<td style="text-align:left">GBR</td>
</tr>
<tr>
<td style="text-align:left">ERR188383</td>
<td style="text-align:left">Male</td>
<td style="text-align:left">GBR</td>
</tr>
<tr>
<td style="text-align:left">ERR204916</td>
<td style="text-align:left">Female</td>
<td style="text-align:left">YRI</td>
</tr>
<tr>
<td style="text-align:left">ERR188234</td>
<td style="text-align:left">Female</td>
<td style="text-align:left">YRI</td>
</tr>
<tr>
<td style="text-align:left">ERR188273</td>
<td style="text-align:left">Female</td>
<td style="text-align:left">YRI</td>
</tr>
<tr>
<td style="text-align:left">ERR188454</td>
<td style="text-align:left">Male</td>
<td style="text-align:left">YRI</td>
</tr>
<tr>
<td style="text-align:left">ERR188104</td>
<td style="text-align:left">Male</td>
<td style="text-align:left">YRI</td>
</tr>
<tr>
<td style="text-align:left">ERR188044</td>
<td style="text-align:left">Male</td>
<td style="text-align:left">YRI</td>
</tr>
</tbody>
</table>
<p>文章作者已经将数据集打包好了，下载链接：<code>ftp://ftp.ccb.jhu.edu/pub/RNAseq_protocol/chrX_data.tar.gz</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 下载</span><br><span class="line">$ nohup wget -c ftp://ftp.ccb.jhu.edu/pub/RNAseq_protocol/chrX_data.tar.gz &gt;download.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"># 解压</span><br><span class="line">$ tar zxvf chrX_data.tar.gz</span><br></pre></td></tr></table></figure>
<p>解压后可以看到数据集的组成<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">|---chrX_data</span><br><span class="line">	|---genes</span><br><span class="line">		|---chrX.gtf</span><br><span class="line">	|---genome</span><br><span class="line">		|---chrX.fa</span><br><span class="line">	|---indexes</span><br><span class="line">		|---chrX_tran.[1-7].ht2</span><br><span class="line">	|---samples</span><br><span class="line">		|---ERR*_chrX_[12].fastq.gz</span><br></pre></td></tr></table></figure></p>
<h4 id="跑RNA-seq分析流程"><a href="#跑RNA-seq分析流程" class="headerlink" title="跑RNA-seq分析流程"></a>跑RNA-seq分析流程</h4><p>RNA-seq分析流程，请参考 <a href="https://github.com/Ming-Lian/NGS-analysis/blob/master/RNA-seq.md" target="_blank" rel="noopener">这里</a></p>
<p>一般第一步是创建hisat2索引，由于数据集中已经提供了（在<code>chrX_data/indexes</code>文件夹下），所以跳过这一步</p>
<ul>
<li><strong>hisat2比对</strong></li>
</ul>
<p>在比对前，先准备好一个保存样本Id的文本<code>samplelist.txt</code>：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">ERR188245</span><br><span class="line">ERR188428</span><br><span class="line">ERR188337</span><br><span class="line">ERR188401</span><br><span class="line">ERR188257</span><br><span class="line">ERR188383</span><br><span class="line">ERR204916</span><br><span class="line">ERR188234</span><br><span class="line">ERR188273</span><br><span class="line">ERR188454</span><br><span class="line">ERR188104</span><br><span class="line">ERR188044</span><br></pre></td></tr></table></figure>
<p>开始执行批量比对</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ mkdir map</span><br><span class="line"></span><br><span class="line">$ nohup cat samplelist.txt | while read i;</span><br><span class="line">do</span><br><span class="line">	hisat2 -p 10 --dta -x indexes/chrX_tran -1 samples/$&#123;i&#125;_chrX_1.fastq.gz -2 samples/$&#123;i&#125;_chrX_2.fastq.gz | \</span><br><span class="line">	samtools sort -@ 8 -O bam -o map/$&#123;i&#125;_chrX.sort.bam 1&gt;map/$&#123;i&#125;_map.log 2&gt;&amp;1</span><br><span class="line">done &amp;</span><br></pre></td></tr></table></figure>
<ul>
<li><strong>stringtie转录本拼接</strong></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">$ mkdir asm</span><br><span class="line"></span><br><span class="line"># 拼接</span><br><span class="line">$ cat cat samplelist.txt | while read i;</span><br><span class="line">do</span><br><span class="line">	stringtie -p 16 -G genes/chrX.gtf -o asm/$&#123;i&#125;_chrX.gtf -l $i map/$&#123;i&#125;_chrX.sort.bam 1&gt;asm/$&#123;i&#125;_strg_assm.log 2&gt;&amp;1</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 准备mergelist.txt文件</span><br><span class="line">$ awk &apos;&#123;print &quot;asm/&quot;$0&quot;_chrX.gtf&quot;&#125;&apos; samplelist.txt &gt;mergelist.txt</span><br><span class="line"></span><br><span class="line"># merge多样本的转录本拼接结果</span><br><span class="line">$ stringtie --merge -p 16 -G genes/chrX.gtf -o asm/merge.gtf mergelist.txt 1&gt;asm/strg_merge.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<ul>
<li><strong>stringtie定量</strong></li>
</ul>
<p>以read count进行定量，作为DESeq2或edgeR的输入</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">$ mkdir quant</span><br><span class="line"></span><br><span class="line"># 转录本定量</span><br><span class="line">$ cat samplelist.txt | while read i;</span><br><span class="line">do</span><br><span class="line">	stringtie -e -p 16 -G asm/merge.gtf -o quant/$&#123;i&#125;_chrX_quant.gtf map/$&#123;i&#125;_chrX.sort.bam 1&gt;quant/$&#123;i&#125;_chrX_quant.log 2&gt;&amp;1</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 整合多样本的转录本定量结果</span><br><span class="line">## 先准备sample_lst.txt文件，格式如下：</span><br><span class="line">		ERR188021 &lt;PATH_TO_ERR188021.gtf&gt;</span><br><span class="line">		ERR188023 &lt;PATH_TO_ERR188023.gtf&gt;</span><br><span class="line">		...</span><br><span class="line">$ perl -ne &apos;chomp;print &quot;$_\tquant/$&#123;_&#125;_chrX_quant.gtf\n&quot;&apos; samplelist.txt &gt;samplelist.quant.txt</span><br><span class="line"></span><br><span class="line">## 执行定量结果整合</span><br><span class="line">$ python2 /Path/To/prepDE.py -i sample_lst.txt</span><br></pre></td></tr></table></figure>
<p>最后会在工作目录下产生两个read count定量文件：</p>
<blockquote>
<ul>
<li>gene定量：gene_count_matrix.csv</li>
<li>transcript定量：transcript_count_matrix.csv</li>
</ul>
</blockquote>
<p>想下载这两个文件，请点 <a href="./txt-supply/">这里</a></p>
<ul>
<li><strong>差异表达分析：DESeq2</strong></li>
</ul>
<p>DESeq2要求输入的表达矩阵是read counts</p>
<p>构建 DESeqDataSet 对象</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># 构建表达矩阵</span><br><span class="line">count_table&lt;-read.csv(&quot;gene_count_matrix.csv&quot;,row.names=&quot;gene_id&quot;)</span><br><span class="line">count_matrix&lt;-as.matrix(count_table)</span><br><span class="line"># 构建分组矩阵</span><br><span class="line">phenodata&lt;-read.csv(&quot;geuvadis_phenodata.csv&quot;,row.names=&quot;ids&quot;)</span><br><span class="line"># 构建 DESeqDataSet 对象</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData = count_matrix,</span><br><span class="line">	colData = phenodata,</span><br><span class="line">	design = ~ population)</span><br></pre></td></tr></table></figure>
<p>差异表达分析</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">dds &lt;- DESeq(dds)</span><br><span class="line">res &lt;- results(dds)</span><br><span class="line">resOrdered &lt;- res[order(res$padj),]</span><br><span class="line">diffResult &lt;- as.data.frame(resOrdered)</span><br><span class="line">write.table(diffResult,&quot;gene_diffResult.txt&quot;,sep=&quot;\t&quot;,quote=F)</span><br></pre></td></tr></table></figure>
<p>这一步最终得到差异表达分析结果文件<code>gene_diffResult.txt</code>，想下载这两个文件，请点 <a href="./txt-supply/">这里</a></p>
<h3 id="将差异表达结果写进MySQL"><a href="#将差异表达结果写进MySQL" class="headerlink" title="将差异表达结果写进MySQL"></a>将差异表达结果写进MySQL</h3><p>过程与 part1 <a href="#use-mysqli">方法一：使用MySQLi</a> 相同</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># 创建表格DiffExp_result</span><br><span class="line"></span><br><span class="line">mysql&gt; CREATE TABLE DiffExp_result (</span><br><span class="line">  `id` int(11) NOT NULL AUTO_INCREMENT,</span><br><span class="line">  `RefSeq` varchar(255) NOT NULL DEFAULT &apos;&apos; COMMENT &apos;RefSeq id&apos;,</span><br><span class="line">  `baseMean` DOUBLE(40,20) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;the base mean over all rows&apos;,</span><br><span class="line">  `log2FoldChange` DOUBLE(40,20) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;log2 fold change (MAP): treatment OHT vs Control&apos;,</span><br><span class="line">  `lfcSE` DOUBLE(40,20) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;standard error: treatment OHT vs Control&apos;,</span><br><span class="line">  `stat` DOUBLE(40,20) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;Wald statistic: treatment OHT vs Control&apos;,</span><br><span class="line">  `pvalue` FLOAT(20) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;Wald test p-value: treatment OHT vs Control&apos;,</span><br><span class="line">  `padj` FLOAT(20) NOT NULL DEFAULT &apos;0&apos; COMMENT &apos;BH adjusted p-values&apos;,</span><br><span class="line">  PRIMARY KEY (`id`) </span><br><span class="line">) ;</span><br></pre></td></tr></table></figure>
<table>
<thead>
<tr>
<th style="text-align:left">数据类型</th>
<th style="text-align:left">描述</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">TINYINT(size)</td>
<td style="text-align:left">-128 到 127 常规。0 到 255 无符号*。在括号中规定最大位数。</td>
</tr>
<tr>
<td style="text-align:left">SMALLINT(size)</td>
<td style="text-align:left">-32768 到 32767 常规。0 到 65535 无符号*。在括号中规定最大位数。</td>
</tr>
<tr>
<td style="text-align:left">MEDIUMINT(size)</td>
<td style="text-align:left">-8388608 到 8388607 普通。0 to 16777215 无符号*。在括号中规定最大位数。</td>
</tr>
<tr>
<td style="text-align:left">INT(size)</td>
<td style="text-align:left">-2147483648 到 2147483647 常规。0 到 4294967295 无符号*。在括号中规定最大位数。</td>
</tr>
<tr>
<td style="text-align:left">BIGINT(size)</td>
<td style="text-align:left">-9223372036854775808 到 9223372036854775807 常规。0 到 18446744073709551615 无符号*。在括号中规定最大位数。</td>
</tr>
<tr>
<td style="text-align:left">FLOAT(size,d)</td>
<td style="text-align:left">带有浮动小数点的小数字。在括号中规定最大位数。在 d 参数中规定小数点右侧的最大位数。</td>
</tr>
<tr>
<td style="text-align:left">DOUBLE(size,d)</td>
<td style="text-align:left">带有浮动小数点的大数字。在括号中规定最大位数。在 d 参数中规定小数点右侧的最大位数。</td>
</tr>
<tr>
<td style="text-align:left">DECIMAL(size,d)</td>
<td style="text-align:left">作为字符串存储的 DOUBLE 类型，允许固定的小数点。</td>
</tr>
</tbody>
</table>
<p>基于 part1 中的php脚本<code>data_batch_import.php</code>稍作修改即可复用</p>
<p>需要改动的位置如下图：</p>
<p><img src="/images/InAction-PHP-MySQL-part2-modify-phpscript-location.jpg" alt></p>
<p>直接手动修改PHP脚本，保存为<code>lincRNA_DiffExp_data_batch_import.php</code></p>
<p>以下只给出修改部分的代码</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">// 预处理及绑定</span><br><span class="line">$stmt=$conn-&gt;prepare(&quot;INSERT INTO DiffExp_result (RefSeq,baseMean,log2FoldChange,lfcSE,stat,pvalue,padj) VALUES (?,?,?,?,?,?,?)&quot;);</span><br><span class="line">$stmt-&gt;bind_param(&quot;sdddddd&quot;,$RefSeq,$baseMean,$log2FoldChange,$lfcSE,$stat,$pvalue,$padj);</span><br><span class="line"></span><br><span class="line">// 打开文件，通过$argv[1]进行脚本传参</span><br><span class="line">$file=fopen(&quot;$argv[1]&quot;,&apos;r&apos;) or exit(&quot;Unable to open file!&quot;);</span><br><span class="line"></span><br><span class="line">// 逐行读取文件，并写入数据库</span><br><span class="line">fgets($file);	// 跳过第一行表头</span><br><span class="line">while(!feof($file))&#123;</span><br><span class="line">	$data=explode(&quot;\t&quot;,fgets($file)); // 以tab为间隔标识将字符串打散</span><br><span class="line">	</span><br><span class="line">	// 设置参数</span><br><span class="line">	$RefSeq=$data[0];</span><br><span class="line">	$baseMean=$data[1];</span><br><span class="line">	$log2FoldChange=$data[2];</span><br><span class="line">	$lfcSE=$data[3];</span><br><span class="line">	$stat=$data[4];</span><br><span class="line">	$pvalue=$data[5];</span><br><span class="line">	$padj=$data[6];</span><br><span class="line">	// 执行</span><br><span class="line">	$stmt-&gt;execute();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>最后执行php脚本，完成数据的批量导入</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ php -f lincRNA_DiffExp_data_batch_import.php &lt;data-to-import&gt;</span><br></pre></td></tr></table></figure>
<h3 id="检索表达谱数据库"><a href="#检索表达谱数据库" class="headerlink" title="检索表达谱数据库"></a>检索表达谱数据库</h3><p>过程类似于 part1 <a href="#query">检索数据库</a></p>
<p>表单部分，提供以下检索筛选条件</p>
<blockquote>
<ul>
<li>log2FoldChange 一般认为其绝对值大于1，差异显著</li>
<li>p-value 一般认为其值小于等于0.05，差异显著</li>
<li>q-value 一般认为其值小于等于0.05，差异显著</li>
</ul>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">&lt;form action=&quot;#&quot; method=&quot;post&quot;&gt;</span><br><span class="line">&lt;table style=&quot;text-align:center; border:0;&quot;&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;|log2FoldChange| &gt;=&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; name=&quot;log2FoldChange&quot; value=&quot;1&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;p-value &lt;=&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; name=&quot;p_value&quot; value=&quot;0.05&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;q-value &lt;=&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;&lt;input type=&quot;text&quot; name=&quot;q_value&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td align=&quot;center&quot;&gt;&lt;input type=&quot;submit&quot; name=&quot;submit&quot; value=&quot;Submit&quot;&gt;&lt;/td&gt;</span><br><span class="line">	&lt;td align=&quot;center&quot;&gt;&lt;input type=&quot;reset&quot; value=&quot;Reset&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;!-- 错误提示信息 --&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td style=&quot;color:red; font-size:10px;&quot;&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">$err=isset($_GET[&quot;err&quot;])?$_GET[&quot;err&quot;]:&quot;&quot;;</span><br><span class="line">switch($err) &#123;</span><br><span class="line">case 1:</span><br><span class="line">	echo &quot;表单填写错误：至少要填写一项！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 2:</span><br><span class="line">	echo &quot;表单填写错误：填写信息必须是数字！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 3:</span><br><span class="line">	echo &quot;未登陆，无权访问！&quot;;</span><br><span class="line">	echo &quot;快去&lt;a href=&apos;login.php&apos;&gt;登录&lt;/a&gt;吧！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 4:</span><br><span class="line">	echo &quot;Update或Delete操作异常！未在数据库中找到对应记录！&quot;;</span><br><span class="line">	break;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;	</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;/table&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<p>php脚本部分</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 获取表单提交数据</span><br><span class="line">$log2FoldChange=isset($_POST[&apos;log2FoldChange&apos;])?$_POST[&apos;log2FoldChange&apos;]:&quot;&quot;;</span><br><span class="line">$p_value=isset($_POST[&apos;p_value&apos;])?$_POST[&apos;p_value&apos;]:&quot;&quot;;</span><br><span class="line">$q_value=isset($_POST[&apos;q_value&apos;])?$_POST[&apos;q_value&apos;]:&quot;&quot;;</span><br><span class="line">$submit=isset($_POST[&apos;submit&apos;])?$_POST[&apos;submit&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">if($submit)&#123;</span><br><span class="line">	// 判断提交的检索信息是否满足要求：三个提交的变量,至少一个不为空，且若不为空则必须是数值</span><br><span class="line">	if(empty($log2FoldChange)&amp;&amp;empty($p_value)&amp;&amp;empty($q_value))&#123;</span><br><span class="line">		header(&quot;Location:databaseQuery_DiffExp.php?err=1&quot;);	</span><br><span class="line">	&#125;elseif((empty($log2FoldChange)||is_numeric($log2FoldChange))&amp;&amp;(empty($p_value)||is_numeric($p_value))&amp;&amp;(empty($q_value)||is_numeric($q_value)))&#123;</span><br><span class="line">		// 连接数据库</span><br><span class="line">		$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">		if($conn-&gt;connect_error)&#123;</span><br><span class="line">			header(&quot;Location:databaseQuery_DiffExp.php?err=3&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">		// 创建SQL查询语句</span><br><span class="line">		$sql=&quot;SELECT * from DiffExp_result where&quot;;</span><br><span class="line">		// 根据实际表单填写情况追加查询条件</span><br><span class="line">		$bool_and=false;</span><br><span class="line">		if(!empty($log2FoldChange))&#123;</span><br><span class="line">			$sql .= &quot; log2FoldChange &gt;= $log2FoldChange or log2FoldChange &lt;= -$log2FoldChange&quot;;</span><br><span class="line">			$bool_and=true;</span><br><span class="line">		&#125;</span><br><span class="line">		if(!empty($p_value))&#123;</span><br><span class="line">			if($bool_and)&#123;</span><br><span class="line">				$sql .= &quot; and&quot;;</span><br><span class="line">			&#125;</span><br><span class="line">			$sql .= &quot; pvalue &lt;= $p_value&quot;;</span><br><span class="line">			$bool_and=true;</span><br><span class="line">		&#125;</span><br><span class="line">		if(!empty($q_value))&#123;</span><br><span class="line">			if($bool_and)&#123;</span><br><span class="line">				$sql .= &quot; and&quot;;</span><br><span class="line">			&#125;</span><br><span class="line">			$sql .= &quot; padj &lt;= $q_value&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">		$sql .= &quot;;&quot;;</span><br><span class="line">		</span><br><span class="line">		// 执行查询语句</span><br><span class="line">		$result=$conn-&gt;query($sql);</span><br><span class="line">		if($result-&gt;num_rows &gt; 0)&#123;</span><br><span class="line">			// 输出查询结果</span><br><span class="line">			echo &quot;&lt;table&gt;&lt;tr&gt;&lt;th&gt;RefSeq&lt;/th&gt;&lt;th&gt;baseMean&lt;/th&gt;&lt;th&gt;log2FoldChange&lt;/th&gt;&lt;th&gt;lfcSE&lt;/th&gt;&lt;th&gt;stat&lt;/th&gt;&lt;th&gt;pvalue&lt;/th&gt;&lt;th&gt;padj&lt;/th&gt;&lt;th&gt;操作&lt;/th&gt;&lt;/tr&gt;&quot;;</span><br><span class="line">			</span><br><span class="line">			while($row = $result-&gt;fetch_array())&#123;</span><br><span class="line">        			echo &quot;&lt;tr&gt;&lt;td&gt;&quot;.$row[&apos;RefSeq&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;baseMean&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;log2FoldChange&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;lfcSE&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;stat&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;pvalue&apos;].&quot;&lt;/td&gt;&lt;td&gt;&quot;.$row[&apos;padj&apos;].&quot;&lt;/td&gt;&lt;td&gt;&lt;a href=\&quot;update_form.php?id=&quot;.$row[&apos;id&apos;].&quot;&amp;dbname=DiffExp_result\&quot; style=\&quot;padding:2px;\&quot;&gt;Update&lt;/a&gt;&lt;a href=\&quot;barplox.php?RefSeq=&quot;.$row[&apos;RefSeq&apos;].&quot;\&quot; style=\&quot;padding:2px;\&quot;&gt;Barplox&lt;/a&gt;&lt;/td&gt;&lt;/tr&gt;\n&quot;;</span><br><span class="line">    		&#125;</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">    			echo &quot;未检索到满足条件的记录!&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">	</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		header(&quot;Location:databaseQuery_DiffExp.php?err=2&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h3 id="画图：单个基因表达谱"><a href="#画图：单个基因表达谱" class="headerlink" title="画图：单个基因表达谱"></a>画图：单个基因表达谱</h3><p>实现思路：</p>
<blockquote>
<ul>
<li>用户在检索页面的记录栏的最后一项点击<code>Barplot</code>以后，向绘图脚本(<code>barplot.php</code>)传递对应记录的RefSeq</li>
<li><p>绘图脚本(<code>barplot.php</code>)根据传递过来的RefSeq，调用绘图用的底层R脚本(<code>barplot.R</code>)，调用的同时传递参数——RefSeq值</p>
<p> <code>barplot.R</code>实现的功能：读取表达谱文件（<a href="#run-rnaseq-pipeline">跑RNA-seq分析流程：stringtie定量</a> 产生的<code>gene_count_matrix.csv</code>文件），根据传递过来的RefSeq值匹配出对应的基因的表达谱，然后用ggplot把图画出来，并将图保存为图像文件</p>
</li>
<li>绘图脚本(<code>barplot.php</code>)将图片用<code>&lt;img&gt;</code>标签呈现给用户</li>
</ul>
</blockquote>
<ul>
<li>绘图脚本(<code>barplot.php</code>)</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 用GET方法获取提交的数据</span><br><span class="line">$RefSeq=isset($_GET[&apos;RefSeq&apos;])?$_GET[&apos;RefSeq&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 登录数据库，检查用户是否有操作权限</span><br><span class="line">$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">if($conn-&gt;connect_error)&#123;</span><br><span class="line">	header(&quot;Location:databaseQuery_DiffExp.php?err=3&quot;);</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	system(&apos;Rscript barplot.R &apos;.$RefSeq);</span><br><span class="line">	// 检查文件是否存在</span><br><span class="line">	if(file_exists(&apos;barplot.jpg&apos;))&#123;</span><br><span class="line">		echo &quot;&lt;p align=\&quot;center\&quot;&gt;&lt;img src=./barplot.jpg width=600 &gt;&lt;/p&gt;&quot;;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		echo &quot;&lt;p style=\&quot;color:red; text-align:center; font-size:50px;\&quot;&gt;绘图失败！&lt;/p&gt;&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li>底层R脚本(<code>barplot.R</code>)</li>
</ul>
<p>注意：</p>
<blockquote>
<p>该脚本底层依赖<code>ggplot2</code>包，若未安装请提前安装好，推荐以下两种安装方法：</p>
<ul>
<li>直接在R的交互环境下，使用<code>install.packages(&quot;ggplot2&quot;)</code></li>
<li>使用conda进行安装</li>
</ul>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line"># 获取传入参数</span><br><span class="line">args&lt;-commandArgs(T)</span><br><span class="line"># 载入全表达谱</span><br><span class="line">profile &lt;- read.csv(&quot;gene_count_matrix.csv&quot;)</span><br><span class="line">profile_matrix &lt;- as.matrix(profile[,-1])</span><br><span class="line">rownames(profile_matrix)&lt;-profile$gene_id</span><br><span class="line"># 获取目标基因的表达谱</span><br><span class="line">geneExp &lt;- data.frame(Sample=colnames(profile_matrix),Expression=profile_matrix[args[1],])</span><br><span class="line">#画图并导出</span><br><span class="line">jpeg(&quot;barplot.jpg&quot;)</span><br><span class="line">ggplot(geneExp)+geom_bar(aes(x=sample,y=expression),stat=&quot;identity&quot;)+theme(axis.text.x = element_text(angle = 60, hjust = 0.5, vjust = 0.5))</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<h2 id="添加批量提交数据功能"><a href="#添加批量提交数据功能" class="headerlink" title="添加批量提交数据功能"></a>添加批量提交数据功能</h2><p>基本的实现思路为：</p>
<blockquote>
<p>提供文件提交的输入框，同时让用户指定向哪个数据库（实际上是不同的表格）提交数据。将用户提交的文件保存为服务器某个目录下的临时文件，然后用 </p>
<ul>
<li>part1 <a href="#use-mysqli">将数据写入数据库中：方法一：使用MySQLi</a>的脚本<code>lincRNA_Ano_data_batch_import.php</code></li>
<li>或者，part2 <a href="#write-profile-data">将差异表达结果写进MySQL</a>的脚本<code>lincRNA_DiffExp_data_batch_import.php</code></li>
</ul>
<p>将临时文件的数据追加导入相应的表格中</p>
</blockquote>
<p>预备知识：</p>
<blockquote>
<ul>
<li>文件上传，请点 <a href="https://github.com/Ming-Lian/Memo/blob/master/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9APHP%E4%B8%80%E5%91%A8%E9%80%9F%E6%88%90.md#upload-file" target="_blank" rel="noopener">这里</a></li>
</ul>
</blockquote>
<p>要实现文件上传，首先要在php脚本所在的文件夹下创建一个权限全开的<code>upload</code>文件夹</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ mkdir upload</span><br><span class="line">$ chmod 777 upload</span><br></pre></td></tr></table></figure>
<p>接着开始编辑php脚本，保存为<code>batch_submit_data.php</code></p>
<p>表单部分</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br></pre></td><td class="code"><pre><span class="line">&lt;form action=&quot;batch_submit_data.php&quot; method=&quot;post&quot; enctype=&quot;multipart/form-data&quot;&gt;</span><br><span class="line">&lt;table&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;选择提交数据库：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;</span><br><span class="line">		&lt;select name=&quot;dbname&quot;&gt;</span><br><span class="line">			&lt;option value=&quot;&quot;&gt;选择数据库:&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;ano&quot;&gt;基因结构注释&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;diffexp&quot;&gt;基因表达谱&lt;/option&gt;</span><br><span class="line">		&lt;/select&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;选择物种（基因结构注释）：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;</span><br><span class="line">		&lt;select name=&quot;specie&quot;&gt;</span><br><span class="line">			&lt;option value=&quot;&quot;&gt;选择一个物种:&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;human&quot;&gt;Homo Sapiens&lt;/option&gt;</span><br><span class="line">			&lt;option value=&quot;mouse&quot;&gt;Mus musculus&lt;/option&gt;</span><br><span class="line">		&lt;/select&gt;</span><br><span class="line">	&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td&gt;选择上传数据文件：&lt;/td&gt;</span><br><span class="line">	&lt;td&gt;</span><br><span class="line">    	&lt;label for=&quot;file&quot;&gt;文件名：&lt;/label&gt;</span><br><span class="line">    	&lt;input type=&quot;file&quot; name=&quot;file&quot; id=&quot;file&quot;&gt;&lt;br&gt;</span><br><span class="line">    &lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td align=&quot;center&quot;&gt;&lt;input type=&quot;submit&quot; name=&quot;submit&quot; value=&quot;提交&quot;&gt;&lt;/td&gt;</span><br><span class="line">	&lt;td align=&quot;center&quot;&gt;&lt;input type=&quot;reset&quot; value=&quot;Reset&quot;&gt;&lt;/td&gt;</span><br><span class="line">&lt;/tr&gt;</span><br><span class="line">&lt;!-- 报错信息 --&gt;</span><br><span class="line">&lt;tr&gt;</span><br><span class="line">	&lt;td style=&quot;color:red; font-size:10px;&quot;&gt;</span><br><span class="line">&lt;?php</span><br><span class="line">$err=isset($_GET[&quot;err&quot;])?$_GET[&quot;err&quot;]:&quot;&quot;;</span><br><span class="line">switch($err) &#123;</span><br><span class="line">case 1:</span><br><span class="line">	echo &quot;表单填写错误：必须选择一个数据库！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 2:</span><br><span class="line">	echo &quot;上传文件格式不符合要求！请参照&lt;a href=\&quot;formatGuid.html\&quot;&gt;格式要求&lt;/a&gt;进行修改！&quot;;</span><br><span class="line">	break;</span><br><span class="line">case 3:</span><br><span class="line">	echo &quot;未登陆，无权访问！&quot;;</span><br><span class="line">	echo &quot;快去&lt;a href=&apos;login.php&apos;&gt;登录&lt;/a&gt;吧！&quot;;</span><br><span class="line">	break;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<p>php脚本</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 创建Ano_check函数，进行上传文本的文件格式检查</span><br><span class="line">function Ano_check($fname)</span><br><span class="line">&#123;</span><br><span class="line">	$file=fopen(&quot;upload/$fname&quot;,&quot;r&quot;) or exit(&quot;Unable to open file!&quot;);</span><br><span class="line">	$bool_mark=1;</span><br><span class="line">	while(!feof($file))&#123;</span><br><span class="line">		$data=explode(&quot;\t&quot;,fgets($file));</span><br><span class="line">		// 检查第一列，是否为1-22或XY</span><br><span class="line">		if(!preg_match(&quot;\d&#123;1,2&#125;|[XY]&quot;,$data[0]))&#123;</span><br><span class="line">			$bool_mark=0;</span><br><span class="line">			break;</span><br><span class="line">		&#125;</span><br><span class="line">		// 检查第二三列，是否都是英文字母</span><br><span class="line">		if(!preg_match(&quot;[a-zA-Z]&quot;,$data[1])||!preg_match(&quot;[a-zA-Z]&quot;,$data[2]))&#123;</span><br><span class="line">			$bool_mark=0;</span><br><span class="line">			break;</span><br><span class="line">		&#125;</span><br><span class="line">		// 检查第四五列，即start和end，是否是数字</span><br><span class="line">		if(!is_numeric($data[3])||!is_numeric($data[4]))&#123;</span><br><span class="line">			$bool_mark=0;</span><br><span class="line">			break;</span><br><span class="line">		&#125;</span><br><span class="line">		// 检查第六八列，即GeneId和TranscriptId,是否以ENSG或ENSG打头</span><br><span class="line">		if(!preg_match(&quot;^ENSG&quot;,$data[5])||!preg_match(&quot;^(ENST)|-&quot;,$data[5]))&#123;</span><br><span class="line">			$bool_mark=0;</span><br><span class="line">			break;</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	return $bool_mark;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">// 创建DiffExp_check函数，进行上传文本的文件格式检查</span><br><span class="line">function DiffExp_check($fname)</span><br><span class="line">&#123;</span><br><span class="line">	$file=fopen(&quot;upload/$fname&quot;,&quot;r&quot;) or exit(&quot;Unable to open file!&quot;);</span><br><span class="line">	$bool_mark=1;</span><br><span class="line">	while(!feof($file))&#123;</span><br><span class="line">		$data=explode(&quot;\t&quot;,fgets($file));</span><br><span class="line">		// 检查第一列，RefSeq，是否以大写字母打头</span><br><span class="line">		if(!preg_match(&quot;^[A-Z]&#123;2,&#125;&quot;,$data[0]))&#123;</span><br><span class="line">			$bool_mark=0;</span><br><span class="line">			break;</span><br><span class="line">		&#125;</span><br><span class="line">		// 检查其他列，是否都是数字</span><br><span class="line">		if(!is_numeric($data[1])||!is_numeric($data[2])||!is_numeric($data[3])||!is_numeric($data[4])||!is_numeric($data[5])||!is_numeric($data[6]))&#123;</span><br><span class="line">			$bool_mark=0;</span><br><span class="line">			break;</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	return $bool_mark;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">// 获取表单提交数据</span><br><span class="line">$dbname=isset($_POST[&apos;dbname&apos;])?$_POST[&apos;dbname&apos;]:&quot;&quot;;</span><br><span class="line">$specie=isset($_POST[&apos;specie&apos;])?$_POST[&apos;specie&apos;]:&quot;&quot;;</span><br><span class="line">$submit=isset($_POST[&apos;submit&apos;])?$_POST[&apos;submit&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">if($submit)&#123;</span><br><span class="line">	// 检查文件上传是否成功</span><br><span class="line">	if ($_FILES[&quot;file&quot;][&quot;error&quot;] &gt; 0)</span><br><span class="line">	&#123;</span><br><span class="line">    	echo &quot;错误：&quot; . $_FILES[&quot;file&quot;][&quot;error&quot;] . &quot;&lt;br&gt;&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	// 进行文件格式检查</span><br><span class="line">	if(empty($dbname))&#123;</span><br><span class="line">		header(&quot;Location:batch_submit_data.php?err=1&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">	elseif($dbname==&quot;ano&quot;)&#123;</span><br><span class="line">		$check_stat=Ano_check($_FILE[&apos;file&apos;][&apos;name&apos;]);</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		$check_stat=DiffExp_check($_FILE[&apos;file&apos;][&apos;name&apos;]);</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	// 判断文件格式是否符合要求</span><br><span class="line">	if(!$check_stat)&#123;</span><br><span class="line">		header(&quot;Location:batch_submit_data.php?err=2&quot;);</span><br><span class="line">	// 判断用户名和密码是否已经设置</span><br><span class="line">	&#125;elseif(!empty($username)&amp;&amp;!empty($password))&#123;</span><br><span class="line">		// 连接数据库</span><br><span class="line">		$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">		if($conn-&gt;connect_error)&#123;</span><br><span class="line">			header(&quot;batch_submit_data.php?err=3&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">		</span><br><span class="line">		// 设置数据库表格名，并执行批量操作</span><br><span class="line">		if($dbname==&quot;ano&quot;)&#123;</span><br><span class="line">			if($specie==&quot;human&quot;)&#123;</span><br><span class="line">				$table=&quot;lincRNA_h&quot;;</span><br><span class="line">			&#125;else&#123;</span><br><span class="line">				$table=&quot;lincRNA_m&quot;;</span><br><span class="line">			&#125;</span><br><span class="line">			system(&quot;php -f lincRNA_Ano_data_batch_import.php &quot;.$table.&quot; upload/&quot;.$_FILE[&apos;file&apos;][&apos;name&apos;].&quot; &amp;&quot;);</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			system(&quot;php -f lincRNA_DiffExp_data_batch_import.php upload/&quot;.$_FILE[&apos;file&apos;][&apos;name&apos;].&quot; &amp;&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">		</span><br><span class="line">		// 调用之前的php脚本进行批量提交操作		</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		header(&quot;Location:batch_submit_data.php?err=3&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<h2 id="添加Update功能"><a href="#添加Update功能" class="headerlink" title="添加Update功能"></a>添加Update功能</h2><p>实现思路：</p>
<blockquote>
<ul>
<li>用户在检索页面的记录栏的最后一项点击<code>update</code>以后，跳转到update表单填写页面(<code>update_form.php</code>脚本)，同时向该php脚本传递对应记录的id</li>
<li>update表单填写页面(<code>update_form.php</code>脚本)根据传递过来的id，把对应记录的信息以表单形式呈现，同时表单的每个输入框可以收集用户的修改信息</li>
<li>用户修改后点击”submit”，执行update操作(<code>updata_operate_Ano.php</code>或<code>updata_operate_DiffExp.php</code>脚本)，同时将操作状态告知用户：是不是成功了？</li>
</ul>
</blockquote>
<ul>
<li>脚本<code>update_form.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 获取通过GET方法向该脚本传递的变量</span><br><span class="line">$id=isset($_GET[&apos;id&apos;])?$_GET[&apos;id&apos;]:&quot;&quot;;</span><br><span class="line">$dbname=isset($_GET[&apos;dbname&apos;])?$_GET[&apos;dbname&apos;]:&quot;&quot;;</span><br><span class="line">$statu=isset($_GET[&apos;statu&apos;])?$_GET[&apos;statu&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 检查用户是否登录</span><br><span class="line">if(!empty($username)&amp;&amp;!empty($password))&#123;</span><br><span class="line">	// 连接数据库</span><br><span class="line">	$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">	if($conn-&gt;connect_error)&#123;</span><br><span class="line">		if($dbname==&quot;lincRNA_h&quot;||$dbname==&quot;lincRNA_m&quot;)&#123;</span><br><span class="line">			header(&quot;Location:databaseQuery_Ano.php?err=3&quot;);</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			header(&quot;Location:databaseQuery_DiffExp.php?err=3&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	// 根据id或dbname，从数据库中检索出相应的记录</span><br><span class="line">	// 构造SQL查询语句</span><br><span class="line">	$sql=&quot;SELECT * from $dbname where id=$id&quot;;</span><br><span class="line">	// 执行查询命令</span><br><span class="line">	$result = $conn-&gt;query($sql);</span><br><span class="line">	</span><br><span class="line">	// 选择对应的updata操作脚本名</span><br><span class="line">	if($dbname==&quot;lincRNA_h&quot;||$dbname==&quot;lincRNA_m&quot;)&#123;</span><br><span class="line">		$phpscript=&quot;update_operate_Ano.php&quot;;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		$phpscript=&quot;update_operate_DiffExp.php&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	// 将查询结果以表单形式呈现</span><br><span class="line">	if($result-&gt;num_rows &gt; 0)&#123;</span><br><span class="line">		echo &quot;&lt;form action=\&quot;&quot;.$phpscript.&quot;?id=&quot;.$id.&quot;&amp;dbname=&quot;.$dbname.&quot;\&quot; method=\&quot;post\&quot;&gt;\n&quot;;</span><br><span class="line">		echo &quot;&lt;table&gt;\n&lt;tr&gt;\n\t&lt;th&gt;Feature&lt;/th&gt;&lt;th&gt;Value&lt;/th&gt;\n&lt;/tr&gt;\n&quot;;</span><br><span class="line">		$row = $result-&gt;fetch_array();</span><br><span class="line">		// 遍历关联数组的每一个元素，并将它们填到表单里</span><br><span class="line">		foreach($row as $key=&gt;$key_value)&#123;</span><br><span class="line">			if(is_numeric($key)||$key==&quot;id&quot;)&#123;</span><br><span class="line">				continue;</span><br><span class="line">			&#125;</span><br><span class="line">			echo &quot;&lt;tr&gt;\n&quot;;</span><br><span class="line">			echo &quot;\t&lt;td&gt;&quot;.$key.&quot;&lt;/td&gt;\n&quot;;</span><br><span class="line">			echo &quot;\t&lt;td&gt;&lt;input type=\&quot;text\&quot; name=\&quot;&quot;.$key.&quot;\&quot; value=\&quot;&quot;.$key_value.&quot;\&quot;&gt;&lt;/td&gt;\n&lt;/tr&gt;&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">		// 表单的submit和reset按钮</span><br><span class="line">		echo &quot;&lt;tr&gt;\n\t&lt;td align=\&quot;center\&quot;&gt;&lt;input type=\&quot;submit\&quot; name=\&quot;submit\&quot; value=\&quot;Submit\&quot;&gt;&lt;/td&gt;\n&quot;;</span><br><span class="line">		echo &quot;\t&lt;td align=\&quot;center\&quot;&gt;&lt;input type=\&quot;reset\&quot; value=\&quot;Reset\&quot;&gt;&lt;/td&gt;\n&lt;/tr&gt;\n&quot;;</span><br><span class="line">		// 提示任务执行状态：成功或失败</span><br><span class="line">		echo &quot;&lt;tr&gt;\n\t&lt;td style=\&quot;color:red; font-size:10px;\&quot;&gt;&quot;;</span><br><span class="line">		switch($statu) &#123;</span><br><span class="line">			case 1:</span><br><span class="line">				echo &quot;该记录已修改成功！&quot;;</span><br><span class="line">			case 2:</span><br><span class="line">				echo &quot;该记录修改失败！&quot;;</span><br><span class="line">			case 3:</span><br><span class="line">				echo &quot;修改数据类型不符合格式要求，请重新填写！&quot;;</span><br><span class="line">			case 4:</span><br><span class="line">				echo &quot;未登陆，没有修改权限！请&lt;a href=\&quot;login.php\&quot;&gt;登陆&lt;/a&gt;&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">		echo &quot;&lt;/td&gt;\n&lt;/tr&gt;\n&quot;;</span><br><span class="line">		echo &quot;&lt;/table&gt;\n&lt;/form&gt;\n&quot;;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		if($dbname==&quot;lincRNA_h&quot;||$dbname==&quot;lincRNA_m&quot;)&#123;</span><br><span class="line">			header(&quot;Location:databaseQuery_Ano.php?err=4&quot;);</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			header(&quot;Location:databaseQuery_DiffExp.php?err=4&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	if($dbname==&quot;lincRNA_h&quot;||$dbname==&quot;lincRNA_m&quot;)&#123;</span><br><span class="line">		header(&quot;Location:databaseQuery_Ano.php?err=3&quot;);</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		header(&quot;Location:databaseQuery_DiffExp.php?err=3&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li>脚本<code>update_operate_Ano.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 获取通过GET方法向该脚本传递的变量</span><br><span class="line">$id=isset($_GET[&apos;id&apos;])?$_GET[&apos;id&apos;]:&quot;&quot;;</span><br><span class="line">$dbname=isset($_GET[&apos;dbname&apos;])?$_GET[&apos;dbname&apos;]:&quot;&quot;;</span><br><span class="line">// 获取通过POST方法向该脚本传递的变量</span><br><span class="line">$chrom=isset($_POST[&apos;chrom&apos;])?$_POST[&apos;chrom&apos;]:&quot;&quot;;</span><br><span class="line">$biotype=isset($_POST[&apos;biotype&apos;])?$_POST[&apos;biotype&apos;]:&quot;&quot;;</span><br><span class="line">$feature=isset($_POST[&apos;feature&apos;])?$_POST[&apos;feature&apos;]:&quot;&quot;;</span><br><span class="line">$start=isset($_POST[&apos;start&apos;])?$_POST[&apos;start&apos;]:&quot;&quot;;</span><br><span class="line">$end=isset($_POST[&apos;end&apos;])?$_POST[&apos;end&apos;]:&quot;&quot;;</span><br><span class="line">$geneid=isset($_POST[&apos;geneid&apos;])?$_POST[&apos;geneid&apos;]:&quot;&quot;;</span><br><span class="line">$genename=isset($_POST[&apos;genename&apos;])?$_POST[&apos;genename&apos;]:&quot;&quot;;</span><br><span class="line">$transcriptid=isset($_POST[&apos;transcriptid&apos;])?$_POST[&apos;transcriptid&apos;]:&quot;&quot;;</span><br><span class="line">$exon=isset($_POST[&apos;exon&apos;])?$_POST[&apos;exon&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 进行提交数据的格式检查</span><br><span class="line">$bool_mark=1;</span><br><span class="line">// 检查chrom，是否为1-22或XY</span><br><span class="line">if(!preg_match(&quot;\d&#123;1,2&#125;|[XY]&quot;,$chrom))&#123;</span><br><span class="line">	$bool_mark=0;</span><br><span class="line">&#125;</span><br><span class="line">// 检查biotype和feature，是否都是英文字母</span><br><span class="line">if(!preg_match(&quot;[a-zA-Z]&quot;,$biotype)||!preg_match(&quot;[a-zA-Z]&quot;,$feature))&#123;</span><br><span class="line">	$bool_mark=0;</span><br><span class="line">&#125;</span><br><span class="line">// 检查start、end和exon，是否是数字</span><br><span class="line">if(!is_numeric($start)||!is_numeric($end)||!is_numeric($exon))&#123;</span><br><span class="line">	$bool_mark=0;</span><br><span class="line">&#125;</span><br><span class="line">// 检查GeneId和TranscriptId,是否以ENSG或ENSG打头</span><br><span class="line">if(!preg_match(&quot;^ENSG&quot;,$geneid)||!preg_match(&quot;^(ENST)|-&quot;,$transcriptid))&#123;</span><br><span class="line">	$bool_mark=0;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">// 若用户提交的修改申请不符合格式要求，则返回update表单填写页面，并提示错误信息</span><br><span class="line">if(!$bool_mark)&#123;</span><br><span class="line">	header(&quot;Location:update_form.php?statu=3&quot;);</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	// 判断用户名和密码是否已经设置</span><br><span class="line">	if(!empty($username)&amp;&amp;!empty($password))&#123;</span><br><span class="line">		// 连接数据库</span><br><span class="line">		$conn=new mysqli(&apos;localhost&apos;,$username,$password,&apos;testdb&apos;);</span><br><span class="line">		if($conn-&gt;connect_error)&#123;</span><br><span class="line">			header(&quot;Location:update_form.php?statu=4&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">		</span><br><span class="line">		// 创建SQL的UPDATE语句</span><br><span class="line">		$sql=&quot;UPDATE $dbname SET chrom=&apos;$chrom&apos;,biotype=&apos;$biotype&apos;,feature=&apos;$feature&apos;,start=$start,end=$end,geneid=&apos;$geneid&apos;,genename=&apos;$genename&apos;,transcriptid=&apos;$transcriptid&apos;,exon=$exon where id=$id;&quot;;</span><br><span class="line">		</span><br><span class="line">		// 执行SQL语句</span><br><span class="line">		if($conn-&gt;query($sql)===TRUE)&#123;</span><br><span class="line">			header(&quot;Location:update_form.php?statu=1&quot;);</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			header(&quot;Location:update_form.php?statu=2&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		header(&quot;Location:update_form.php?statu=3&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">?&gt;</span><br></pre></td></tr></table></figure>
<ul>
<li>脚本<code>updata_operate_DiffExp.php</code></li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&lt;?php</span><br><span class="line">//开启session</span><br><span class="line">session_start();</span><br><span class="line">// 获取用户名和密码</span><br><span class="line">$username=isset($_SESSION[&apos;user&apos;])?$_SESSION[&apos;user&apos;]:&quot;&quot;;</span><br><span class="line">$password=isset($_SESSION[&apos;password&apos;])?$_SESSION[&apos;password&apos;]:&quot;&quot;;</span><br><span class="line"></span><br><span class="line">// 获取通过GET方法向该脚本传递的变量</span><br></pre></td></tr></table></figure>
<h1 id="番外篇：前端优化——html-css"><a href="#番外篇：前端优化——html-css" class="headerlink" title="番外篇：前端优化——html+css"></a>番外篇：前端优化——html+css</h1><h2 id="登录界面优化"><a href="#登录界面优化" class="headerlink" title="登录界面优化"></a>登录界面优化</h2><p>实现效果：</p>
<blockquote>
<p>山寨主流网站的登录界面，即将页面分为左右两半，左侧放置图片，右侧放置登录信息输入框</p>
</blockquote>
<p><img src="/images/InAction-PHP-MySQL-part3-login-ui.png" alt></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">&lt;head&gt;</span><br><span class="line">&lt;title&gt;登录&lt;/title&gt;</span><br><span class="line">&lt;style&gt;</span><br><span class="line">/* 设置背景色：碧绿色 */</span><br><span class="line">body&#123;</span><br><span class="line">	background-color:#7FFFAA;</span><br><span class="line">	margin:50px;</span><br><span class="line">&#125;</span><br><span class="line">form &#123;</span><br><span class="line">	background-color:white;</span><br><span class="line">	width:100%;</span><br><span class="line">	margin:auto;</span><br><span class="line">	padding:70px 0;</span><br><span class="line">&#125;</span><br><span class="line">&lt;/style&gt;</span><br><span class="line">&lt;/head&gt;</span><br><span class="line">&lt;body&gt;</span><br><span class="line">&lt;div&gt;</span><br><span class="line">	&lt;div style=&quot;float:left; width:50%; height:100%;&quot;&gt;</span><br><span class="line">		&lt;img src=./picture/lncRNAdb-logo.jpg width=100% /&gt;</span><br><span class="line">	&lt;/div&gt;</span><br><span class="line">	&lt;div style=&quot;float:right; width:50%; height:100%;&quot;&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	&lt;/div&gt;</span><br><span class="line">&lt;/div&gt;</span><br><span class="line">&lt;/body&gt;</span><br></pre></td></tr></table></figure>
<h2 id="导航条菜单的制作"><a href="#导航条菜单的制作" class="headerlink" title="导航条菜单的制作"></a>导航条菜单的制作</h2><p>实现效果：</p>
<blockquote>
<p>将页面分为左右部分，左侧放置垂直导航栏，右侧放置主面板</p>
</blockquote>
<p><img src="/images/InAction-PHP-MySQL-part3-page-with-navigation.png" alt></p>
<p>使用外部css制定页面渲染风格，css文件保存为<code>private.css</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br></pre></td><td class="code"><pre><span class="line">/* 导航栏样式 */</span><br><span class="line">/* 页面左侧主导航栏（垂直导航栏），占据页面左侧10%的空间，且位置固定 */</span><br><span class="line">.nav_vertical &#123;</span><br><span class="line">	margin:0;</span><br><span class="line">	padding:0;</span><br><span class="line">	width:10%;</span><br><span class="line">	list-style-type:none;</span><br><span class="line">	background-color: #f1f1f1;</span><br><span class="line">	position: fixed;</span><br><span class="line">	height: 100%;</span><br><span class="line">&#125;</span><br><span class="line">.nav_vertical li a&#123;</span><br><span class="line">	display:block;</span><br><span class="line">	padding: 8px 16px;</span><br><span class="line">	text-decoration:none;</span><br><span class="line">	color: #000;	/*  字体颜色：灰色 */</span><br><span class="line">&#125;</span><br><span class="line">.nav_vertical li a:hover&#123;	/* 鼠标悬停时链接块颜色：橙色 */</span><br><span class="line">	background-color:#F60;</span><br><span class="line">	color:#fff</span><br><span class="line">&#125;</span><br><span class="line">.nav_vertical li a.active&#123;	</span><br><span class="line">	background-color: #4CAF50;	/*  激活的链接块颜色：绿色 */</span><br><span class="line">	color:#fff;</span><br><span class="line">&#125;</span><br><span class="line">/* 页面右侧子导航栏（水平导航栏），用于在进行数据库检索时，在不同数据库间切换 */</span><br><span class="line">.nav_horizontal &#123;</span><br><span class="line">	margin-left:10px;</span><br><span class="line">	padding:0;</span><br><span class="line">	list-style-type:none;</span><br><span class="line">&#125;</span><br><span class="line">.nav_horizontal li &#123;</span><br><span class="line">	display:inline;</span><br><span class="line">&#125;</span><br><span class="line">.nav_horizontal li a&#123;</span><br><span class="line">	padding: 8px 16px;</span><br><span class="line">	text-decoration:none;</span><br><span class="line">	background-color: #f1f1f1;</span><br><span class="line">	color: #000;	/*  字体颜色：灰色 */</span><br><span class="line">&#125;</span><br><span class="line">.nav_horizontal li a:hover&#123;	/* 鼠标悬停时链接块颜色：橙色 */</span><br><span class="line">	background-color:#F60;</span><br><span class="line">	color:#fff</span><br><span class="line">&#125;</span><br><span class="line">.nav_horizontal li a.active&#123;	</span><br><span class="line">	background-color: #4CAF50;	/*  激活的链接块颜色：绿色 */</span><br><span class="line">	color:#fff;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">/* 链接样式 */</span><br><span class="line">a:link &#123;color:#0000FF;text-decoration:none;&#125; /* 蓝色 */</span><br><span class="line">a:visited &#123;color:#778899;text-decoration:none;&#125; /* 灰色 */</span><br><span class="line">a:hover &#123;color:#000000;text-decoration:none;&#125; /* 黑色 */</span><br><span class="line">a:active &#123;color:#FF704D;text-decoration:none;&#125;</span><br><span class="line"></span><br><span class="line">/* 表格样式 */</span><br><span class="line">/* 检索表单部分表格样式 */</span><br><span class="line">.result &#123;border-collapse:collapse;&#125;</span><br><span class="line">table,th,td &#123;border:1px solid black;&#125;</span><br><span class="line">th &#123;width:100%;height:50px;&#125;</span><br></pre></td></tr></table></figure>
<p>则在每个php脚本前加入以下代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">&lt;head&gt;</span><br><span class="line">&lt;link rel=&quot;stylesheet&quot; type=&quot;text/css&quot; href=&quot;private.css&quot;&gt;</span><br><span class="line">&lt;/head&gt;</span><br><span class="line">&lt;body&gt;</span><br><span class="line">&lt;ul class=&quot;nav_vertical&quot;&gt;</span><br><span class="line">  	&lt;li&gt;&lt;a class=&quot;active&quot; href=&quot;loginsucc.php&quot;&gt;主页&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">  	&lt;li&gt;&lt;a href=&quot;databaseQuery_Ano.php&quot;&gt;检索&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">  	&lt;li&gt;&lt;a href=&quot;batch_submit_data.php&quot;&gt;提交数据&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">	&lt;li&gt;&lt;a href=&quot;#&quot;&gt;BLAST&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">	&lt;li&gt;&lt;a href=&quot;help.html&quot;&gt;帮助&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">	&lt;li&gt;&lt;a href=&quot;logout.php&quot;&gt;退出&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">&lt;/ul&gt;</span><br><span class="line">&lt;div style=&quot;margin-left:10%; padding:10px; height:100%;&quot;&gt;</span><br><span class="line">&lt;!-- 若为检索部分，请添加以下部分代码 --&gt;</span><br><span class="line">&lt;!-- &lt;ul class=&quot;nav_horizontal&quot;&gt;</span><br><span class="line">  	&lt;li&gt;&lt;a class=&quot;active&quot; href=&quot;databaseQuery_Ano.php&quot;&gt;基因结构注释&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">  	&lt;li&gt;&lt;a href=&quot;databaseQuery_DiffExp.php&quot;&gt;基因表达谱&lt;/a&gt;&lt;/li&gt;</span><br><span class="line">&lt;/ul&gt; --&gt;</span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">&lt;/div&gt;</span><br><span class="line">&lt;/body&gt;</span><br></pre></td></tr></table></figure>
<hr>
<p>参考资料：</p>
<p>(1) <a href="https://www.fenxd.com/111.html" target="_blank" rel="noopener">PHP+MySQLi实现注册和登陆</a></p>
<p><a name="nat-pro">(2) Pertea M, Kim D, Pertea G M, et al. Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown[J]. Nature Protocols, 2016, 11(9):1650.</a></p>
<p>(3) <a href="https://github.com/Ming-Lian/NGS-analysis/blob/master/RNA-seq.md" target="_blank" rel="noopener">Analysis pipeline for RNA-seq</a></p>
<p>(4) <a href="http://www.w3school.com.cn/sql/sql_datatypes.asp" target="_blank" rel="noopener">W3School：SQL 数据类型</a></p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/09/Load-and-resolve-MarkdownDoc/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/09/Load-and-resolve-MarkdownDoc/" class="post-title-link" itemprop="url">前端小技巧：加载并解析Markdown文档</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-09 00:28:49" itemprop="dateCreated datePublished" datetime="2019-02-09T00:28:49+00:00">2019-02-09</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-02-17 06:12:08" itemprop="dateModified" datetime="2019-02-17T06:12:08+00:00">2019-02-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/前后端技术/" itemprop="url" rel="index"><span itemprop="name">前后端技术</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="问题描述与分析"><a href="#问题描述与分析" class="headerlink" title="问题描述与分析"></a>问题描述与分析</h2><p>作为程序猿，应该多多少少都用过Markdown，或者至少读过别人用Markdown语法写的一些文档，比如在GitHub有一个你要用的开源程序，而你又是第一回用它，那么你一般会在这个仓库的Readme里读一读开发者提供的工具说明和使用的相关信息，这部分文档一般就是用Markdown的语法写的</p>
<p><img src="/images/MarkDown2HTML-markdown-example.png" alt></p>
<p align="center">此处以李恒大神的BWA的Readme为例</p>

<p>简单来说，Markdown就是简化阉割过的HTML，优点是语法简单高效，缺点就是HTML中一些稍微高级复杂一点的效果，比如文本居中，Markdown就无法实现，所以Markdown一般被用来写对页面排版要求不高，以文字为主的笔记和文档</p>
<p>如果你一开始用Markdown写了文档，想要把它放到你的网页上去，并以解析后的形式呈现</p>
<p><img src="/images/MarkDown2HTML-transformation.png" alt></p>
<p align="center">左边是原始Markdown文档，右边是解析后的网页效果</p>

<p>那么你有两种实现途径：</p>
<ul>
<li>将Markdown文档以HTML形式导出，在网页上加载之前导出的HTML文件即可</li>
<li>原始网页获取Markdown文档，用js将Markdown文本解析成HTML文本，动态修改原始网页对应标签节点的内容，就实现了Markdown文本的解析和展示</li>
</ul>
<p>第一种方法：</p>
<blockquote>
<p>技术难度最小，也很容易理解，但是操作起来比较繁琐：你的Markdown文档一般是要经常修改更新的，那么每修改一次，你就需要重新导出，然后将之前导出的那一版HTML替换掉</p>
</blockquote>
<p>第二种方法：</p>
<blockquote>
<ul>
<li>技术难度上要高一些，但是目前已经有可以调用的JS框架来实现Markdown2HTML的解析</li>
<li>最大的优点就是，每一次修改更新很方便，只需要对Markdown文档就可以了</li>
</ul>
</blockquote>
<p>下面我们对<strong>第二种方法</strong>的实现过程进行详细的说明</p>
<h2 id="用AJAX获取Markdown文档"><a href="#用AJAX获取Markdown文档" class="headerlink" title="用AJAX获取Markdown文档"></a>用AJAX获取Markdown文档</h2><p>上一部分已经提到，我们需要先让原始网页请求服务器中的Markdown文档</p>
<p>这一步需要使用AJAX（中文音译为阿甲克斯），这里先对AJAX作一个简单的科普：</p>
<blockquote>
<p>AJAX = Asynchronous JavaScript and XML（异步的 JavaScript 和 XML），AJAX 是一种用于创建快速动态网页的技术</p>
<p>首先要说清楚什么是动态网页。与动态网页相对的是静态网页，也就是大家都知道的用HTML写的网页，这种网页的一个鲜明的特点就是它的内容一旦确定就无法改变，在互联网发展的最早期的时候大家用的都是这种静态网页。动态网页要求页面的内容能进行更新。那么怎么让原先的静态网页动起来成为动态网页呢？也有两种方法：</p>
<ul>
<li>重载整个网页面来更新页面内容</li>
<li>上一个页面状态与下一个页面状态，它们之间有相同不变的部分，也有发生了改变的部分，那么我只需要把那些需要发生变化的部分进行修改就行了。这意味着可以在不重新加载整个网页的情况下，对网页的某部分进行更新</li>
</ul>
<p>第一种方法的不经济显而易见，不管各个页面节点前后是否发生了改变，统一都进行刷新</p>
<p>AJAX就是采用第二种页面刷新方法：通过在后台与服务器进行少量数据交换，向服务器请求数据，解析请求的数据，然后对原页面需要刷新的部分进行刷新</p>
</blockquote>
<p><img src="/images/ajax-process.gif" alt></p>
<p align="center">AJAX工作原理</p>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">// 获取表单变量</span><br><span class="line">var f=form;</span><br><span class="line">// 创建XMLHttpRequest对象</span><br><span class="line">var xmlhttp;</span><br><span class="line">if (window.XMLHttpRequest)</span><br><span class="line">&#123;</span><br><span class="line">	// IE7+, Firefox, Chrome, Opera, Safari 浏览器执行代码</span><br><span class="line">	xmlhttp=new XMLHttpRequest();</span><br><span class="line">&#125;</span><br><span class="line">else</span><br><span class="line">&#123;</span><br><span class="line">	// IE6, IE5 浏览器执行代码</span><br><span class="line">	xmlhttp=new ActiveXObject(&quot;Microsoft.XMLHTTP&quot;);</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">// 得到服务器响应后，对得到的Markdown文档进行解析</span><br><span class="line">xmlhttp.onreadystatechange=function()</span><br><span class="line">&#123;</span><br><span class="line">	if (xmlhttp.readyState==4 &amp;&amp; xmlhttp.status==200)</span><br><span class="line">	&#123;</span><br><span class="line">		// 这里调用了marked框架中的marked函数实现Markdown2HTML的解析</span><br><span class="line">		document.getElementById(&quot;content&quot;).innerHTML=marked(xmlhttp.responseText);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">// 向服务器发送请求，获取你需要的Markdown文档</span><br><span class="line">xmlhttp.open(&quot;GET&quot;,f.q.value,true);</span><br><span class="line">xmlhttp.send();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>对上面的脚本继续简单的说明：</p>
<blockquote>
<p>这个是AJAX的基本语法框架</p>
<p>完成三步操作：</p>
<ol>
<li>创建http请求对象，告诉服务器我要哪个Markdown文档</li>
<li>向服务器发起http请求</li>
<li>收到服务器的反馈数据，得到请求的Markdown文档，进行数据的解析，实现Markdown2HTML</li>
</ol>
</blockquote>
<p>那么这里有几个问题需要解答：</p>
<p><strong>1. 怎么指定我想要的Markdown文档呢？</strong></p>
<blockquote>
<p>看上面的代码部分的最后一小部分，有这样一行命令：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; xmlhttp.open(&quot;GET&quot;,f.q.value,true);</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<p>这个函数的第二个参数<code>f.q.value</code>，指定了你要的Markdown文档的地址，所以如果想要指定Markdown文档就需要对这个变量进行设定</p>
<p>怎么设定这个变量？</p>
<p>用表单传参来实现，这里的f就是表单变量，表单部分怎么实现请点 <a href="#specify-markdown-doc">这里</a></p>
</blockquote>
<h2 id="用表单指定Markdown文档"><a href="#用表单指定Markdown文档" class="headerlink" title="用表单指定Markdown文档"></a>用表单指定Markdown文档</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">&lt;form name=&quot;form&quot; action=&quot;&quot; method=&quot;post&quot;&gt;</span><br><span class="line">	&lt;select name=&quot;q&quot;&gt;</span><br><span class="line">		&lt;option class=&quot;active&quot; alue=&quot;&quot;&gt;选择一个笔记:&lt;/option&gt;</span><br><span class="line">		&lt;option value=&quot;A.md&quot;&gt;第一个Markdown文档&lt;/option&gt;</span><br><span class="line">		&lt;option value=&quot;B.md&quot;&gt;第二个Markdown文档&lt;/option&gt;</span><br><span class="line">		&lt;option value=&quot;C.md&quot;&gt;第三个Markdown文档&lt;/option&gt;</span><br><span class="line">		&lt;option value=&quot;Evalue-TranscriptomeData.md&quot;&gt;第四个Markdown文档&lt;/option&gt;</span><br><span class="line">	&lt;/select&gt;</span><br><span class="line">	&lt;input type=&quot;button&quot; value=&quot;显示&quot; onclick=&quot;showMarkdown()&quot;&gt;</span><br><span class="line">&lt;/form&gt;</span><br></pre></td></tr></table></figure>
<p>表单的显示效果如下：</p>
<p><img src="/images/MarkDown2HTML-form.png" alt></p>
<p>点击表单中的“显示”按钮后会执行<code>showMarkdown( )</code>函数，即 <a href="#get-markdown-doc">用AJAX获取Markdown文档</a> 部分的那个函数，并且将表单选择的信息通过<code>form</code>变量传递给了<code>showMarkdown( )</code>函数中的<code>f</code>变量</p>
<p>这样就通过表单设定了用户指定的Markdown文档</p>
<h2 id="引用marked框架解析Markdown文档"><a href="#引用marked框架解析Markdown文档" class="headerlink" title="引用marked框架解析Markdown文档"></a>引用marked框架解析Markdown文档</h2><p>这里采用的是GitHub上的名为<code>marked</code>的JS框架，链接：<code>https://github.com/markedjs/marked</code></p>
<p>要想使用这个框架，需要在html脚本的头信息中引用这个框架：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">&lt;html&gt;</span><br><span class="line">&lt;head&gt;</span><br><span class="line">	...</span><br><span class="line">	&lt;script src=&quot;https://cdn.jsdelivr.net/npm/marked/marked.min.js&quot;&gt;&lt;/script&gt;</span><br><span class="line">	...</span><br><span class="line">&lt;/head&gt;</span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">.</span><br></pre></td></tr></table></figure>
<p>引用这个框架后就可以使用里面定义的<code>marked( )</code>函数进行Markdown文本解析了</p>
<h2 id="附：完整的实现脚本"><a href="#附：完整的实现脚本" class="headerlink" title="附：完整的实现脚本"></a>附：完整的实现脚本</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br></pre></td><td class="code"><pre><span class="line">&lt;!DOCTYPE html&gt;</span><br><span class="line">&lt;html&gt;</span><br><span class="line">&lt;head&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	// 在这里对marked框架进行引用</span><br><span class="line">	&lt;script src=&quot;https://cdn.jsdelivr.net/npm/marked/marked.min.js&quot;&gt;&lt;/script&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	// 在这里编写AJAX代码</span><br><span class="line">	&lt;script&gt;</span><br><span class="line">  function showMarkdown()</span><br><span class="line">&#123;</span><br><span class="line">	// 获取表单变量</span><br><span class="line">	var f=form;</span><br><span class="line">	// 创建XMLHttpRequest对象</span><br><span class="line">	var xmlhttp;</span><br><span class="line">	if (window.XMLHttpRequest)</span><br><span class="line">	&#123;</span><br><span class="line">		// IE7+, Firefox, Chrome, Opera, Safari 浏览器执行代码</span><br><span class="line">		xmlhttp=new XMLHttpRequest();</span><br><span class="line">	&#125;</span><br><span class="line">	else</span><br><span class="line">	&#123;</span><br><span class="line">		// IE6, IE5 浏览器执行代码</span><br><span class="line">		xmlhttp=new ActiveXObject(&quot;Microsoft.XMLHTTP&quot;);</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	</span><br><span class="line">	// 得到服务器响应后，对得到的Markdown文档进行解析</span><br><span class="line">	xmlhttp.onreadystatechange=function()</span><br><span class="line">	&#123;</span><br><span class="line">		if (xmlhttp.readyState==4 &amp;&amp; xmlhttp.status==200)</span><br><span class="line">		&#123;</span><br><span class="line">			// 这里调用了marked框架中的marked函数实现Markdown2HTML的解析</span><br><span class="line">			document.getElementById(&quot;content&quot;).innerHTML=marked(xmlhttp.responseText);</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	</span><br><span class="line">	// 向服务器发送请求，获取你需要的Markdown文档</span><br><span class="line">	xmlhttp.open(&quot;GET&quot;,f.q.value,true);</span><br><span class="line">	xmlhttp.send();</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">	&lt;/script&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">&lt;/head&gt;</span><br><span class="line">&lt;body&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	// 在这里编辑表单代码</span><br><span class="line">	&lt;div class=&quot;container&quot; style=&quot;margin-top:30px&quot;&gt;</span><br><span class="line">		&lt;form name=&quot;form&quot; action=&quot;&quot; method=&quot;post&quot;&gt;</span><br><span class="line">			&lt;select name=&quot;q&quot;&gt;</span><br><span class="line">				&lt;option class=&quot;active&quot; alue=&quot;&quot;&gt;选择一个笔记:&lt;/option&gt;</span><br><span class="line">				&lt;option value=&quot;A.md&quot;&gt;第一个Markdown文档&lt;/option&gt;</span><br><span class="line">				&lt;option value=&quot;B.md&quot;&gt;第二个Markdown文档&lt;/option&gt;</span><br><span class="line">				&lt;option value=&quot;C.md&quot;&gt;第三个Markdown文档&lt;/option&gt;</span><br><span class="line">				&lt;option value=&quot;Evalue-TranscriptomeData.md&quot;&gt;第四个Markdown文档&lt;/option&gt;</span><br><span class="line">			&lt;/select&gt;</span><br><span class="line">			&lt;input type=&quot;button&quot; value=&quot;显示&quot; onclick=&quot;showMarkdown()&quot;&gt;</span><br><span class="line">		&lt;/form&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	// 在这里编辑空的div节点，用于后面展示解析后的内容</span><br><span class="line">	&lt;div id=&quot;content&quot;&gt;&lt;/div&gt;</span><br><span class="line">	&lt;/div&gt;</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">&lt;/body&gt;</span><br><span class="line">&lt;/html&gt;</span><br></pre></td></tr></table></figure>
<hr>
<p>参考资料：</p>
<p>(1) <a href="https://github.com/Ming-Lian/Setup-Database/blob/master/AJAX%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0.md#content" target="_blank" rel="noopener">本人github笔记：AJAX学习笔记</a></p>
<p>(2) <a href="https://github.com/markedjs/marked" target="_blank" rel="noopener">marked说明文档</a></p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/09/InAction-ML-XGS/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/09/InAction-ML-XGS/" class="post-title-link" itemprop="url">西瓜书带学训练营·实战任务</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-09 00:17:04" itemprop="dateCreated datePublished" datetime="2019-02-09T00:17:04+00:00">2019-02-09</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-02-24 14:23:38" itemprop="dateModified" datetime="2019-02-24T14:23:38+00:00">2019-02-24</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Machine-Learning/" itemprop="url" rel="index"><span itemprop="name">Machine-Learning</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="1-“达观杯”文本智能处理挑战赛：sklearn包中logistic算法的使用"><a href="#1-“达观杯”文本智能处理挑战赛：sklearn包中logistic算法的使用" class="headerlink" title="1. “达观杯”文本智能处理挑战赛：sklearn包中logistic算法的使用"></a>1. “达观杯”文本智能处理挑战赛：sklearn包中logistic算法的使用</h2><h3 id="1-1-任务描述"><a href="#1-1-任务描述" class="headerlink" title="1.1. 任务描述"></a>1.1. 任务描述</h3><p>具体任务可以到 <a href="http://www.dcjingsai.com/common/cmpt/%E2%80%9C%E8%BE%BE%E8%A7%82%E6%9D%AF%E2%80%9D%E6%96%87%E6%9C%AC%E6%99%BA%E8%83%BD%E5%A4%84%E7%90%86%E6%8C%91%E6%88%98%E8%B5%9B_%E8%B5%9B%E4%BD%93%E4%B8%8E%E6%95%B0%E6%8D%AE.html" target="_blank" rel="noopener">‘达观杯’文本智能处理挑战赛官网</a> 查看</p>
<p><strong>任务</strong>：建立模型通过长文本数据正文(article)，预测文本对应的类别(class)</p>
<p>数据：</p>
<blockquote>
<p>数据包含2个csv文件：</p>
<ul>
<li><p><strong>train_set.csv</strong></p>
<p>  此数据集用于训练模型，每一行对应一篇文章。文章分别在“字”和“词”的级别上做了脱敏处理。共有四列：</p>
<p>  第一列是文章的索引(id)，第二列是文章正文在“字”级别上的表示，即字符相隔正文(article)；第三列是在“词”级别上的表示，即词语相隔正文(word_seg)；第四列是这篇文章的标注(class)。</p>
<p>  注：每一个数字对应一个“字”，或“词”，或“标点符号”。“字”的编号与“词”的编号是独立的！</p>
</li>
<li><p><strong>test_set.csv</strong> </p>
<p>  此数据用于测试。数据格式同train_set.csv，但不包含class</p>
<p>  注：test_set与train_test中文章id的编号是独立的。</p>
</li>
</ul>
</blockquote>
<h3 id="1-2-编程实现"><a href="#1-2-编程实现" class="headerlink" title="1.2. 编程实现"></a>1.2. 编程实现</h3><p>使用<strong>logistic回归</strong>来实现这个多元分类任务</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><span class="line">import pandas as pd</span><br><span class="line">from sklearn.feature_extract.txt import TfidfVectorizer</span><br><span class="line">from sklearn.decomposition import TruncatedSVD</span><br><span class="line">from sklearn.linear_model import LogisticRegression</span><br><span class="line">from sklearn.model_select import train_test_split</span><br><span class="line">from sklearn.externals import joblib</span><br><span class="line">import time</span><br><span class="line"></span><br><span class="line">t_start = time.time()</span><br><span class="line"></span><br><span class="line"># ===========================================================</span><br><span class="line"># 1. 载人数据与数据预处理</span><br><span class="line">df_train = pd.read_csv(&apos;data/train_set.csv&apos;)</span><br><span class="line">df_train = df_train.drop(columns=&apos;article&apos;,inplace=True)</span><br><span class="line">df_test = pd.read_csv(&apos;data/test_set.csv&apos;)</span><br><span class="line">df_test = df_test.drop(columns=&apos;article&apos;,inplace=True)</span><br><span class="line">y_train = (df_train[&apos;class&apos;] - 1).values</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># ===========================================================</span><br><span class="line"># 2. 特征工程，这里先使用经典的文本特征提取方法TFIDF，提取的TFIDF特征</span><br><span class="line">vectorizer = TfidfVectorizer(ngram_range=(1, 2), min_df=3, max_df=0.9, sublinear_tf=True)</span><br><span class="line">vectorizer.fit(df_train[&apos;word_seg&apos;])</span><br><span class="line">X_train = vectorizer.transform(df_train[&apos;word_seg&apos;])</span><br><span class="line">X_test = vectorizer.transform(df_test[&apos;word_seg&apos;])</span><br><span class="line"></span><br><span class="line"># 可以将得到的TFIDF特征保存至本地</span><br><span class="line">&apos;&apos;&apos;</span><br><span class="line">data = (X_train,y_train,X_test)</span><br><span class="line">f = open(&apos;data/data_tfidf.pkl&apos;,&apos;wb&apos;)</span><br><span class="line">pickle.dump(data,f)</span><br><span class="line">f.close()</span><br><span class="line">&apos;&apos;&apos;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># ===========================================================</span><br><span class="line"># 3. 特征降维，将上一步提取的TFIDF特征使用lsa方法进行特征降维</span><br><span class="line">lsa = TruncatedSVD(n_components=200)</span><br><span class="line">X_train = lsa.transform(X_train)</span><br><span class="line">X_test = lsa.transform(X_test)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># ===========================================================</span><br><span class="line"># 4. 训练分类器</span><br><span class="line"></span><br><span class="line"># 划分训练集和验证集</span><br><span class="line">X_train,X_vali,y_train,y_vali = train_test_split(X_train,y_train,test_size=0.1,random_state=0)</span><br><span class="line"></span><br><span class="line">##  multi_class:分类方式选择参数，有&quot;ovr(默认)&quot;和&quot;multinomial&quot;两个值可选择，在二元逻辑回归中无区别</span><br><span class="line">##  solver:优化算法选择参数，当penalty为&quot;l1&quot;时，参数只能是&quot;liblinear(坐标轴下降法)&quot;；&quot;lbfgs&quot;和&quot;cg&quot;都是关于目标函数的二阶泰勒展开</span><br><span class="line">##  当penalty为&quot;l2&quot;时，参数可以是&quot;lbfgs(拟牛顿法)&quot;,&quot;newton_cg(牛顿法变种)&quot;,&quot;seg(minibactch随机平均梯度下降)&quot;</span><br><span class="line">lr = LogisticRegression(multi_class=&quot;ovr&quot;,penalty=&quot;l2&quot;,solver=&quot;lbfgs&quot;)</span><br><span class="line">lr.fit(X_train,y_train)</span><br><span class="line"></span><br><span class="line"># 模型的保存与持久化</span><br><span class="line">joblib.dump(lr,&quot;logistic_lr.model&quot;)</span><br><span class="line">joblib.load(&quot;logistic_lr.model&quot;) #加载模型,会保存该model文件</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># ===========================================================</span><br><span class="line"># 5. 在验证集上评估模型</span><br><span class="line">pre_vali = lr.predict(X_vali)</span><br><span class="line">pre_score = f1_score(y_true=y_vali,y_pred=pre_vali,average=&apos;macro&apos;)</span><br><span class="line">print(&quot;验证集分数：&#123;&#125;&quot;.format(score_vali))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># ===========================================================</span><br><span class="line"># 6. 对测试集进行预测</span><br><span class="line">y_test = lr.predict(X_test) + 1</span><br></pre></td></tr></table></figure>
<h3 id="1-3-sklearn包中logistic算法的使用"><a href="#1-3-sklearn包中logistic算法的使用" class="headerlink" title="1.3. sklearn包中logistic算法的使用"></a>1.3. sklearn包中logistic算法的使用</h3><p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-1.png" alt></p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-2.png" alt></p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-3.png" alt></p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-4.png" alt></p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-5.png" alt></p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-6.png" alt></p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-lr-7.png" alt></p>
<h2 id="2-鸢尾花数据分类：sklearn包中决策树算法类库的使用"><a href="#2-鸢尾花数据分类：sklearn包中决策树算法类库的使用" class="headerlink" title="2. 鸢尾花数据分类：sklearn包中决策树算法类库的使用"></a>2. 鸢尾花数据分类：sklearn包中决策树算法类库的使用</h2><h3 id="2-1-DecisionTreeClassifier实例"><a href="#2-1-DecisionTreeClassifier实例" class="headerlink" title="2.1. DecisionTreeClassifier实例"></a>2.1. DecisionTreeClassifier实例</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br></pre></td><td class="code"><pre><span class="line">from itertools import product</span><br><span class="line"></span><br><span class="line">import numpy as np</span><br><span class="line">import matplotlib.pyplot as plt</span><br><span class="line"></span><br><span class="line">from sklearn import datasets</span><br><span class="line">from sklearn.tree import DecisionTreeClassifier</span><br><span class="line">from sklearn.model_selection import train_test_split</span><br><span class="line">from sklearn.preprocessing import MinMaxScaler</span><br><span class="line">from sklearn.feature_selection import SelectKBest</span><br><span class="line">from sklearn.feature_selection import chi2</span><br><span class="line">from sklearn.decomposition import PCA</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># 使用自带的iris数据</span><br><span class="line">iris = datasets.load_iris()</span><br><span class="line">X = iris.data[:, np.arange(0,4)]</span><br><span class="line">y = iris.target</span><br><span class="line"></span><br><span class="line"># 划分训练集与测试集</span><br><span class="line">x_train, x_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=14)</span><br><span class="line">print(&quot;训练数据集样本总数:%d;测试数据集样本总数:%d&quot; %(x_train.shape[0],x_test.shape[0]))</span><br><span class="line"></span><br><span class="line"># 对数据集进行标准化</span><br><span class="line">ss = MinMaxScaler()</span><br><span class="line">x_train = ss.fit_transform(x_train,y_train)</span><br><span class="line">x_test = ss.transform(x_test)</span><br><span class="line"></span><br><span class="line"># 特征选择：从已有的特征属性中选择出影响目标最大的特征属性</span><br><span class="line"># 常用方法：</span><br><span class="line">#	离散属性：F统计量、卡方系数、互信息mutual_info_classif</span><br><span class="line">#	连续属性：皮尔逊相关系数、F统计量、互信息mutual_info_classif&#125;</span><br><span class="line"># 这里使用离散属性的卡方系数，实现函数为SelectKBest，用SelectKBest方法从四个原始特征属性中选择出最能影响目标的3个特征属性</span><br><span class="line">ch2 = SelectKBest(chi2,k=3) # k默认为10，指定后会返回想要的特征个数</span><br><span class="line">ch2.fit(x_train,y_train)</span><br><span class="line">x_train = ch2.transform(x_train)</span><br><span class="line">x_test = ch2.transform(x_test)</span><br><span class="line"></span><br><span class="line"># 特征降维，这里使用PCA方法</span><br><span class="line">pca = PCA(n_components=2)   # 构建一个PCA对象，设置最终维度为2维。这里为了后边画图方便，将数据维度设置为 2，一般用默认不设置就可以</span><br><span class="line">x_train = pca.fit_transform(x_train) # 训练与转换，也可以拆分成两步</span><br><span class="line">x_test = pca.transform(x_test)</span><br><span class="line"></span><br><span class="line"># 训练模型</span><br><span class="line">#	criterion：指定特征选择标准，可以使用&quot;gini&quot;或者&quot;entropy&quot;，前者代表基尼系数，后者代表信息增益</span><br><span class="line">#	max_depth：限制树的最大深度4</span><br><span class="line">clf = DecisionTreeClassifier(criterion=&quot;entropy&quot;,max_depth=4)</span><br><span class="line">clf.fit(X, y) # 拟合模型</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># 画图</span><br><span class="line">x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1</span><br><span class="line">y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1</span><br><span class="line"># 生成网格采样点</span><br><span class="line">xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.1),</span><br><span class="line">                     np.arange(y_min, y_max, 0.1))</span><br><span class="line"></span><br><span class="line">Z = clf.predict(np.c_[xx.ravel(), yy.ravel()])</span><br><span class="line">Z = Z.reshape(xx.shape)</span><br><span class="line"></span><br><span class="line">plt.contourf(xx, yy, Z, alpha=0.4)</span><br><span class="line">plt.scatter(X[:, 0], X[:, 1], c=y, alpha=0.8)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-decision-tree-1.png" alt></p>
<h3 id="2-2-可视化决策树"><a href="#2-2-可视化决策树" class="headerlink" title="2.2. 可视化决策树"></a>2.2. 可视化决策树</h3><p>scikit-learn中决策树的可视化一般需要安装graphviz，主要包括graphviz的安装和python的graphviz插件的安装</p>
<blockquote>
<ul>
<li><p>1) 安装graphviz。下载地址在：<code>http://www.graphviz.org</code>/。如果你是linux，可以用apt-get或者yum的方法安装。如果是windows，就在官网下载msi文件安装。无论是linux还是windows，装完后都要设置环境变量，将graphviz的bin目录加到PATH，比如我是windows，将<code>C:/Program Files (x86)/Graphviz2.38/bin/</code>加入了<code>PATH</code></p>
</li>
<li><p>2) 安装python插件graphviz： <code>pip install graphviz</code></p>
</li>
</ul>
<ul>
<li>3) 安装python插件pydotplus: <code>pip install pydotplus</code></li>
</ul>
</blockquote>
<p>这样环境就搭好了，有时候python会很笨，仍然找不到graphviz，这时，可以在代码里面加入这一行：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">os.environ[&quot;PATH&quot;] += os.pathsep + &apos;C:/Program Files (x86)/Graphviz2.38/bin/&apos;</span><br></pre></td></tr></table></figure>
<p>可视化决策树的代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">from IPython.display import Image  </span><br><span class="line">from sklearn import tree</span><br><span class="line">import pydotplus </span><br><span class="line">dot_data = tree.export_graphviz(clf, out_file=None, </span><br><span class="line">                         feature_names=iris.feature_names,  </span><br><span class="line">                         class_names=iris.target_names,  </span><br><span class="line">                         filled=True, rounded=True,  </span><br><span class="line">                         special_characters=True)  </span><br><span class="line">graph = pydotplus.graph_from_dot_data(dot_data)  </span><br><span class="line">Image(graph.create_png())</span><br></pre></td></tr></table></figure>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-decision-tree-2.png" alt></p>
<h2 id="3-sklearn中测试数据：sklearn包中SVM算法库的使用"><a href="#3-sklearn中测试数据：sklearn包中SVM算法库的使用" class="headerlink" title="3. sklearn中测试数据：sklearn包中SVM算法库的使用"></a>3. sklearn中测试数据：sklearn包中SVM算法库的使用</h2><h3 id="3-1-SVM相关知识点回顾"><a href="#3-1-SVM相关知识点回顾" class="headerlink" title="3.1. SVM相关知识点回顾"></a>3.1. SVM相关知识点回顾</h3><h4 id="3-1-1-SVM与SVR"><a href="#3-1-1-SVM与SVR" class="headerlink" title="3.1.1. SVM与SVR"></a>3.1.1. SVM与SVR</h4><ul>
<li><p><strong>SVM分类算法</strong></p>
<p>  其原始形式是：</p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-1.png" alt></p>
<p>  其中m为样本个数，我们的样本为(x<sub>1</sub>,y<sub>1</sub>),(x<sub>2</sub>,y<sub>2</sub>),…,(x<sub>m</sub>,y<sub>m</sub>)。w,b是我们的分离超平面的w∙ϕ(x<sub>i</sub>)+b=0系数, ξ<sub>i</sub>为第i个样本的松弛系数， C为惩罚系数。ϕ(x<sub>i</sub>)为低维到高维的映射函数</p>
<p>  通过拉格朗日函数以及对偶化后的形式为：</p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-2.png" alt></p>
</li>
<li><p><strong>SVR回归算法</strong></p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-3.png" alt></p>
<p>  其中m为样本个数，我们的样本为(x<sub>1</sub>,y<sub>1</sub>),(x<sub>2</sub>,y<sub>2</sub>),…,(x<sub>m</sub>,y<sub>m</sub>)。w,b是我们的回归超平面的w∙x<sub>i</sub>+b=0系数, ξ<sup>∨</sup><sub>i</sub>，ξ<sup>∧</sup><sub>i</sub>为第i个样本的松弛系数， C为惩罚系数，ϵ为损失边界，到超平面距离小于ϵ的训练集的点没有损失。ϕ(x<sub>i</sub>)为低维到高维的映射函数。</p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-4.png" alt></p>
</li>
</ul>
<h4 id="3-1-2-核函数"><a href="#3-1-2-核函数" class="headerlink" title="3.1.2. 核函数"></a>3.1.2. 核函数</h4><p>在scikit-learn中，内置的核函数一共有4种：</p>
<ul>
<li><p><strong>线性核函数</strong>（Linear Kernel）表达式为：K(x,z)=x∙z，就是普通的内积</p>
</li>
<li><p><strong>多项式核函数</strong>（Polynomial Kernel）是线性不可分SVM常用的核函数之一，表达式为：K(x,z)=（γx∙z+r)<sup>d</sup> ，其中，γ,r,d都需要自己调参定义</p>
</li>
<li><p><strong>高斯核函数</strong>（Gaussian Kernel），在SVM中也称为径向基核函数（Radial Basis Function,RBF），它是 libsvm 默认的核函数，当然也是 scikit-learn 默认的核函数。表达式为：K(x,z)=exp(−γ||x−z||<sup>2</sup>)， 其中，γ大于0，需要自己调参定义</p>
</li>
<li><p><strong>Sigmoid核函数</strong>（Sigmoid Kernel）也是线性不可分SVM常用的核函数之一，表达式为：K(x,z)=tanh（γx∙z+r)， 其中，γ，r都需要自己调参定义</p>
</li>
</ul>
<p>一般情况下，对非线性数据使用默认的高斯核函数会有比较好的效果，如果你不是SVM调参高手的话，建议使用高斯核来做数据分析。　</p>
<h3 id="3-2-sklearn中SVM相关库的简介"><a href="#3-2-sklearn中SVM相关库的简介" class="headerlink" title="3.2. sklearn中SVM相关库的简介"></a>3.2. sklearn中SVM相关库的简介</h3><p>scikit-learn SVM算法库封装了libsvm 和 liblinear 的实现，仅仅重写了算法了接口部分</p>
<h4 id="3-2-1-分类库与回归库"><a href="#3-2-1-分类库与回归库" class="headerlink" title="3.2.1. 分类库与回归库"></a>3.2.1. 分类库与回归库</h4><ul>
<li><p><strong>分类算法库</strong></p>
<p>  包括SVC， NuSVC，和LinearSVC 3个类</p>
<p>  对于SVC， NuSVC，和LinearSVC 3个分类的类，SVC和 NuSVC差不多，区别仅仅在于对损失的度量方式不同，而LinearSVC从名字就可以看出，他是线性分类，也就是不支持各种低维到高维的核函数，仅仅支持线性核函数，对线性不可分的数据不能使用</p>
</li>
<li><p><strong>回归算法库</strong></p>
<p>  包括SVR， NuSVR，和LinearSVR 3个类</p>
<p>  同样的，对于SVR， NuSVR，和LinearSVR 3个回归的类， SVR和NuSVR差不多，区别也仅仅在于对损失的度量方式不同。LinearSVR是线性回归，只能使用线性核函数</p>
</li>
</ul>
<h4 id="3-2-2-高斯核调参"><a href="#3-2-2-高斯核调参" class="headerlink" title="3.2.2. 高斯核调参"></a>3.2.2. 高斯核调参</h4><h5 id="3-2-2-1-需要调节的参数"><a href="#3-2-2-1-需要调节的参数" class="headerlink" title="3.2.2.1. 需要调节的参数"></a>3.2.2.1. 需要调节的参数</h5><ul>
<li><p><strong>SVM分类模型</strong></p>
<p>  如果是SVM分类模型，这两个超参数分别是<strong>惩罚系数C</strong>和<strong>RBF核函数的系数γ</strong></p>
<p>  <strong>惩罚系数C</strong></p>
<blockquote>
<p>它在优化函数里主要是平衡支持向量的复杂度和误分类率这两者之间的关系，可以理解为正则化系数</p>
<ul>
<li><p>当C比较大时，我们的损失函数也会越大，这意味着我们不愿意放弃比较远的离群点。这样我们会有更加多的支持向量，也就是说支持向量和超平面的模型也会变得越复杂，也容易过拟合</p>
</li>
<li><p>当C比较小时，意味我们不想理那些离群点，会选择较少的样本来做支持向量，最终的支持向量和超平面的模型也会简单</p>
</li>
</ul>
<p>scikit-learn中默认值是1</p>
</blockquote>
<p>  C越大，泛化能力越差，易出现过拟合现象；C越小，泛化能力越好，易出现过欠拟合现象</p>
<p>  <strong>BF核函数的参数γ</strong></p>
<blockquote>
<p>RBF 核函数K(x,z)=exp(−γ||x−z||<sup>2</sup>) γ&gt;0</p>
<p>γ主要定义了单个样本对整个分类超平面的影响</p>
<ul>
<li><p>当γ比较小时，单个样本对整个分类超平面的影响比较小，不容易被选择为支持向量</p>
</li>
<li><p>当γ比较大时，单个样本对整个分类超平面的影响比较大，更容易被选择为支持向量，或者说整个模型的支持向量也会多</p>
</li>
</ul>
<p>scikit-learn中默认值是 <code>1/样本特征数</code></p>
</blockquote>
<p>  γ越大，训练集拟合越好，泛化能力越差，易出现过拟合现象</p>
<p>  如果把惩罚系数C和RBF核函数的系数γ一起看，当C比较大， γ比较大时，我们会有更多的支持向量，我们的模型会比较复杂，容易过拟合一些。如果C比较小 ， γ比较小时，模型会变得简单，支持向量的个数会少</p>
</li>
<li><p><strong>SVM回归模型</strong></p>
<p>  SVM回归模型的RBF核比分类模型要复杂一点，因为此时我们除了惩罚系数C和RBF核函数的系数γ之外，还多了一个<strong>损失距离度量ϵ</strong></p>
<blockquote>
<p>对于损失距离度量ϵ，它决定了样本点到超平面的距离损失</p>
<ul>
<li><p>当 ϵ 比较大时，损失较小，更多的点在损失距离范围之内，而没有损失,模型较简单</p>
</li>
<li><p>当 ϵ 比较小时，损失函数会较大，模型也会变得复杂</p>
</li>
</ul>
<p>scikit-learn中默认值是0.1</p>
</blockquote>
<p>  如果把惩罚系数C，RBF核函数的系数γ和损失距离度量ϵ一起看，当C比较大， γ比较大，ϵ比较小时，我们会有更多的支持向量，我们的模型会比较复杂，容易过拟合一些。如果C比较小 ， γ比较小，ϵ比较大时，模型会变得简单，支持向量的个数会少</p>
</li>
</ul>
<h5 id="3-2-2-2-调参方法：网格搜索"><a href="#3-2-2-2-调参方法：网格搜索" class="headerlink" title="3.2.2.2. 调参方法：网格搜索"></a>3.2.2.2. 调参方法：网格搜索</h5><p>对于SVM的RBF核，我们主要的调参方法都是交叉验证。具体在scikit-learn中，主要是使用网格搜索，即GridSearchCV类</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">from sklearn.model_selection import GridSearchCV</span><br><span class="line">grid = GridSearchCV(SVC(), param_grid=&#123;&quot;C&quot;:[0.1, 1, 10], &quot;gamma&quot;: [1, 0.1, 0.01]&#125;, cv=4)</span><br><span class="line">grid.fit(X, y)</span><br></pre></td></tr></table></figure>
<p>将GridSearchCV类用于SVM RBF调参时要注意的参数有：</p>
<blockquote>
<p>1) <strong>estimator</strong>：即我们的模型，此处我们就是带高斯核的SVC或者SVR</p>
<p>2) <strong>param_grid</strong>：即我们要调参的参数列表。 比如我们用SVC分类模型的话，那么param_grid可以定义为{“C”:[0.1, 1, 10], “gamma”: [0.1, 0.2, 0.3]}，这样我们就会有9种超参数的组合来进行网格搜索，选择一个拟合分数最好的超平面系数</p>
<p>3) <strong>cv</strong>：S折交叉验证的折数，即将训练集分成多少份来进行交叉验证。默认是3。如果样本较多的话，可以适度增大cv的值</p>
</blockquote>
<h3 id="3-3-编程实现"><a href="#3-3-编程实现" class="headerlink" title="3.3. 编程实现"></a>3.3. 编程实现</h3><ol>
<li><p>生成测试数据</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">from sklearn.datasets import make_circles</span><br><span class="line">from sklearn.preprocessing import StandardScaler</span><br><span class="line"></span><br><span class="line"># 生成一些随机数据用于后续分类</span><br><span class="line">X, y = make_circles(noise=0.2, factor=0.5, random_state=1) # 生成时加入了一些噪声</span><br><span class="line">X = StandardScaler().fit_transform(X) # 把数据归一化</span><br></pre></td></tr></table></figure>
</li>
</ol>
<pre><code>生成的随机数据可视化结果如下：
</code></pre><p><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-5.png" alt></p>
<ol start="2">
<li><p>调参</p>
<p> 接着采用网格搜索的策略进行RBF核函数参数搜索</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">from sklearn.model_selection import GridSearchCV</span><br><span class="line">grid = GridSearchCV(SVC(), param_grid=&#123;&quot;C&quot;:[0.1, 1, 10], &quot;gamma&quot;: [1, 0.1, 0.01]&#125;, cv=4) # 总共有9种参数组合的搜索空间</span><br><span class="line">grid.fit(X, y)</span><br><span class="line">print(&quot;The best parameters are %s with a score of %0.2f&quot;</span><br><span class="line">      % (grid.best_params_, grid.best_score_))</span><br><span class="line"></span><br><span class="line">输出为：</span><br><span class="line">The best parameters are &#123;&apos;C&apos;: 10, &apos;gamma&apos;: 0.1&#125; with a score of 0.91</span><br></pre></td></tr></table></figure>
<p> 可以对9种参数组合训练的结果进行可视化，观察分类的效果：</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1</span><br><span class="line">y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1</span><br><span class="line">xx, yy = np.meshgrid(np.arange(x_min, x_max,0.02),</span><br><span class="line">                     np.arange(y_min, y_max, 0.02))</span><br><span class="line"></span><br><span class="line">for i, C in enumerate((0.1, 1, 10)):</span><br><span class="line">    for j, gamma in enumerate((1, 0.1, 0.01)):</span><br><span class="line">        plt.subplot()       </span><br><span class="line">        clf = SVC(C=C, gamma=gamma)</span><br><span class="line">        clf.fit(X,y)</span><br><span class="line">        Z = clf.predict(np.c_[xx.ravel(), yy.ravel()])</span><br><span class="line"></span><br><span class="line">        # Put the result into a color plot</span><br><span class="line">        Z = Z.reshape(xx.shape)</span><br><span class="line">        plt.contourf(xx, yy, Z, cmap=plt.cm.coolwarm, alpha=0.8)</span><br><span class="line"></span><br><span class="line">        # Plot also the training points</span><br><span class="line">        plt.scatter(X[:, 0], X[:, 1], c=y, cmap=plt.cm.coolwarm)</span><br><span class="line"></span><br><span class="line">        plt.xlim(xx.min(), xx.max())</span><br><span class="line">        plt.ylim(yy.min(), yy.max())</span><br><span class="line">        plt.xticks(())</span><br><span class="line">        plt.yticks(())</span><br><span class="line">        plt.xlabel(&quot; gamma=&quot; + str(gamma) + &quot; C=&quot; + str(C))</span><br><span class="line">        plt.show()</span><br></pre></td></tr></table></figure>
</li>
</ol>
<table>
<thead>
<tr>
<th style="text-align:center"><code>C \ gamma</code></th>
<th style="text-align:center">1</th>
<th style="text-align:center">0.1</th>
<th style="text-align:center">0.001</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">0.1</td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-6.png" alt></td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-7.png" alt></td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-8.png" alt></td>
</tr>
<tr>
<td style="text-align:center">1</td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-9.png" alt></td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-10.png" alt></td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-11.png" alt></td>
</tr>
<tr>
<td style="text-align:center">10</td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-12.png" alt></td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-13.png" alt></td>
<td style="text-align:center"><img src="/images/InAction-MachineLearning-XGS-sklearn-SVM-14.png" alt></td>
</tr>
</tbody>
</table>
<h2 id="4-sklearn包中朴素贝叶斯库的使用"><a href="#4-sklearn包中朴素贝叶斯库的使用" class="headerlink" title="4.sklearn包中朴素贝叶斯库的使用"></a>4.sklearn包中朴素贝叶斯库的使用</h2><h3 id="4-1-朴素贝叶斯相关知识点回顾"><a href="#4-1-朴素贝叶斯相关知识点回顾" class="headerlink" title="4.1. 朴素贝叶斯相关知识点回顾"></a>4.1. 朴素贝叶斯相关知识点回顾</h3><h4 id="4-1-1-什么是朴素贝叶斯分类器"><a href="#4-1-1-什么是朴素贝叶斯分类器" class="headerlink" title="4.1.1. 什么是朴素贝叶斯分类器"></a>4.1.1. 什么是朴素贝叶斯分类器</h4><p>判别式模型（discriminative models)：像决策树、BP神经网络、支持向量机等，都可以归入判别式模型，它们都是直接学习出输出Y与特征X的关系，如：</p>
<ul>
<li>决策函数 Y=f(X)</li>
<li>条件概率 P(Y|X)</li>
</ul>
<p>生成式模型 (gernerative models)：先对联合概率分布 P(X,Y) 进行建模，然后再由此获得P(Y|X) = P(X,Y)/P(X)</p>
<p>贝叶斯学派的思想：</p>
<blockquote>
<p>贝叶斯学派的思想可以概括为先验概率+数据=后验概率。也就是说我们在实际问题中需要得到的后验概率，可以通过先验概率和数据一起综合得到。数据大家好理解，被频率学派攻击的是先验概率，一般来说先验概率就是我们对于数据所在领域的历史经验，但是这个经验常常难以量化或者模型化，于是贝叶斯学派大胆的假设先验分布的模型，比如正态分布，beta分布等。这个假设一般没有特定的依据，因此一直被频率学派认为很荒谬。虽然难以从严密的数学逻辑里推出贝叶斯学派的逻辑，但是在很多实际应用中，贝叶斯理论很好用，比如垃圾邮件分类，文本分类</p>
</blockquote>
<p>如何对P(X,Y)进行建模？</p>
<blockquote>
<p>假如我们的分类模型样本是：</p>
<p>(x<sup>(1)</sup><sub>1</sub>,x<sup>(1)</sup><sub>2</sub>,…x<sup>(1)</sup><sub>n</sub>,y<sub>1</sub>), (x<sup>(2)</sup><sub>1</sub>,x<sup>(2)</sup><sub>2</sub>,…x<sup>(2)</sup><sub>n</sub>,y<sub>2</sub>),…(x<sup>(m)</sup><sub>1</sub>,x<sup>(m)</sup><sub>2</sub>, …, x<sup>(m)</sup><sub>n</sub>,y<sub>m</sub>)</p>
<p>则</p>
<p>P(X, Y) = P(Y) * P( X = (x<sub>1</sub>, x<sub>2</sub>, …, x<sub>n</sub>) | Y ) </p>
<p>其中</p>
<p>P( X = (x<sub>1</sub>, x<sub>2</sub>, …, x<sub>n</sub>) | Y )  =  P( x<sub>1</sub> | Y) * P( x<sub>2</sub> | Y, x<sub>1</sub>) * … P( x<sub>n</sub> | Y, x<sub>1</sub>, … , x<sub>n-1</sub>)</p>
</blockquote>
<p>这是一个超级复杂的有n个维度的条件分布，很难求出</p>
<p>朴素贝叶斯模型在这里做了一个大胆的假设，即X的n个维度之间相互独立，这样就可以得出：</p>
<p>P( X = (x<sub>1</sub>, x<sub>2</sub>, …, x<sub>n</sub>) | Y ) = P( x<sub>1</sub> | Y) * P( x<sub>2</sub> | Y) * … * P( x<sub>n</sub> | Y) </p>
<h4 id="4-1-2-朴素贝叶斯推断"><a href="#4-1-2-朴素贝叶斯推断" class="headerlink" title="4.1.2. 朴素贝叶斯推断"></a>4.1.2. 朴素贝叶斯推断</h4><p><img src="/images/InAction-MachineLearning-XGS-sklearn-naive-bayes-1.png" alt></p>
<h4 id="4-1-3-朴素贝叶斯学习"><a href="#4-1-3-朴素贝叶斯学习" class="headerlink" title="4.1.3. 朴素贝叶斯学习"></a>4.1.3. 朴素贝叶斯学习</h4><p>需要从训练样本中学习到以下两个参数：</p>
<ul>
<li><p><strong>先验概率</strong> P(c)</p>
<p>  P(c)表示了样本空间中各类样本所占的比例</p>
<p>  根据大数定律，当训练集中包含充足的独立同分布样本时，P(c)可根据各类样本出现的频率来估计</p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-naive-bayes-2.png" alt></p>
</li>
<li><p><strong>类条件概率</strong>（又称为似然） P(x<sub>i</sub> | c)</p>
<p>  （1）如果 x<sub>i</sub> 是离散的，可以假设 x<sub>i</sub>符合多项式分布，这样得到 P(x<sub>i</sub> | c) 是在样本类别 c 中，特征 x<sub>i</sub> 出现的频率</p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-naive-bayes-3.png" alt></p>
<p>  （2）如果 x<sub>i</sub> 是连续属性，可以假设 P(x<sub>i</sub> | c) ~ N( μ<sub>c,i</sub> , σ<sup>2</sup><sub>c,i</sub> )</p>
<p>  <img src="/images/InAction-MachineLearning-XGS-sklearn-naive-bayes-4.png" alt></p>
</li>
</ul>
<h3 id="4-2-sklearn中朴素贝叶斯类库的简介"><a href="#4-2-sklearn中朴素贝叶斯类库的简介" class="headerlink" title="4.2. sklearn中朴素贝叶斯类库的简介"></a>4.2. sklearn中朴素贝叶斯类库的简介</h3><p>在scikit-learn中，一共有3个朴素贝叶斯的分类算法类</p>
<blockquote>
<ul>
<li><p>GaussianNB：先验为高斯分布的朴素贝叶斯</p>
</li>
<li><p>MultinomialNB：先验为多项式分布的朴素贝叶斯</p>
</li>
<li><p>BernoulliNB：先验为伯努利分布的朴素贝叶斯</p>
</li>
</ul>
</blockquote>
<p>这三个类适用的分类场景各不相同</p>
<blockquote>
<p>一般来说，如果样本特征的分布大部分是<strong>连续值</strong>，使用GaussianNB会比较好</p>
<p>如果如果样本特征的分大部分是<strong>多元离散值</strong>，使用MultinomialNB比较合适</p>
<p>如果样本特征是<strong>二元离散值</strong>或者<strong>很稀疏的多元离散值</strong>，应该使用BernoulliNB。</p>
</blockquote>
<h4 id="4-2-1-GaussianNB类"><a href="#4-2-1-GaussianNB类" class="headerlink" title="4.2.1. GaussianNB类"></a>4.2.1. GaussianNB类</h4><p>GaussianNB类的主要参数仅有一个，即先验概率priors</p>
<p>这个值默认不给出，如果不给出此时P(Y=c)= m<sub>c</sub> / m，如果给出的话就以priors 为准</p>
<p>在使用GaussianNB的fit方法拟合数据后，我们可以进行预测。此时预测有三种方法</p>
<blockquote>
<ul>
<li><p>predict方法：就是我们最常用的预测方法，直接给出测试集的预测类别输出；</p>
</li>
<li><p>predict_proba方法：给出测试集样本在各个类别上预测的概率；</p>
</li>
<li><p>predict_log_proba方法：和predict_proba类似，它会给出测试集样本在各个类别上预测的概率的一个对数转化；</p>
</li>
</ul>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">from sklearn.naive_bayes import GaussianNB</span><br><span class="line"></span><br><span class="line">clf = GaussianNB()</span><br><span class="line">#拟合数据</span><br><span class="line">clf.fit(X, Y)</span><br><span class="line"></span><br><span class="line">#进行预测</span><br><span class="line">clf.predict([[-0.8, -1]])</span><br></pre></td></tr></table></figure>
<h4 id="4-2-2-MultinomialNB类"><a href="#4-2-2-MultinomialNB类" class="headerlink" title="4.2.2. MultinomialNB类"></a>4.2.2. MultinomialNB类</h4><p>MultinomialNB假设特征的先验概率为多项式分布，即如下式：</p>
<p><img src="/images/InAction-MachineLearning-XGS-sklearn-naive-bayes-5.png" alt></p>
<p>MultinomialNB参数比GaussianNB多，但是一共也只有仅仅3个</p>
<ul>
<li><p><strong>参数alpha</strong>：为上面的常数λ。如果你没有特别的需要，用默认的1即可。如果发现拟合的不好，需要调优时，可以选择稍大于1或者稍小于1的数</p>
</li>
<li><p><strong>参数fit_prior</strong>：是否要考虑先验概率，如果是false，则所有的样本类别输出都有相同的类别先验概率；否则可以自己用第三个参数class_prior输入先验概率，或者不输入第三个参数class_prior让MultinomialNB自己从训练集样本来计算先验概率</p>
</li>
<li><p><strong>参数class_prior</strong>：输入先验概率，若不输入第三个参数class_prior让MultinomialNB自己从训练集样本来计算先验概率</p>
</li>
</ul>
<h4 id="4-2-3-BernoulliNB类"><a href="#4-2-3-BernoulliNB类" class="headerlink" title="4.2.3. BernoulliNB类"></a>4.2.3. BernoulliNB类</h4><p><img src="/images/InAction-MachineLearning-XGS-sklearn-naive-bayes-6.png" alt></p>
<p>其中，x <sub>i</sub> 只能取0或1</p>
<p>BernoulliNB一共有4个参数，其中3个参数的名字和意义和MultinomialNB完全相同</p>
<p>唯一增加的一个参数是binarize，这个参数主要是用来帮BernoulliNB处理二项分布的。如果不输入，则BernoulliNB认为每个数据特征都已经是二元的。否则的话，小于binarize的会归为一类，大于binarize的会归为另外一类</p>
<hr>
<p>参考资料：</p>
<p>(1) <a href="https://github.com/MLjian/TextClassificationImplement" target="_blank" rel="noopener">【GitHub】MLjian/TextClassificationImplement</a></p>
<p>(2) <a href="https://www.cnblogs.com/pinard/p/6035872.html" target="_blank" rel="noopener">刘建平Pinard《scikit-learn 逻辑回归类库使用小结》</a></p>
<p>(3) <a href="https://blog.csdn.net/loveliuzz/article/details/78708359" target="_blank" rel="noopener">loveliuzz《机器学习sklearn19.0——Logistic回归算法》</a></p>
<p>(4) <a href="https://www.cnblogs.com/pinard/p/6056319.html" target="_blank" rel="noopener">刘建平Pinard《scikit-learn决策树算法类库使用小结》</a></p>
<p>(5) <a href="https://blog.csdn.net/loveliuzz/article/details/78739438" target="_blank" rel="noopener">loveliuzz《机器学习sklearn19.0——决策树算法》</a></p>
<p>(6) <a href="https://www.cnblogs.com/pinard/p/6117515.html" target="_blank" rel="noopener">刘建平Pinard《scikit-learn 支持向量机算法库使用小结》</a></p>
<p>(7) <a href="https://www.cnblogs.com/pinard/p/6126077.html" target="_blank" rel="noopener">刘建平Pinard《支持向量机高斯核调参小结》</a></p>
<p>(8) <a href="https://blog.csdn.net/loveliuzz/article/details/78768063" target="_blank" rel="noopener">loveliuzz《机器学习sklearn19.0——SVM算法》</a></p>
<p>(9) 周志华《机器学习》</p>
<p>(10) <a href="https://www.cnblogs.com/pinard/p/6069267.html" target="_blank" rel="noopener">刘建平Pinard《朴素贝叶斯算法原理小结》</a></p>
<p>(11) <a href="https://www.cnblogs.com/pinard/p/6074222.html" target="_blank" rel="noopener">刘建平Pinard《scikit-learn 朴素贝叶斯类库使用小结》</a></p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/09/Helpdoc-for-different-languages/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/09/Helpdoc-for-different-languages/" class="post-title-link" itemprop="url">在Perl、Shell和Python中传参与输出帮助文档</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-09 00:03:24" itemprop="dateCreated datePublished" datetime="2019-02-09T00:03:24+00:00">2019-02-09</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-02-17 06:12:08" itemprop="dateModified" datetime="2019-02-17T06:12:08+00:00">2019-02-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/编程/" itemprop="url" rel="index"><span itemprop="name">编程</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>基于本人对多种编程语言的粗浅了解，不论是哪种编程语言它的参数传递方式主要分为下面两类：</p>
<blockquote>
<ul>
<li><strong>直接传递</strong>（以Perl为例进行说明）</li>
</ul>
<p>在调用脚本时，直接传递参数，如：<code>./script.pl a b c</code></p>
<p>然后在脚本中用<code>@ARGV</code>变量获取这些参数</p>
<ul>
<li><strong>getopt方法</strong></li>
</ul>
<p>这种方法是大多数专业程序中都会用到的方法，使用的时候在参数名前加连接符，后面接上参数的实际赋值，例如：<code>-a 1</code></p>
<p>不过getopt方法的参数解析方式略显复杂，下面会在具体的语言中进行逐一说明</p>
</blockquote>
<p>直接传递的传参方式的优点是编写和使用起来很方便，但缺点很明显，参数的顺序是固定的，不能随意改变，每回使用时都需要确定各个参数分别是什么，而且一般采用这种传参方式的人是不会编写帮助文档的，所以一旦忘了只能查看源代码</p>
<p>getopt方法的优点是，传参方式灵活，而且采用这种传参方式的程序员一般都会在程序中添加帮助文档，因此这种传参方式对用户是非常友好的，但是对于程序员来说，则意味着他或她不得不多写好几行代码——所以一个好的程序员头顶凉凉是可以理解的~</p>
<p>以下我们只介绍第二种传参方法</p>
<h2 id="Perl"><a href="#Perl" class="headerlink" title="Perl"></a>Perl</h2><h3 id="Perl中getopt传参"><a href="#Perl中getopt传参" class="headerlink" title="Perl中getopt传参"></a>Perl中getopt传参</h3><p>Perl中的这个功能需要通过调用<code>Getopt::Long</code>模块实现</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">use Getopt::Long;</span><br></pre></td></tr></table></figure>
<p>然后使用GetOptions函数承接传递的参数：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">my ($var1,$var2,$var3,$var4); # 若使用&quot;use strict&quot;模式，则需要提前定义变量</span><br><span class="line">GetOptions(</span><br><span class="line">	&quot;i:s&quot;=&gt;\$var1,</span><br><span class="line">	&quot;o:s&quot;=&gt;\$var2,</span><br><span class="line">	&quot;n:i&quot;=&gt;\$var3,</span><br><span class="line">	&quot;m:i&quot;=&gt;\$var4</span><br><span class="line">	);</span><br></pre></td></tr></table></figure>
<p>这样，你就可以通过以下的方式进行灵活的Perl脚本参数传递了：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl perlscript.pl -i var1 -o var2 ...</span><br></pre></td></tr></table></figure>
<h3 id="Perl中输出帮助文档"><a href="#Perl中输出帮助文档" class="headerlink" title="Perl中输出帮助文档"></a>Perl中输出帮助文档</h3><p>可以使用POD文档实现在Perl中输出帮助文档，想了解更多关于POD文档的知识，请点 <a href="http://www.runoob.com/perl/perl-embedded-documentation.html" target="_blank" rel="noopener">这里</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">=head1 part1</span><br><span class="line"></span><br><span class="line">	doc in part1</span><br><span class="line"></span><br><span class="line">=head2 part2</span><br><span class="line"></span><br><span class="line">	doc in part2</span><br><span class="line"></span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line"></span><br><span class="line">=cut	# pod文档结束的标志</span><br></pre></td></tr></table></figure>
<p>注意：每个=标签上下必须隔一行，否则就会错误解析。</p>
<p>用<code>pod2doc $0</code>可以将程序中的文档打印出来，不过一般用在程序内部，当程序参数设定错误时打印pod文档：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">die `pod2doc $0` if (...);</span><br></pre></td></tr></table></figure>
<h3 id="实现实例"><a href="#实现实例" class="headerlink" title="实现实例"></a>实现实例</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line">use POSIX;</span><br><span class="line"></span><br><span class="line"># 帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">	This script is used to split fasta file, which is too large with thosands of sequence</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">	$0 -i &lt;input&gt; -o &lt;output_dir&gt; [-n &lt;seq_num_per_file&gt;] [-m &lt;output_file_num&gt;]</span><br><span class="line">	</span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">	-i	[str]	Input raw fasta file</span><br><span class="line">	-o	[str]	Output file to which directory</span><br><span class="line">	-n	[int]	Sequence number per file, alternate chose paramerter &quot;-n&quot; or &quot;-m&quot;, if set &quot;-n&quot; and &quot;-m&quot; at the same time, only take &quot;-n&quot; parameter</span><br><span class="line">	-m	[int]	Output file number (default:100)</span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($input,$output_dir,$seq_num,$file_num);</span><br><span class="line">GetOptions(</span><br><span class="line">	&quot;i:s&quot;=&gt;\$input,</span><br><span class="line">	&quot;o:s&quot;=&gt;\$output_dir,</span><br><span class="line">	&quot;n:i&quot;=&gt;\$seq_num,</span><br><span class="line">	&quot;m:i&quot;=&gt;\$file_num</span><br><span class="line">	);</span><br><span class="line"></span><br><span class="line">die `pod2text $0` if ((!$input) or (!$output_dir));</span><br><span class="line"></span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">.</span><br></pre></td></tr></table></figure>
<h2 id="Shell"><a href="#Shell" class="headerlink" title="Shell"></a>Shell</h2><h3 id="Shell中的getopt传参"><a href="#Shell中的getopt传参" class="headerlink" title="Shell中的getopt传参"></a>Shell中的getopt传参</h3><p>Shell中的这个功能可以通过getopts函数实现</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">getopts [option[:]] [DESCPRITION] VARIABLE</span><br></pre></td></tr></table></figure>
<blockquote>
<p><code>option</code>：表示为某个脚本可以使用的选项</p>
<p><code>&quot;:&quot;</code>：如果某个选项（option）后面出现了冒号（”:”），则表示这个选项后面可以接参数（即一段描述信息DESCPRITION）</p>
<p><code>VARIABLE</code>：表示将某个选项保存在变量VARIABLE中</p>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">while getopts &quot;:a:b:c:&quot; opt</span><br><span class="line">do</span><br><span class="line">    case $opt in</span><br><span class="line">        a)</span><br><span class="line">        echo &quot;参数a的值$OPTARG&quot;</span><br><span class="line">        ;;</span><br><span class="line">        b)</span><br><span class="line">        echo &quot;参数b的值$OPTARG&quot;</span><br><span class="line">        ;;</span><br><span class="line">        c)</span><br><span class="line">        echo &quot;参数c的值$OPTARG&quot;</span><br><span class="line">        ;;</span><br><span class="line">        ?)</span><br><span class="line">        echo &quot;未知参数&quot;</span><br><span class="line">        exit 1;;</span><br><span class="line">    esac</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<h3 id="Shell中输出帮助文档"><a href="#Shell中输出帮助文档" class="headerlink" title="Shell中输出帮助文档"></a>Shell中输出帮助文档</h3><p>在Shell中编辑一个<code>helpdoc( )</code>的函数即可实现输出帮助文档</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">helpdoc()&#123;</span><br><span class="line">    cat &lt;&lt;EOF</span><br><span class="line">Description:</span><br><span class="line"></span><br><span class="line">    .</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line"></span><br><span class="line">Usage:</span><br><span class="line"></span><br><span class="line">    $0 -a &lt;argv1&gt; -b &lt;argv2&gt; -c &lt;argv3&gt; ...</span><br><span class="line"></span><br><span class="line">Option:</span><br><span class="line"></span><br><span class="line">    .</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">EOF</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>将你想要打印出来的帮助信息写在<code>cat &lt;&lt;EOF</code>和<code>EOF</code>之间</p>
<p>之所以使用<code>EOF</code>来编写是因为，只是一种所见即所得的文本编辑形式（这个不好解释，请自行百度）</p>
<p>当你要打印帮助文档时，直接调用执行<code>helpdoc( )</code>函数即可</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 当没有指定参数时，即参数个数为0时，输出帮助文档并退出程序执行</span><br><span class="line">if [ $# = 0 ]</span><br><span class="line">then</span><br><span class="line">	helpdoc()</span><br><span class="line">	exit 1</span><br><span class="line">fi</span><br></pre></td></tr></table></figure>
<h3 id="实现实例-1"><a href="#实现实例-1" class="headerlink" title="实现实例"></a>实现实例</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br></pre></td><td class="code"><pre><span class="line">helpdoc()&#123;</span><br><span class="line">    cat &lt;&lt;EOF</span><br><span class="line">Description:</span><br><span class="line"></span><br><span class="line">    This shellscript is used to run the pipeline to call snp using GATK4</span><br><span class="line">    - Data merge: merge multi-BAM files coresponding to specified strain</span><br><span class="line">    - Data pre-processing: Mark Duplicates + Base (Quality Score) Recalibration</span><br><span class="line">    - Call variants per-sample</span><br><span class="line">    - Filter Variants: hard-filtering</span><br><span class="line"></span><br><span class="line">Usage:</span><br><span class="line"></span><br><span class="line">    $0 -S &lt;strain name&gt; -R &lt;bwa index&gt; -k &lt;known-site&gt; -i &lt;intervals list&gt;</span><br><span class="line"></span><br><span class="line">Option:</span><br><span class="line"></span><br><span class="line">    -S    strain name, if exist character &quot;/&quot;, place &quot;/&quot; with &quot;_&quot; (Required)</span><br><span class="line">    -R    the path of bwa index (Required)</span><br><span class="line">    -k    known-sites variants VCF file</span><br><span class="line">    -i    intervals list file,must be sorted (Required)</span><br><span class="line">EOF</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">workdir=&quot;/work_dir&quot;</span><br><span class="line"></span><br><span class="line"># 若无指定任何参数则输出帮助文档</span><br><span class="line">if [ $# = 0 ]</span><br><span class="line">then</span><br><span class="line">    helpdoc</span><br><span class="line">    exit 1</span><br><span class="line">fi</span><br><span class="line"></span><br><span class="line">while getopts &quot;hS:k:R:i:&quot; opt</span><br><span class="line">do</span><br><span class="line">    case $opt in</span><br><span class="line">        h)</span><br><span class="line">            helpdoc</span><br><span class="line">            exit 0</span><br><span class="line">            ;;</span><br><span class="line">        S)</span><br><span class="line">            strain=$OPTARG</span><br><span class="line">            # 检测输入的strain名是否合格：是否含有非法字符&quot;/&quot;</span><br><span class="line">            if [[ $strain =~ &quot;/&quot; ]]</span><br><span class="line">            then</span><br><span class="line">                echo &quot;Error in specifing strain name, if exist character \&quot;/\&quot;, place \&quot;/\&quot; with \&quot;_\&quot;&quot;</span><br><span class="line">                helpdoc</span><br><span class="line">                exit 1</span><br><span class="line">            fi</span><br><span class="line">            if [ ! -d $workdir/SAM/$strain ]</span><br><span class="line">            then</span><br><span class="line">                echo &quot;There is no such folder coresponding to $strain&quot;</span><br><span class="line">                helpdoc</span><br><span class="line">                exit 1</span><br><span class="line">            fi</span><br><span class="line">            ;;</span><br><span class="line">        R)</span><br><span class="line">            index=$OPTARG</span><br><span class="line">            ;;</span><br><span class="line">        k)</span><br><span class="line">            vcf=$OPTARG</span><br><span class="line">            if [ ! -f $vcf ]</span><br><span class="line">            then</span><br><span class="line">                echo &quot;No such file: $vcf&quot;</span><br><span class="line">                helpdoc</span><br><span class="line">                exit 1</span><br><span class="line">            fi</span><br><span class="line">            ;;    </span><br><span class="line">        i)</span><br><span class="line">            intervals=$OPTARG</span><br><span class="line">            if [ ! -f $bed ]</span><br><span class="line">            then</span><br><span class="line">                echo &quot;No such file: $intervals&quot;</span><br><span class="line">                helpdoc</span><br><span class="line">                exit 1</span><br><span class="line">            fi</span><br><span class="line">            ;;</span><br><span class="line">        ?)</span><br><span class="line">            echo &quot;Unknown option: $opt&quot;</span><br><span class="line">            helpdoc</span><br><span class="line">            exit 1</span><br><span class="line">            ;;</span><br><span class="line">    esac</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">.</span><br></pre></td></tr></table></figure>
<h2 id="Python"><a href="#Python" class="headerlink" title="Python"></a>Python</h2><h3 id="Python中的getopt传参"><a href="#Python中的getopt传参" class="headerlink" title="Python中的getopt传参"></a>Python中的getopt传参</h3><p>Python中的这种功能需要通过<code>getopt</code>模块实现</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">import getopt</span><br></pre></td></tr></table></figure>
<p>Python脚本获得成对的参数名和参数值后，会分别把它们保存在一个字典变量中，参数名为key，参数值为value</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">opts,args = getopt.getopt(argv,&quot;hi:o:t:n:&quot;,[&quot;ifile=&quot;,&quot;ofile=&quot;,&quot;time=&quot;])</span><br></pre></td></tr></table></figure>
<p>getopt函数的使用说明：</p>
<blockquote>
<p><code>argv</code>：使用<code>argv</code>过滤掉第一个参数（它是执行脚本的名字，不应算作参数的一部分）</p>
<p><code>&quot;hi:o:t:n:&quot;</code>：使用短格式分析串，当一个选项只是表示开关状态时，即后面不带附加参数时，在分析串中写入选项字符。当选项后面是带一个附加参数时，在分析串中写入选项字符同时后面加一个”:” 号</p>
<p><code>[&quot;ifile=&quot;,&quot;ofile=&quot;,&quot;time=&quot;]</code>： 使用长格式分析串列表，长格式串也可以有开关状态，即后面不跟”=” 号。如果跟一个等号则表示后面还应有一个参数 </p>
</blockquote>
<p>然后通过条件判断的方法对参数进行解析：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">for opt,arg in opts:</span><br><span class="line">	if opt in (&quot;-h&quot;,&quot;--help&quot;):</span><br><span class="line">		print(helpdoc)</span><br><span class="line">		sys.exit()</span><br><span class="line">	elif opt in (&quot;-i&quot;,&quot;--ifile&quot;):</span><br><span class="line">		infile = arg</span><br><span class="line">	elif opt in (&quot;-t&quot;,&quot;--time&quot;):</span><br><span class="line">		sleep_time = int(arg)</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br></pre></td></tr></table></figure>
<h3 id="Python中的argparse传参"><a href="#Python中的argparse传参" class="headerlink" title="Python中的argparse传参"></a>Python中的argparse传参</h3><p>该示例例来自<strong>黄树嘉</strong>大佬的github项目 <a href="https://github.com/ShujiaHuang/cmdbtools/blob/master/cmdbtools/cmdbtools.py" target="_blank" rel="noopener">cmdbtools</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">import argparse    //导入命令行解析的库文件</span><br><span class="line"></span><br><span class="line">// 为了别人执行代码的时候用--help看出来怎么使用这些代码</span><br><span class="line">argparser = argparse.ArgumentParser(description=&apos;Manage authentication for CMDB API and do querying from command line.&apos;) </span><br><span class="line">// 添加子命令</span><br><span class="line">commands = argparser.add_subparsers(dest=&apos;command&apos;, title=&apos;Commands&apos;)</span><br><span class="line"></span><br><span class="line">// 分别定义各个子命令，并未各个子命令设置参数</span><br><span class="line"></span><br><span class="line">// 1. 定义子命令login</span><br><span class="line">login_command = commands.add_parser(&apos;login&apos;, help=&apos;Authorize access to CMDB API.&apos;)</span><br><span class="line">login_command.add_argument(&apos;-k&apos;, &apos;--token&apos;, type=str, required=True, dest=&apos;token&apos;,help=&apos;CMDB API access key(Token).&apos;) // 给子命令添加&apos;-k&apos;参数</span><br><span class="line"></span><br><span class="line">// 2. 定义子命令logout</span><br><span class="line">logout_command = commands.add_parser(&apos;logout&apos;, help=&apos;Logout CMDB.&apos;)</span><br><span class="line"></span><br><span class="line">// 3. 定义子命令token</span><br><span class="line">token_command = commands.add_parser(&apos;print-access-token&apos;, help=&apos;Display access token for CMDB API.&apos;)</span><br><span class="line"></span><br><span class="line">// 4. 定义子命令annotate</span><br><span class="line">annotate_command = commands.add_parser(&apos;annotate&apos;, help=&apos;Annotate input VCF.&apos;,</span><br><span class="line">		description=&apos;Input VCF file. Multi-allelic variant records in input VCF must be split into multiple bi-allelic variant records.&apos;)</span><br><span class="line">annotate_command.add_argument(&apos;-i&apos;, &apos;--vcffile&apos;, metavar=&apos;VCF_FILE&apos;, type=str, required=True, dest=&apos;in_vcffile&apos;,help=&apos;input VCF file.&apos;)</span><br><span class="line"></span><br><span class="line">// 5. 定义子命令query_variant</span><br><span class="line">query_variant_command = commands.add_parser(&apos;query-variant&apos;,</span><br><span class="line">											help=&apos;Query variant by variant identifier or by chromosome name and chromosomal position.&apos;,</span><br><span class="line">											description=&apos;Query variant by identifier chromosome name and chromosomal position.&apos;)</span><br><span class="line">query_variant_command.add_argument(&apos;-c&apos;, &apos;--chromosome&apos;, metavar=&apos;name&apos;, type=str, dest=&apos;chromosome&apos;,help=&apos;Chromosome name.&apos;, default=None)</span><br><span class="line">query_variant_command.add_argument(&apos;-p&apos;, &apos;--position&apos;, metavar=&apos;genome-position&apos;, type=int, dest=&apos;position&apos;,help=&apos;Genome position.&apos;, default=None)</span><br><span class="line">query_variant_command.add_argument(&apos;-l&apos;, &apos;--positions&apos;, metavar=&apos;File-contain-a-list-of-genome-positions&apos;,</span><br><span class="line">									type=str, dest=&apos;positions&apos;,</span><br><span class="line">									help=&apos;Genome positions list in a file. One for each line. You can input single &apos;</span><br><span class="line">										&apos;position by -c and -p or using -l for multiple poisitions in a single file, &apos;</span><br><span class="line">										&apos;could be .gz file&apos;,</span><br><span class="line">									default=None)</span><br></pre></td></tr></table></figure>
<p>输出的主命令的帮助文档如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">$ cmdbtools --help</span><br><span class="line">usage: cmdbtools [-h]</span><br><span class="line">                &#123;login,logout,print-access-token,annotate,query-variant&#125; ...</span><br><span class="line"></span><br><span class="line">Manage authentication for CMDB API and do querying from command line.</span><br><span class="line"></span><br><span class="line">optional arguments:</span><br><span class="line"> -h, --help            show this help message and exit</span><br><span class="line"></span><br><span class="line">Commands:</span><br><span class="line"> &#123;login,logout,print-access-token,annotate,query-variant&#125;</span><br><span class="line">   login               Authorize access to CMDB API.</span><br><span class="line">   logout              Logout CMDB.</span><br><span class="line">   print-access-token  Display access token for CMDB API.</span><br><span class="line">   annotate            Annotate input VCF.</span><br><span class="line">   query-variant       Query variant by variant identifier or by chromosome</span><br><span class="line">                       name and chromosomal position.</span><br></pre></td></tr></table></figure>
<h3 id="Python中输出帮助文档"><a href="#Python中输出帮助文档" class="headerlink" title="Python中输出帮助文档"></a>Python中输出帮助文档</h3><p>在Python中创建一个字符串变量<code>helpdoc</code>即可实现输出帮助文档</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">helpdoc = &apos;&apos;&apos;</span><br><span class="line">Description</span><br><span class="line"></span><br><span class="line">	...</span><br><span class="line"></span><br><span class="line">Usage</span><br><span class="line"></span><br><span class="line">	python pyscript.py -i/--ifile &lt;input file&gt; -o/--ofile &lt;output file&gt; -t/--time &lt;int&gt; ...</span><br><span class="line"></span><br><span class="line">Parameters</span><br><span class="line"></span><br><span class="line">	-h/--help</span><br><span class="line">		Print helpdoc</span><br><span class="line">	-i/--ifile</span><br><span class="line">		Input file, including only one column with sampleId</span><br><span class="line">	-o/--ofile</span><br><span class="line">		Output file, including two columns, the 1st column is sampleId, the 2nd column is attribute information</span><br><span class="line">	-t/--time</span><br><span class="line">		Time for interval (seconds, default 5s)</span><br><span class="line">	...</span><br><span class="line">&apos;&apos;&apos;</span><br></pre></td></tr></table></figure>
<p>在需要时将这个变量打印出来即可：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">try:</span><br><span class="line">	opts,args = getopt.getopt(argv,&quot;hi:o:t:n:&quot;,[&quot;ifile=&quot;,&quot;ofile=&quot;,&quot;time=&quot;])</span><br><span class="line">	if len(opts) == 0:</span><br><span class="line">		print(&quot;Options Error!\n\n&quot;+helpdoc)</span><br><span class="line">		sys.exit(2)</span><br><span class="line">except getopt.GetoptError:</span><br><span class="line">	print(&quot;Options Error!\n\n&quot;+helpdoc)</span><br><span class="line">	sys.exit(2)</span><br></pre></td></tr></table></figure>
<h3 id="实现实例-2"><a href="#实现实例-2" class="headerlink" title="实现实例"></a>实现实例</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><span class="line">import getopt</span><br><span class="line"></span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line">.</span><br><span class="line"></span><br><span class="line">if __name__ == &apos;__main__&apos;:</span><br><span class="line">	...</span><br><span class="line">	helpdoc = &apos;&apos;&apos;</span><br><span class="line">Description</span><br><span class="line"></span><br><span class="line">    This script is used to grab SRA sample attributes information based on SampleId</span><br><span class="line"></span><br><span class="line">Usage</span><br><span class="line"></span><br><span class="line">    python webspider_ncbiBiosample.py -i/--ifile &lt;input file&gt; -o/--ofile &lt;output file&gt; -t/--time &lt;int&gt; -n/--requests-number &lt;int&gt;</span><br><span class="line"></span><br><span class="line">Parameters</span><br><span class="line"></span><br><span class="line">    -h/--help</span><br><span class="line">        Print helpdoc</span><br><span class="line">    -i/--ifile</span><br><span class="line">        Input file, including only one column with sampleId</span><br><span class="line">    -o/--ofile</span><br><span class="line">        Output file, including two columns, the 1st column is sampleId, the 2nd column is attribute information</span><br><span class="line">    -t/--time</span><br><span class="line">        Time for interval (seconds, default 5s)</span><br><span class="line">    -n/--requests-number</span><br><span class="line">        Setting the requests number between interval (default 10)</span><br><span class="line">&apos;&apos;&apos;</span><br><span class="line"></span><br><span class="line">    # 获取命令行参数</span><br><span class="line">    try:</span><br><span class="line">        opts,args = getopt.getopt(argv,&quot;hi:o:t:n:&quot;,[&quot;ifile=&quot;,&quot;ofile=&quot;,&quot;time=&quot;])</span><br><span class="line">        if len(opts) == 0:</span><br><span class="line">            print(&quot;Options Error!\n\n&quot;+helpdoc)</span><br><span class="line">            sys.exit(2)</span><br><span class="line">    except getopt.GetoptError:</span><br><span class="line">        print(&quot;Options Error!\n\n&quot;+helpdoc)</span><br><span class="line">        sys.exit(2)</span><br><span class="line"></span><br><span class="line">    # 设置参数</span><br><span class="line">    for opt,arg in opts:</span><br><span class="line">        if opt in (&quot;-h&quot;,&quot;--help&quot;):</span><br><span class="line">            print(helpdoc)</span><br><span class="line">            sys.exit()</span><br><span class="line">        elif opt in (&quot;-i&quot;,&quot;--ifile&quot;):</span><br><span class="line">            infile = arg</span><br><span class="line">        elif opt in (&quot;-o&quot;,&quot;--ofile&quot;):</span><br><span class="line">            outfile = arg</span><br><span class="line">            # 若指定的输出文件已经存在，让用户决定覆盖该文件，还是直接退出程序</span><br><span class="line">            if os.path.exists(outfile):</span><br><span class="line">                keyin = input(&quot;The output file you specified exists, rewrite it?([y]/n: &quot;)</span><br><span class="line">                if keyin in (&quot;y&quot;,&quot;Y&quot;,&quot;&quot;):</span><br><span class="line">                    os.remove(outfile)</span><br><span class="line">                elif keyin in (&quot;n&quot;,&quot;N&quot;):</span><br><span class="line">                    print(&quot;The output file existed!\n&quot;)</span><br><span class="line">                    sys.exit(2)</span><br><span class="line">                else:</span><br><span class="line">                    print(&quot;Input error!\n&quot;)</span><br><span class="line">                    sys.exit(2)</span><br><span class="line">        elif opt in (&quot;-t&quot;,&quot;--time&quot;):</span><br><span class="line">            sleep_time = int(arg)</span><br><span class="line">        elif opt in (&quot;-n&quot;,&quot;--requests-number&quot;):</span><br><span class="line">            requestNum = int(arg)</span><br><span class="line">	</span><br><span class="line">	.</span><br><span class="line">	.</span><br><span class="line">	.</span><br></pre></td></tr></table></figure>
<hr>
<p>参考资料：</p>
<p>(1) <a href="https://github.com/Ming-Lian/Bioinformatics-skills/blob/master/Perl%E8%BF%9B%E9%98%B6%E7%AC%94%E8%AE%B0.md" target="_blank" rel="noopener">本人github笔记：Perl进阶笔记</a></p>
<p>(2) <a href="https://github.com/Ming-Lian/Bioinformatics-skills/blob/master/%E5%AE%9E%E7%94%A8%E5%B0%8F%E8%84%9A%E6%9C%AC.md" target="_blank" rel="noopener">本人github笔记：实用小脚本</a></p>
<p>(3) <a href="https://github.com/Ming-Lian/Hello-RaspberryPi/blob/master/Linux%E6%93%8D%E4%BD%9C%E8%BF%9B%E9%98%B6.md#shell-programing" target="_blank" rel="noopener">本人github笔记：Linux (Raspbian) 操作进阶——Shell编程</a></p>
<p>(4) <a href="https://www.cnblogs.com/kex1n/p/5975722.html" target="_blank" rel="noopener">Python 命令行参数和getopt模块详解</a></p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/08/RNA-seq/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/08/RNA-seq/" class="post-title-link" itemprop="url">多套RNA-seq分析流程</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-08 23:29:11" itemprop="dateCreated datePublished" datetime="2019-02-08T23:29:11+00:00">2019-02-08</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-02-17 06:12:08" itemprop="dateModified" datetime="2019-02-17T06:12:08+00:00">2019-02-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p><img src="/images/hisat2-stringtie-ballgown.png" alt></p>
<h2 id="1-测序数据下载"><a href="#1-测序数据下载" class="headerlink" title="1. 测序数据下载"></a>1. 测序数据下载</h2><p>参见： <a href="https://github.com/Ming-Lian/Memo/blob/master/ChIP-seq-pipeline.md#get-data" target="_blank" rel="noopener">https://github.com/Ming-Lian/Memo/blob/master/ChIP-seq-pipeline.md#get-data</a></p>
<h2 id="2-比对与定量"><a href="#2-比对与定量" class="headerlink" title="2. 比对与定量"></a>2. 比对与定量</h2><h3 id="2-1-Salmon流程"><a href="#2-1-Salmon流程" class="headerlink" title="2.1. Salmon流程"></a>2.1. Salmon流程</h3><p>不需要比对，直接对转录水平进行定量</p>
<h4 id="2-1-1-创建索引"><a href="#2-1-1-创建索引" class="headerlink" title="2.1.1. 创建索引"></a>2.1.1. 创建索引</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ salmon index -t Arabidopsis_thaliana.TAIR10.28.cdna.all.fa.gz -i athal_index_salmon</span><br></pre></td></tr></table></figure>
<h4 id="2-1-2-定量"><a href="#2-1-2-定量" class="headerlink" title="2.1.2. 定量"></a>2.1.2. 定量</h4><p>salmon quant 有两种运行模式：</p>
<blockquote>
<ul>
<li><strong>Salmon’s quasi-mapping-based mode</strong>： using raw reads</li>
</ul>
<p>输入以下命令查看该模式下的help文档</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; salmon quant --help-reads</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<blockquote>
<ul>
<li><strong>Salmon’s alignment-based mode</strong>： makes use of already-aligned reads (in BAM/SAM format)</li>
</ul>
<p>当使用参数<code>-a</code>时，启用该模式，否则默认使用quasi-mapping-based mode</p>
<p>输入以下命令查看该模式下的help文档</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; salmon quant --help-alignment</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">#! /bin/bash</span><br><span class="line">index=~/rna_seq_practice_2/data/ref/athal_index_salmon ## 指定索引文件夹</span><br><span class="line">quant=~/rna_seq_practice_2/work/quant_salmon # 指定定量文件输出文件夹</span><br><span class="line">for fn in ERR1698&#123;194..209&#125;;</span><br><span class="line">do</span><br><span class="line">	sample=`basename $&#123;fn&#125;` # basename命令用于去掉路径信息，返回纯粹的文件名，如果指定的文件的扩展名，则将扩展名也一并去掉。</span><br><span class="line">	echo &quot;Processin sample $&#123;sampe&#125;&quot;</span><br><span class="line">	salmon quant -i $index -l A \</span><br><span class="line">		-1 $&#123;sample&#125;_1.fastq.gz \</span><br><span class="line">		-2 $&#123;sample&#125;_2.fastq.gz \</span><br><span class="line">		-p 5 -o $quant/$&#123;sample&#125;_quant</span><br><span class="line">done</span><br><span class="line"># quant_salmon.sh</span><br></pre></td></tr></table></figure>
<p>salmon quant 参数：</p>
<blockquote>
<ul>
<li>-i Salmon index</li>
<li>-l Format string describing the library type. <code>-l A</code> tells salmon that it should automatically determine the library type of the sequencing reads (e.g. stranded vs. unstranded etc.)</li>
<li>-1 File containing the #1 mates</li>
<li>-2 File containing the #2 mates</li>
<li>-p The number of threads to use</li>
<li>-o Output quantification file.</li>
</ul>
</blockquote>
<h3 id="2-2-subread流程"><a href="#2-2-subread流程" class="headerlink" title="2.2. subread流程"></a>2.2. subread流程</h3><h4 id="2-2-1-创建索引"><a href="#2-2-1-创建索引" class="headerlink" title="2.2.1. 创建索引"></a>2.2.1. 创建索引</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ gunzip Arabidopsis_thaliana.TAIR10.28.dna.genome.fa.gz</span><br><span class="line">$ subread-buildindex -o athal_index_subread   Arabidopsis_thaliana.TAIR10.28.dna.genome.fa</span><br></pre></td></tr></table></figure>
<h4 id="2-2-2-比对"><a href="#2-2-2-比对" class="headerlink" title="2.2.2. 比对"></a>2.2.2. 比对</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">#! /bin/bash</span><br><span class="line">index=~/rna_seq_practice_2/data/ref/athal_index_subread</span><br><span class="line">map=~/rna_seq_practice_2/work/map</span><br><span class="line">for fn in ERR1698&#123;194..209&#125;;</span><br><span class="line">do</span><br><span class="line">	sample=`basename $&#123;fn&#125;`</span><br><span class="line">	echo &quot;Processin sample $&#123;sampe&#125;&quot; </span><br><span class="line">	subjunc -i $index \</span><br><span class="line">		-r $&#123;sample&#125;_1.fastq.gz \</span><br><span class="line">		-R $&#123;sample&#125;_2.fastq.gz \</span><br><span class="line">		-T 5 -o $map/$&#123;sample&#125;_subjunc.bam</span><br><span class="line"># 比对的sam自动转为bam，但是并不按照参考基因组坐标排序</span><br><span class="line">done</span><br><span class="line"># map_subjunc.sh</span><br></pre></td></tr></table></figure>
<h4 id="2-2-3-定量"><a href="#2-2-3-定量" class="headerlink" title="2.2.3. 定量"></a>2.2.3. 定量</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">featureCounts=~/anaconda2/bin</span><br><span class="line"># gff3=~/rna_seq_practice_2/data/ref/Arabidopsis_thaliana.TAIR10.28.gff3.gz</span><br><span class="line">gtf=~/rna_seq_practice_2/data/ref/Arabidopsis_thaliana.TAIR10.28.gtf</span><br><span class="line">count=~/rna_seq_practice_2/work/quant_subjunc</span><br><span class="line">nohup $featureCounts/featureCounts  -T 5 -p -t exon -g gene_name -a $gtf -o  $count/counts.txt   *.bam &amp;</span><br><span class="line">nohup $featureCounts/featureCounts  -T 5 -p -t exon -g gene_id -a $gtf -o  $count/counts_id.txt   *.bam &amp;</span><br></pre></td></tr></table></figure>
<h3 id="2-3-hisat2-stringtie流程"><a href="#2-3-hisat2-stringtie流程" class="headerlink" title="2.3. hisat2-stringtie流程"></a>2.3. hisat2-stringtie流程</h3><h4 id="2-3-1-hisat2创建索引"><a href="#2-3-1-hisat2创建索引" class="headerlink" title="2.3.1. hisat2创建索引"></a>2.3.1. hisat2创建索引</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"># build reference index</span><br><span class="line">##  using the python scripts included in the HISAT2 package, extract splice-site and exon information from the gene</span><br><span class="line">annotation fle</span><br><span class="line">$ extract_splice_sites.py chrX_data/genes/chrX.gtf &gt;chrX.ss</span><br><span class="line">$ extract_exons.py chrX_data/genes/chrX.gtf &gt;chrX.exon</span><br><span class="line">##  build a HISAT2 index</span><br><span class="line">$ hisat2-build --ss chrX.ss --exon chrX.exon chrX_data/genome/chrX.fa chrX_tran</span><br></pre></td></tr></table></figure>
<h4 id="2-3-2-hisat2比对"><a href="#2-3-2-hisat2比对" class="headerlink" title="2.3.2. hisat2比对"></a>2.3.2. hisat2比对</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ hisat2 -p 10 --dta -x chrX_tran -1 reads1_1.fastq -2 reads1_2.fastq | samtools sort -@ 8 -O bam -o reads1.sort.bam 1&gt;map.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<p><code>Usage: hisat2 [options]* -x &lt;ht2-idx&gt; {-1 &lt;m1&gt; -2 &lt;m2&gt; | -U &lt;r&gt;} [-S &lt;sam&gt;]</code></p>
<blockquote>
<ul>
<li>-p Number of threads to use</li>
<li>–dta reports alignments tailored for transcript assemblers</li>
<li>-x Hisat2 index</li>
<li>-1 The 1st input fastq file of paired-end reads</li>
<li>-2 The 2nd input fastq file of paired-end reads</li>
<li>-S File for SAM output (default: stdout)</li>
</ul>
</blockquote>
<h4 id="2-3-3-stringtie转录本拼接"><a href="#2-3-3-stringtie转录本拼接" class="headerlink" title="2.3.3. stringtie转录本拼接"></a>2.3.3. stringtie转录本拼接</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ stringtie -p 16 -G Ref/hg19/grch37_tran/Homo_sapiens.GRCh37.75.gtf -o Asm/read1.gtf -l prefix Map/read1.bam 1&gt;Asm/read1_strg_assm.log 2&gt;&amp;1</span><br><span class="line"># 样本间转录本合并</span><br><span class="line">$ stringtie --merge -p 16 -G Ref/hg19/grch37_tran/Homo_sapiens.GRCh37.75.gtf -o Asm/merge.gtf Asm/mergelist.txt 1&gt;Asm/strg_merge.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<p><code>Transcript merge usage mode: stringtie --merge [Options] { gtf_list | strg1.gtf ...}</code></p>
<blockquote>
<ul>
<li>-p number of threads (CPUs) to use</li>
<li>-G reference annotation to include in the merging (GTF/GFF3)</li>
<li>-o output path/file name for the assembled transcripts GTF</li>
<li>-l name prefix for output transcripts (default: STRG)</li>
</ul>
</blockquote>
<h4 id="2-3-4-stringtie定量"><a href="#2-3-4-stringtie定量" class="headerlink" title="2.3.4. stringtie定量"></a>2.3.4. stringtie定量</h4><ol>
<li><p>以ballgown格式输出</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ stringtie -e -B -p 16 -G Asm/merge.gtf -o quant/read1/read1.gtf \</span><br><span class="line">	Map/read1.bam 1&gt;quant/read1/read1_strg_quant.log 2&gt;&amp;1</span><br></pre></td></tr></table></figure>
<blockquote>
<ul>
<li>-e only estimate the abundance of given reference transcripts (requires -G)</li>
<li>-B enable output of Ballgown table files which will be created in the same directory as the output GTF (requires -G, -o recommended)</li>
</ul>
</blockquote>
</li>
<li><p>以read count进行定量，作为DESeq2或edgeR的输入</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ stringtie -e -p 16 -G Asm/merge.gtf -o quant/read1/read1.gtf \</span><br><span class="line">	Map/read1.bam 1&gt;quant/read1/read1_strg_quant.log 2&gt;&amp;1</span><br><span class="line">$ python prepDE.py -i sample_lst.txt</span><br></pre></td></tr></table></figure>
</li>
</ol>
<pre><code>注意：
- stringtie的用法与上面相同，除了少了一个参数`-B`
- `prepDE.py`脚本需要到stringtie官网下载：http://ccb.jhu.edu/software/stringtie/dl/prepDE.py ，注意该脚本是用**python2**编写的
- `prepDE.py` 会以csv格式，分别输出基因和转录本的count matrices

prepDE.py参数

&gt; - -i the parent directory of the sample sub-directories or a textfile listing the paths to GTF files [default: ballgown]
&gt; - -g where to output the gene count matrix [default: gene_count_matrix.csv]
&gt; - -t where to output the transcript count matrix [default: transcript_count_matrix.csv]

samplelist textfile 格式如下：

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">ERR188021 &lt;PATH_TO_ERR188021.gtf&gt;</span><br><span class="line">ERR188023 &lt;PATH_TO_ERR188023.gtf&gt;</span><br><span class="line">ERR188024 &lt;PATH_TO_ERR188024.gtf&gt;</span><br><span class="line">ERR188025 &lt;PATH_TO_ERR188025.gtf&gt;</span><br><span class="line">ERR188027 &lt;PATH_TO_ERR188027.gtf&gt;</span><br><span class="line">ERR188028 &lt;PATH_TO_ERR188028.gtf&gt;</span><br><span class="line">ERR188030 &lt;PATH_TO_ERR188030.gtf&gt;</span><br><span class="line">ERR188033 &lt;PATH_TO_ERR188033.gtf&gt;</span><br><span class="line">ERR188034 &lt;PATH_TO_ERR188034.gtf&gt;</span><br><span class="line">ERR188037 &lt;PATH_TO_ERR188037.gtf&gt;</span><br></pre></td></tr></table></figure>
</code></pre><h3 id="2-4-RSEM流程"><a href="#2-4-RSEM流程" class="headerlink" title="2.4. RSEM流程"></a>2.4. RSEM流程</h3><p>RSEM属于Alignment-based transcript quantification的转录本定量工具的一种，也就是先比对后定量</p>
<p>RSEM最早被广泛应用于无参转录组的定量分析，因为无参转录组需要对reads进行拼接，然后将reads比对至拼接的转录本上，再通过定量获得其表达丰度</p>
<p>RSEM是在2010年发表的，最新更新是在2016年，下载地址则 <code>http://deweylab.github.io/RSEM/</code>，下载后编译下即可</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ make</span><br><span class="line">$ make install # 默认安装在/usr/local/bin目录下</span><br><span class="line"># 自定义安装路径</span><br><span class="line"># 若要修改安装路径需要设置DESRDIR和prefix参数，默认情况下DESRDIR=&apos;&apos;，prefix=/usr/local，最终的安装目录为 $DESTDIR/$prefix/bin</span><br><span class="line">make install DESRDIR=/Path/To/Destdir prefix=</span><br></pre></td></tr></table></figure>
<p>RSEM整体上来说是属于定量软件，但其支持调用其他比对软件，如Bowtie，Bowtie2和STAR，来将reads比对至转录本上，所以必须至少预先安装上述3款比对软件中的一种，且要将它们的安装路径加入到环境变量<code>PATH</code>中</p>
<h4 id="2-4-1-创建索引"><a href="#2-4-1-创建索引" class="headerlink" title="2.4.1. 创建索引"></a>2.4.1. 创建索引</h4><p>这步可以理解为是对转录本建索引，RSEM支持两种方式：</p>
<blockquote>
<ol>
<li><p><strong>提供参考基因组fa序列和GTF注释文件</strong></p>
<p> RSEM则会通过GTF注释文件从参考基因组序列中提取出各个转录本的序列，然后利用Bowtie2 or STAR等软件来建索引</p>
 <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; 	$ ~/biosoft/rsem/RSEM-1.3.0/rsem-prepare-reference \</span><br><span class="line">&gt; 		-gtf ~/annotation/hg38/gencode.v27.annotation.gtf \</span><br><span class="line">&gt; 		--bowtie2 \</span><br><span class="line">&gt; 		~/reference/genome/hg38/GRCh38.p10.genome.fa ~/reference/index/RSEM/hg38/hg38</span><br><span class="line">&gt;</span><br></pre></td></tr></table></figure>
</li>
</ol>
</blockquote>
<blockquote>
<ol start="2">
<li><p><strong>直接提供转录本序列，然后再建索引</strong></p>
<p> 无参转录组一般是这样的</p>
<p> 参数与上面的情况相似，只是不需要设置<code>-gtf</code>参数，若用户不指定<code>-gtf</code>参数，则RSEM会将后面提供的参考序列作为转录本</p>
<p> 如果还想有基于基因水平的定量结果，则需再加–transcript-to-gene-map参数，用于导入转录本和基因的对应关系的文件（一列基因ID，一列对应的转录本ID）</p>
</li>
</ol>
</blockquote>
<p>在~/reference/index/RSEM/hg38/目录下包含有从参考基因组提取出来的转录本的序列文件以及bowtie2的索引文件（以.bt2结尾）等</p>
<h4 id="2-4-2-转录本定量"><a href="#2-4-2-转录本定量" class="headerlink" title="2.4.2. 转录本定量"></a>2.4.2. 转录本定量</h4><p>用RSEM的<code>rsem-calculate-expression</code>命令来对reads进行bowtie2比对以及表达水平的定量</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">$ ~/biosoft/rsem/RSEM-1.3.0/rsem-calculate-expression \</span><br><span class="line">	-p 6 \</span><br><span class="line">	--bowtie2 \</span><br><span class="line">	--append-names \</span><br><span class="line">	--output-genome-bam \</span><br><span class="line">	--paired-end sample_R1.fastq sample_R2.fastq \</span><br><span class="line">	~/reference/index/RSEM/hg38/ \</span><br><span class="line">	RSEM/sample</span><br></pre></td></tr></table></figure>
<p>该命令有以下三种命令行形式：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 1. Single-end，比对与定量</span><br><span class="line">rsem-calculate-expression [options] upstream_read_file(s) reference_name sample_name</span><br><span class="line"># 2. Paired-end，比对与定量</span><br><span class="line">rsem-calculate-expression [options] --paired-end upstream_read_file(s) downstream_read_file(s) reference_name sample_name</span><br><span class="line"># 3. 用户提供SAM/BAM/CRAM格式的比对文件，只进行定量</span><br><span class="line">rsem-calculate-expression [options] --alignments [--paired-end] input reference_name sample_name</span><br></pre></td></tr></table></figure>
<p>参数说明：</p>
<blockquote>
<ul>
<li><code>–paired-end</code> 表明是双端测序数据</li>
<li><code>–bowtie2</code> 指定bowtie2来用于reads比对</li>
<li><code>–append-names</code> 用于在结果中附加上gene name和transcript name</li>
<li><code>–output-genome-bam</code> 结果中输出基于基因水平的bam文件（默认只有转录水平的bam文件）</li>
<li><code>–estimate-rspd</code> 文档中的例子还加了这个参数，用于estimate the read start position distribution，看是否有positional biases</li>
</ul>
</blockquote>
<p>RSEM通过调用bowtie2来完成mapping，其调用bowtie2时使用的默认参数为：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">$ bowtie2 \</span><br><span class="line">	-q \</span><br><span class="line">	--phred33 \</span><br><span class="line">	--sensitive \</span><br><span class="line">	--dpad 0 \</span><br><span class="line">	--gbar 99999999 \	# 不支持有gap的比对</span><br><span class="line">	--mp 1,1 \</span><br><span class="line">	--np 1 \</span><br><span class="line">	--score-min L,0,-0.1 \</span><br><span class="line">	-I 1 -X 1000 --no-mixed \</span><br><span class="line">	--no-discordant \	# 不支持paired reads discordant的比对</span><br><span class="line">	-p 6 \</span><br><span class="line">	-k 200 \	# 输出最佳的200个比对结果</span><br><span class="line">	-x /home/anlan/reference/index/RSEM/hg38/ \</span><br><span class="line">	-1 SRR6269049.clean_1.fastq \</span><br><span class="line">	-2 SRR6269049.clean_2.fastq | samtools view -S -b -o RSEM/SRR6269049.temp/SRR6269049.bam</span><br></pre></td></tr></table></figure>
<p>也可以自己用bowtie2进行mapping将mapping得到的文件作为RSEM定量的输入，但是必须保证：</p>
<blockquote>
<ul>
<li>不支持有gap的比对</li>
<li>不支持paired reads discordant的比对</li>
<li>输出最佳的200个比对结果</li>
</ul>
</blockquote>
<p>否则RSEM会报错！</p>
<p><strong>RSEM对multiple mapping结果的处理是这款工具的亮点</strong>：</p>
<blockquote>
<p>对于RSEM如何处理multimapping reads，RSEM官方教程特意讲了一点</p>
<p>以Ccl6基因的3个转录本的比对结果为例，其第2个转录本全长与第1个转录本重叠，所以在这重叠区域会产生大量的multimapping reads，黑色表示uniquely aligned reads，红色表示multi-mapping reads</p>
<p>RSEM的处理方法是这样的：RSEM先尽量让第1个转录本在其重叠区域的reads（包括unique reads和multimapping reads）分布趋于平滑？因此再将在区域的剩下multimapping reads算在第2个转录本上；至于第3个转录本，其没有reads比对上，所以就空的；总体上如下图所示</p>
</blockquote>
<p><img src="/images/RNA-seq-RSEM.png" alt></p>
<h2 id="3-差异表达分析"><a href="#3-差异表达分析" class="headerlink" title="3. 差异表达分析"></a>3. 差异表达分析</h2><h3 id="3-1-使用DESeq2进行差异分析"><a href="#3-1-使用DESeq2进行差异分析" class="headerlink" title="3.1. 使用DESeq2进行差异分析"></a>3.1. 使用DESeq2进行差异分析</h3><p>DESeq2要求输入的表达矩阵是<strong>read counts</strong></p>
<p>构建 DESeqDataSet 对象</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"># 构建表达矩阵</span><br><span class="line">count_table&lt;-read.csv(&quot;/path/to/gene_count_matrix.csv&quot;,row.names=&quot;gene_id&quot;)</span><br><span class="line">count_matrix&lt;-as.matrix(count_table)</span><br><span class="line"># 构建分组矩阵</span><br><span class="line">group_list&lt;-factor(rep(c(&quot;control&quot;,&quot;treat&quot;),c(3,3)))</span><br><span class="line">colData &lt;- data.frame(row.names=colnames(count_matrix), group_list=group_list)</span><br><span class="line"># 构建 DESeqDataSet 对象</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData = count_matrix,</span><br><span class="line">	colData = colData,</span><br><span class="line">	design = ~ group_list)</span><br></pre></td></tr></table></figure>
<p>表达矩阵的格式为：</p>
<table>
<thead>
<tr>
<th style="text-align:center"><code></code></th>
<th style="text-align:center">SRR1039508</th>
<th style="text-align:center">SRR1039509</th>
<th style="text-align:center">SRR1039512</th>
<th style="text-align:center">SRR1039513</th>
<th style="text-align:center">SRR1039516</th>
<th style="text-align:center">SRR1039517</th>
<th style="text-align:center">SRR1039520</th>
<th style="text-align:center">SRR1039521</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">ENSG00000000003</td>
<td style="text-align:center">679</td>
<td style="text-align:center">448</td>
<td style="text-align:center">873</td>
<td style="text-align:center">408</td>
<td style="text-align:center">1138</td>
<td style="text-align:center">1047</td>
<td style="text-align:center">770</td>
<td style="text-align:center">572</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000000005</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000000419</td>
<td style="text-align:center">467</td>
<td style="text-align:center">515</td>
<td style="text-align:center">621</td>
<td style="text-align:center">365</td>
<td style="text-align:center">587</td>
<td style="text-align:center">799</td>
<td style="text-align:center">417</td>
<td style="text-align:center">508</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000000457</td>
<td style="text-align:center">260</td>
<td style="text-align:center">211</td>
<td style="text-align:center">263</td>
<td style="text-align:center">164</td>
<td style="text-align:center">245</td>
<td style="text-align:center">331</td>
<td style="text-align:center">233</td>
<td style="text-align:center">229</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000000460</td>
<td style="text-align:center">60</td>
<td style="text-align:center"></td>
<td style="text-align:center">55</td>
<td style="text-align:center">40</td>
<td style="text-align:center">35</td>
<td style="text-align:center">78</td>
<td style="text-align:center">63</td>
<td style="text-align:center">76</td>
<td>60</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000000938</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">2</td>
<td style="text-align:center">0</td>
<td style="text-align:center">1</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
<td style="text-align:center">0</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000000971</td>
<td style="text-align:center">3251</td>
<td style="text-align:center">3679</td>
<td style="text-align:center">6177</td>
<td style="text-align:center">4252</td>
<td style="text-align:center">6721</td>
<td style="text-align:center">11027</td>
<td style="text-align:center">5176</td>
<td style="text-align:center">7995</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001036</td>
<td style="text-align:center">1433</td>
<td style="text-align:center">1062</td>
<td style="text-align:center">1733</td>
<td style="text-align:center">881</td>
<td style="text-align:center">1424</td>
<td style="text-align:center">1439</td>
<td style="text-align:center">1359</td>
<td style="text-align:center">1109</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001084</td>
<td style="text-align:center">519</td>
<td style="text-align:center">380</td>
<td style="text-align:center">595</td>
<td style="text-align:center">493</td>
<td style="text-align:center">820</td>
<td style="text-align:center">714</td>
<td style="text-align:center">696</td>
<td style="text-align:center">704</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001167</td>
<td style="text-align:center">394</td>
<td style="text-align:center">236</td>
<td style="text-align:center">464</td>
<td style="text-align:center">175</td>
<td style="text-align:center">658</td>
<td style="text-align:center">584</td>
<td style="text-align:center">360</td>
<td style="text-align:center">269</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001460</td>
<td style="text-align:center">172</td>
<td style="text-align:center">168</td>
<td style="text-align:center">264</td>
<td style="text-align:center">118</td>
<td style="text-align:center">241</td>
<td style="text-align:center">210</td>
<td style="text-align:center">155</td>
<td style="text-align:center">177</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001461</td>
<td style="text-align:center">2112</td>
<td style="text-align:center">1867</td>
<td style="text-align:center">5137</td>
<td style="text-align:center">2657</td>
<td style="text-align:center">2735</td>
<td style="text-align:center">2751</td>
<td style="text-align:center">2467</td>
<td style="text-align:center">2905</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001497</td>
<td style="text-align:center">524</td>
<td style="text-align:center">488</td>
<td style="text-align:center">638</td>
<td style="text-align:center">357</td>
<td style="text-align:center">676</td>
<td style="text-align:center">806</td>
<td style="text-align:center">493</td>
<td style="text-align:center">475</td>
</tr>
<tr>
<td style="text-align:center">ENSG00000001561</td>
<td style="text-align:center">71</td>
<td style="text-align:center">51</td>
<td style="text-align:center">211</td>
<td style="text-align:center">156</td>
<td style="text-align:center">23</td>
<td style="text-align:center">38</td>
<td style="text-align:center">134</td>
<td style="text-align:center">172</td>
</tr>
</tbody>
</table>
<p>分组矩阵的格式为：</p>
<table>
<thead>
<tr>
<th style="text-align:center"><code></code></th>
<th style="text-align:center">dex</th>
<th style="text-align:center">SampleName</th>
<th style="text-align:center">cell</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">SRR1039508</td>
<td style="text-align:center">untrt</td>
<td style="text-align:center">GSM1275862</td>
<td style="text-align:center">N61311</td>
</tr>
<tr>
<td style="text-align:center">SRR1039509</td>
<td style="text-align:center">trt</td>
<td style="text-align:center">GSM1275863</td>
<td style="text-align:center">N61311</td>
</tr>
<tr>
<td style="text-align:center">SRR1039512</td>
<td style="text-align:center">untrt</td>
<td style="text-align:center">GSM1275866</td>
<td style="text-align:center">N052611</td>
</tr>
<tr>
<td style="text-align:center">SRR1039513</td>
<td style="text-align:center">trt</td>
<td style="text-align:center">GSM1275867</td>
<td style="text-align:center">N052611</td>
</tr>
<tr>
<td style="text-align:center">SRR1039516</td>
<td style="text-align:center">untrt</td>
<td style="text-align:center">GSM1275870</td>
<td style="text-align:center">N080611</td>
</tr>
<tr>
<td style="text-align:center">SRR1039517</td>
<td style="text-align:center">trt</td>
<td style="text-align:center">GSM1275871</td>
<td style="text-align:center">N080611</td>
</tr>
<tr>
<td style="text-align:center">SRR1039520</td>
<td style="text-align:center">untrt</td>
<td style="text-align:center">GSM1275874</td>
<td style="text-align:center">N061011</td>
</tr>
<tr>
<td style="text-align:center">SRR1039521</td>
<td style="text-align:center">trt</td>
<td style="text-align:center">GSM1275875</td>
<td style="text-align:center">N061011</td>
</tr>
</tbody>
</table>
<p>差异表达分析</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">## 数据过滤</span><br><span class="line"># dds &lt;- dds[ rowSums(counts(dds)) &gt; 1 ,]</span><br><span class="line">dds &lt;- DESeq(dds)</span><br><span class="line"># plotDispEsts(dds, main=&quot;Dispersion plot&quot;)</span><br><span class="line"># rld &lt;- rlogTransformation(dds)</span><br><span class="line"># xprMatrix_rlog=assay(rld)</span><br><span class="line">## 若有多组，可以用参数contrast提取指定两组的比较结果</span><br><span class="line">res &lt;- results(dds, contrast=c(&quot;group_list&quot;,&apos;control&apos;,&apos;treat&apos;))</span><br><span class="line">## 将比较结果依据p-adjust进行排序</span><br><span class="line">resOrdered &lt;- res[order(res$padj),] </span><br><span class="line">res_Day1_Day0=as.data.frame(resOrdered)</span><br></pre></td></tr></table></figure>
<h3 id="3-2-使用Ballgown进行差异分析"><a href="#3-2-使用Ballgown进行差异分析" class="headerlink" title="3.2. 使用Ballgown进行差异分析"></a>3.2. 使用Ballgown进行差异分析</h3><p>紧接着stringtie的定量结果进行分析</p>
<p>stringtie的定量结果提供多种表达值的表示方法，有read counts, RPKM/FPKM, TPM 。其中read counts是原始reads计算，RPKM/FPKM 和 TPM 都是基因表达值的归一化后的，因为本身的某些缺点，主流科学家强烈要求它就被TPM取代了。</p>
<p>下面对 RPKM/FPKM 和 TPM 的说明摘抄自健明大大的简书：<code>https://www.jianshu.com/p/e9d5d7206124</code> ，说得通俗易懂</p>
<blockquote>
<p><strong>TPM是什么？</strong></p>
<p>我不喜欢看公式，直接说事情，我有一个基因A，它在这个样本的转录组数据中被测序而且mapping到基因组了 5000个的reads，而这个基因A长度是10K，我们总测序文库是50M，所以这个基因A的RPKM值是 5000除以10，再除以50，为10. 就是把基因的reads数量根据基因长度和样本测序文库来normalization 。</p>
<p>那么它的TPM值是多少呢？ 这个时候这些信息已经不够了，需要知道该样本其它基因的RPKM值是多少，加上该样本有3个基因，另外两个基因的RPKM值是5和35，那么我们的基因A的RPKM值为10需要换算成TPM值就是 1,000,000 *10/(5+10+35)=200,000， 看起来是不是有点大呀！其实主要是因为我们假设的基因太少了，一般个体里面都有两万多个基因的，总和会大大的增加，这样TPM值跟RPKM值差别不会这么恐怖的。</p>
</blockquote>
<ul>
<li>载入R包</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">require(ballgown)</span><br><span class="line">require(dplyr)</span><br><span class="line">require(genefilter)</span><br><span class="line">setwd(&quot;/share/disk5/lianm&quot;)</span><br></pre></td></tr></table></figure>
<ul>
<li>载入stringtie输出的表达数据，并设置表型信息（即分组信息）</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">bg_Z1Z4&lt;-ballgown(dataDir=&quot;SingleCell_process/Cleandata/Expression&quot;,samplePattern=“Z[14]T&quot;,meas=&quot;FPKM&quot;)</span><br><span class="line">pData(bg_Z1Z4)&lt;-data.frame(id=sampleNames(bg_Z1Z4),group=c(rep(1,num_group1),rep(0,num_group2)))</span><br></pre></td></tr></table></figure>
<ul>
<li>过滤低丰度的基因</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bg_Z1Z4_filt&lt;-subset(bg_Z1Z4,&quot;rowVars(texpr(bg_Z1Z4))&gt;1&quot;,genomesubset=T)</span><br></pre></td></tr></table></figure>
<ul>
<li>差异表达基因分析</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">result_genes&lt;-stattest(bg_Z1Z4_filt,feature=&quot;gene&quot;,covariate=&quot;group&quot;,getFC=T)</span><br><span class="line">result_genes&lt;-data.frame(geneNames=geneNames(bg_Z1Z4_filt)[match(result_genes$id,geneIDs(bg_Z1Z4_filt))],geneIDs=geneIDs(bg_Z1Z4_filt)[match(result_genes$id,geneIDs(bg_Z1Z4_filt))],result_genes)</span><br><span class="line">result_genes_sort&lt;-arrange(result_genes,pval)</span><br><span class="line">write.csv(result_genes_sort,file=paste(&quot;SingleCell_process/Cleandata/Expression/&quot;,name_group1,&quot;_VS_&quot;,name_group2,&quot;_geneDiff_results.csv&quot;,sep=&quot;&quot;),row.names=F)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>分组设置对差异表达分析的影响：</p>
<ul>
<li>FC = group_1/group_0，所以分组标签互换后FC会变为原来的倒数</li>
</ul>
</blockquote>
<ul>
<li>差异转录本分析</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">result_trans&lt;-stattest(bg_Z1Z4_filt,feature=&quot;transcript&quot;,covariate=&quot;group&quot;,getFC=T)</span><br><span class="line">result_trans&lt;-data.frame(geneNames=geneNames(bg_Z1Z4_filt),geneIDs=geneIDs(bg_Z1Z4_filt),result_trans)</span><br><span class="line">result_trans_sort&lt;-arrange(result_transs,pval)</span><br><span class="line">write.csv(result_trans_sort,file=paste(&quot;SingleCell_process/Cleandata/Expression/&quot;,name_group1,&quot;_VS_&quot;,name_group2,&quot;_transDiff_results.csv&quot;,sep=&quot;&quot;),row.names=F)</span><br></pre></td></tr></table></figure>
<h2 id="4-几点思考"><a href="#4-几点思考" class="headerlink" title="4. 几点思考"></a>4. 几点思考</h2><ol>
<li>为什么要进行转录本拼接？不是都有参考转录组了吗？</li>
</ol>
<ul>
<li><strong>无参考基因组</strong>：对于无参考转录组的物种来说，往往需要通过RNA-Seq的数据自己拼出参考转录组，然后再进行下游的数据分析。所以，对于无参分析来说，往往需要自己拼装转录本，生成自己的参考转录组信息</li>
<li><strong>注释完备的基因组</strong>：例如human，不同的细胞条件可能不同，一些永生化的细胞系往往都具有“癌症”的特征，这些细胞的转录组，基因组或多或少都有结构的变异，以及转录本的差异。</li>
</ul>
<hr>
<p>参考资料：</p>
<p>(1) <a href="https://mp.weixin.qq.com/s/74xmKy30ecQ2ciCqOq8MdQ" target="_blank" rel="noopener">【PANDA姐的转录组入门】拟南芥实战项目-定量、差异表达、功能分析</a></p>
<p>(2) <a href="https://github.com/jmzeng1314/my-R/tree/master/DEG_scripts" target="_blank" rel="noopener">一步法差异分析</a></p>
<p>(3) Pertea M, Kim D, Pertea G M, et al. Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown.[J]. Nature Protocols, 2016, 11(9):1650.</p>
<p>(4) <a href="http://www.bioinfo-scrounger.com/archives/482" target="_blank" rel="noopener">Alignment-based的转录本定量-RSEM</a></p>
<p>(5) <a href="https://github.com/bli25broad/RSEM_tutorial" target="_blank" rel="noopener">RSEM官方教程</a></p>
<p>(6) <a href="https://zhuanlan.zhihu.com/p/36065699" target="_blank" rel="noopener">知乎《高通量测序技术》专栏：生物信息学100个基础问题 —— 第25题 GTF/GFF的注释是怎么来的，应该从哪里下载？</a></p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/08/Stat-on-RNAseq/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                
                <a href="/2019/02/08/Stat-on-RNAseq/" class="post-title-link" itemprop="url">RNA-seq中的那些统计学问题</a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-02-08 00:38:20" itemprop="dateCreated datePublished" datetime="2019-02-08T00:38:20+00:00">2019-02-08</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-02-17 06:22:29" itemprop="dateModified" datetime="2019-02-17T06:22:29+00:00">2019-02-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h2><h3 id="过滤低表达的基因"><a href="#过滤低表达的基因" class="headerlink" title="过滤低表达的基因"></a>过滤低表达的基因</h3><blockquote>
<p>所有数据集将包括表达的基因和不表达的基因的组合。 虽然检查在一种条件下表达但不在另一种条件下表达的基因是有意义的，但是一些基因在所有样品中都未表达</p>
</blockquote>
<p><img src="/images/DiffExpAna-filtData.png" alt></p>
<h3 id="标准化"><a href="#标准化" class="headerlink" title="标准化"></a>标准化</h3><p><img src="/images/DiffExpAna-normalization.png" alt></p>
<p>由于不同文库测序深度不同，比较前当然要进行均一化！用总reads进行均一化可能最简单，其基于以下两个基本假设：</p>
<ul>
<li>绝大多数的gene表达量不变；</li>
<li>高表达量的gene表达量不发生改变；</li>
</ul>
<p>但在转录组中，通常一小部分极高丰度基因往往会贡献很多reads，如果这些“位高权重”的基因还是差异表达的，则会影响所有其它基因分配到的reads数，而且，两个样本总mRNA量完全相同的前提假设也过于理想了。那如何比较呢，各个方家使出浑身解数，有用中位数的，有用75分位数的，有用几何平均数的，有用TMM(trimmed mean of Mvalues)的等等，总之要<strong>找一个更稳定的参考</strong>值。</p>
<h4 id="House-keeping-gene-s"><a href="#House-keeping-gene-s" class="headerlink" title="House-keeping gene(s)"></a>House-keeping gene(s)</h4><p>矫正的思路很简单，就是在变化的样本中寻找不变的量</p>
<p>那么在不同RNA-seq样本中，那些是不变的量呢？一个很容易想到的就是<strong>管家基因</strong> (House-keeping gene(s)) </p>
<p>那么 Human 常用的 House-keeping gene 怎么确定？</p>
<p>目前大家用的比较多的一个human housekeeping gene list 来源于下面这篇文章，是2013年发表在 Cell系列的 Trends in Genetics 部分的一篇文章</p>
<p><img src="/images/DiffExpAna-normalization-House-keeping-genes.jpg" alt></p>
<h4 id="spike-in"><a href="#spike-in" class="headerlink" title="spike-in"></a>spike-in</h4><p>使用Housekeeping gene的办法来进行相对定量，这种办法在一定程度上能够解决我们遇到的问题。但其实这种办法有一个<strong>非常强的先验假设</strong>：housekeeping gene的表达量不怎么发生变化。其实housekeeping gene list有几千个，这几千个基因有一定程度上的变化是有可能的</p>
<p><strong>spike-in方法</strong>：在RNA-Seq建库的过程中掺入一些预先知道序列信息以及序列绝对数量的内参。这样在进行RNA-Seq测序的时候就可以通过不同样本之间内参（spike-in）的量来做一条标准曲线，就可以非常准确地对不同样本之间的表达量进行矫正</p>
<p>比较常用的spike-in类型：ERCC Control RNA</p>
<blockquote>
<p>ERCC = External RNA Controls Consortium</p>
<p>ERCC就是一个专门为了定制一套spike-in RNA而成立的组织，这个组织早在2003年的时候就已经宣告成立。主要的工作就是设计了一套非常好用的spike-in RNA，方便microarray，以及RNA-Seq进行内参定量</p>
<p><img src="/images/DiffExpAna-normalization-spike-in.jpg" alt></p>
</blockquote>
<h4 id="CPM"><a href="#CPM" class="headerlink" title="CPM"></a>CPM</h4><p>CPM(count-per-million)</p>
<p><img src="/images/DiffExpAna-normalization-CPM.png" alt></p>
<h4 id="TCS-Total-Count-Scaling"><a href="#TCS-Total-Count-Scaling" class="headerlink" title="TCS (Total Count Scaling)"></a>TCS (Total Count Scaling)</h4><p>简单来说，就是找出多个样本中library size为中位数的样本，作为参考样本，将所有的样本的library size按比例缩放到参考样本的水平</p>
<p>选择一个library size为中位数的sample，以它为baseline，计算出其它sample对于baseline的normalization factor，即一个缩放因子：</p>
<p><img src="/images/DiffExpAna-normalization-TCS-1.png" alt></p>
<p>然后基于该缩放因子对特定的sample中的每个基因的read count进行标准化（缩放）：</p>
<p><img src="/images/DiffExpAna-normalization-TCS-2.png" alt></p>
<h4 id="Quantile"><a href="#Quantile" class="headerlink" title="Quantile"></a>Quantile</h4><p>简单来说，就是排序后求平均，然后再回序</p>
<p><img src="/images/DiffExpAna-normalization-quantile.png" alt></p>
<p>在R里面，推荐用preprocessCore 包来做quantile normalization，不需要自己造轮子啦！<br>但是需要明白什么时候该用quantile normalization，什么时候不应该用，就复杂很多了</p>
<h4 id="Median-of-Ratio-DESeq2"><a href="#Median-of-Ratio-DESeq2" class="headerlink" title="Median of Ratio (DESeq2)"></a>Median of Ratio (DESeq2)</h4><p>该方法基于的假设是，即使处在不同条件下的不同个样本，大多数基因的表达是不存在差异的，实际存在差异的基因只占很小的部分那么我们只需要将这些稳定的部分找出来，作为标准化的内参，依据内参算出各个样本的标准化因子</p>
<p>（1）对每个基因计算几何平均数，得到一个假设的参考样本(pseudo-reference sample)</p>
<table>
<thead>
<tr>
<th style="text-align:left">gene</th>
<th style="text-align:left">sampleA</th>
<th style="text-align:left">sampleB</th>
<th style="text-align:left">pseudo-reference sample</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">EF2A</td>
<td style="text-align:left">1489</td>
<td style="text-align:left">906</td>
<td style="text-align:left">sqrt(1489 * 906) = 1161.5</td>
</tr>
<tr>
<td style="text-align:left">ABCD1</td>
<td style="text-align:left">22</td>
<td style="text-align:left">13</td>
<td style="text-align:left">sqrt(22 * 13) = 17.7</td>
</tr>
<tr>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
</tr>
</tbody>
</table>
<p>（2）对每个样本的每个基因对于参考样本计算Fold Change</p>
<table>
<thead>
<tr>
<th style="text-align:left">gene</th>
<th style="text-align:left">sampleA</th>
<th style="text-align:left">sampleB</th>
<th style="text-align:left">pseudo-reference sample</th>
<th style="text-align:left">ratio of sampleA/ref</th>
<th style="text-align:left">ratio of sampleB/ref</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">EF2A</td>
<td style="text-align:left">1489</td>
<td style="text-align:left">906</td>
<td style="text-align:left">1161.5</td>
<td style="text-align:left">1489/1161.5 = 1.28</td>
<td style="text-align:left">906/1161.5 = 0.78</td>
</tr>
<tr>
<td style="text-align:left">ABCD1</td>
<td style="text-align:left">22</td>
<td style="text-align:left">13</td>
<td style="text-align:left">16.9</td>
<td style="text-align:left">22/16.9 = 1.30</td>
<td style="text-align:left">13/16.9 = 0.77</td>
</tr>
<tr>
<td style="text-align:left">MEFV</td>
<td style="text-align:left">793</td>
<td style="text-align:left">410</td>
<td style="text-align:left">570.2</td>
<td style="text-align:left">793/570.2 = 1.39</td>
<td style="text-align:left">410/570.2 = 0.72</td>
</tr>
<tr>
<td style="text-align:left">BAG1</td>
<td style="text-align:left">76</td>
<td style="text-align:left">42</td>
<td style="text-align:left">56.5</td>
<td style="text-align:left">76/56.5 = 1.35</td>
<td style="text-align:left">42/56.5 = 0.74</td>
</tr>
<tr>
<td style="text-align:left">MOV10</td>
<td style="text-align:left">521</td>
<td style="text-align:left">1196</td>
<td style="text-align:left">883.7</td>
<td style="text-align:left">521/883.7 = 0.590</td>
<td style="text-align:left">1196/883.7 = 1.35</td>
</tr>
<tr>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
<td style="text-align:left">…</td>
</tr>
</tbody>
</table>
<p><img src="/images/DiffExpAna-normalization-DESeq2.png" alt></p>
<p>（3）获取每个样本中Fold Change的中位数，我们就得到了非DE基因代表的Fold Change，该基因就是我们选择的该样本的内参基因，它的Fold Change就是该样本的标准化因子</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">normalization_factor_sampleA &lt;- median(c(1.28, 1.3, 1.39, 1.35, 0.59))</span><br><span class="line"></span><br><span class="line">normalization_factor_sampleB &lt;- median(c(0.78, 0.77, 0.72, 0.74, 1.35))</span><br></pre></td></tr></table></figure>
<h4 id="TMM-Trimmed-Mean-of-M-value-edgeR"><a href="#TMM-Trimmed-Mean-of-M-value-edgeR" class="headerlink" title="TMM (Trimmed Mean of M value, edgeR)"></a>TMM (Trimmed Mean of M value, edgeR)</h4><p>该方法的思想与DESeq2的Median of Ratio相同，假设前提都是：大多数基因的表达是不存在差异的</p>
<p>它与DESeq2的不同之处在于对内参的选择上：</p>
<blockquote>
<ul>
<li><p>DESeq2选择一个内参基因，它的Ratio/Fold-Change就是标准化因子</p>
</li>
<li><p>edgeR选择一组内参基因集合，它们对标准化因子的计算均有贡献：加权平均</p>
</li>
</ul>
</blockquote>
<p>（1）移除所有未表达基因</p>
<p>（2）从众多样本中找出一个数据趋势较为平均的样本作为参考样本</p>
<blockquote>
<ul>
<li><p>对所有样本求总Read数；</p>
</li>
<li><p>各样本除以各自的总Read数，得到修正Read数；</p>
</li>
<li><p>求出各自样本修正Read数的Q3值（第3个四分位数）；</p>
</li>
<li><p>所有的Q3值求平均，与平均Q3相差最小的样本即是参考样本。</p>
</li>
</ul>
</blockquote>
<p><img src="/images/DiffExpAna-normalization-TMM-1.jpg" alt></p>
<p>（3）找出每个样本中的代表基因集，参考这些代表基因集的fold change，计算出该样本的标准化因子</p>
<p>寻找样本的代表基因集：依据基因的<strong>偏倚程度</strong>和<strong>Reads数</strong>大小选出——偏倚程度小、reads数居中的基因</p>
<blockquote>
<ul>
<li><strong>衡量偏倚度的量：LFC (log fold change)</strong></li>
</ul>
<p><img src="/images/DiffExpAna-normalization-TMM-2.png" alt></p>
<p>LFC过大或过小都表示具有偏倚性，LFC越大表示reads数在sample<sub>i</sub>中越高，即偏向sample<sub>i</sub>；LFC越小表示reads数在ref中越高，即偏向ref</p>
<ul>
<li><strong>衡量reads数的量：read的几何平均数 (read geometric mean, RGM)</strong></li>
</ul>
<p>RGM越大表示基因reads越多，RGM越小表示基因reads越少</p>
<p>结合偏倚度、read数画出散点图：</p>
<p><img src="/images/DiffExpAna-normalization-TMM-3.jpg" alt></p>
<p>偏倚度小、表达量居中的那些基因落在图中的红线附近</p>
</blockquote>
<p>由参考代表基因集计算样本的标准化因子：</p>
<blockquote>
<p>对这些代表基因集计算加权平均数：</p>
<p><img src="/images/DiffExpAna-normalization-TMM-4.png" alt></p>
<p>该加权平均数就能代表该样本的标准化因子，只是经过了log变换，所以需要恢复为它的正值：</p>
<p align="center">Scaling-Factor = 2 <sup>weight-average</sup></p>
</blockquote>
<h3 id="为什么说FPKM和RPKM都错了？"><a href="#为什么说FPKM和RPKM都错了？" class="headerlink" title="为什么说FPKM和RPKM都错了？"></a>为什么说FPKM和RPKM都错了？</h3><h4 id="FPKM和RPKM分别是什么"><a href="#FPKM和RPKM分别是什么" class="headerlink" title="FPKM和RPKM分别是什么"></a>FPKM和RPKM分别是什么</h4><p>FPKM和RPKM分别是什么</p>
<blockquote>
<p><strong>RPKM</strong>是Reads Per Kilobase per Million的缩写，它的计算方程非常简单：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-1.png" alt></p>
<p><strong>FPKM</strong>是Fregments Per Kilobase per Million的缩写，它的计算与RPKM极为类似，如下：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-2.png" alt></p>
<p>与RPKM唯一的区别为：F是fragments，R是reads，如果是PE(Pair-end)测序，每个fragments会有两个reads，FPKM只计算两个reads能比对到同一个转录本的fragments数量，而RPKM计算的是可以比对到转录本的reads数量而不管PE的两个reads是否能比对到同一个转录本上。如果是SE(single-end)测序，那么FPKM和RPKM计算的结果将是一致的。</p>
</blockquote>
<p>这两个量的计算方式的目的是为了解决计算RNA-seq转录本丰度时的两个bias：</p>
<ul>
<li>相同表达丰度的转录本，往往会由于其基因长度上的差异，导致测序获得的Read（Fregment）数不同。总的来说，越长的转录本，测得的Read（Fregment）数越多；</li>
<li>由测序文库的不同大小而引来的差异。即同一个转录本，其测序深度越深，通过测序获得的Read（Fregment）数就越多。</li>
</ul>
<h4 id="什么样才算好的统计量"><a href="#什么样才算好的统计量" class="headerlink" title="什么样才算好的统计量"></a>什么样才算好的统计量</h4><p>首先，到底什么是RNA转录本的表达丰度这个问题</p>
<p>对于样本X，其有一个基因g被转录了mRNA_g次，同时样本X中所有基因的转录总次数假定是mRNA_total, 那么正确描述基因g转录丰度的值应该是：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-3.png" alt></p>
<p>则一个样本中基因表达丰度的均值为</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-4.png" alt></p>
<p>而</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-5.png" alt></p>
<p>所以</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-6.png" alt></p>
<p>这个期望值竟然和测序状态无关！仅仅由样本中基因的总数所决定的</p>
<p>也就是说，对于同一个物种，不管它的样本是哪种组织（正常的或病变的），也不管有多少个不同的样本，只要它们都拥有相同数量的基因，那么它们的r_mean都将是一致的</p>
<p>由于上面的结果是在理论情况下推导出来的，实际上我们无法直接计算这个r，那么我们可以尝试通过其他方法来近似估计r，<strong>只要这些近似统计量可以隐式地包含这一恒等关系即可</strong></p>
<h4 id="FPKM和RPKM犯的错"><a href="#FPKM和RPKM犯的错" class="headerlink" title="FPKM和RPKM犯的错"></a>FPKM和RPKM犯的错</h4><p>实际数据来证明</p>
<blockquote>
<p>假定有两个来自同一个个体不同组织的样本X和Y，这个个体只有5个基因，分别为A、B、C、D和E它们的长度分别如下：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-7.png" alt></p>
<p>r_mean值是:</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-8.png" alt></p>
<p>如果FPKM或RPKM是一个合适的统计量的话，那么，样本X和Y的平均FPKM（或RPKM）应该相等。</p>
<p>以下这个表格列出的分别是样本X和Y在这5个基因分别比对上的fregment数和各自总的fregment数量：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-9.jpg" alt></p>
<p>可以算出：样本X在这5个基因上的FPKM均值FPKM_mean = 5,680;样本Y在这5个基因上的FPKM均值FPKM_mean = 161,840</p>
<p>很明显，它们根本不同，而且差距相当大</p>
</blockquote>
<p>究竟为什么会有如此之大的差异？</p>
<p>可以从其公式上找到答案</p>
<blockquote>
<p>首先，我们可以把FPKM的计算式拆分成如下两部分。</p>
<p>第一部分的统计量是对一个基因转录本数量的一个等价描述（虽然严格来讲也没那么等价）：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-10.png" alt></p>
<p>第二部分的统计量是测序获得的总有效Fregment数量的百万分之一：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-11.png" alt></p>
<p>这么一拆，就可以看出这个公式的问题了：逻辑上根本说不通嘛！</p>
<p>尤其是第二部分（N/10^6），本来式子的第一部分是为了描述一个基因的转录本数量，那么正常来讲，第二部分就应该是样本的转录本总数量（或至少是其总数量的等价描述）才能形成合理的比例关系，而且可以看出来FPKM/RPMK是有此意的，这本来就是这个统计量的目的。</p>
<p>但是N/10^6并不能描述样本的转录本总数量。N/10^6的大小其实是由RNA-seq测序深度所决定的，并且是一个和总转录本数量无直接线性关系的统计量——N与总转录本数量之间的关系还受转录本的长度分布所决定，而这个分布往往在不同样本中是有差异的！</p>
</blockquote>
<h4 id="TPM是一个合适的选择"><a href="#TPM是一个合适的选择" class="headerlink" title="TPM是一个合适的选择"></a>TPM是一个合适的选择</h4><p>这个统计量在2012年所发表的一篇讨论RPKM的文章（RPKM measure is inconsistent among samples. Wagner GP, Kin K, Lynch VJ. Theory Biosci. 2012.）中就被提出来了，称之为TPM —— Transcripts Per Million，它的计算是：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-12.png" alt></p>
<p>简单计算之后我们就可以发现TPM的均值是一个独立于样本之外的恒定值，它等于：</p>
<p><img src="/images/DiffExpAna-normalization-analysis-FPKM-RPKM-13.png" alt></p>
<p>这个值刚好是r_mean的一百万倍，满足等价描述的关系。</p>
<h2 id="统计学原理"><a href="#统计学原理" class="headerlink" title="统计学原理"></a>统计学原理</h2><h3 id="转录组数据统计推断的难题"><a href="#转录组数据统计推断的难题" class="headerlink" title="转录组数据统计推断的难题"></a>转录组数据统计推断的难题</h3><p>在RNA-seq中进行两组间的差异分析是最正常不过的了。</p>
<p>我们在其它实验中同样会遇到类似的分析，通常，我们可以用方差分析判定两组“分布”数据间是否存在显著差异。原理是：当组间方差大于组内方差（误差效应），并且统计学显著时，则认为组间处理是可以引起差异的。</p>
<p>有伙伴肯定要问，转录组数据到底有什么了不起的？它们为什么不能用我们熟悉的算法简单地进行计算？</p>
<p>其实统计学家也很无奈啊，看看我们转录组实验得到的这些数据吧：我们的实验只进行少得可怜的生物学重复（n&lt;10），而且，任何基因的表达量都不能是负数，这些数据并不符合正态分布，用于表征表达量的counts是非连续的（芯片信号是连续的），RNA-seq数据的离散通常是高度扭曲的，方差往往会大于均值……，就这些奇怪的特征，使得准确估计方差并没有想象的那么容易。</p>
<p>我们面临两个核心问题：</p>
<ul>
<li>基因表达数据适合用什么统计学分布进行差异显著性检验？</li>
<li>如何利用少量生物学重复数据估算基因表达的标准差？</li>
</ul>
<h3 id="泊松分布-or-负二项分布？"><a href="#泊松分布-or-负二项分布？" class="headerlink" title="泊松分布 or 负二项分布？"></a>泊松分布 or 负二项分布？</h3><p>从统计学的角度出发，进行差异分析肯定会需要假设检验，通常对于分布已知的数据，运用参数检验结果的假阳性率会更低。转录组数据中，raw count值符合什么样的分布呢？</p>
<p>count值本质是reads的数目，是一个非零整数，而且是离散的，其分布肯定也是离散型分布。对于转录组数据，学术界常用的分布包括<strong>泊松分布 (poisson)</strong>和<strong>负二项分布 (negative binomial)</strong>两种。</p>
<h4 id="为什么泊松分布不行？"><a href="#为什么泊松分布不行？" class="headerlink" title="为什么泊松分布不行？"></a>为什么泊松分布不行？</h4><p>首先有必要简单地介绍一下泊松分布</p>
<blockquote>
<p>泊松分布适合于描述单位时间（或空间）内随机事件发生的次数（事件发生的次数只能是离散的整数）。如某一服务设施在一定时间内到达的人数，电话交换机接到呼叫的次数，汽车站台的候客人数，机器出现的故障数，自然灾害发生的次数，一块产品上的缺陷数，显微镜下单位分区内的细菌分布数等等。</p>
</blockquote>
<blockquote>
<p><img src="/images/DiffExpAna-statistic-principle-4.jpg" alt></p>
<p>泊松分布大概长这样：</p>
<p><img src="/images/BioStat-introduction-common-distributions-poisson-distribution-curve.png" alt></p>
<p>λ是波松分布所依赖的唯一参数。 λ值愈小分布愈偏倚， 随着λ的增大 ， 分布趋于对称。 当λ=20时分布接近于正态分布；当λ=50时， 可以认为波松分布呈正态分布。</p>
</blockquote>
<p>在数据分析的早期，确实有学者采用泊松分布进行差异分析，但是发展到现在，几乎全部都是基于负二项分布了，究竟是什么因素导致了这种现象呢？为了解释这个问题，我们必须提到一个概念 <strong>overdispersion</strong>。</p>
<p>dispersion指的是离散程度，研究一个数据分布的离散程度，我们常用方差这个指标。<strong>对于泊松分布而言，其均值和方差是相等的，但是我们的数据确不符合这样的规律</strong>。通过计算所有基因的均值和方差，可以绘制如下的图片：</p>
<p><img src="/images/DiffExpAna-statistic-principle-1.png" alt></p>
<p>横坐标为基因在所有样本中的均值，纵坐标为基因在所有样本中的方差，直线的斜率为1，代表泊松分布的均值和方差的分布。可以看到，真实数据的分布是偏离了泊松分布的，方差明显比均值要大。</p>
<p>如果假定总体分布为泊松分布， 根据我们的定量数据是无法估计出一个合理的参数，能够符合上图中所示分布的，这样的现象就称之为overdispersion。</p>
<p>由于真实数据与泊松分布之间的overdispersion，<strong>选择泊松分布分布作为总体的分布是不合理</strong>。</p>
<p>以上只证明了泊松分布是个不太恰当的分布估计，那<strong>怎么证明负二项分布就是合适的分布估计呢？</strong></p>
<h4 id="为什么负二项分布行？"><a href="#为什么负二项分布行？" class="headerlink" title="为什么负二项分布行？"></a>为什么负二项分布行？</h4><p>主要是从均值与方差之间的关系去证明</p>
<p>同样的，也先简单介绍一下负二项分布：</p>
<blockquote>
<p>二项分布描述的是n重伯努利实验，在n重贝努利试验中，事件A恰好发生x(0≤x≤n)次的概率为：</p>
<p><img src="/images/DiffExpAna-statistic-principle-binomial-distribution.png" alt></p>
<p>它的概率分布图如下：</p>
<p><img src="/images/DiffExpAna-statistic-principle-binomial-distribution-curve.jpg" alt></p>
<p>负二项分布描述的<strong>也是伯努利实验</strong>，不过它的目标事件变成了：对于Bernoulli过程，我们设定，当某个结果出现固定次数的时候，整个过程的数量，比如我们生产某个零件，假设每个零件的合格与否都是相互独立的，且分布相同，那么当我们生产出了五个不合格零件时，一共生产了多少合格的零件，这个数量就是一个<strong>负二项分布</strong>，公式如下：</p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-1.png" alt></p>
<p>该公式描述的是，在合格率为p的一堆产品中，进行连续有放回的抽样，当抽到r个次品时，停止抽样，此时抽到的正品正好为k个的概率</p>
<p>它的概率分布如下：</p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-2.png" alt="p=0.5, r=5"></p>
</blockquote>
<p>负二项分布的均值和方差分别为：</p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-expectation.png" alt></p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-diff.png" alt></p>
<p>将p用μ表示，得到：</p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-3.png" alt></p>
<p>将上一步推出的p和1-p带入到方差的表达式中，得到：</p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-4.png" alt></p>
<p>记<code>1/r=α</code>，则</p>
<p><img src="/images/DiffExpAna-statistic-principle-negative-binomial-distribution-5.png" alt></p>
<p>从上面的式子可以看出，均值是方差的二次函数，方差随着均值的增加而进行二次函数形式的递增，正好符合上文 <a href="#reason-to-deny-poisson-distribution">2.2.1. 为什么泊松分布不行？</a>部分均值与方差分布图的情况</p>
<p>其中<code>α</code>和<code>r</code>被称为<strong>dispersion parameter</strong></p>
<p>负二项分布与泊松分布的关系，可以用<code>α</code>或<code>r</code>推出：</p>
<blockquote>
<p>当 <code>r -∞</code> 时，<code>α -0</code>，此时 σ<sup>2&lt;/sup= μ，为泊松分布；</sup></p>
<p>当 <code>r -&gt; 0</code> 时，<code>α -&gt; ∞</code>，此时overdispersion</p>
</blockquote>
<h3 id="方差估计"><a href="#方差估计" class="headerlink" title="方差估计"></a>方差估计</h3><p>在生物学重复很少时，我们是很难准确计算每个基因表达的标准差的（相当于这个数据集的离散程度）。我们<strong>很可能会低估数据的离散程度</strong>。</p>
<p>被逼无奈的科学家提出了一个假设：表达丰度相似的基因，在总体上标准差应该也是相似的。我们把不同生物学重复中表达丰度相同的基因的总标准差取个平均值，低于这个值的都用这个值，高于这个值的就用算出来的值。</p>
<p><img src="/images/DiffExpAna-statistic-principle-2.png" alt></p>
<p><img src="/images/DiffExpAna-statistic-principle-3.png" alt="（以上图来自 H. J. Pimentel, et al. Differential analysis of RNA-Seq incorporatingquantification uncertainty. bioRxiv, 2016）"></p>
<hr>
<p>参考资料：</p>
<p>(1) <a href="http://www.bio-info-trainee.com/2043.html" target="_blank" rel="noopener">【生信菜鸟团】quantile normalization到底对数据做了什么？</a></p>
<p>(2) <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html" target="_blank" rel="noopener">Introduction to DGE</a></p>
<p>(3) <a href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247485369&amp;idx=1&amp;sn=791cb8c26b19a1181ceccf586787f078&amp;scene=21#wechat_redirect" target="_blank" rel="noopener">生信菜鸟团：StatQuest生物统计学专题 - library normalization进阶之edgeR的标准化方法 </a></p>
<p>(4) <a href="https://www.jianshu.com/p/35e861b76486" target="_blank" rel="noopener">【简书】为什么说FPKM和RPKM都错了？</a></p>
<p>(5) <a href="https://mp.weixin.qq.com/s/m2ydqpKofYo2bK61A9hZWw" target="_blank" rel="noopener">【生信修炼手册】负二项分布在差异分析中的应用</a></p>
<p>(6) <a href="https://mp.weixin.qq.com/s/VcjnvI5FqwOFEC9wSUfdSw" target="_blank" rel="noopener">【 生信百科】转录组差异表达筛选的真相</a></p>
<p>(7) <a href="https://mp.weixin.qq.com/s/UTmSzCgDIFYbG2WByzaqQQ" target="_blank" rel="noopener">【生信媛】RNA-seq分析中的dispersion，你知道吗？</a></p>
<p>(8) H. J. Pimentel, et al. Differential analysis of RNA-Seq incorporatingquantification uncertainty. bioRxiv, 2016</p>

          
        
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><a class="extend next" rel="next" href="/page/2/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <div class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">Lianm</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">14</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">5</span>
                    <span class="site-state-item-name">categories</span>
                  
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">21</span>
                    <span class="site-state-item-name">tags</span>
                  
                </div>
              
            </nav>
          

          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://github.com/Ming-Lian" title="GitHub &rarr; https://github.com/Ming-Lian" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="mailto:lianming17m@big.ac.cn" title="E-Mail &rarr; mailto:lianming17m@big.ac.cn" rel="noopener" target="_blank"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                </span>
              
            </div>
          

          

          
          

          
            
          
          

        </div>
      </div>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Lianm</span>

  

  
</div>


  <div class="powered-by">Powered by <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> v3.8.0</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Muse</a> v7.0.0</div>




        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>


  


  <script src="/js/src/utils.js?v=7.0.0"></script>

  <script src="/js/src/motion.js?v=7.0.0"></script>



  
  


  <script src="/js/src/schemes/muse.js?v=7.0.0"></script>



  

  


  <script src="/js/src/bootstrap.js?v=7.0.0"></script>



  



  




  

  

  

  

  

  

  

  

  

  

  

  

  

</body>
</html>
