<!DOCTYPE html>












  


<html class="theme-next mist use-motion" lang="zh-CN">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">


























<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2">

<link rel="stylesheet" href="/css/main.css?v=7.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/Favico.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/Favico.png?v=7.0.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/Favico.png?v=7.0.0">


  <link rel="mask-icon" href="/images/logo.svg?v=7.0.0" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '7.0.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="Perl1. 拆分FASTA文件将FASTA文件按照用户指定的序列条数进行拆分，即每n条序列写到一个文件中，或者按照指定的输出文件数进行拆分，则每个文件中包含的序列数相同 脚本名：splitFasta.pl1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253">
<meta name="keywords" content="Bioinformatics,R,Python,Perl">
<meta property="og:type" content="article">
<meta property="og:title" content="实用小脚本">
<meta property="og:url" content="http://yoursite.com/2019/02/17/Practical-Scripts/index.html">
<meta property="og:site_name" content="Lianm&#39;s Blog">
<meta property="og:description" content="Perl1. 拆分FASTA文件将FASTA文件按照用户指定的序列条数进行拆分，即每n条序列写到一个文件中，或者按照指定的输出文件数进行拆分，则每个文件中包含的序列数相同 脚本名：splitFasta.pl1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253">
<meta property="og:locale" content="zh-CN">
<meta property="og:updated_time" content="2019-04-28T12:00:44.780Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="实用小脚本">
<meta name="twitter:description" content="Perl1. 拆分FASTA文件将FASTA文件按照用户指定的序列条数进行拆分，即每n条序列写到一个文件中，或者按照指定的输出文件数进行拆分，则每个文件中包含的序列数相同 脚本名：splitFasta.pl1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253">






  <link rel="canonical" href="http://yoursite.com/2019/02/17/Practical-Scripts/">



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>实用小脚本 | Lianm's Blog</title>
  












  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Lianm's Blog</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">Focus on Bioinformatics and Machine-Learning</p>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="切换导航栏">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>首页</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br>关于</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">

    
    
    
      
    

    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br>标签</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">

    
    
    
      
    

    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br>分类</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>归档</a>

  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>搜索</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    
  
  
  
  

  

  <a href="https://github.com/Ming-Lian" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="80" height="80" viewbox="0 0 250 250" style="fill: #222; color: #fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a>



    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/02/17/Practical-Scripts/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Lianm">
      <meta itemprop="description" content="中国科学院北京基因组研究所研究生">
      <meta itemprop="image" content="/images/avatar.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Lianm's Blog">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">实用小脚本

              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2019-02-17 08:14:59" itemprop="dateCreated datePublished" datetime="2019-02-17T08:14:59+08:00">2019-02-17</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2019-04-28 20:00:44" itemprop="dateModified" datetime="2019-04-28T20:00:44+08:00">2019-04-28</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/编程/" itemprop="url" rel="index"><span itemprop="name">编程</span></a></span>

                
                
              
            </span>
          

          
            
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="Perl"><a href="#Perl" class="headerlink" title="Perl"></a>Perl</h2><h3 id="1-拆分FASTA文件"><a href="#1-拆分FASTA文件" class="headerlink" title="1. 拆分FASTA文件"></a>1. 拆分FASTA文件</h3><p>将FASTA文件按照用户指定的序列条数进行拆分，即每n条序列写到一个文件中，或者按照指定的输出文件数进行拆分，则每个文件中包含的序列数相同</p>
<p>脚本名：<code>splitFasta.pl</code><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line">use POSIX;</span><br><span class="line"></span><br><span class="line"># 帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">	This script is used to split fasta file, which is too large with thosands of sequence</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">	$0 -i &lt;input&gt; -o &lt;output_dir&gt; [-n &lt;seq_num_per_file&gt;] [-m &lt;output_file_num&gt;]</span><br><span class="line">	</span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">	-i	[str]	Input raw fasta file</span><br><span class="line">	-o	[str]	Output file to which directory</span><br><span class="line">	-n	[int]	Sequence number per file, alternate chose paramerter &quot;-n&quot; or &quot;-m&quot;, if set &quot;-n&quot; and &quot;-m&quot; at the same time, only take &quot;-n&quot; parameter</span><br><span class="line">	-m	[int]	Output file number (default:100)</span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($input,$output_dir,$seq_num,$file_num);</span><br><span class="line">GetOptions(</span><br><span class="line">	&quot;i:s&quot;=&gt;\$input,</span><br><span class="line">	&quot;o:s&quot;=&gt;\$output_dir,</span><br><span class="line">	&quot;n:i&quot;=&gt;\$seq_num,</span><br><span class="line">	&quot;m:i&quot;=&gt;\$file_num</span><br><span class="line">	);</span><br><span class="line"></span><br><span class="line">die `pod2text $0` if ((!$input) or (!$output_dir));</span><br><span class="line"></span><br><span class="line"># 设置每个文件的序列条数</span><br><span class="line">if(!defined($seq_num))&#123;</span><br><span class="line">	if(!defined($file_num))&#123;</span><br><span class="line">		$file_num=100;</span><br><span class="line">		my $total_seq_num=`awk &apos;BEGIN&#123;n=0&#125; /^&gt;/&#123;n++&#125; END&#123;print n&#125;&apos; $input`;</span><br><span class="line">		chomp $total_seq_num;</span><br><span class="line">		$seq_num=ceil($total_seq_num/$file_num);</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		my $total_seq_num=`awk &apos;BEGIN&#123;n=0&#125; /^&gt;/&#123;n++&#125; END&#123;print n&#125;&apos; $input`;</span><br><span class="line">		chomp $total_seq_num;</span><br><span class="line">		$seq_num=ceil($total_seq_num/$file_num);</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">open IN,&quot;&lt;$input&quot; or die &quot;Cann&apos;t open $input\n&quot;;</span><br><span class="line"></span><br><span class="line">my $n_seq=0;	# 该变量用于记录当前扫描到的序列数</span><br><span class="line">my $n_file=1;	# 该变量用于记录当前真正写入的文件的计数</span><br><span class="line">my $input_base=`basename $input`;</span><br><span class="line">chomp $input_base;</span><br><span class="line"></span><br><span class="line">open OUT,&quot;&gt;$output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;&quot; or die &quot;Cann&apos;t create $output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;\n&quot;;</span><br><span class="line"></span><br><span class="line">while(&lt;IN&gt;)&#123;</span><br><span class="line">	next if (/^\s+$/);	# 跳过空行</span><br><span class="line">	chomp;</span><br><span class="line">	if (/^&gt;/)&#123;</span><br><span class="line">		$n_seq++;</span><br><span class="line">		# 判断目前已经扫描到的序列数，若大于设定的split的序列数，则创建新文件</span><br><span class="line">		if ($n_seq&gt;$seq_num)&#123;</span><br><span class="line">			$n_seq=1;</span><br><span class="line">			$n_file++;</span><br><span class="line">			close OUT;</span><br><span class="line">			open OUT,&quot;&gt;$output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;&quot; or die &quot;Cann&apos;t create $output_dir/$&#123;input_base&#125;_$&#123;n_file&#125;\n&quot;;</span><br><span class="line">			print OUT &quot;$_\n&quot;;</span><br><span class="line">		&#125;else&#123;</span><br><span class="line">			print OUT &quot;$_\n&quot;;</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		print OUT &quot;$_\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close IN;</span><br></pre></td></tr></table></figure></p>
<p>执行方法： </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl splitFasta.pl -i &lt;input&gt; -o &lt;output_dir&gt; [-n &lt;seq_num_per_file&gt;] [-m &lt;output_file_num&gt;]</span><br></pre></td></tr></table></figure>
<p>具体的参数使用说明可以执行 <code>perl splitFasta.pl</code>，查看脚本使用文档</p>
<h3 id="2-从双端FASTQ文件中抽取指定数据量-bp-的序列"><a href="#2-从双端FASTQ文件中抽取指定数据量-bp-的序列" class="headerlink" title="2. 从双端FASTQ文件中抽取指定数据量(bp)的序列"></a>2. 从双端FASTQ文件中抽取指定数据量(bp)的序列</h3><p>按照用户指定的数据量，即多少bp，从原始的双端FASTQ文件中随机抽取序列，要求双端FASTQ文件中的序列必须配对</p>
<p>脚本名：<code>extractFastq.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line"></span><br><span class="line"># 脚本帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">	Thise script is used to extract a number of fastq recorders from original input fastq file</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">	$0 -n &lt;totalReads&gt; -e &lt;base-pairs to extract&gt; -l &lt;readLength&gt; -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br><span class="line"></span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">	-n	[int]	Total reads number of input.fastq</span><br><span class="line">	-e	[int]	Base-pairs per end you want to extract</span><br><span class="line">	-l	[int]	The reads length</span><br><span class="line">	-1	[str]	Input 1st-end fastq file</span><br><span class="line">	-2	[str]	Input 2nd-end fastq file</span><br><span class="line">	-o	[str]	Output directory [default: current folder]</span><br><span class="line"></span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($totalReads,$bpNum,$length,$Input1,$Input2,$Outdir);</span><br><span class="line">GetOptions(</span><br><span class="line">	&quot;n:i&quot;=&gt;\$totalReads,</span><br><span class="line">	&quot;e:i&quot;=&gt;\$bpNum,</span><br><span class="line">	&quot;l:i&quot;=&gt;\$length,</span><br><span class="line">	&quot;1:s&quot;=&gt;\$Input1,</span><br><span class="line">	&quot;2:s&quot;=&gt;\$Input2,</span><br><span class="line">	&quot;o:s&quot;=&gt;\$Outdir</span><br><span class="line">	);</span><br><span class="line"></span><br><span class="line">$Outdir=`pwd` unless (defined($Outdir));</span><br><span class="line">die `pod2text $0` if ((!$totalReads) or (!$bpNum)) or (!$length) or (!$Input1) or (!$Input2);</span><br><span class="line"></span><br><span class="line">open FQ1,&quot;&lt;$Input1&quot; or die &quot;$!\n&quot;;</span><br><span class="line">open FQ2,&quot;&lt;$Input2&quot; or die &quot;$!\n&quot;;</span><br><span class="line"></span><br><span class="line">my $Input1_basename=`basename $Input1`;</span><br><span class="line">chomp $Input1_basename;</span><br><span class="line">my $Input2_basename=`basename $Input2`;</span><br><span class="line">chomp $Input2_basename;</span><br><span class="line">open OUT1,&quot;&gt;$Outdir/$&#123;Input1_basename&#125;.extract&quot; or die &quot;$!\n&quot;;</span><br><span class="line">open OUT2,&quot;&gt;$Outdir/$&#123;Input2_basename&#125;.extract&quot; or die &quot;$!\n&quot;;</span><br><span class="line"></span><br><span class="line">my $readsRemain=$bpNum/$length;</span><br><span class="line"></span><br><span class="line">my $remainCount=0;</span><br><span class="line"></span><br><span class="line">while(! eof($FQ1))&#123;</span><br><span class="line">	# 读入1st-end fastq 文件的四行</span><br><span class="line">	my $fq1_1=&lt;FQ1&gt;;</span><br><span class="line">	my $fq1_2=&lt;FQ1&gt;;</span><br><span class="line">	my $fq1_3=&lt;FQ1&gt;;</span><br><span class="line">	my $fq1_4=&lt;FQ1&gt;;</span><br><span class="line">	chomp($fq1_1,$fq1_2,$fq1_3,$fq1_4);</span><br><span class="line">	# 读入2nd-end fastq 文件的四行</span><br><span class="line">	my $fq2_1=&lt;FQ2&gt;;</span><br><span class="line">	my $fq2_2=&lt;FQ2&gt;;</span><br><span class="line">	my $fq2_3=&lt;FQ2&gt;;</span><br><span class="line">	my $fq2_4=&lt;FQ2&gt;;</span><br><span class="line">	chomp($fq2_1,$fq2_2,$fq2_3,$fq2_4);</span><br><span class="line"></span><br><span class="line">	# 随机抽取</span><br><span class="line">	if (rand()&lt;$readsRemain/$totalReads)&#123;</span><br><span class="line">		$remainCount++;</span><br><span class="line">		print OUT1 &quot;$fq1_1\n$fq1_2\n$fq1_3\n$fq1_4\n&quot;;</span><br><span class="line">		print OUT2 &quot;$fq2_1\n$fq2_2\n$fq2_3\n$fq2_4\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">print &quot;Total reads: $totalReads\n&quot;;</span><br><span class="line">print &quot;Theorical remained reads: $readsRemain\n&quot;;</span><br><span class="line">print &quot;Practical remained reads: $remainCount\n&quot;;</span><br><span class="line"></span><br><span class="line">close FQ1;</span><br><span class="line">close FQ2;</span><br><span class="line">close OUT1;</span><br><span class="line">close OUT2;</span><br></pre></td></tr></table></figure>
<p>执行方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl extractFastq.pl -n &lt;totalReads&gt; -e &lt;base-pairs to extract&gt; -l &lt;readLength&gt; -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br></pre></td></tr></table></figure>
<p>具体的参数使用说明可以执行 <code>perl extractFastq.pl</code>，查看脚本使用文档</p>
<h3 id="3-双端FASTQ文件进行双端配对"><a href="#3-双端FASTQ文件进行双端配对" class="headerlink" title="3. 双端FASTQ文件进行双端配对"></a>3. 双端FASTQ文件进行双端配对</h3><p>脚本名：<code>PairsMate.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line">use Getopt::Long;</span><br><span class="line"></span><br><span class="line"># 脚本帮助文档</span><br><span class="line">=head1 Description</span><br><span class="line"></span><br><span class="line">Thise script is used to match the pair-end reads from one sample</span><br><span class="line"></span><br><span class="line">=head1 Usage</span><br><span class="line"></span><br><span class="line">$0 -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br><span class="line"></span><br><span class="line">=head1 Parameters</span><br><span class="line"></span><br><span class="line">-1	[str]	Input 1st-end fastq file</span><br><span class="line">-2	[str]	Input 2nd-end fastq file</span><br><span class="line">-o	[str]	Output directory [default: current folder]</span><br><span class="line"></span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">my ($Input1,$Input2,$Outdir);</span><br><span class="line">GetOptions(</span><br><span class="line">&quot;1:s&quot;=&gt;\$Input1,</span><br><span class="line">&quot;2:s&quot;=&gt;\$Input2,</span><br><span class="line">&quot;o:s&quot;=&gt;\$Outdir</span><br><span class="line">);</span><br><span class="line"></span><br><span class="line">$Outdir=`pwd` unless (defined($Outdir));</span><br><span class="line">die `pod2text $0` if (!$Input1) or (!$Input2);</span><br><span class="line"></span><br><span class="line">open FQ1,&quot;&lt;$Input1&quot; or die &quot;Cann&apos;t open $Input1\n&quot;;</span><br><span class="line">open FQ2,&quot;&lt;$Input2&quot; or die &quot;Cann&apos;t open $Input2\n&quot;;</span><br><span class="line"></span><br><span class="line">my $Input1_basename=`basename $Input1`;</span><br><span class="line">chomp $Input1_basename;</span><br><span class="line">my $Input2_basename=`basename $Input2`;</span><br><span class="line">chomp $Input2_basename;</span><br><span class="line">open OUT1,&quot;&gt;$Outdir/$&#123;Input1_basename&#125;.match&quot; or die &quot;Cann&apos;t create $Outdir/$&#123;Input1_basename&#125;.match\n&quot;;</span><br><span class="line">open OUT2,&quot;&gt;$Outdir/$&#123;Input2_basename&#125;.match&quot; or die &quot;Cann&apos;t create $Outdir/$&#123;Input2_basename&#125;.match\n&quot;;</span><br><span class="line"></span><br><span class="line"># 载入两个fastq文件，保存成哈希</span><br><span class="line">my (%fq1_seq,%fq1_qua,%fq2_seq,%fq2_qua);</span><br><span class="line">while (!eof(FQ1))&#123;</span><br><span class="line">        my ($id,)=split(/\s/,&lt;FQ1&gt;);</span><br><span class="line">        $fq1_seq&#123;$id&#125;=&lt;FQ1&gt;;</span><br><span class="line">        &lt;FQ1&gt;;</span><br><span class="line">        $fq1_qua&#123;$id&#125;=&lt;FQ1&gt;;</span><br><span class="line">        chomp ($fq1_seq&#123;$id&#125;,$fq1_qua&#123;$id&#125;);</span><br><span class="line">&#125;</span><br><span class="line">while (!eof(FQ2))&#123;</span><br><span class="line">        my ($id,)=split(/\s/,&lt;FQ2&gt;);</span><br><span class="line">        $fq2_seq&#123;$id&#125;=&lt;FQ2&gt;;</span><br><span class="line">        &lt;FQ2&gt;;</span><br><span class="line">        $fq2_qua&#123;$id&#125;=&lt;FQ2&gt;;</span><br><span class="line">        chomp ($fq2_seq&#123;$id&#125;,$fq2_qua&#123;$id&#125;);</span><br><span class="line">&#125;</span><br><span class="line">close FQ1;</span><br><span class="line">close FQ2;</span><br><span class="line"></span><br><span class="line"># 双端配对</span><br><span class="line">foreach my $key (sort keys %fq1_seq)&#123;</span><br><span class="line">        if(defined($fq2_seq&#123;$key&#125;))&#123;</span><br><span class="line">                print OUT1 &quot;$key\n$fq1_seq&#123;$key&#125;\n+\n$fq1_qua&#123;$key&#125;\n&quot;;</span><br><span class="line">                print OUT2 &quot;$key\n$fq2_seq&#123;$key&#125;\n+\n$fq2_qua&#123;$key&#125;\n&quot;;</span><br><span class="line">        &#125;</span><br><span class="line">&#125;</span><br><span class="line">close OUT1;</span><br><span class="line">close OUT2;</span><br></pre></td></tr></table></figure>
<p>执行方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl PairsMate.pl -1 &lt;input1.fastq&gt; -2 &lt;input2.fastq&gt; [-o &lt;outdir&gt;]</span><br></pre></td></tr></table></figure>
<p>具体的参数使用说明可以执行 <code>perl PairsMate.pl</code>，查看脚本使用文档</p>
<h3 id="4-根据序列Id提取FASTA序列"><a href="#4-根据序列Id提取FASTA序列" class="headerlink" title="4. 根据序列Id提取FASTA序列"></a>4. 根据序列Id提取FASTA序列</h3><p>提供FASTA文件，和包含序列Id的文件（有多列，用制表符隔开，其中一列为序列Id），提取该序列Id对应的FASTA序列</p>
<p>脚本名：<code>extractSeqFromFasta.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">use strict;</span><br><span class="line">use warnings;</span><br><span class="line"></span><br><span class="line">my geneList = $ARGV[0];</span><br><span class="line">my fasta = $ARGV[1];</span><br><span class="line">my out = $ARGV[2];</span><br><span class="line"></span><br><span class="line">my %Hash_fasta;</span><br><span class="line">my $seqId;</span><br><span class="line"></span><br><span class="line"># 读入fasta文件，存为哈希</span><br><span class="line">open FA,&quot;&lt;$fasta&quot; or die &quot;$!\n&quot;;</span><br><span class="line">while(&lt;FA&gt;)&#123;</span><br><span class="line">	chomp;</span><br><span class="line">	next if(/^\s?$/);</span><br><span class="line">	if(/^&gt;(.+)$/)&#123;</span><br><span class="line">		$seqId = $1;</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		$Hash_fasta&#123;$seqId&#125; .= $_;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">close(FA);</span><br><span class="line"></span><br><span class="line"># 逐行读入geneList文件，并从上一步的哈希中将该序列提取出来</span><br><span class="line">open LIST,&quot;&lt;$geneList&quot; or die &quot;$!\n&quot;;</span><br><span class="line">open OUT,&quot;&gt;$out&quot; or die &quot;$!\n&quot;;</span><br><span class="line">while(&lt;LIST&gt;)&#123;</span><br><span class="line">	chomp;</span><br><span class="line">	next if(/^\s?$/);</span><br><span class="line">	my $gene = $_;</span><br><span class="line">	if($Hash_fasta&#123;$gene&#125;)&#123;</span><br><span class="line">		print OUT &quot;&gt;$gene\n$Hash_fasta&#123;$gene&#125;\n&quot;;</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">close(LIST);</span><br><span class="line">close(OUT);</span><br></pre></td></tr></table></figure>
<p>使用方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl extractSeqFromFasta.pl &lt;in.fasta&gt; &lt;gene list&gt; &lt;out.fasta&gt;</span><br></pre></td></tr></table></figure>
<h3 id="5-统计FastQC输出"><a href="#5-统计FastQC输出" class="headerlink" title="5. 统计FastQC输出"></a>5. 统计FastQC输出</h3><p>统计FastQC结果，注意脚本需要在FastQC输出结果所在目录下运行。脚本的统计项目及输出如下：</p>
<table>
<thead>
<tr>
<th style="text-align:left">Sample</th>
<th style="text-align:left">Total Reads</th>
<th style="text-align:left">GC Content</th>
<th style="text-align:left">Q20</th>
<th style="text-align:left">Q30</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">NP006_RRS05401_1.clean.fq.extract_fastqc</td>
<td style="text-align:left">39998924</td>
<td style="text-align:left">49</td>
<td style="text-align:left">0.999430935</td>
<td style="text-align:left">0.946251454</td>
</tr>
<tr>
<td style="text-align:left">NP006_RRS05401_2.clean.fq.extract_fastqc</td>
<td style="text-align:left">39998924</td>
<td style="text-align:left">49</td>
<td style="text-align:left">0.986292131</td>
<td style="text-align:left">0.883245809</td>
</tr>
<tr>
<td style="text-align:left">NP007_RRS05402_1.clean.fq.extract_fastqc</td>
<td style="text-align:left">40003323</td>
<td style="text-align:left">49</td>
<td style="text-align:left">0.999429311</td>
<td style="text-align:left">0.946197936</td>
</tr>
</tbody>
</table>
<p>脚本名：<code>fastqc_stat.pl</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">opendir (DIR, &quot;./&quot;) or die &quot;can&apos;t open the directory!&quot;;</span><br><span class="line">@dir = readdir DIR;</span><br><span class="line">foreach $file ( sort  @dir) </span><br><span class="line">&#123;</span><br><span class="line"></span><br><span class="line"># 跳过不需要的文件/文件夹，留下需要的文件夹</span><br><span class="line">next unless -d $file;</span><br><span class="line">next if $file eq &apos;.&apos;;</span><br><span class="line">next if $file eq &apos;..&apos;;</span><br><span class="line"></span><br><span class="line"># 提取total reads</span><br><span class="line">$total_reads=  `grep &apos;^Total&apos; ./$file/fastqc_data.txt`;</span><br><span class="line">$total_reads=(split(/\s+/,$total_reads))[2];</span><br><span class="line"># 提取%GC</span><br><span class="line">$GC= `grep &apos;%GC&apos; ./$file/fastqc_data.txt`;</span><br><span class="line">$GC=(split(/\s+/,$GC))[1];</span><br><span class="line">chomp $GC;</span><br><span class="line"></span><br><span class="line"># 提取Q20，Q30</span><br><span class="line">## 读入Per sequence quality scores部分的信息，保存成哈希</span><br><span class="line">open FH , &quot;&lt;./$file/fastqc_data.txt&quot;;</span><br><span class="line">while (&lt;FH&gt;)</span><br><span class="line">    &#123;</span><br><span class="line">    next unless /#Quality/;</span><br><span class="line">    while (&lt;FH&gt;)</span><br><span class="line">        &#123;</span><br><span class="line">        @F=split;</span><br><span class="line">        $hash&#123;$F[0]&#125;=$F[1];</span><br><span class="line">        last if /&gt;&gt;END_MODULE/;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">## 统计Q20，Q30</span><br><span class="line">$all=0;$Q20=0;$Q30=0;</span><br><span class="line">$all+=$hash&#123;$_&#125; foreach keys %hash;</span><br><span class="line">$Q20+=$hash&#123;$_&#125; foreach 0..20;</span><br><span class="line">$Q30+=$hash&#123;$_&#125; foreach 0..30;</span><br><span class="line">$Q20=1-$Q20/$all;</span><br><span class="line">$Q30=1-$Q30/$all;</span><br><span class="line">print &quot;$file\t$total_reads\t$GC\t$Q20\t$Q30\n&quot;;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>执行方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl fastqc_stat.pl</span><br></pre></td></tr></table></figure>
<h2 id="Python"><a href="#Python" class="headerlink" title="Python"></a>Python</h2><h3 id="1-格式化FASTA文件"><a href="#1-格式化FASTA文件" class="headerlink" title="1. 格式化FASTA文件"></a>1. 格式化FASTA文件</h3><p>在规范的Fasta文件中，你看到的一条序列记录包括两部分：</p>
<ul>
<li>以<code>&gt;</code>起始的序列名称，占一行；</li>
<li>由核苷酸ATCGN（核酸序列）或氨基酸字符组成的字符串，一般每60个字符一行，若一条序列很长，那么它可能会占多行；</li>
</ul>
<p>有的时候因为一些原因（一般都是自己在上游分析时图方便生成的）你得到的fasta文件中的序列部分没有按照60个字符一行的形式进行组织，而是将整条序列放在一行里，虽然一般来说这并不会对你的分析产生太大的影响，但是进行查看的时候会有一些不方便，比如</p>
<p>这个时候如果想将它调整组成规范的格式，要怎么实现呢？</p>
<p>用BioPython将原始FASTA文件读入，然后在写出，就能得到你想要的效果</p>
<p>这个脚本很简单，只有四行代码</p>
<p>脚本名：<code>formatFasta.py</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">from Bio import SeqIO</span><br><span class="line">import sys</span><br><span class="line"></span><br><span class="line">Seq = [ seq for seq in SeqIO.parse(sys.argv[1],&apos;fasta&apos;)]</span><br><span class="line"></span><br><span class="line">SeqIO.write(Seq,sys.argv[2],&apos;fasta&apos;)</span><br></pre></td></tr></table></figure>
<p>用法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ python formatFasta.py &lt;in.fa&gt; &lt;out.fa&gt;</span><br></pre></td></tr></table></figure>
<h2 id="R"><a href="#R" class="headerlink" title="R"></a>R</h2><h3 id="1-合并多个样本的定量结果成一个大矩阵-profile-matrix"><a href="#1-合并多个样本的定量结果成一个大矩阵-profile-matrix" class="headerlink" title="1. 合并多个样本的定量结果成一个大矩阵 (profile matrix)"></a>1. 合并多个样本的定量结果成一个大矩阵 (profile matrix)</h3><p>在RNAseq或者metagenome shotgun中，需要对每个样本进行逐一定量，生成每个样本各种的定量结果，然后需要合并每个样本的定量结果，形成一个包含所有样本定量的profile matrix</p>
<p>我们这个脚本的要实现的就是合并多个样本的定量结果成一个大矩阵 (profile matrix)</p>
<p>实现思路：</p>
<blockquote>
<ul>
<li><p>想要保证所有样本的定量结果文件都保存在一个文件夹下，且每个文件的命名形式为：<code>sample+suffix</code>，即<code>样本名+固定后缀</code>；</p>
</li>
<li><p>每个文件中只有两列信息，第一列为feature（对于RNAseq，它的feature是gene或transcript；对于metagenome shotgun，它的feature是各个物种分类层级），第二列是定量结果（quant，对于RNAseq，它的定量可以是reads count、FPKM/RPKM、TPM等；对于metagenome shotgun，它的定量一般是一个取值在[0,1]的比例值）；</p>
</li>
<li><p>根据指定的文件夹（下面记作<code>dir</code>）和文件后缀（下面记作<code>pattern</code>），将文件路径为<code>dir/*pattern</code>的文件逐一读入，获取它们的第一列，即features的列表，从而得到total unique features的列表，同时记下样本列表（样本名为：<code>文件名-pattern</code>）；</p>
</li>
<li><p>根据上一步得到的total unique features列表和样本列表，创建一个行数等于total unique features列表长度，列数等于样本列表的矩阵（记作$\text{MergeMatrix}_{ij}$，i表示$\text{feature}_i$，j表示$\text{sample}^j$），行名对应total unique features列表，列名对应样本列表，矩阵中每个元素的值初始化为0；</p>
</li>
<li><p>再一次将文件路径为<code>dir/*pattern</code>的文件逐一读入，更新MergeMatrix对应列的值：若当前读入的文件名为<code>sample_j.pattern</code>，则当前样本为$\text{sample}^j$，则根据$\text{MergeMatrix}_{ij}=\text{sample}<em>i^j$，对矩阵$\text{MergeMatrix}</em>{ij}$进行更新；</p>
</li>
</ul>
</blockquote>
<p>脚本名：<code>Merge2ProfileMatrix.R</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">loadMatrix &lt;- function(dir,pattern)&#123;</span><br><span class="line">	files &lt;- list.files(dir)</span><br><span class="line">	clones &lt;- vector()	# 用于保存所有样本中出现的克隆</span><br><span class="line">	samples &lt;- vector()	# 用于保存所有样本的id</span><br><span class="line">	# 获取所有样本中出现的unique克隆，与所有样本的id</span><br><span class="line">	for (file in files) &#123;</span><br><span class="line">		if (grepl(pattern,file))&#123;</span><br><span class="line">			data &lt;- read.table(paste(dir,file,sep=&quot;/&quot;),header=F,sep=&quot;\t&quot;)</span><br><span class="line">			clones &lt;- c(clones,as.character(data$V1))</span><br><span class="line">			samples &lt;- c(samples,sub(pattern,&quot;&quot;,file))</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	# 创建用于保存所有样本整合的表达谱，以0填充</span><br><span class="line">	clones &lt;- unique(clones)</span><br><span class="line">	dataMat &lt;- matrix(rep(0,length(samples)*length(clones)),nrow = length(clones),ncol = length(samples))</span><br><span class="line">	rownames(dataMat) &lt;- clones</span><br><span class="line">	colnames(dataMat) &lt;- samples</span><br><span class="line">	# 逐一读入样本的count文件，对matrix中对应的元素进行赋值</span><br><span class="line">	for (file in files) &#123;</span><br><span class="line">		if (grepl(pattern,file))&#123;</span><br><span class="line">			data &lt;- read.table(paste(dir,file,sep=&quot;/&quot;),header=F,sep=&quot;\t&quot;)</span><br><span class="line">			sample &lt;- as.character(sub(pattern,&quot;&quot;,file))</span><br><span class="line">			index &lt;- match(data$V1,rownames(dataMat))</span><br><span class="line">			dataMat[index,sample] &lt;- data$V2</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	dataMat</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>该脚本只定义了一个函数，可以在R交互模式下，载人这个脚本中的函数，然后调用它，或者直接对这个脚本进行修改，然后用<code>Rscript</code>直接执行修改后的脚本</p>
<h3 id="2-对profiles进行wilcox检验（并行化）"><a href="#2-对profiles进行wilcox检验（并行化）" class="headerlink" title="2. 对profiles进行wilcox检验（并行化）"></a>2. 对profiles进行wilcox检验（并行化）</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br></pre></td><td class="code"><pre><span class="line"># 四个参数，按顺序分别为：</span><br><span class="line"># - profiles matrix file</span><br><span class="line"># - sample group file </span><br><span class="line"># - outfile name</span><br><span class="line"># - threads number（默认为服务器总线程-10）</span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(T)</span><br><span class="line"></span><br><span class="line">library(parallel)</span><br><span class="line"></span><br><span class="line"># Initiate cluster</span><br><span class="line">if(is.na(Args[4])) &#123;</span><br><span class="line">	no_cores &lt;- detectCores() - 10</span><br><span class="line">&#125; else &#123;</span><br><span class="line">	no_cores &lt;- as.integer(Args[4])</span><br><span class="line">&#125;</span><br><span class="line">cl &lt;- makeCluster(no_cores)</span><br><span class="line"></span><br><span class="line">data &lt;- read.table(Args[1],head=T,row.names=1,sep=&apos;\t&apos;)</span><br><span class="line">group &lt;- read.table(Args[2],head=T,sep=&apos;\t&apos;)</span><br><span class="line"></span><br><span class="line"># 找出profiles中每列代表的样本所属的组</span><br><span class="line">col_group1 &lt;- colnames(data) %in% sub(&apos;-&apos;,&apos;.&apos;,paste(group$Sample[group$Group==1],&apos;.rpkm&apos;,sep=&apos;&apos;))</span><br><span class="line">col_group2 &lt;- colnames(data) %in% sub(&apos;-&apos;,&apos;.&apos;,paste(group$Sample[group$Group==2],&apos;.rpkm&apos;,sep=&apos;&apos;))</span><br><span class="line"></span><br><span class="line"># 对单个基因进行wilcox检验，并输出统计检验结果，输出格式如下：</span><br><span class="line"># 	geneName,pvalue,group1Median,group2Median,direction</span><br><span class="line">wilcox_fun &lt;- function(data,col_group1,col_group2)&#123;</span><br><span class="line">	g1 &lt;- as.numeric(data[col_group1])</span><br><span class="line">	g2 &lt;- as.numeric(data[col_group2])</span><br><span class="line">	# 执行wilcox检验</span><br><span class="line">	stat &lt;- wilcox.test(g1,g2,paired = FALSE, exact=NULL, correct=TRUE, alternative=&quot;two.sided&quot;)</span><br><span class="line">	if(is.na(stat$p.value))&#123;</span><br><span class="line">		stat$p.value &lt;- 1</span><br><span class="line">	&#125;</span><br><span class="line">	# 对有统计学意义的基因进行判断，是上调&quot;up&quot;还是下调&quot;down&quot;或者是不变&quot;-&quot;（对于组2，即不吃药组）</span><br><span class="line">	if(stat$p.value &lt; 0.1  &amp; median(g1) &lt; median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;down&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else if(stat$p.value &lt; 0.1  &amp; median(g1) &gt; median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;up&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else if(stat$p.value &lt; 0.1  &amp; median(g1) == median(g2))&#123;</span><br><span class="line">		G12 &lt;- c(ifelse(is.na(stat$p.value),1,stat$p.value),median(g1),median(g2),&apos;-&apos;)</span><br><span class="line">	&#125;</span><br><span class="line">	else&#123;</span><br><span class="line">		G12 &lt;- c()</span><br><span class="line">	&#125;</span><br><span class="line">	G12</span><br><span class="line">&#125;</span><br><span class="line">	</span><br><span class="line">statOut &lt;- parApply(cl,data,1,wilcox_fun,col_group1,col_group2)</span><br><span class="line">stopCluster(cl)</span><br><span class="line"># 删除为NULL的列表元素</span><br><span class="line">for(i in names(statOut))&#123;</span><br><span class="line">	if(is.null(statOut[[i]]))&#123;</span><br><span class="line">		statOut[[i]] &lt;- NULL</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"># 将结果写入文件中</span><br><span class="line">if(!is.null(statOut))&#123;</span><br><span class="line">	# 转换成数据框</span><br><span class="line">	statOut &lt;- as.data.frame(t(as.matrix(as.data.frame(statOut))))</span><br><span class="line">	colnames(statOut) &lt;- c(&apos;pvalue&apos;,&apos;group1Median&apos;,&apos;group2Median&apos;,&apos;direction&apos;)</span><br><span class="line">	# 写入文件</span><br><span class="line">	write.table(statOut,Args[3],sep = &apos;\t&apos;,row.names = T,col.names = T,quote = F)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<h3 id="3-对差异显著的subprofile进行可视化——箱线图与热图"><a href="#3-对差异显著的subprofile进行可视化——箱线图与热图" class="headerlink" title="3. 对差异显著的subprofile进行可视化——箱线图与热图"></a>3. 对差异显著的subprofile进行可视化——箱线图与热图</h3><p>想法：</p>
<blockquote>
<p>从上游的分析中得到两种样本profile的差异统计检验结果，其中有p-value和p-adjust，可以根据p-value或p-adjust设置阈值，筛选出有统计学显著性的差异observation</p>
<p>根据这些observation的Id从原始profile中将这些差异observation的部分提取了得到subprofile</p>
<p>对subprofile进行可视化——箱线图（boxplot）与热图（heatmap）</p>
</blockquote>
<p>需要提供的输入：</p>
<blockquote>
<ul>
<li><p>上游的分析中得到两种样本profile的差异统计检验结果</p>
</li>
<li><p>原始profile</p>
</li>
<li><p>样本的分组信息</p>
</li>
</ul>
</blockquote>
<p>脚本名：<code>plotDiffSubprofile.R</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br></pre></td><td class="code"><pre><span class="line"># 实现功能说明：导入差异分析的结果与原始profile，提取原始profile中提取</span><br><span class="line">#	差异显著（通过第3个参数设置筛选的阈值）的集合绘制热图/箱线图</span><br><span class="line">#</span><br><span class="line">#	考虑到可能直接得到的画图输出不满足用户的需求，所以提供了一个参数来控</span><br><span class="line">#	制是否输出画图用的数据，以便用户可以直接载人绘图数据，根据自己的需求</span><br><span class="line">#	灵活画图</span><br><span class="line">#########################################################################</span><br><span class="line"># 必须参数说明：</span><br><span class="line"># - output file from wilcox stat</span><br><span class="line"># - profiles matrix file</span><br><span class="line"># - padj threshold</span><br><span class="line"># - min memmbers for subprofile</span><br><span class="line"># - sample group file, 2 columns, 1st column corresponding to sample, 2nd</span><br><span class="line">#	 column corresponding to group</span><br><span class="line"># - logic value, False or True, whether to output prepared data frame for</span><br><span class="line">#	 ggplot-boxplot and filted subprofile for heatmap, saved as raw txt and</span><br><span class="line">#	 Rdata(subprofile_melt--boxplot, subprofile_matrix,anno_col--heatmap) formats</span><br><span class="line">#########################################################################</span><br><span class="line"></span><br><span class="line">library(reshape2)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(pheatmap)</span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(T)</span><br><span class="line"></span><br><span class="line">stat &lt;- read.table(Args[1],header=T,sep=&apos;\t&apos;)</span><br><span class="line">profile &lt;- read.table(Args[2],head=T,sep=&apos;\t&apos;)</span><br><span class="line">padj &lt;- as.numeric(Args[3])</span><br><span class="line">num_min &lt;- as.numeric(Args[4])</span><br><span class="line">group &lt;- read.table(Args[5],head=T,sep=&apos;\t&apos;)</span><br><span class="line"></span><br><span class="line">stat_filt &lt;- stat[stat$pAdj&lt;=padj,]</span><br><span class="line"></span><br><span class="line"># 若差异的集合太小则不进行后续的绘图操作</span><br><span class="line">if(dim(stat_filt)[1]&gt;=num_min)&#123;</span><br><span class="line">	# 提取出差异的subprofile</span><br><span class="line">	subprofile &lt;- profile[profile[,1] %in% rownames(stat_filt),]</span><br><span class="line">	</span><br><span class="line">	# 将数据框的短格式展开为长格式</span><br><span class="line">	subprofile_melt &lt;- melt(subprofile,id.vars=1,variable.name = &apos;sample&apos;,value.name = &apos;abundance&apos;)</span><br><span class="line">	colnames(subprofile_melt) &lt;- c(&apos;ID&apos;,&apos;sample&apos;,&apos;abundance&apos;)</span><br><span class="line">	</span><br><span class="line">	# 添加分组信息</span><br><span class="line">	subprofile_melt$group &lt;- NULL</span><br><span class="line">	group_factor &lt;- unique(group[,2])	# 获得所有组别及其表示符</span><br><span class="line">	for(i in group_factor)&#123;</span><br><span class="line">		subprofile_melt$group[subprofile_melt$sample %in% sub(&apos;-&apos;,&apos;.&apos;,group[group[,2]==i,1])] &lt;- i</span><br><span class="line">	&#125;</span><br><span class="line">	subprofile_melt$group &lt;- as.factor(subprofile_melt$group) # 分组变量必须为factor</span><br><span class="line">	</span><br><span class="line">	# 画boxplot</span><br><span class="line">	# png(paste(Args[2],&quot;.boxplotDiff.png&quot;,sep=&apos;&apos;))</span><br><span class="line">	p_box &lt;- ggplot(subprofile_melt)</span><br><span class="line">	p_box + geom_boxplot(aes(x=ID,y=abundance,fill=group))</span><br><span class="line">	# dev.off()</span><br><span class="line">	ggsave(paste(Args[2],&quot;.boxplotDiff.png&quot;,sep=&apos;&apos;))</span><br><span class="line">	</span><br><span class="line">	# 画heatmap</span><br><span class="line">	## 整理绘图数据</span><br><span class="line">	subprofile_matrix &lt;- as.matrix(subprofile[,-1])</span><br><span class="line">	rownames(subprofile_matrix) &lt;- subprofile[,1]</span><br><span class="line">	## 准备heatmap的注释信息</span><br><span class="line">	anno_col &lt;- data.frame(group=as.factor(group[,2]))</span><br><span class="line">	rownames(anno_col) &lt;- sub(&apos;-&apos;,&apos;.&apos;,group[,1])</span><br><span class="line">	png(paste(Args[2],&quot;.heatmapDiff.png&quot;,sep=&apos;&apos;))</span><br><span class="line">	pheatmap(subprofile_matrix,scale=&quot;row&quot;,cluster_rows=T,cluster_cols=T,annotation_col=anno_col,show_rownames = F)</span><br><span class="line">	dev.off()</span><br><span class="line">	</span><br><span class="line">	# 若最后一个参数为True则写出数据</span><br><span class="line">	if(Args[6])&#123;</span><br><span class="line">		write.table(subprofile_melt,paste(Args[2],&quot;.boxplotDiff.data&quot;,sep=&apos;&apos;),row.names=F,col.names=T,sep=&apos;\t&apos;,quote=F) # 画箱线图的数据</span><br><span class="line">		write.table(subprofile_matrix,paste(Args[2],&quot;.heatmapDiff.data&quot;,sep=&apos;&apos;),row.names=T,col.names=T,sep=&apos;\t&apos;,quote=F) # 画热图的数据</span><br><span class="line">		write.table(anno_col,paste(Args[2],&quot;.heatmapDiff.anno&quot;,sep=&apos;&apos;),row.names=T,col.names=T,sep=&apos;\t&apos;,quote=F) # 画热图的注释数据</span><br><span class="line">		save(subprofile_melt,subprofile_matrix,anno_col,file=paste(Args[2],&quot;.Rdata&quot;,sep=&apos;&apos;))</span><br><span class="line">	&#125;</span><br><span class="line">&#125;else&#123;</span><br><span class="line">	print(&quot;Not enough memmbers of subprofile passing the padj threshold!&quot;)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<h3 id="4-对大矩阵计算Jaccard-Index（并行化-分治）"><a href="#4-对大矩阵计算Jaccard-Index（并行化-分治）" class="headerlink" title="4. 对大矩阵计算Jaccard Index（并行化+分治）"></a>4. 对大矩阵计算Jaccard Index（并行化+分治）</h3><p>在对一个大矩阵执行相关性计算或Jaccard Index的计算时，其实执行的是矩阵任意两行（这里假设要进行分析的对象是矩阵的每个行）之间的两两的计算，若这个矩阵的规模非常庞大，有n行时，计算的时间复杂度就是$O(n^2)$，这个时候可以采用并行化策略来加速这个进程（参考上文的 <a href="#r-parallel">2. R中的并行化方法</a>）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">StatOut &lt;- parApply(cl, data, 1, fun, data)</span><br></pre></td></tr></table></figure>
<p>这样就会实现将一个 <code>n vs. n</code> 的问题拆分成 n 个可以并行解决的 <code>1 vs. n</code> 的问题，当然通过设置线程数为$m,\,(m\le n)$，使得每次只并行执行m个 <code>1 vs. n</code> 的问题</p>
<p>然后再在函数<code>fun</code>内部再定义一个并行化计算来进一步并行化加速上面产生的 <code>1 vs. n</code> 的计算：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">fun &lt;- function(vec, data)&#123;</span><br><span class="line">	...</span><br><span class="line">	parApply(cl, data, 1, fun2, vec)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>在这个函数内部实现了将一个 <code>1 vs. n</code> 的问题拆分成 n 个可以并行解决的 <code>1 vs. 1</code> 的问题</p>
<p>这样就实现了两步并行化，这样在保证硬件条件满足的情况下，的确能显著加快分析速度</p>
<p>但是并行化技术会带来一个问题是，虽然时间开销减少了，但是<strong>空间开销显著增加了</strong></p>
<blockquote>
<p>比如，第一次并行化，将一个 <code>n vs. n</code> 的问题拆分成 $\frac{n}{m}$ 次可以并行解决的 m个 <code>1 vs. n</code> 的问题，则需要在每一次并行化任务中拷贝一个 <code>1 vs. n</code> 的计算对象，即原始有n行的矩阵被拷贝了m次，则相应的缓存空间也增加了m倍，很显然内存的占用大大增加了</p>
</blockquote>
<p>空间开销显著增加带来的后果就是，很容易导致运行内存不足程序运行中断的问题，那该怎么解决这个问题呢？</p>
<p>可以采用分治方法（Divide-Conquer)，将原始大矩阵，按照行拆分成多个小的子块，对每个子块执行计算，从而得到每个子块的运算结果，最后再讲每个子块的结果进行合并：</p>
<p>脚本名：<code>JaccardIndex.R</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">###########################################################</span><br><span class="line"># 	该脚本用于对Observed矩阵（行为Feature，列为Sample）</span><br><span class="line"># 对feature（即行）计算Jaccard Index，可以进行并行化计算</span><br><span class="line"># 	考虑到原始矩阵可能过大（主要是行太多），如果直接对</span><br><span class="line"># 它执行并行化计算的话，非常容易导致内存溢出，为了解决</span><br><span class="line"># 这个问题，引入了分治方法：</span><br><span class="line"># 	将原始矩阵按照行等分n份（平均每份有abs(N/n)，最后一</span><br><span class="line"># 份可能无法进行等分，最后一份为N%n），然后对这n份的每两</span><br><span class="line"># 份之间分别执行并行化计算，则总共有n^2次，得到n^2个子结</span><br><span class="line"># 果进行合并</span><br><span class="line">###########################################################</span><br><span class="line"></span><br><span class="line"># 参数说明：</span><br><span class="line"># （1）Observed Matrix，每行表示一个feature，每列表示一个样本</span><br><span class="line"># （2）并行化使用的线程数</span><br><span class="line"># （3）切分的分数（默认不切块）</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">library(parallel)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Args &lt;- commandArgs(TRUE)</span><br><span class="line"></span><br><span class="line">inFile &lt;- Args[1]</span><br><span class="line"></span><br><span class="line"># 开启并行化</span><br><span class="line">if(!is.na(Args[2]))&#123;</span><br><span class="line">	no_cores &lt;- as.integer(Args[2])</span><br><span class="line">	cl &lt;- makeCluster(no_cores)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">data &lt;- read.table(inFile,header=T,row.names=1)</span><br><span class="line"></span><br><span class="line">print(&quot;Load Observed Matrix Successfully&quot;)</span><br><span class="line"></span><br><span class="line"># 计算Jaccard Index</span><br><span class="line">JaccardIndexSer &lt;- function(SourceVec, TargetMatrix)&#123;</span><br><span class="line">	JaccardIndexOne &lt;- function(SourceVec, TargetVec)&#123;</span><br><span class="line">		sum(SourceVec == TargetVec)/length(SourceVec)</span><br><span class="line">	&#125;</span><br><span class="line">	apply(TargetMatrix, 1, JaccardIndexOne, SourceVec)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"># 执行分治方法</span><br><span class="line">if(!is.na(Args[3]))&#123;</span><br><span class="line">	n_row &lt;- nrow(data)</span><br><span class="line">	nblock &lt;- Args[3]</span><br><span class="line">	nrow_block &lt;- ceiling(n_row/nblock)</span><br><span class="line">	StatOutList &lt;- vector(&quot;list&quot;, nblock)</span><br><span class="line">	</span><br><span class="line">	# 1. 开始进行分块计算</span><br><span class="line">	print(&quot;[Divide-Conquer]Start carry out Divide-Conquer for Large Observed Matrix&quot;)</span><br><span class="line">	print(&quot;##################################################&quot;)</span><br><span class="line">	for(i in 1:nblock)&#123;</span><br><span class="line">		for(j in 1:nblock)&#123;</span><br><span class="line">			nrow_start &lt;- (i-1)*nrow_block+1</span><br><span class="line">			nrow_end &lt;- i*nrow_block</span><br><span class="line">			# 并行化计算</span><br><span class="line">			if(!is.na(Args[2]))&#123;</span><br><span class="line">				print(paste(&quot;[Divide-Conquer]Start carry out statistic Jaccard Index parallel for block: &quot;,i,&quot;-&quot;,j,sep=&apos;&apos;))</span><br><span class="line">				StatOutList[[i]] &lt;- append(StatOutList[[i]], parApply(cl, data[nrow_start:nrow_end,], 1 , JaccardIndexSer, data))</span><br><span class="line">				print(paste(&quot;[Divide-Conquer]Finish run parallel for block: &quot;,i,&quot;-&quot;,j,sep=&apos;&apos;))</span><br><span class="line">			# 串行计算</span><br><span class="line">			&#125;else&#123;</span><br><span class="line">				print(paste(&quot;[Divide-Conquer]Start carry out statistic Jaccard Index serially for block: &quot;,i,&quot;-&quot;,j,sep=&apos;&apos;))</span><br><span class="line">				StatOutList[[i]] &lt;- append(StatOutList[[i]], apply(data, 1 , JaccardIndexSer, data))</span><br><span class="line">				print(paste(&quot;[Divide-Conquer]Finish run serially for block: &quot;,i,&quot;-&quot;,j,sep=&apos;&apos;))</span><br><span class="line">			&#125;</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	# 2. 结束分治方法的分块计算</span><br><span class="line">	if(!is.na(Args[2]))&#123;</span><br><span class="line">		print(&quot;##################################################&quot;)</span><br><span class="line">		print(&quot;[Divide-Conquer]Finish parallel running for statistic Jaccard Index!&quot;)</span><br><span class="line">		stopCluster(cl)</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		print(&quot;##################################################&quot;)</span><br><span class="line">		print(&quot;[Divide-Conquer]Finish serial running for statistic Jaccard Index!&quot;)</span><br><span class="line">	&#125;</span><br><span class="line">	# 3. 开始进行子块结果的合并</span><br><span class="line">	print(&quot;[Divide-Conquer]Start bind sub-block statout&quot;)</span><br><span class="line">	StatOut &lt;- vector(&quot;list&quot;, nblock)</span><br><span class="line">	# 先对列进行合并</span><br><span class="line">	for(i in 1:nblock)&#123;</span><br><span class="line">		for(block in StatOutList[[i]])&#123;</span><br><span class="line">			StatOut[[i]] &lt;- cbind(StatOut[[i]], block)</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	# 再对行进行合并</span><br><span class="line">	StatOutMerge &lt;- data.frame()</span><br><span class="line">	for(block in StatOut)&#123;</span><br><span class="line">		StatOutMerge &lt;- rbind(StatOutMerge, block)</span><br><span class="line">	&#125;</span><br><span class="line">	StatOut &lt;- StatOutMerge</span><br><span class="line"># 不执行分治方法</span><br><span class="line">&#125;else&#123;	</span><br><span class="line"># 并行化计算</span><br><span class="line">	if(!is.na(Args[2]))&#123;</span><br><span class="line">		print(&quot;[Common]Start carry out statistic Jaccard Index parallel&quot;)</span><br><span class="line">		StatOut &lt;- parApply(cl, data, 1 , JaccardIndexSer, data)</span><br><span class="line">		print(&quot;[Common]Finish run parallel&quot;)</span><br><span class="line">		stopCluster(cl)</span><br><span class="line">	# 串行计算</span><br><span class="line">	&#125;else&#123;</span><br><span class="line">		print(&quot;[Common]Start carry out statistic Jaccard Index serially&quot;)</span><br><span class="line">		StatOut &lt;- apply(data, 1 , JaccardIndexSer, data)</span><br><span class="line">		print(&quot;[Common]Finish run serially&quot;)</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line">save(StatOut,&apos;JaccardIndex.Rdata&apos;)</span><br></pre></td></tr></table></figure>
<h3 id="5-合并两个矩阵"><a href="#5-合并两个矩阵" class="headerlink" title="5. 合并两个矩阵"></a>5. 合并两个矩阵</h3><p>在什么情节下，我们需要合并两个矩阵？</p>
<p>对于两个批次的数据，我们分别用上面 <a href="#1-合并多个样本的定量结果成一个大矩阵-profile-matrix">1. 合并多个样本的定量结果成一个大矩阵 (profile matrix)</a>得到了两个profile matrix，这两个matrix的行表示feature，列表示sample，样本一般不会有重复，而features大部分是重叠的，也有少部分是不重叠的，那么将这两个矩阵合并之后，这个新矩阵的行数为两个矩阵features的并集的大小，列数为两个矩阵列数的和</p>
<p>脚本名：<code>MergeTwoMat.R</code></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">Args &lt;- commandArgs(T)</span><br><span class="line"></span><br><span class="line">Matrix1 &lt;- read.table(Args[1], header=T, row.names=1)</span><br><span class="line">Matrix2 &lt;- read.table(Args[2], header=T, row.names=1)</span><br><span class="line"></span><br><span class="line"># 获取两个矩阵feature的并集</span><br><span class="line">MergeFeatureList &lt;- as.character(union(rownames(Matrix1),rownames(Matrix2)))</span><br><span class="line"># 获取两个矩阵的sample的并集（默认没有重叠）</span><br><span class="line">ColList &lt;- as.character(c(colnames(Matrix1), colnames(Matrix2)))</span><br><span class="line"></span><br><span class="line"># 初始化合并后的新矩阵，行数为feature并集大小，列数为sample并集大小</span><br><span class="line">MergeMatrix &lt;- matrix(rep(0,length(MergeFeatureList)*length(ColList)), nrow = length(MergeFeatureList), ncol = length(ColList))</span><br><span class="line">rownames(MergeMatrix) &lt;- MergeCloneList</span><br><span class="line">colnames(MergeMatrix) &lt;- ColList</span><br><span class="line"></span><br><span class="line"># 对第一个矩阵中对应的列（即样本）进行更新</span><br><span class="line">index &lt;- match(rownames(Matrix1), MergeCloneList)</span><br><span class="line">for(col in colnames(Matrix1))&#123;</span><br><span class="line">	MergeMatrix[index, col] &lt;- Matrix1[,col]</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"># 对第二个矩阵中对应的列（即样本）进行更新</span><br><span class="line">index &lt;- match(rownames(Matrix2), MergeCloneList)</span><br><span class="line">for(col in colnames(Matrix2))&#123;</span><br><span class="line">	MergeMatrix[index, col] &lt;- Matrix2[,col]</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>用法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ Rscript MergeTwoMat.R &lt;matrix1&gt; &lt;matrix2&gt;</span><br></pre></td></tr></table></figure>
      
    </div>

    

    
    
    

    

    
      
    
    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioinformatics/" rel="tag"># Bioinformatics</a>
          
            <a href="/tags/R/" rel="tag"># R</a>
          
            <a href="/tags/Python/" rel="tag"># Python</a>
          
            <a href="/tags/Perl/" rel="tag"># Perl</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/02/17/MeRIP-seq/" rel="next" title="MeRIP/ChIP-seq分析流程">
                <i class="fa fa-chevron-left"></i> MeRIP/ChIP-seq分析流程
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/02/24/Follow-LYL-learning-math-physics/" rel="prev" title="跟着李永乐老师学数学/物理">
                跟着李永乐老师学数学/物理 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      
        <div id="gitment-container"></div>
      
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/images/avatar.jpg" alt="Lianm">
            
              <p class="site-author-name" itemprop="name">Lianm</p>
              <p class="site-description motion-element" itemprop="description">中国科学院北京基因组研究所研究生</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">44</span>
                    <span class="site-state-item-name">日志</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">9</span>
                    <span class="site-state-item-name">分类</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">36</span>
                    <span class="site-state-item-name">标签</span>
                  </a>
                </div>
              
            </nav>
          

          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://github.com/Ming-Lian" title="GitHub &rarr; https://github.com/Ming-Lian" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="mailto:lianming17m@big.ac.cn" title="E-Mail &rarr; mailto:lianming17m@big.ac.cn" rel="noopener" target="_blank"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://www.jianshu.com/u/42e711970fa8" title="简书 &rarr; https://www.jianshu.com/u/42e711970fa8" rel="noopener" target="_blank"><i class="fa fa-fw fa-jianshu"></i>简书</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://www.zhihu.com/people/understorm_lianm" title="知乎 &rarr; https://www.zhihu.com/people/understorm_lianm" rel="noopener" target="_blank"><i class="fa fa-fw fa-zhihu"></i>知乎</a>
                </span>
              
            </div>
          

          

          
          

          
            
          
          

        </div>
      </div>

      
      <!--noindex-->
        <div class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
            
            
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#Perl"><span class="nav-text">Perl</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-拆分FASTA文件"><span class="nav-text">1. 拆分FASTA文件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-从双端FASTQ文件中抽取指定数据量-bp-的序列"><span class="nav-text">2. 从双端FASTQ文件中抽取指定数据量(bp)的序列</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-双端FASTQ文件进行双端配对"><span class="nav-text">3. 双端FASTQ文件进行双端配对</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4-根据序列Id提取FASTA序列"><span class="nav-text">4. 根据序列Id提取FASTA序列</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-统计FastQC输出"><span class="nav-text">5. 统计FastQC输出</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Python"><span class="nav-text">Python</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-格式化FASTA文件"><span class="nav-text">1. 格式化FASTA文件</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#R"><span class="nav-text">R</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-合并多个样本的定量结果成一个大矩阵-profile-matrix"><span class="nav-text">1. 合并多个样本的定量结果成一个大矩阵 (profile matrix)</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-对profiles进行wilcox检验（并行化）"><span class="nav-text">2. 对profiles进行wilcox检验（并行化）</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-对差异显著的subprofile进行可视化——箱线图与热图"><span class="nav-text">3. 对差异显著的subprofile进行可视化——箱线图与热图</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4-对大矩阵计算Jaccard-Index（并行化-分治）"><span class="nav-text">4. 对大矩阵计算Jaccard Index（并行化+分治）</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-合并两个矩阵"><span class="nav-text">5. 合并两个矩阵</span></a></li></ol></li></ol></div>
            

          </div>
        </div>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Lianm</span>

  

  
</div>


  <div class="powered-by">由 <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> 强力驱动 v3.8.0</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Mist</a> v7.0.0</div>




        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>














  
    
    
  
  <script color="0,0,255" opacity="0.5" zindex="-1" count="99" src="/lib/canvas-nest/canvas-nest.min.js"></script>



  
  











  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script src="/lib/three/three.min.js"></script>

  
  <script src="/lib/three/three-waves.min.js"></script>


  


  <script src="/js/src/utils.js?v=7.0.0"></script>

  <script src="/js/src/motion.js?v=7.0.0"></script>



  
  


  <script src="/js/src/schemes/muse.js?v=7.0.0"></script>




  
  <script src="/js/src/scrollspy.js?v=7.0.0"></script>
<script src="/js/src/post-details.js?v=7.0.0"></script>



  


  <script src="/js/src/bootstrap.js?v=7.0.0"></script>



  


  
    <!-- LOCAL: You can save these files to your site and update links -->

  
  <script src="https://cdn.jsdelivr.net/gh/theme-next/theme-next-gitment@1/gitment.browser.js"></script>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/theme-next/theme-next-gitment@1/default.css">
<!-- END LOCAL -->

<style>
#gitment-container a {
  border-bottom: none;
}

</style>

<script>
  function renderGitment() {
    var gitment = new Gitment({
      id: window.location.pathname,
      owner: 'Ming-Lian',
      repo: 'Gitment-repo',
      
      oauth: {
      
      
        client_secret: '3dc34ff32bbb0ee876b487f13a800b08857e0a10',
      
        client_id: '0e14870b1fb232903713'
      }
    });
    gitment.render('gitment-container');
  }

  
    renderGitment();
  
</script>

  


  
  <script>
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x"></i></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x"></i></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  

  
  

  


  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js'; 
      }
      else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
      }
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  

  

  

  

  

  

  

</body>
</html>
